{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Extended RNA-Seq Analysis Training Demo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For simplicity and time, The short tutorial workflow uses truncated and partial run data from the Cushman et al., project.\n",
    "\n",
    "The tutorial repeats the short tutorial, but with the full fastq files and includes some extra steps, such as how to download and prepare the transcriptome files used by salmon, alternate ways to navigate the NCBI databases for annotation or reference files you might need, and how to combine salmon outputs at the end into a single genecount file.\n",
    "\n",
    "Full fastq files can be rather large, and so the downloading, extracting, and analysis of them means this tutorial can take over 1 hour 45 minutes to run the code fully. This is part of the reason we have a short and easy introductory tutorial, and this longer more full tutorial for those interested.\n",
    "\n",
    "If this is too lengthy feel free to move on to the snakemake tutorial or the DEG analysis tutorial -- all the files used in the DEG tutorial were created using this extended tutorial workflow.\n",
    "\n",
    "![RNA-Seq workflow](images/rnaseq-workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 1: Install Mambaforge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First install Mambaforge.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "100 86.8M  100 86.8M    0     0  97.5M      0 --:--:-- --:--:-- --:--:-- 97.5M\n",
      "PREFIX=/home/ec2-user/mambaforge\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/mambaforge/envs/_virtual_specs_checks\n",
      "\n",
      "  All requested packages already installed\n",
      "\n",
      "Dry run. Not executing the transaction.\n",
      "Unpacking payload ...\n",
      "Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2\n",
      "Extracting ca-certificates-2024.8.30-hbcca054_0.conda\n",
      "Extracting ld_impl_linux-64-2.40-hf3520f5_7.conda\n",
      "Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2\n",
      "Extracting python_abi-3.12-5_cp312.conda\n",
      "Extracting tzdata-2024a-h8827d51_1.conda\n",
      "Extracting libgomp-14.1.0-h77fa898_1.conda\n",
      "Extracting _openmp_mutex-4.5-2_gnu.tar.bz2\n",
      "Extracting libgcc-14.1.0-h77fa898_1.conda\n",
      "Extracting libgcc-ng-14.1.0-h69a702a_1.conda\n",
      "Extracting libstdcxx-14.1.0-hc0a3c3a_1.conda\n",
      "Extracting bzip2-1.0.8-h4bc722e_7.conda\n",
      "Extracting c-ares-1.32.3-h4bc722e_0.conda\n",
      "Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2\n",
      "Extracting libev-4.33-hd590300_2.conda\n",
      "Extracting libexpat-2.6.2-h59595ed_0.conda\n",
      "Extracting libffi-3.4.2-h7f98852_5.tar.bz2\n",
      "Extracting libiconv-1.17-hd590300_2.conda\n",
      "Extracting libnsl-2.0.1-hd590300_0.conda\n",
      "Extracting libstdcxx-ng-14.1.0-h4852527_1.conda\n",
      "Extracting libuuid-2.38.1-h0b41bf4_0.conda\n",
      "Extracting libxcrypt-4.4.36-hd590300_1.conda\n",
      "Extracting libzlib-1.3.1-h4ab18f5_1.conda\n",
      "Extracting lzo-2.10-hd590300_1001.conda\n",
      "Extracting ncurses-6.5-he02047a_1.conda\n",
      "Extracting openssl-3.3.1-hb9d3cd8_3.conda\n",
      "Extracting reproc-14.2.4.post0-hd590300_1.conda\n",
      "Extracting xz-5.2.6-h166bdaf_0.tar.bz2\n",
      "Extracting fmt-10.2.1-h00ab1b0_0.conda\n",
      "Extracting icu-75.1-he02047a_0.conda\n",
      "Extracting libedit-3.1.20191231-he28a2e2_2.tar.bz2\n",
      "Extracting libnghttp2-1.58.0-h47da74e_1.conda\n",
      "Extracting libsolv-0.7.30-h3509ff9_0.conda\n",
      "Extracting libsqlite-3.46.1-hadc24fc_0.conda\n",
      "Extracting libssh2-1.11.0-h0841786_0.conda\n",
      "Extracting lz4-c-1.9.4-hcb278e6_0.conda\n",
      "Extracting readline-8.2-h8228510_1.conda\n",
      "Extracting reproc-cpp-14.2.4.post0-h59595ed_1.conda\n",
      "Extracting tk-8.6.13-noxft_h4845f30_101.conda\n",
      "Extracting yaml-cpp-0.8.0-h59595ed_0.conda\n",
      "Extracting zstd-1.5.6-ha6fb4c9_0.conda\n",
      "Extracting krb5-1.21.3-h659f571_0.conda\n",
      "Extracting libxml2-2.12.7-he7c6b58_4.conda\n",
      "Extracting python-3.12.5-h2ad013b_0_cpython.conda\n",
      "Extracting libarchive-3.7.4-hfca40fe_0.conda\n",
      "Extracting libcurl-8.9.1-hdb1bdb2_0.conda\n",
      "Extracting menuinst-2.1.2-py312h7900ff3_1.conda\n",
      "Extracting archspec-0.2.3-pyhd8ed1ab_0.conda\n",
      "Extracting boltons-24.0.0-pyhd8ed1ab_0.conda\n",
      "Extracting brotli-python-1.1.0-py312h2ec8cdc_2.conda\n",
      "Extracting certifi-2024.8.30-pyhd8ed1ab_0.conda\n",
      "Extracting charset-normalizer-3.3.2-pyhd8ed1ab_0.conda\n",
      "Extracting colorama-0.4.6-pyhd8ed1ab_0.tar.bz2\n",
      "Extracting distro-1.9.0-pyhd8ed1ab_0.conda\n",
      "Extracting frozendict-2.4.4-py312h9a8786e_0.conda\n",
      "Extracting hpack-4.0.0-pyh9f0ad1d_0.tar.bz2\n",
      "Extracting hyperframe-6.0.1-pyhd8ed1ab_0.tar.bz2\n",
      "Extracting idna-3.8-pyhd8ed1ab_0.conda\n",
      "Extracting jsonpointer-3.0.0-py312h7900ff3_1.conda\n",
      "Extracting libmamba-1.5.9-h4cc3d14_0.conda\n",
      "Extracting packaging-24.1-pyhd8ed1ab_0.conda\n",
      "Extracting platformdirs-4.2.2-pyhd8ed1ab_0.conda\n",
      "Extracting pluggy-1.5.0-pyhd8ed1ab_0.conda\n",
      "Extracting pycosat-0.6.6-py312h98912ed_0.conda\n",
      "Extracting pycparser-2.22-pyhd8ed1ab_0.conda\n",
      "Extracting pysocks-1.7.1-pyha2e5f31_6.tar.bz2\n",
      "Extracting ruamel.yaml.clib-0.2.8-py312h98912ed_0.conda\n",
      "Extracting setuptools-73.0.1-pyhd8ed1ab_0.conda\n",
      "Extracting truststore-0.9.2-pyhd8ed1ab_0.conda\n",
      "Extracting wheel-0.44.0-pyhd8ed1ab_0.conda\n",
      "Extracting cffi-1.17.0-py312h06ac9bb_1.conda\n",
      "Extracting h2-4.1.0-pyhd8ed1ab_0.tar.bz2\n",
      "Extracting jsonpatch-1.33-pyhd8ed1ab_0.conda\n",
      "Extracting libmambapy-1.5.9-py312h7fb9e8e_0.conda\n",
      "Extracting pip-24.2-pyh8b19718_1.conda\n",
      "Extracting ruamel.yaml-0.18.6-py312h98912ed_0.conda\n",
      "Extracting tqdm-4.66.5-pyhd8ed1ab_0.conda\n",
      "Extracting zstandard-0.23.0-py312hef9b889_1.conda\n",
      "Extracting conda-package-streaming-0.10.0-pyhd8ed1ab_0.conda\n",
      "Extracting urllib3-2.2.2-pyhd8ed1ab_1.conda\n",
      "Extracting conda-package-handling-2.3.0-pyh7900ff3_0.conda\n",
      "Extracting requests-2.32.3-pyhd8ed1ab_0.conda\n",
      "Extracting conda-24.7.1-py312h7900ff3_0.conda\n",
      "Extracting conda-libmamba-solver-24.7.0-pyhd8ed1ab_0.conda\n",
      "Extracting mamba-1.5.9-py312h9460a1c_0.conda\n",
      "!!!!!! Mambaforge is now deprecated !!!!!\n",
      "Future Miniforge releases will NOT build Mambaforge installers.\n",
      "We advise you switch to Miniforge at your earliest convenience.\n",
      "More details at https://conda-forge.org/news/2024/07/29/sunsetting-mambaforge/.\n",
      "If you are unable to switch to Miniforge, you may pin your installer version to one found in \n",
      "https://github.com/conda-forge/miniforge/releases/tag/24.5.0-1\n",
      "or if you lack the system requirements (Linux glibc >= 2.17, or macOS + x86-64bit >= 10.13)\n",
      "you may pin your installer to one older version found in \n",
      "https://github.com/conda-forge/miniforge/releases/tag/24.3.0-0\n",
      "\n",
      "Installing base environment...\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/mambaforge\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - conda-forge/linux-64::_libgcc_mutex==0.1=conda_forge[md5=d7c89558ba9fa0495403155b64376d81]\n",
      "   - conda-forge/linux-64::ca-certificates==2024.8.30=hbcca054_0[md5=c27d1c142233b5bc9ca570c6e2e0c244]\n",
      "   - conda-forge/linux-64::ld_impl_linux-64==2.40=hf3520f5_7[md5=b80f2f396ca2c28b8c14c437a4ed1e74]\n",
      "   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab_3[md5=878f923dd6acc8aeb47a75da6c4098be]\n",
      "   - conda-forge/linux-64::python_abi==3.12=5_cp312[md5=0424ae29b104430108f5218a66db7260]\n",
      "   - conda-forge/noarch::tzdata==2024a=h8827d51_1[md5=8bfdead4e0fff0383ae4c9c50d0531bd]\n",
      "   - conda-forge/linux-64::libgomp==14.1.0=h77fa898_1[md5=23c255b008c4f2ae008f81edcabaca89]\n",
      "   - conda-forge/linux-64::_openmp_mutex==4.5=2_gnu[md5=73aaf86a425cc6e73fcf236a5a46396d]\n",
      "   - conda-forge/linux-64::libgcc==14.1.0=h77fa898_1[md5=002ef4463dd1e2b44a94a4ace468f5d2]\n",
      "   - conda-forge/linux-64::libgcc-ng==14.1.0=h69a702a_1[md5=1efc0ad219877a73ef977af7dbb51f17]\n",
      "   - conda-forge/linux-64::libstdcxx==14.1.0=hc0a3c3a_1[md5=9dbb9699ea467983ba8a4ba89b08b066]\n",
      "   - conda-forge/linux-64::bzip2==1.0.8=h4bc722e_7[md5=62ee74e96c5ebb0af99386de58cf9553]\n",
      "   - conda-forge/linux-64::c-ares==1.32.3=h4bc722e_0[md5=7624e34ee6baebfc80d67bac76cc9d9d]\n",
      "   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf_0[md5=30186d27e2c9fa62b45fb1476b7200e3]\n",
      "   - conda-forge/linux-64::libev==4.33=hd590300_2[md5=172bf1cd1ff8629f2b1179945ed45055]\n",
      "   - conda-forge/linux-64::libexpat==2.6.2=h59595ed_0[md5=e7ba12deb7020dd080c6c70e7b6f6a3d]\n",
      "   - conda-forge/linux-64::libffi==3.4.2=h7f98852_5[md5=d645c6d2ac96843a2bfaccd2d62b3ac3]\n",
      "   - conda-forge/linux-64::libiconv==1.17=hd590300_2[md5=d66573916ffcf376178462f1b61c941e]\n",
      "   - conda-forge/linux-64::libnsl==2.0.1=hd590300_0[md5=30fd6e37fe21f86f4bd26d6ee73eeec7]\n",
      "   - conda-forge/linux-64::libstdcxx-ng==14.1.0=h4852527_1[md5=bd2598399a70bb86d8218e95548d735e]\n",
      "   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4_0[md5=40b61aab5c7ba9ff276c41cfffe6b80b]\n",
      "   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300_1[md5=5aa797f8787fe7a17d1b0821485b5adc]\n",
      "   - conda-forge/linux-64::libzlib==1.3.1=h4ab18f5_1[md5=57d7dc60e9325e3de37ff8dffd18e814]\n",
      "   - conda-forge/linux-64::lzo==2.10=hd590300_1001[md5=ec7398d21e2651e0dcb0044d03b9a339]\n",
      "   - conda-forge/linux-64::ncurses==6.5=he02047a_1[md5=70caf8bb6cf39a0b6b7efc885f51c0fe]\n",
      "   - conda-forge/linux-64::openssl==3.3.1=hb9d3cd8_3[md5=6c566a46baae794daf34775d41eb180a]\n",
      "   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300_1[md5=82ca53502dfd5a64a80dee76dae14685]\n",
      "   - conda-forge/linux-64::xz==5.2.6=h166bdaf_0[md5=2161070d867d1b1204ea749c8eec4ef0]\n",
      "   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0_0[md5=35ef8bc24bd34074ebae3c943d551728]\n",
      "   - conda-forge/linux-64::icu==75.1=he02047a_0[md5=8b189310083baabfb622af68fd9d3ae3]\n",
      "   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2_2[md5=4d331e44109e3f0e19b4cb8f9b82f3e1]\n",
      "   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e_1[md5=700ac6ea6d53d5510591c4344d5c989a]\n",
      "   - conda-forge/linux-64::libsolv==0.7.30=h3509ff9_0[md5=02539b77d25aa4f65b20246549e256c3]\n",
      "   - conda-forge/linux-64::libsqlite==3.46.1=hadc24fc_0[md5=36f79405ab16bf271edb55b213836dac]\n",
      "   - conda-forge/linux-64::libssh2==1.11.0=h0841786_0[md5=1f5a58e686b13bcfde88b93f547d23fe]\n",
      "   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6_0[md5=318b08df404f9c9be5712aaa5a6f0bb0]\n",
      "   - conda-forge/linux-64::readline==8.2=h8228510_1[md5=47d31b792659ce70f470b5c82fdfb7a4]\n",
      "   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed_1[md5=715e1d720ec1a03715bebd237972fca5]\n",
      "   - conda-forge/linux-64::tk==8.6.13=noxft_h4845f30_101[md5=d453b98d9c83e71da0741bb0ff4d76bc]\n",
      "   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed_0[md5=965eaacd7c18eb8361fd12bb9e7a57d7]\n",
      "   - conda-forge/linux-64::zstd==1.5.6=ha6fb4c9_0[md5=4d056880988120e29d75bfff282e0f45]\n",
      "   - conda-forge/linux-64::krb5==1.21.3=h659f571_0[md5=3f43953b7d3fb3aaa1d0d0723d91e368]\n",
      "   - conda-forge/linux-64::libxml2==2.12.7=he7c6b58_4[md5=08a9265c637230c37cb1be4a6cad4536]\n",
      "   - conda-forge/linux-64::python==3.12.5=h2ad013b_0_cpython[md5=9c56c4df45f6571b13111d8df2448692]\n",
      "   - conda-forge/linux-64::libarchive==3.7.4=hfca40fe_0[md5=32ddb97f897740641d8d46a829ce1704]\n",
      "   - conda-forge/linux-64::libcurl==8.9.1=hdb1bdb2_0[md5=7da1d242ca3591e174a3c7d82230d3c0]\n",
      "   - conda-forge/linux-64::menuinst==2.1.2=py312h7900ff3_1[md5=c6575ae996f2bc0369c73b632db5ca61]\n",
      "   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab_0[md5=192278292e20704f663b9c766909d67b]\n",
      "   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab_0[md5=61de176bd62041f9cd5bd4fcd09eb0ff]\n",
      "   - conda-forge/linux-64::brotli-python==1.1.0=py312h2ec8cdc_2[md5=b0b867af6fc74b2a0aa206da29c0f3cf]\n",
      "   - conda-forge/noarch::certifi==2024.8.30=pyhd8ed1ab_0[md5=12f7d00853807b0531775e9be891cb11]\n",
      "   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab_0[md5=7f4a9e3fcff3f6356ae99244a014da6a]\n",
      "   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab_0[md5=3faab06a954c2a04039983f2c4a50d99]\n",
      "   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab_0[md5=bbdb409974cd6cb30071b1d978302726]\n",
      "   - conda-forge/linux-64::frozendict==2.4.4=py312h9a8786e_0[md5=ff14ec1103a0817d45e7cf012742ce60]\n",
      "   - conda-forge/noarch::hpack==4.0.0=pyh9f0ad1d_0[md5=914d6646c4dbb1fd3ff539830a12fd71]\n",
      "   - conda-forge/noarch::hyperframe==6.0.1=pyhd8ed1ab_0[md5=9f765cbfab6870c8435b9eefecd7a1f4]\n",
      "   - conda-forge/noarch::idna==3.8=pyhd8ed1ab_0[md5=99e164522f6bdf23c177c8d9ae63f975]\n",
      "   - conda-forge/linux-64::jsonpointer==3.0.0=py312h7900ff3_1[md5=6b51f7459ea4073eeb5057207e2e1e3d]\n",
      "   - conda-forge/linux-64::libmamba==1.5.9=h4cc3d14_0[md5=896cece5b883ad86e9dd88b1f4d23c99]\n",
      "   - conda-forge/noarch::packaging==24.1=pyhd8ed1ab_0[md5=cbe1bb1f21567018ce595d9c2be0f0db]\n",
      "   - conda-forge/noarch::platformdirs==4.2.2=pyhd8ed1ab_0[md5=6f6cf28bf8e021933869bae3f84b8fc9]\n",
      "   - conda-forge/noarch::pluggy==1.5.0=pyhd8ed1ab_0[md5=d3483c8fc2dc2cc3f5cf43e26d60cabf]\n",
      "   - conda-forge/linux-64::pycosat==0.6.6=py312h98912ed_0[md5=8f1c372e7b843167be885dc8229931c1]\n",
      "   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab_0[md5=844d9eb3b43095b031874477f7d70088]\n",
      "   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31_6[md5=2a7de29fb590ca14b5243c4c812c8025]\n",
      "   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py312h98912ed_0[md5=05f31c2a79ba61df8d6d903ce4a4ce7b]\n",
      "   - conda-forge/noarch::setuptools==73.0.1=pyhd8ed1ab_0[md5=f0b618d7673d1b2464f600b34d912f6f]\n",
      "   - conda-forge/noarch::truststore==0.9.2=pyhd8ed1ab_0[md5=f14e46d1bf271e748ff556d8b872e28a]\n",
      "   - conda-forge/noarch::wheel==0.44.0=pyhd8ed1ab_0[md5=d44e3b085abcaef02983c6305b84b584]\n",
      "   - conda-forge/linux-64::cffi==1.17.0=py312h06ac9bb_1[md5=db9bdbaee0f524ead0471689f002781e]\n",
      "   - conda-forge/noarch::h2==4.1.0=pyhd8ed1ab_0[md5=b748fbf7060927a6e82df7cb5ee8f097]\n",
      "   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab_0[md5=bfdb7c5c6ad1077c82a69a8642c87aff]\n",
      "   - conda-forge/linux-64::libmambapy==1.5.9=py312h7fb9e8e_0[md5=ccaeeb6e3caaf0c744480393791aa366]\n",
      "   - conda-forge/noarch::pip==24.2=pyh8b19718_1[md5=6c78fbb8ddfd64bcb55b5cbafd2d2c43]\n",
      "   - conda-forge/linux-64::ruamel.yaml==0.18.6=py312h98912ed_0[md5=a99a06a875138829ef65f44bbe2c30ca]\n",
      "   - conda-forge/noarch::tqdm==4.66.5=pyhd8ed1ab_0[md5=c6e94fc2b2ec71ea33fe7c7da259acb4]\n",
      "   - conda-forge/linux-64::zstandard==0.23.0=py312hef9b889_1[md5=8b7069e9792ee4e5b4919a7a306d2e67]\n",
      "   - conda-forge/noarch::conda-package-streaming==0.10.0=pyhd8ed1ab_0[md5=3480386e00995f7a1dfb3b9aa2fe70fd]\n",
      "   - conda-forge/noarch::urllib3==2.2.2=pyhd8ed1ab_1[md5=e804c43f58255e977093a2298e442bb8]\n",
      "   - conda-forge/noarch::conda-package-handling==2.3.0=pyh7900ff3_0[md5=0a7dce281ae2be81acab0aa963e6bb99]\n",
      "   - conda-forge/noarch::requests==2.32.3=pyhd8ed1ab_0[md5=5ede4753180c7a550a443c430dc8ab52]\n",
      "   - conda-forge/linux-64::conda==24.7.1=py312h7900ff3_0[md5=e1bf59014e88eaff036101358f63a496]\n",
      "   - conda-forge/noarch::conda-libmamba-solver==24.7.0=pyhd8ed1ab_0[md5=857c9e25f0a77c0bd7eb622d46d9418f]\n",
      "   - conda-forge/linux-64::mamba==1.5.9=py312h9460a1c_0[md5=a8525c8a1647b4f5967fa6b552722851]\n",
      "\n",
      "\n",
      "  Package                         Version  Build               Channel         Size\n",
      "─────────────────────────────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "─────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[32m+ _libgcc_mutex          \u001b[0m           0.1  conda_forge         conda-forge         \n",
      "  \u001b[32m+ ca-certificates        \u001b[0m     2024.8.30  hbcca054_0          conda-forge         \n",
      "  \u001b[32m+ ld_impl_linux-64       \u001b[0m          2.40  hf3520f5_7          conda-forge         \n",
      "  \u001b[32m+ pybind11-abi           \u001b[0m             4  hd8ed1ab_3          conda-forge         \n",
      "  \u001b[32m+ python_abi             \u001b[0m          3.12  5_cp312             conda-forge         \n",
      "  \u001b[32m+ tzdata                 \u001b[0m         2024a  h8827d51_1          conda-forge         \n",
      "  \u001b[32m+ libgomp                \u001b[0m        14.1.0  h77fa898_1          conda-forge         \n",
      "  \u001b[32m+ _openmp_mutex          \u001b[0m           4.5  2_gnu               conda-forge         \n",
      "  \u001b[32m+ libgcc                 \u001b[0m        14.1.0  h77fa898_1          conda-forge         \n",
      "  \u001b[32m+ libgcc-ng              \u001b[0m        14.1.0  h69a702a_1          conda-forge         \n",
      "  \u001b[32m+ libstdcxx              \u001b[0m        14.1.0  hc0a3c3a_1          conda-forge         \n",
      "  \u001b[32m+ bzip2                  \u001b[0m         1.0.8  h4bc722e_7          conda-forge         \n",
      "  \u001b[32m+ c-ares                 \u001b[0m        1.32.3  h4bc722e_0          conda-forge         \n",
      "  \u001b[32m+ keyutils               \u001b[0m         1.6.1  h166bdaf_0          conda-forge         \n",
      "  \u001b[32m+ libev                  \u001b[0m          4.33  hd590300_2          conda-forge         \n",
      "  \u001b[32m+ libexpat               \u001b[0m         2.6.2  h59595ed_0          conda-forge         \n",
      "  \u001b[32m+ libffi                 \u001b[0m         3.4.2  h7f98852_5          conda-forge         \n",
      "  \u001b[32m+ libiconv               \u001b[0m          1.17  hd590300_2          conda-forge         \n",
      "  \u001b[32m+ libnsl                 \u001b[0m         2.0.1  hd590300_0          conda-forge         \n",
      "  \u001b[32m+ libstdcxx-ng           \u001b[0m        14.1.0  h4852527_1          conda-forge         \n",
      "  \u001b[32m+ libuuid                \u001b[0m        2.38.1  h0b41bf4_0          conda-forge         \n",
      "  \u001b[32m+ libxcrypt              \u001b[0m        4.4.36  hd590300_1          conda-forge         \n",
      "  \u001b[32m+ libzlib                \u001b[0m         1.3.1  h4ab18f5_1          conda-forge         \n",
      "  \u001b[32m+ lzo                    \u001b[0m          2.10  hd590300_1001       conda-forge         \n",
      "  \u001b[32m+ ncurses                \u001b[0m           6.5  he02047a_1          conda-forge         \n",
      "  \u001b[32m+ openssl                \u001b[0m         3.3.1  hb9d3cd8_3          conda-forge         \n",
      "  \u001b[32m+ reproc                 \u001b[0m  14.2.4.post0  hd590300_1          conda-forge         \n",
      "  \u001b[32m+ xz                     \u001b[0m         5.2.6  h166bdaf_0          conda-forge         \n",
      "  \u001b[32m+ fmt                    \u001b[0m        10.2.1  h00ab1b0_0          conda-forge         \n",
      "  \u001b[32m+ icu                    \u001b[0m          75.1  he02047a_0          conda-forge         \n",
      "  \u001b[32m+ libedit                \u001b[0m  3.1.20191231  he28a2e2_2          conda-forge         \n",
      "  \u001b[32m+ libnghttp2             \u001b[0m        1.58.0  h47da74e_1          conda-forge         \n",
      "  \u001b[32m+ libsolv                \u001b[0m        0.7.30  h3509ff9_0          conda-forge         \n",
      "  \u001b[32m+ libsqlite              \u001b[0m        3.46.1  hadc24fc_0          conda-forge         \n",
      "  \u001b[32m+ libssh2                \u001b[0m        1.11.0  h0841786_0          conda-forge         \n",
      "  \u001b[32m+ lz4-c                  \u001b[0m         1.9.4  hcb278e6_0          conda-forge         \n",
      "  \u001b[32m+ readline               \u001b[0m           8.2  h8228510_1          conda-forge         \n",
      "  \u001b[32m+ reproc-cpp             \u001b[0m  14.2.4.post0  h59595ed_1          conda-forge         \n",
      "  \u001b[32m+ tk                     \u001b[0m        8.6.13  noxft_h4845f30_101  conda-forge         \n",
      "  \u001b[32m+ yaml-cpp               \u001b[0m         0.8.0  h59595ed_0          conda-forge         \n",
      "  \u001b[32m+ zstd                   \u001b[0m         1.5.6  ha6fb4c9_0          conda-forge         \n",
      "  \u001b[32m+ krb5                   \u001b[0m        1.21.3  h659f571_0          conda-forge         \n",
      "  \u001b[32m+ libxml2                \u001b[0m        2.12.7  he7c6b58_4          conda-forge         \n",
      "  \u001b[32m+ python                 \u001b[0m        3.12.5  h2ad013b_0_cpython  conda-forge         \n",
      "  \u001b[32m+ libarchive             \u001b[0m         3.7.4  hfca40fe_0          conda-forge         \n",
      "  \u001b[32m+ libcurl                \u001b[0m         8.9.1  hdb1bdb2_0          conda-forge         \n",
      "  \u001b[32m+ menuinst               \u001b[0m         2.1.2  py312h7900ff3_1     conda-forge         \n",
      "  \u001b[32m+ archspec               \u001b[0m         0.2.3  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ boltons                \u001b[0m        24.0.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ brotli-python          \u001b[0m         1.1.0  py312h2ec8cdc_2     conda-forge         \n",
      "  \u001b[32m+ certifi                \u001b[0m     2024.8.30  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ charset-normalizer     \u001b[0m         3.3.2  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ colorama               \u001b[0m         0.4.6  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ distro                 \u001b[0m         1.9.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ frozendict             \u001b[0m         2.4.4  py312h9a8786e_0     conda-forge         \n",
      "  \u001b[32m+ hpack                  \u001b[0m         4.0.0  pyh9f0ad1d_0        conda-forge         \n",
      "  \u001b[32m+ hyperframe             \u001b[0m         6.0.1  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ idna                   \u001b[0m           3.8  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ jsonpointer            \u001b[0m         3.0.0  py312h7900ff3_1     conda-forge         \n",
      "  \u001b[32m+ libmamba               \u001b[0m         1.5.9  h4cc3d14_0          conda-forge         \n",
      "  \u001b[32m+ packaging              \u001b[0m          24.1  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ platformdirs           \u001b[0m         4.2.2  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ pluggy                 \u001b[0m         1.5.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ pycosat                \u001b[0m         0.6.6  py312h98912ed_0     conda-forge         \n",
      "  \u001b[32m+ pycparser              \u001b[0m          2.22  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ pysocks                \u001b[0m         1.7.1  pyha2e5f31_6        conda-forge         \n",
      "  \u001b[32m+ ruamel.yaml.clib       \u001b[0m         0.2.8  py312h98912ed_0     conda-forge         \n",
      "  \u001b[32m+ setuptools             \u001b[0m        73.0.1  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ truststore             \u001b[0m         0.9.2  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ wheel                  \u001b[0m        0.44.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ cffi                   \u001b[0m        1.17.0  py312h06ac9bb_1     conda-forge         \n",
      "  \u001b[32m+ h2                     \u001b[0m         4.1.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ jsonpatch              \u001b[0m          1.33  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ libmambapy             \u001b[0m         1.5.9  py312h7fb9e8e_0     conda-forge         \n",
      "  \u001b[32m+ pip                    \u001b[0m          24.2  pyh8b19718_1        conda-forge         \n",
      "  \u001b[32m+ ruamel.yaml            \u001b[0m        0.18.6  py312h98912ed_0     conda-forge         \n",
      "  \u001b[32m+ tqdm                   \u001b[0m        4.66.5  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ zstandard              \u001b[0m        0.23.0  py312hef9b889_1     conda-forge         \n",
      "  \u001b[32m+ conda-package-streaming\u001b[0m        0.10.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ urllib3                \u001b[0m         2.2.2  pyhd8ed1ab_1        conda-forge         \n",
      "  \u001b[32m+ conda-package-handling \u001b[0m         2.3.0  pyh7900ff3_0        conda-forge         \n",
      "  \u001b[32m+ requests               \u001b[0m        2.32.3  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ conda                  \u001b[0m        24.7.1  py312h7900ff3_0     conda-forge         \n",
      "  \u001b[32m+ conda-libmamba-solver  \u001b[0m        24.7.0  pyhd8ed1ab_0        conda-forge         \n",
      "  \u001b[32m+ mamba                  \u001b[0m         1.5.9  py312h9460a1c_0     conda-forge         \n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 85 packages\n",
      "\n",
      "  Total download: 0 B\n",
      "\n",
      "─────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\n",
      "Transaction starting\n",
      "\u001b[?25l\u001b[2K\u001b[0G\u001b[?25h\n",
      "Transaction finished\n",
      "\n",
      "To activate this environment, use:\n",
      "\n",
      "    micromamba activate /home/ec2-user/mambaforge\n",
      "\n",
      "Or to execute a single command in this environment, use:\n",
      "\n",
      "    micromamba run -p /home/ec2-user/mambaforge mycommand\n",
      "\n",
      "installation finished.\n",
      "04:51:50\n"
     ]
    }
   ],
   "source": [
    "!curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-$(uname)-$(uname -m).sh\n",
    "!bash Mambaforge-$(uname)-$(uname -m).sh -b -u -p $HOME/mambaforge\n",
    "!date +\"%T\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, using mambaforge and bioconda, install the tools that will be used in this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Looking for: ['trimmomatic', 'fastqc', 'multiqc', 'salmon', 'gsutil', 'sql-magic', 'entrez-direct', 'gffread', 'parallel-fastq-dump', 'sra-tools', 'sql-magic', 'pyathena']\n",
      "\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/497deca9.json\" was modified by another program\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/09cdf8bf.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "bioconda/linux-64 (check zst) \u001b[33m━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "bioconda/linux-64 (check zst) \u001b[33m━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.2s\n",
      "bioconda/linux-64 (check zst) \u001b[33m━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/linux-64 (check zst)                       Checked  0.2s\n",
      "\u001b[?25h\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "bioconda/noarch (check zst) \u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/noarch (check zst)                         Checked  0.0s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse state file: Could not load cache state: [json.exception.type_error.302] type must be string, but is null\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghttps://aws-ml-conda.s3.us-west-2.amazonaws.com/..  0.2s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse state file: Could not load cache state: [json.exception.type_error.302] type must be string, but is null\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[33m━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━\u001b[0m   0.0 B  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghttps://aws-ml-conda.s3.us-west-2.amazonaws.com/..  0.1s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/c9ddbd6b.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "nvidia/linux-64 (check zst) \u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gnvidia/linux-64 (check zst)                         Checked  0.0s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/b121c3e7.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "nvidia/noarch (check zst) \u001b[90m━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gnvidia/noarch (check zst)                           Checked  0.0s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/ee0ed9e9.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "pytorch/linux-64 (check zst) \u001b[90m╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gpytorch/linux-64 (check zst)                        Checked  0.0s\n",
      "\u001b[?25h\u001b[33m\u001b[1mwarning  libmamba\u001b[m Cache file \"/home/ec2-user/anaconda3/pkgs/cache/edb1952f.json\" was modified by another program\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "pytorch/noarch (check zst) \u001b[90m━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━\u001b[0m   0.0 B @  ??.?MB/s Checking  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gpytorch/noarch (check zst)                          Checked  0.0s\n",
      "\u001b[?25h\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "conda-forge/linux-64                               \u001b[33m━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m   0.0 B  0.0s\n",
      "conda-forge/noarch                                 \u001b[33m━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m   0.0 B  0.0s\n",
      "bioconda/linux-64                                  \u001b[33m━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━\u001b[0m   0.0 B  0.0s\n",
      "bioconda/noarch                                    \u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m   0.0 B  0.0s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[90m━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━\u001b[0m   0.0 B  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "conda-forge/linux-64                               \u001b[90m━━━━━━━━━━━━━━━\u001b[0m 605.5kB  0.1s\n",
      "conda-forge/noarch                                 \u001b[90m━━━━━━━━━━━━━━━\u001b[0m  13.7kB  0.1s\n",
      "bioconda/linux-64                                  \u001b[90m━━━━━━━━━━━━━━━\u001b[0m  14.2kB  0.1s\n",
      "bioconda/noarch                                    \u001b[90m━━━━━━━━━━━━━━━\u001b[0m  97.3kB  0.1s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[90m━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━\u001b[0m   0.0 B  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.2s\n",
      "conda-forge/linux-64                               \u001b[90m━━━━━━━━━━━━━━━\u001b[0m   2.5MB  0.2s\n",
      "conda-forge/noarch                                 \u001b[90m━━━━━━━━━━━━━━━\u001b[0m 879.9kB  0.2s\n",
      "bioconda/linux-64                                  ━━╸\u001b[90m━━━━━━━━━━━━\u001b[0m 997.7kB  0.2s\n",
      "bioconda/noarch                                    ━━━━╸\u001b[90m━━━━━━━━━━\u001b[0m   1.5MB  0.2s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. ━━━━━╸\u001b[90m━━━━━━━━━\u001b[0m  69.2kB  0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghttps://aws-ml-conda.s3.us-west-2.amazonaws.com/..  0.3s\n",
      "[+] 0.3s\n",
      "conda-forge/linux-64 ━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━\u001b[0m   3.7MB /  38.1MB @  12.5MB/s  0.3s\n",
      "conda-forge/noarch   ━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━\u001b[0m   1.6MB /  16.5MB @   6.4MB/s  0.3s\n",
      "bioconda/linux-64    ━━━━━━━╸\u001b[90m━━━━━━━━━━━━━━━\u001b[0m   1.7MB /   4.6MB @   6.8MB/s  0.3s\n",
      "bioconda/noarch      ━━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━\u001b[0m   2.6MB /   4.3MB @   9.4MB/s  0.3s\n",
      "pytorch/linux-64     \u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.4s\n",
      "conda-forge/linux-64 ━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━━\u001b[0m   4.3MB /  38.1MB @  11.6MB/s  0.4s\n",
      "conda-forge/noarch   ━━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━\u001b[0m   2.7MB /  16.5MB @   7.1MB/s  0.4s\n",
      "bioconda/linux-64    ━━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━\u001b[0m   2.8MB /   4.6MB @   7.4MB/s  0.4s\n",
      "bioconda/noarch      ━━━━━━━━━━━━━━━━╸\u001b[90m━━━━━━\u001b[0m   3.3MB /   4.3MB @   9.1MB/s  0.4s\n",
      "pytorch/linux-64     \u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/noarch                                      4.3MB @   9.8MB/s  0.4s\n",
      "pytorch/linux-64                                   184.8kB @ 409.8kB/s  0.2s\n",
      "nvidia/noarch                                       14.9kB @  30.1kB/s  0.1s\n",
      "[+] 0.5s\n",
      "conda-forge/linux-64 ━━╸\u001b[90m━━━━━━━━━━━━━━━━━━━━\u001b[0m   5.9MB /  38.1MB @  12.2MB/s  0.5s\n",
      "conda-forge/noarch   ━━━━╸\u001b[90m━━━━━━━━━━━━━━━━━━\u001b[0m   4.3MB /  16.5MB @   8.8MB/s  0.5s\n",
      "bioconda/linux-64    ━━━━━━━━━━━━━━━━━━━━━╸\u001b[90m━\u001b[0m   4.6MB /   4.6MB @   9.3MB/s  0.5s\n",
      "nvidia/linux-64      \u001b[90m━━━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.0s\n",
      "pytorch/noarch       \u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m   0.0 B /  ??.?MB @  ??.?MB/s  0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gbioconda/linux-64                                    4.6MB @   9.0MB/s  0.5s\n",
      "pytorch/noarch                                       9.9kB @  19.2kB/s  0.1s\n",
      "nvidia/linux-64                                    185.1kB @ 332.4kB/s  0.1s\n",
      "[+] 0.6s\n",
      "conda-forge/linux-64                               ━━╸\u001b[90m━━━━━━━━━━━━\u001b[0m   7.7MB  0.6s\n",
      "conda-forge/noarch                                 ━━━━╸\u001b[90m━━━━━━━━━━\u001b[0m   6.1MB  0.6s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[90m━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━\u001b[0m   0.0 B  0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.7s\n",
      "conda-forge/linux-64                               ━━╸\u001b[90m━━━━━━━━━━━━\u001b[0m   9.8MB  0.7s\n",
      "conda-forge/noarch                                 ━━━━━╸\u001b[90m━━━━━━━━━\u001b[0m   7.0MB  0.7s\n",
      "https://aws-ml-conda.s3.us-west-2.amazonaws.com/.. \u001b[33m━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m   0.0 B  0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghttps://aws-ml-conda.s3.us-west-2.amazonaws.com/..  0.2s\n",
      "[+] 0.8s\n",
      "conda-forge/linux-64 ━━━━━╸\u001b[90m━━━━━━━━━━━━━━━━━\u001b[0m  10.9MB /  38.1MB @  14.6MB/s  0.8s\n",
      "conda-forge/noarch   ━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━━\u001b[0m   9.3MB /  16.5MB @  12.3MB/s  0.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.9s\n",
      "conda-forge/linux-64 ━━━━━━━╸\u001b[90m━━━━━━━━━━━━━━━\u001b[0m  13.4MB /  38.1MB @  15.6MB/s  0.9s\n",
      "conda-forge/noarch   ━━━━━━━━━━━━━━━╸\u001b[90m━━━━━━━\u001b[0m  11.8MB /  16.5MB @  13.6MB/s  0.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1.0s\n",
      "conda-forge/linux-64 ━━━━━━━━╸\u001b[90m━━━━━━━━━━━━━━\u001b[0m  16.7MB /  38.1MB @  17.4MB/s  1.0s\n",
      "conda-forge/noarch   ━━━━━━━━━━━━━━━━━━━╸\u001b[90m━━━\u001b[0m  15.1MB /  16.5MB @  15.7MB/s  1.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gconda-forge/noarch                                  16.5MB @  16.4MB/s  1.0s\n",
      "[+] 1.1s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━╸\u001b[90m━━━━━━━━━━━\u001b[0m  21.2MB /  38.1MB @  19.9MB/s  1.1s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1.2s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━╸\u001b[90m━━━━━━━━\u001b[0m  25.7MB /  38.1MB @  22.0MB/s  1.2s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1.3s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━━━━━╸\u001b[90m━━━━\u001b[0m  31.9MB /  38.1MB @  24.9MB/s  1.3s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1.4s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━━━━━━╸\u001b[90m━━━\u001b[0m  34.8MB /  38.1MB @  25.1MB/s  1.4s\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1.5s\n",
      "conda-forge/linux-64 ━━━━━━━━━━━━━━━━━━━━━╸\u001b[90m━\u001b[0m  37.7MB /  38.1MB @  25.2MB/s  1.5s\u001b[2K\u001b[1A\u001b[2K\u001b[0Gconda-forge/linux-64                                38.1MB @  25.4MB/s  1.5s\n",
      "\u001b[?25h\n",
      "Pinned packages:\n",
      "  - python 3.10.*\n",
      "\n",
      "\n",
      "Transaction\n",
      "\n",
      "  Prefix: /home/ec2-user/anaconda3/envs/tensorflow2_p310\n",
      "\n",
      "  Updating specs:\n",
      "\n",
      "   - trimmomatic\n",
      "   - fastqc\n",
      "   - multiqc\n",
      "   - salmon\n",
      "   - gsutil\n",
      "   - sql-magic\n",
      "   - entrez-direct\n",
      "   - gffread\n",
      "   - parallel-fastq-dump\n",
      "   - sra-tools\n",
      "   - sql-magic\n",
      "   - pyathena\n",
      "   - ca-certificates\n",
      "   - certifi\n",
      "   - openssl\n",
      "\n",
      "\n",
      "  Package                        Version  Build                Channel           Size\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "  Install:\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[32m+ colormath                \u001b[0m      3.0.0  py_2                 conda-forge       35kB\n",
      "  \u001b[32m+ httplib2                 \u001b[0m     0.20.4  pyhd8ed1ab_0         conda-forge       98kB\n",
      "  \u001b[32m+ boto                     \u001b[0m     2.49.0  py_0                 conda-forge      858kB\n",
      "  \u001b[32m+ pyu2f                    \u001b[0m      0.1.5  pyhd8ed1ab_0         conda-forge       32kB\n",
      "  \u001b[32m+ pyaml-env                \u001b[0m      1.2.1  pyhd8ed1ab_0         conda-forge       14kB\n",
      "  \u001b[32m+ humanize                 \u001b[0m     4.10.0  pyhd8ed1ab_0         conda-forge       65kB\n",
      "  \u001b[32m+ importlib-metadata       \u001b[0m      8.5.0  pyha770c72_0         conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ packaging                \u001b[0m       24.1  pyhd8ed1ab_0         conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ rich-click               \u001b[0m      1.8.3  pyhd8ed1ab_0         conda-forge       34kB\n",
      "  \u001b[32m+ annotated-types          \u001b[0m      0.7.0  pyhd8ed1ab_0         conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ botocore                 \u001b[0m    1.35.21  pyge310_1234567_0    conda-forge        7MB\n",
      "  \u001b[32m+ argcomplete              \u001b[0m      3.5.0  pyhd8ed1ab_0         conda-forge       41kB\n",
      "  \u001b[32m+ fasteners                \u001b[0m     0.17.3  pyhd8ed1ab_0         conda-forge       20kB\n",
      "  \u001b[32m+ monotonic                \u001b[0m        1.5  pyhd8ed1ab_1         conda-forge       13kB\n",
      "  \u001b[32m+ retry_decorator          \u001b[0m      1.1.1  pyh9f0ad1d_0         conda-forge        7kB\n",
      "  \u001b[32m+ cachetools               \u001b[0m      5.5.0  pyhd8ed1ab_0         conda-forge       15kB\n",
      "  \u001b[32m+ pyasn1-modules           \u001b[0m      0.4.1  pyhd8ed1ab_0         conda-forge       96kB\n",
      "  \u001b[32m+ findspark                \u001b[0m      2.0.1  pyhd8ed1ab_0         conda-forge        8kB\n",
      "  \u001b[32m+ sqlparse                 \u001b[0m      0.5.1  pyhd8ed1ab_0         conda-forge       40kB\n",
      "  \u001b[32m+ humanfriendly            \u001b[0m       10.0  pyhd8ed1ab_6         conda-forge       73kB\n",
      "  \u001b[32m+ spectra                  \u001b[0m     0.0.11  py_1                 conda-forge       19kB\n",
      "  \u001b[32m+ google-reauth            \u001b[0m      0.1.1  pyhd3deb0d_0         conda-forge       18kB\n",
      "  \u001b[32m+ typeguard                \u001b[0m      4.3.0  pyhd8ed1ab_1         conda-forge       35kB\n",
      "  \u001b[32m+ markdown                 \u001b[0m        3.6  pyhd8ed1ab_0         conda-forge       78kB\n",
      "  \u001b[32m+ boto3                    \u001b[0m    1.35.21  pyhd8ed1ab_0         conda-forge       81kB\n",
      "  \u001b[32m+ oauth2client             \u001b[0m      4.1.3  py_0                 conda-forge       67kB\n",
      "  \u001b[32m+ google-auth              \u001b[0m     2.17.0  pyh1a96a4e_0         conda-forge       99kB\n",
      "  \u001b[32m+ coloredlogs              \u001b[0m     15.0.1  pyhd8ed1ab_3         conda-forge       41kB\n",
      "  \u001b[32m+ pyathena                 \u001b[0m      3.9.0  pyhd8ed1ab_0         conda-forge       55kB\n",
      "  \u001b[32m+ google-apitools          \u001b[0m     0.5.32  pyhd8ed1ab_0         conda-forge       96kB\n",
      "  \u001b[32m+ google-auth-httplib2     \u001b[0m      0.2.0  pyhd8ed1ab_0         conda-forge       15kB\n",
      "  \u001b[32m+ gcs-oauth2-boto-plugin   \u001b[0m        3.2  pyhd8ed1ab_0         conda-forge       30kB\n",
      "  \u001b[32m+ libgcc                   \u001b[0m     14.1.0  h77fa898_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ perl                     \u001b[0m   5.22.2.1  0                    conda-forge       16MB\n",
      "  \u001b[32m+ libboost-headers         \u001b[0m     1.85.0  ha770c72_2           conda-forge       14MB\n",
      "  \u001b[32m+ mathjax                  \u001b[0m      2.7.7  ha770c72_3           conda-forge       22MB\n",
      "  \u001b[32m+ crcmod                   \u001b[0m        1.7  py310ha75aee5_1011   conda-forge       36kB\n",
      "  \u001b[32m+ pydantic-core            \u001b[0m     2.23.4  py310h505e2c1_0      conda-forge        2MB\n",
      "  \u001b[32m+ libstdcxx                \u001b[0m     14.1.0  hc0a3c3a_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libunistring             \u001b[0m     0.9.10  h7f98852_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pandas                   \u001b[0m      2.2.2  py310hf9f9076_1      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ossuuid                  \u001b[0m      1.6.2  hf484d3e_1000        conda-forge       57kB\n",
      "  \u001b[32m+ libjemalloc              \u001b[0m      5.3.0  hcb278e6_0           conda-forge        2MB\n",
      "  \u001b[32m+ libidn2                  \u001b[0m      2.3.7  hd590300_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ sqlite                   \u001b[0m     3.46.0  h6d4b2fc_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libboost                 \u001b[0m     1.85.0  hba137d9_2           conda-forge        3MB\n",
      "  \u001b[32m+ libboost-devel           \u001b[0m     1.85.0  h00ab1b0_2           conda-forge       41kB\n",
      "  \u001b[32m+ kaleido-core             \u001b[0m      0.2.1  h3644ca4_0           conda-forge       62MB\n",
      "  \u001b[32m+ wget                     \u001b[0m     1.21.4  hda4d442_0           conda-forge      770kB\n",
      "  \u001b[32m+ xorg-libxt               \u001b[0m      1.3.0  hd590300_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ boost-cpp                \u001b[0m     1.85.0  h44aadfe_2           conda-forge       18kB\n",
      "  \u001b[32m+ openjdk                  \u001b[0m     21.0.2  haa376d0_0           conda-forge      181MB\n",
      "  \u001b[32m+ perl-threaded            \u001b[0m     5.32.1  hdfd78af_1           bioconda           6kB\n",
      "  \u001b[32m+ fastqc                   \u001b[0m     0.12.1  hdfd78af_0           bioconda          12MB\n",
      "  \u001b[32m+ trimmomatic              \u001b[0m       0.39  hdfd78af_2           bioconda         147kB\n",
      "  \u001b[32m+ gsutil                   \u001b[0m       5.30  pyhd8ed1ab_0         conda-forge        2MB\n",
      "  \u001b[32m+ pydantic                 \u001b[0m      2.9.2  pyhd8ed1ab_0         conda-forge      301kB\n",
      "  \u001b[32m+ sql-magic                \u001b[0m      0.0.4  pyhd8ed1ab_0         conda-forge       15kB\n",
      "  \u001b[32m+ python-kaleido           \u001b[0m      0.2.1  pyhd8ed1ab_0         conda-forge       18kB\n",
      "  \u001b[32m+ ncbi-vdb                 \u001b[0m      3.1.1  h4ac6f70_2           bioconda          11MB\n",
      "  \u001b[32m+ gffread                  \u001b[0m     0.12.7  hdcf5f25_4           bioconda         134kB\n",
      "  \u001b[32m+ entrez-direct            \u001b[0m       22.4  he881be0_0           bioconda          15MB\n",
      "  \u001b[32m+ salmon                   \u001b[0m     1.10.3  h6dccd9a_2           bioconda           6MB\n",
      "  \u001b[32m+ perl-uri                 \u001b[0m       1.71  0                    bioconda          38kB\n",
      "  \u001b[32m+ perl-xml-sax-base        \u001b[0m       1.08  0                    bioconda          13kB\n",
      "  \u001b[32m+ perl-xml-namespacesupport\u001b[0m       1.11  0                    bioconda           6kB\n",
      "  \u001b[32m+ perl-xml-sax             \u001b[0m       0.99  0                    bioconda          25kB\n",
      "  \u001b[32m+ perl-xml-libxml          \u001b[0m     2.0124  0                    bioconda         220kB\n",
      "  \u001b[32m+ sra-tools                \u001b[0m      3.1.1  h4304569_2           bioconda          61MB\n",
      "  \u001b[32m+ multiqc                  \u001b[0m       1.25  pyhdfd78af_0         bioconda           4MB\n",
      "  \u001b[32m+ parallel-fastq-dump      \u001b[0m      0.6.7  pyhdfd78af_0         bioconda          10kB\n",
      "\n",
      "  Change:\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[31m- libgomp                  \u001b[0m     14.1.0  h77fa898_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgomp                  \u001b[0m     14.1.0  h77fa898_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- libgcc-ng                \u001b[0m     14.1.0  h77fa898_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgcc-ng                \u001b[0m     14.1.0  h69a702a_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- libstdcxx-ng             \u001b[0m     14.1.0  hc0a3c3a_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libstdcxx-ng             \u001b[0m     14.1.0  h4852527_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- libxml2                  \u001b[0m     2.12.7  h4c95cb1_3           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libxml2                  \u001b[0m     2.12.7  hc051c1a_1           conda-forge      705kB\n",
      "  \u001b[31m- aws-checksums            \u001b[0m     0.1.18  he027950_7           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-checksums            \u001b[0m     0.1.18  h83b837d_6           conda-forge       50kB\n",
      "  \u001b[31m- aws-c-compression        \u001b[0m     0.2.18  he027950_7           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-compression        \u001b[0m     0.2.18  h83b837d_6           conda-forge       19kB\n",
      "  \u001b[31m- aws-c-sdkutils           \u001b[0m     0.1.16  he027950_3           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-sdkutils           \u001b[0m     0.1.16  h83b837d_2           conda-forge       55kB\n",
      "  \u001b[31m- xorg-libx11              \u001b[0m      1.8.9  hb711507_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xorg-libx11              \u001b[0m      1.8.9  h8ee46fc_0           conda-forge      828kB\n",
      "  \u001b[31m- mysql-common             \u001b[0m      8.3.0  h70512c7_5           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ mysql-common             \u001b[0m      8.3.0  hf1915f5_4           conda-forge      785kB\n",
      "  \u001b[31m- libxkbcommon             \u001b[0m      1.7.0  h2c5496b_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libxkbcommon             \u001b[0m      1.7.0  h662e7e4_0           conda-forge      594kB\n",
      "  \u001b[31m- xcb-util-image           \u001b[0m      0.4.0  hb711507_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xcb-util-image           \u001b[0m      0.4.0  h8ee46fc_1           conda-forge       24kB\n",
      "  \u001b[31m- mysql-libs               \u001b[0m      8.3.0  ha479ceb_5           conda-forge        2MB\n",
      "  \u001b[32m+ mysql-libs               \u001b[0m      8.3.0  hca2cd23_4           conda-forge        2MB\n",
      "  \u001b[31m- imagecodecs              \u001b[0m   2024.6.1  py310h51fded0_2      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ imagecodecs              \u001b[0m   2024.6.1  py310h6158a3f_0      conda-forge        2MB\n",
      "  \u001b[31m- pytables                 \u001b[0m      3.9.2  py310h73b55d5_3      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pytables                 \u001b[0m      3.9.2  py310hd76cd5d_2      conda-forge        1MB\n",
      "  \u001b[31m- cairo                    \u001b[0m     1.18.0  hbb29018_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cairo                    \u001b[0m     1.18.0  h3faef2a_0           conda-forge      982kB\n",
      "  \u001b[31m- aws-c-event-stream       \u001b[0m      0.4.2  h7671281_15          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-event-stream       \u001b[0m      0.4.2  ha47c788_12          conda-forge       54kB\n",
      "  \u001b[31m- aws-c-mqtt               \u001b[0m     0.10.4  hcd6a914_8           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-mqtt               \u001b[0m     0.10.4  h759edc4_4           conda-forge      164kB\n",
      "  \u001b[31m- aws-c-auth               \u001b[0m     0.7.22  hbd3ac97_10          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-auth               \u001b[0m     0.7.22  h96bc93b_2           conda-forge      106kB\n",
      "  \u001b[31m- pango                    \u001b[0m     1.54.0  h4c5309f_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pango                    \u001b[0m     1.54.0  h84a9a3c_0           conda-forge      448kB\n",
      "  \u001b[31m- qt-main                  \u001b[0m     5.15.8  h320f8da_24          conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ qt-main                  \u001b[0m     5.15.8  hc9dc06e_21          conda-forge       61MB\n",
      "  \u001b[31m- aws-sdk-cpp              \u001b[0m   1.11.329  h46c3b66_9           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-sdk-cpp              \u001b[0m   1.11.329  hba8bd5f_3           conda-forge        4MB\n",
      "  \u001b[31m- qt-webengine             \u001b[0m     5.15.8  h8f589be_8           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ qt-webengine             \u001b[0m     5.15.8  h3e791b3_6           conda-forge       59MB\n",
      "\n",
      "  Upgrade:\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[31m- certifi                  \u001b[0m   2024.7.4  pyhd8ed1ab_0         conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ certifi                  \u001b[0m  2024.8.30  pyhd8ed1ab_0         conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- ca-certificates          \u001b[0m   2024.7.4  hbcca054_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ ca-certificates          \u001b[0m  2024.8.30  hbcca054_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- openssl                  \u001b[0m      3.3.1  h4bc722e_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ openssl                  \u001b[0m      3.3.2  hb9d3cd8_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "\n",
      "  Downgrade:\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "  \u001b[31m- libzlib                  \u001b[0m      1.3.1  h4ab18f5_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libzlib                  \u001b[0m     1.2.13  h4ab18f5_6           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- aws-c-common             \u001b[0m     0.9.23  h4ab18f5_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-common             \u001b[0m     0.9.19  h4ab18f5_0           conda-forge      226kB\n",
      "  \u001b[31m- libxcb                   \u001b[0m       1.16  hd590300_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libxcb                   \u001b[0m       1.15  h0b41bf4_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- zlib-ng                  \u001b[0m      2.2.1  he02047a_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ zlib-ng                  \u001b[0m      2.0.7  h0b41bf4_0           conda-forge       95kB\n",
      "  \u001b[31m- nss                      \u001b[0m      3.102  h593d115_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ nss                      \u001b[0m      3.100  hca3bf56_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- llvmlite                 \u001b[0m     0.43.0  py310h4c7c693_0      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ llvmlite                 \u001b[0m     0.42.0  py310h1b8f574_1      conda-forge        3MB\n",
      "  \u001b[31m- llvm-openmp              \u001b[0m     18.1.8  hf5423f3_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ llvm-openmp              \u001b[0m     18.1.7  ha31de31_0           conda-forge       59MB\n",
      "  \u001b[31m- zlib                     \u001b[0m      1.3.1  h4ab18f5_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ zlib                     \u001b[0m     1.2.13  h4ab18f5_6           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- pcre2                    \u001b[0m      10.44  h0f59acf_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pcre2                    \u001b[0m      10.43  hcad00b1_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- blosc                    \u001b[0m     1.21.6  hef167b5_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ blosc                    \u001b[0m     1.21.5  hc2324a3_1           conda-forge       49kB\n",
      "  \u001b[31m- xcb-util-wm              \u001b[0m      0.4.2  hb711507_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xcb-util-wm              \u001b[0m      0.4.1  h8ee46fc_1           conda-forge       52kB\n",
      "  \u001b[31m- xcb-util-keysyms         \u001b[0m      0.4.1  hb711507_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xcb-util-keysyms         \u001b[0m      0.4.0  h8ee46fc_1           conda-forge       14kB\n",
      "  \u001b[31m- xcb-util                 \u001b[0m      0.4.1  hb711507_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xcb-util                 \u001b[0m      0.4.0  hd590300_1           conda-forge       20kB\n",
      "  \u001b[31m- pillow                   \u001b[0m     10.4.0  py310hebfe307_0      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pillow                   \u001b[0m     10.3.0  py310hf73ecf8_0      conda-forge       42MB\n",
      "  \u001b[31m- xcb-util-renderutil      \u001b[0m     0.3.10  hb711507_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ xcb-util-renderutil      \u001b[0m      0.3.9  hd590300_1           conda-forge       17kB\n",
      "  \u001b[31m- libcurl                  \u001b[0m      8.9.0  hdb1bdb2_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libcurl                  \u001b[0m      8.8.0  hca28451_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- s2n                      \u001b[0m     1.4.17  he19d79f_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ s2n                      \u001b[0m     1.4.15  he19d79f_0           conda-forge      350kB\n",
      "  \u001b[31m- aws-c-cal                \u001b[0m      0.7.1  h87b94db_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-cal                \u001b[0m     0.6.14  h88a6e22_1           conda-forge       47kB\n",
      "  \u001b[31m- c-blosc2                 \u001b[0m     2.15.0  h6d6b9e4_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ c-blosc2                 \u001b[0m     2.14.4  hb4ffafa_1           conda-forge      337kB\n",
      "  \u001b[31m- numba                    \u001b[0m     0.60.0  py310h5dc88bb_0      conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ numba                    \u001b[0m     0.59.1  py310h7dc5dd1_0      conda-forge        4MB\n",
      "  \u001b[31m- libllvm18                \u001b[0m     18.1.8  h8b73ec9_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libllvm18                \u001b[0m     18.1.7  hb77312f_0           conda-forge       38MB\n",
      "  \u001b[31m- libglib                  \u001b[0m     2.80.3  h8a4344b_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libglib                  \u001b[0m     2.80.2  hf974151_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[31m- libgoogle-cloud          \u001b[0m     2.26.0  h26d7fe4_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgoogle-cloud          \u001b[0m     2.24.0  h2736e30_0           conda-forge        1MB\n",
      "  \u001b[31m- curl                     \u001b[0m      8.9.0  h18eb788_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ curl                     \u001b[0m      8.8.0  he654da7_1           conda-forge      166kB\n",
      "  \u001b[31m- cmake                    \u001b[0m     3.30.1  hf8c4bd3_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ cmake                    \u001b[0m     3.29.4  h91dbaaa_0           conda-forge       19MB\n",
      "  \u001b[31m- aws-c-io                 \u001b[0m    0.14.10  h826b7d6_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-io                 \u001b[0m     0.14.8  h21d4f22_5           conda-forge      158kB\n",
      "  \u001b[31m- libclang13               \u001b[0m     18.1.8  default_h9def88c_1   conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libclang13               \u001b[0m     18.1.7  default_h087397f_0   conda-forge       11MB\n",
      "  \u001b[31m- glib-tools               \u001b[0m     2.80.3  h73ef956_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ glib-tools               \u001b[0m     2.80.2  hb6ce0ca_0           conda-forge      114kB\n",
      "  \u001b[31m- libgoogle-cloud-storage  \u001b[0m     2.26.0  ha262f82_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libgoogle-cloud-storage  \u001b[0m     2.24.0  h3d9a0c8_0           conda-forge      752kB\n",
      "  \u001b[31m- aws-c-http               \u001b[0m      0.8.2  he17ee6b_6           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-http               \u001b[0m      0.8.1  h29d6fba_17          conda-forge      195kB\n",
      "  \u001b[31m- glib                     \u001b[0m     2.80.3  h8a4344b_1           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ glib                     \u001b[0m     2.80.2  hf974151_0           conda-forge      600kB\n",
      "  \u001b[31m- harfbuzz                 \u001b[0m      9.0.0  hfac3d4d_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ harfbuzz                 \u001b[0m      8.5.0  hfac3d4d_0           conda-forge        2MB\n",
      "  \u001b[31m- gstreamer                \u001b[0m     1.24.6  haf2f30d_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ gstreamer                \u001b[0m     1.24.4  haf2f30d_0           conda-forge        2MB\n",
      "  \u001b[31m- aws-c-s3                 \u001b[0m      0.6.0  h365ddd8_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-c-s3                 \u001b[0m      0.5.9  h594631b_3           conda-forge      110kB\n",
      "  \u001b[31m- gst-plugins-base         \u001b[0m     1.24.6  hbaaba92_0           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ gst-plugins-base         \u001b[0m     1.24.4  h9ad1361_0           conda-forge        3MB\n",
      "  \u001b[31m- aws-crt-cpp              \u001b[0m     0.27.3  hda66527_2           conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ aws-crt-cpp              \u001b[0m     0.26.9  he3a8b3b_0           conda-forge      340kB\n",
      "  \u001b[31m- libarrow                 \u001b[0m     17.0.0  h4b47046_3_cpu       conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libarrow                 \u001b[0m     16.1.0  hcb6531f_6_cpu       conda-forge        8MB\n",
      "  \u001b[31m- pyarrow-core             \u001b[0m     17.0.0  py310hda9dd24_0_cpu  conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ pyarrow-core             \u001b[0m     16.1.0  py310h6f79a3a_1_cpu  conda-forge        4MB\n",
      "  \u001b[31m- libarrow-acero           \u001b[0m     17.0.0  he02047a_3_cpu       conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libarrow-acero           \u001b[0m     16.1.0  hac33072_6_cpu       conda-forge      599kB\n",
      "  \u001b[31m- libparquet               \u001b[0m     17.0.0  h9e5060d_3_cpu       conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libparquet               \u001b[0m     16.1.0  h6a7eafb_6_cpu       conda-forge        1MB\n",
      "  \u001b[31m- libarrow-dataset         \u001b[0m     17.0.0  he02047a_3_cpu       conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libarrow-dataset         \u001b[0m     16.1.0  hac33072_6_cpu       conda-forge      580kB\n",
      "  \u001b[31m- libarrow-substrait       \u001b[0m     17.0.0  hc9a23c6_3_cpu       conda-forge\u001b[32m     Cached\u001b[0m\n",
      "  \u001b[32m+ libarrow-substrait       \u001b[0m     16.1.0  h7e0c224_6_cpu       conda-forge      549kB\n",
      "  \u001b[31m- pyarrow                  \u001b[0m     17.0.0  py310hb7f781d_0      conda-forge       26kB\n",
      "  \u001b[32m+ pyarrow                  \u001b[0m     16.1.0  py310h17c5347_1      conda-forge       27kB\n",
      "\n",
      "  Summary:\n",
      "\n",
      "  Install: 71 packages\n",
      "  Change: 22 packages\n",
      "  Upgrade: 3 packages\n",
      "  Downgrade: 43 packages\n",
      "\n",
      "  Total download: 759MB\n",
      "\n",
      "───────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\n",
      "\u001b[?25l\u001b[2K\u001b[0G[+] 0.0s\n",
      "Downloading      \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   0.0 B                            0.0s\n",
      "Extracting       \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m       0                            0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 0.1s\n",
      "Downloading  (5) \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   0.0 B boto                       0.0s\n",
      "Extracting       \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m       0                            0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghttplib2                                            98.0kB @ 894.2kB/s  0.1s\n",
      "pyaml-env                                           14.4kB @ 106.2kB/s  0.1s\n",
      "colormath                                           35.3kB @ 259.1kB/s  0.1s\n",
      "pyu2f                                               31.9kB @ 234.1kB/s  0.1s\n",
      "monotonic                                           13.3kB @  85.3kB/s  0.0s\n",
      "boto                                               857.7kB @   5.2MB/s  0.2s\n",
      "sqlparse                                            40.4kB @ 239.6kB/s  0.0s\n",
      "findspark                                            8.4kB @  49.3kB/s  0.0s\n",
      "pyasn1-modules                                      95.9kB @ 533.7kB/s  0.0s\n",
      "[+] 0.2s\n",
      "Downloading  (5) \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m   1.2MB aws-c-common               0.1s\n",
      "Extracting   (9) \u001b[90m━━━━━━━━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━\u001b[0m       0 boto                       0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ggcs-oauth2-boto-plugin                              30.1kB @ 149.8kB/s  0.0s\n",
      "pyathena                                            55.1kB @ 270.0kB/s  0.0s\n",
      "oauth2client                                        67.5kB @ 318.8kB/s  0.1s\n",
      "aws-c-common                                       226.5kB @ 977.0kB/s  0.1s\n",
      "aws-checksums                                       50.1kB @ 204.7kB/s  0.0s\n",
      "libjemalloc                                          1.6MB @   5.8MB/s  0.1s\n",
      "xcb-util-wm                                         52.1kB @ 187.2kB/s  0.0s\n",
      "[+] 0.3s\n",
      "Downloading  (5) \u001b[33m━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━\u001b[0m  10.6MB c-blosc2                   0.2s\n",
      "Extracting  (16) \u001b[90m━━━━━━━━━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━\u001b[0m       0 boto                       0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gmysql-common                                       784.8kB @   2.5MB/s  0.0s\n",
      "c-blosc2                                           337.1kB @   1.0MB/s  0.1s\n",
      "libxkbcommon                                       593.5kB @   1.6MB/s  0.0s\n",
      "[+] 0.4s\n",
      "Downloading  (5) ╸\u001b[33m━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━\u001b[0m  49.1MB libgoogle-cloud            0.3s\n",
      "Extracting  (19) \u001b[33m━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m       0 boto                       0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gnumba                                                4.3MB @  10.6MB/s  0.1s\n",
      "mathjax                                             22.3MB @  53.5MB/s  0.2s\n",
      "aws-c-io                                           158.0kB @ 355.7kB/s  0.0s\n",
      "boost-cpp                                           17.9kB @  40.1kB/s  0.0s\n",
      "libgoogle-cloud                                      1.2MB @   2.7MB/s  0.1s\n",
      "aws-c-mqtt                                         164.2kB @ 331.3kB/s  0.0s\n",
      "[+] 0.5s\n",
      "Downloading  (5) ━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m  78.5MB cairo                      0.4s\n",
      "Extracting  (23) \u001b[33m━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━\u001b[0m       1 boto                       0.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gaws-c-http                                         194.8kB @ 391.7kB/s  0.1s\n",
      "cairo                                              982.4kB @   1.9MB/s  0.1s\n",
      "pango                                              448.5kB @ 821.7kB/s  0.0s\n",
      "[+] 0.6s\n",
      "Downloading  (5) ━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m  87.8MB aws-sdk-cpp                0.5s\n",
      "Extracting  (26) \u001b[33m━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━\u001b[0m       2 c-blosc2                   0.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gaws-sdk-cpp                                          3.6MB @   5.9MB/s  0.1s\n",
      "libarrow-dataset                                   580.0kB @ 858.8kB/s  0.1s\n",
      "[+] 0.7s\n",
      "Downloading  (5) ━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m 107.0MB llvm-openmp                0.6s\n",
      "Extracting  (28) \u001b[33m━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━\u001b[0m       2 c-blosc2                   0.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gpyarrow-core                                         4.4MB @   6.3MB/s  0.2s\n",
      "perl-threaded                                        5.5kB @   7.1kB/s  0.1s\n",
      "[+] 0.8s\n",
      "Downloading  (5) ━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━━\u001b[0m 118.2MB llvm-openmp                0.7s\n",
      "Extracting  (30) \u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m       2 c-blosc2                   0.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ggsutil                                               2.3MB @   2.9MB/s  0.1s\n",
      "python-kaleido                                      18.3kB @  22.9kB/s  0.0s\n",
      "perl-xml-sax-base                                   12.9kB @  15.4kB/s  0.0s\n",
      "perl-xml-libxml                                    219.6kB @ 249.9kB/s  0.0s\n",
      "pillow                                              41.8MB @  47.0MB/s  0.6s\n",
      "[+] 0.9s\n",
      "Downloading  (5) ━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━━\u001b[0m 158.0MB llvm-openmp                0.8s\n",
      "Extracting  (34) \u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m       2 c-blosc2                   0.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gparallel-fastq-dump                                  9.9kB @  10.9kB/s  0.0s\n",
      "retry_decorator                                      6.6kB @   7.0kB/s  0.0s\n",
      "llvm-openmp                                         58.7MB @  60.5MB/s  0.8s\n",
      "entrez-direct                                       14.7MB @  15.2MB/s  0.2s\n",
      "spectra                                             19.4kB @  19.8kB/s  0.0s\n",
      "[+] 1.0s\n",
      "Downloading  (5) ━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━\u001b[0m 198.3MB botocore                   0.9s\n",
      "Extracting  (40) \u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m       2 cairo                      0.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gmarkdown                                            78.3kB @  78.1kB/s  0.0s\n",
      "coloredlogs                                         40.6kB @  40.4kB/s  0.0s\n",
      "pydantic-core                                        1.6MB @   1.5MB/s  0.1s\n",
      "[+] 1.1s\n",
      "Downloading  (5) ━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━━\u001b[0m 207.4MB botocore                   1.0s\n",
      "Extracting  (43) \u001b[90m╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m       2 cairo                      0.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibboost                                             2.9MB @   2.6MB/s  0.1s\n",
      "blosc                                               48.7kB @  44.0kB/s  0.0s\n",
      "xcb-util-keysyms                                    14.2kB @  12.2kB/s  0.1s\n",
      "xcb-util                                            19.7kB @  17.0kB/s  0.1s\n",
      "botocore                                             7.0MB @   6.0MB/s  0.3s\n",
      "[+] 1.2s\n",
      "Downloading  (5) ━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━━\u001b[0m 221.2MB kaleido-core               1.1s\n",
      "Extracting  (47) \u001b[90m━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m       3 cairo                      1.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibboost-devel                                      41.0kB @  34.2kB/s  0.0s\n",
      "xcb-util-image                                      24.5kB @  20.1kB/s  0.1s\n",
      "libgoogle-cloud-storage                            752.4kB @ 584.8kB/s  0.1s\n",
      "[+] 1.3s\n",
      "Downloading  (4) ━━━━━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━━\u001b[0m 230.6MB kaleido-core               1.2s\n",
      "Extracting  (49) \u001b[90m━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m       4 cairo                      1.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gpytables                                             1.4MB @   1.0MB/s  0.1s\n",
      "aws-c-s3                                           109.7kB @  80.5kB/s  0.0s\n",
      "harfbuzz                                             1.6MB @   1.1MB/s  0.1s\n",
      "[+] 1.4s\n",
      "Downloading  (5) ━━━━━━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m 245.7MB kaleido-core               1.3s\n",
      "Extracting  (51) \u001b[90m━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m       4 coloredlogs                1.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gperl                                                15.7MB @  11.1MB/s  0.4s\n",
      "libarrow-acero                                     599.3kB @ 420.0kB/s  0.0s\n",
      "pyarrow                                             26.7kB @  18.4kB/s  0.0s\n",
      "pydantic                                           300.6kB @ 205.2kB/s  0.0s\n",
      "[+] 1.5s\n",
      "Downloading  (5) ━━━━━━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m 265.0MB kaleido-core               1.4s\n",
      "Extracting  (56) \u001b[90m━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━\u001b[0m       4 coloredlogs                1.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ggffread                                            133.6kB @  86.9kB/s  0.1s\n",
      "perl-xml-namespacesupport                            6.4kB @   4.2kB/s  0.1s\n",
      "[+] 1.6s\n",
      "Downloading  (5) ━━━━━━━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 277.7MB multiqc                    1.5s\n",
      "Extracting  (57) \u001b[90m━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m       5 coloredlogs                1.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ghumanize                                            64.8kB @  40.3kB/s  0.1s\n",
      "humanfriendly                                       73.4kB @  44.1kB/s  0.0s\n",
      "[+] 1.7s\n",
      "Downloading  (5) ━━━━━━━╸\u001b[33m━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 286.0MB multiqc                    1.6s\n",
      "Extracting  (59) \u001b[90m━━━━━━╸\u001b[0m\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m       5 coloredlogs                1.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gboto3                                               80.9kB @  47.2kB/s  0.0s\n",
      "multiqc                                              3.7MB @   2.2MB/s  0.2s\n",
      "google-auth-httplib2                                15.0kB @   8.5kB/s  0.0s\n",
      "[+] 1.8s\n",
      "Downloading  (5) ━━━━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 312.4MB kaleido-core               1.7s\n",
      "Extracting  (61) ╸\u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m       6 colormath                  1.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gllvmlite                                             3.3MB @   1.8MB/s  0.1s\n",
      "[+] 1.9s\n",
      "Downloading  (5) ━━━━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 322.1MB kaleido-core               1.8s\n",
      "Extracting  (62) ╸\u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m       6 colormath                  1.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gxorg-libx11                                        828.1kB @ 430.5kB/s  0.1s\n",
      "aws-c-cal                                           46.7kB @  23.8kB/s  0.0s\n",
      "[+] 2.0s\n",
      "Downloading  (5) ━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 339.9MB kaleido-core               1.9s\n",
      "Extracting  (63) ╸\u001b[33m━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━━\u001b[0m       7 colormath                  1.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gmysql-libs                                           1.5MB @ 757.2kB/s  0.1s\n",
      "libboost-headers                                    14.0MB @   6.7MB/s  0.3s\n",
      "imagecodecs                                          2.1MB @ 986.6kB/s  0.1s\n",
      "[+] 2.1s\n",
      "Downloading  (5) ━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m 363.7MB kaleido-core               2.0s\n",
      "Extracting  (64) ╸\u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m       7 colormath                  1.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gaws-c-event-stream                                  54.0kB @  25.0kB/s  0.1s\n",
      "glib                                               600.4kB @ 276.0kB/s  0.1s\n",
      "[+] 2.2s\n",
      "Downloading  (5) ━━━━━━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 375.3MB libarrow-substrait         2.1s\n",
      "Extracting  (66) ╸\u001b[33m━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━━\u001b[0m       9 entrez-direct              2.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibarrow-substrait                                 549.5kB @ 244.7kB/s  0.1s\n",
      "sql-magic                                           15.1kB @   6.6kB/s  0.0s\n",
      "[+] 2.3s\n",
      "Downloading  (5) ━━━━━━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 385.6MB kaleido-core               2.2s\n",
      "Extracting  (67) ╸\u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━\u001b[0m       9 entrez-direct              2.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gperl-uri                                            37.7kB @  16.2kB/s  0.0s\n",
      "rich-click                                          33.8kB @  14.3kB/s  0.0s\n",
      "[+] 2.4s\n",
      "Downloading  (5) ━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 405.5MB kaleido-core               2.3s\n",
      "Extracting  (70) ╸\u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━━\u001b[0m       9 entrez-direct              2.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gcachetools                                          14.7kB @   6.1kB/s  0.0s\n",
      "[+] 2.5s\n",
      "Downloading  (4) ━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 414.3MB kaleido-core               2.4s\n",
      "Extracting  (70) ━╸\u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m      11 entrez-direct              2.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gossuuid                                             56.9kB @  22.4kB/s  0.0s\n",
      "aws-c-compression                                   19.3kB @   7.5kB/s  0.0s\n",
      "[+] 2.6s\n",
      "Downloading  (5) ━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 424.3MB kaleido-core               2.5s\n",
      "Extracting  (72) ━╸\u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m      11 findspark                  2.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gs2n                                                349.5kB @ 133.5kB/s  0.0s\n",
      "[+] 2.7s\n",
      "Downloading  (4) ━━━━━━━━━━━━╸\u001b[33m━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 443.7MB openjdk                    2.6s\n",
      "Extracting  (72) ━╸\u001b[33m━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m      13 findspark                  2.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gglib-tools                                         114.4kB @  41.4kB/s  0.0s\n",
      "[+] 2.8s\n",
      "Downloading  (5) ━━━━━━━━━━━━╸\u001b[33m━━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 454.4MB openjdk                    2.7s\n",
      "Extracting  (73) ━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m      13 findspark                  2.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gkaleido-core                                        62.1MB @  21.5MB/s  1.7s\n",
      "gst-plugins-base                                     2.8MB @ 962.9kB/s  0.1s\n",
      "[+] 2.9s\n",
      "Downloading  (5) ━━━━━━━━━━━━━╸\u001b[33m━━╸\u001b[0m\u001b[90m━━━━━━\u001b[0m 471.4MB openjdk                    2.8s\n",
      "Extracting  (72) ━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m      14 findspark                  2.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibparquet                                           1.2MB @ 403.1kB/s  0.1s\n",
      "[+] 3.0s\n",
      "Downloading  (5) ━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m 480.1MB openjdk                    2.9s\n",
      "Extracting  (75) ━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m      14 gcs-oauth2-boto-plugin     2.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 3.1s\n",
      "Downloading  (5) ━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m 498.3MB qt-main                    3.0s\n",
      "Extracting  (74) ━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m      15 gcs-oauth2-boto-plugin     2.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gqt-main                                             61.3MB @  19.4MB/s  1.8s\n",
      "[+] 3.2s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━╸\u001b[33m━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m 509.0MB fastqc                     3.1s\n",
      "Extracting  (75) ━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━━\u001b[0m      15 gcs-oauth2-boto-plugin     3.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ggoogle-reauth                                       17.8kB @   5.5kB/s  0.0s\n",
      "crcmod                                              36.1kB @  11.1kB/s  0.0s\n",
      "aws-c-sdkutils                                      55.0kB @  16.7kB/s  0.0s\n",
      "[+] 3.3s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━╸\u001b[0m\u001b[90m━━\u001b[0m 520.0MB fastqc                     3.2s\n",
      "Extracting  (77) ━╸\u001b[33m━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      15 gcs-oauth2-boto-plugin     3.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gfastqc                                              11.7MB @   3.5MB/s  0.4s\n",
      "aws-c-auth                                         105.9kB @  31.5kB/s  0.0s\n",
      "[+] 3.4s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━\u001b[0m 533.5MB libarrow                   3.3s\n",
      "Extracting  (80) ━╸\u001b[33m━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      15 glib                       3.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 3.5s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━\u001b[0m 544.1MB libarrow                   3.4s\n",
      "Extracting  (79) ━━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      16 glib                       3.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 3.6s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━\u001b[0m 548.4MB libarrow                   3.5s\n",
      "Extracting  (78) ━━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      17 glib                       3.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 3.7s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━╸\u001b[0m\u001b[90m━━\u001b[0m 554.3MB libarrow                   3.6s\n",
      "Extracting  (78) ━━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      17 glib                       3.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibarrow                                             8.3MB @   2.2MB/s  0.4s\n",
      "[+] 3.8s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━╸\u001b[0m\u001b[90m━━\u001b[0m 572.5MB libllvm18                  3.7s\n",
      "Extracting  (78) ━━╸\u001b[33m━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━━\u001b[0m      17 glib-tools                 3.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 3.9s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━╸\u001b[0m\u001b[90m━━\u001b[0m 591.4MB libllvm18                  3.8s\n",
      "Extracting  (79) ━━╸\u001b[33m━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m      17 glib-tools                 3.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gsalmon                                               6.3MB @   1.6MB/s  0.1s\n",
      "fasteners                                           20.0kB @   5.0kB/s  0.0s\n",
      "[+] 4.0s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━╸\u001b[0m\u001b[90m━━\u001b[0m 623.3MB libllvm18                  3.9s\n",
      "Extracting  (80) ━━╸\u001b[33m━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━━\u001b[0m      17 glib-tools                 3.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gqt-webengine                                        58.5MB @  14.5MB/s  1.9s\n",
      "zlib-ng                                             94.6kB @  23.4kB/s  0.0s\n",
      "wget                                               770.4kB @ 189.2kB/s  0.0s\n",
      "[+] 4.1s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━╸\u001b[0m\u001b[90m━━\u001b[0m 652.1MB ncbi-vdb                   4.0s\n",
      "Extracting  (83) ━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━\u001b[0m      18 google-auth                3.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Ggstreamer                                            2.0MB @ 488.0kB/s  0.1s\n",
      "[+] 4.2s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━╸\u001b[0m\u001b[90m━━\u001b[0m 659.5MB ncbi-vdb                   4.1s\n",
      "Extracting  (83) ━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━\u001b[0m      19 google-auth                4.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gtypeguard                                           34.5kB @   8.2kB/s  0.1s\n",
      "xcb-util-renderutil                                 17.0kB @   4.0kB/s  0.1s\n",
      "[+] 4.3s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m╸\u001b[0m\u001b[90m━━\u001b[0m 672.6MB ncbi-vdb                   4.2s\n",
      "Extracting  (84) ━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━━\u001b[0m      20 google-auth                4.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gaws-crt-cpp                                        339.6kB @  78.6kB/s  0.0s\n",
      "argcomplete                                         41.1kB @   9.4kB/s  0.0s\n",
      "[+] 4.4s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━╸\u001b[0m\u001b[90m━\u001b[0m 685.9MB ncbi-vdb                   4.3s\n",
      "Extracting  (86) ━━╸\u001b[33m━━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m      20 google-auth                4.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 4.5s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━╸\u001b[0m\u001b[90m━\u001b[0m 693.1MB openjdk                    4.4s\n",
      "Extracting  (84) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m      22 google-auth-httplib2       4.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibllvm18                                           38.4MB @   8.6MB/s  1.2s\n",
      "perl-xml-sax                                        24.7kB @   5.4kB/s  0.1s\n",
      "ncbi-vdb                                            11.2MB @   2.4MB/s  0.5s\n",
      "[+] 4.6s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m╸\u001b[0m\u001b[90m━\u001b[0m 701.7MB openjdk                    4.5s\n",
      "Extracting  (87) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m      22 google-auth-httplib2       4.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gtrimmomatic                                        147.4kB @  31.9kB/s  0.0s\n",
      "[+] 4.7s\n",
      "Downloading  (5) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m╸\u001b[0m\u001b[90m━\u001b[0m 707.2MB openjdk                    4.6s\n",
      "Extracting  (88) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━╸\u001b[0m\u001b[90m━\u001b[0m      22 google-auth-httplib2       4.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Glibxml2                                            705.0kB @ 149.4kB/s  0.1s\n",
      "google-apitools                                     96.4kB @  20.2kB/s  0.1s\n",
      "[+] 4.8s\n",
      "Downloading  (4) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m 716.4MB openjdk                    4.7s\n",
      "Extracting  (90) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      22 google-auth-httplib2       4.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 4.9s\n",
      "Downloading  (4) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m 728.7MB sra-tools                  4.8s\n",
      "Extracting  (90) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      22 google-reauth              4.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.0s\n",
      "Downloading  (4) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m 741.0MB sra-tools                  4.9s\n",
      "Extracting  (89) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      23 google-reauth              4.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gsra-tools                                           60.7MB @  12.1MB/s  2.1s\n",
      "libclang13                                          11.1MB @   2.2MB/s  0.5s\n",
      "[+] 5.1s\n",
      "Downloading  (2) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m 746.1MB cmake                      5.0s\n",
      "Extracting  (91) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      23 google-reauth              4.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0Gcmake                                               19.0MB @   3.7MB/s  0.8s\n",
      "openjdk                                            180.9MB @  35.0MB/s  4.7s\n",
      "[+] 5.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (93) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      23 google-reauth              5.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (93) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      23 gst-plugins-base           5.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (92) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      24 gst-plugins-base           5.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (91) ━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━━\u001b[0m      25 gst-plugins-base           5.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (90) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      26 gst-plugins-base           5.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (90) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      26 gstreamer                  5.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (90) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      26 gstreamer                  5.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 5.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (89) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      27 gstreamer                  5.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (88) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      28 gstreamer                  5.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (87) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      29 gsutil                     5.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (86) ━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━━\u001b[0m      30 gsutil                     6.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (85) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      31 gsutil                     6.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (84) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      32 gsutil                     6.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (83) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      33 harfbuzz                   6.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (83) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      33 harfbuzz                   6.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (82) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      34 harfbuzz                   6.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (82) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      34 harfbuzz                   6.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 6.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (81) ━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━━\u001b[0m      35 humanfriendly              6.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (80) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      36 humanfriendly              6.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (79) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      37 humanfriendly              6.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (78) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      38 humanfriendly              7.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 humanize                   7.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 humanize                   7.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 humanize                   7.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 humanize                   7.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 kaleido-core               7.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 kaleido-core               7.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 7.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 kaleido-core               7.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 kaleido-core               7.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow                   7.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow                   8.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow                   8.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow                   8.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow-acero             8.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (77) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      39 libarrow-acero             8.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-acero             8.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-acero             8.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 8.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-dataset           8.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-dataset           8.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-dataset           8.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-dataset           9.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-substrait         9.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (76) ━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━━\u001b[0m      40 libarrow-substrait         9.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (75) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      41 libarrow-substrait         9.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (75) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      41 libarrow-substrait         9.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (74) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      42 libboost                   9.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (73) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      43 libboost                   9.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 9.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost                   9.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost                   9.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost-devel             9.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost-devel            10.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost-devel            10.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost-devel            10.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (72) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      44 libboost-headers          10.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (71) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      45 libboost-headers          10.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (71) ━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━━\u001b[0m      45 libboost-headers          10.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (70) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      46 libboost-headers          10.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 10.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (70) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      46 libclang13                10.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (69) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      47 libclang13                10.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (68) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      48 libclang13                10.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (68) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      48 libclang13                11.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (68) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      48 libllvm18                 11.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (67) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      49 libllvm18                 11.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (67) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      49 libllvm18                 11.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (66) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      50 libllvm18                 11.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (66) ━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━━\u001b[0m      50 libparquet                11.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (65) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      51 libparquet                11.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 11.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (65) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      51 libparquet                11.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (65) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      51 libparquet                11.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (64) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      52 libxml2                   11.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (64) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      52 libxml2                   12.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (64) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      52 libxml2                   12.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (63) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      53 libxml2                   12.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (62) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      54 llvm-openmp               12.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 llvm-openmp               12.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 llvm-openmp               12.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 llvm-openmp               12.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 12.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 markdown                  12.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 markdown                  12.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 markdown                  12.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 markdown                  13.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mathjax                   13.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mathjax                   13.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mathjax                   13.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mathjax                   13.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mysql-libs                13.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mysql-libs                13.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 13.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mysql-libs                13.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 mysql-libs                13.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ncbi-vdb                  13.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ncbi-vdb                  14.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ncbi-vdb                  14.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ncbi-vdb                  14.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 openjdk                   14.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 openjdk                   14.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 openjdk                   14.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 openjdk                   14.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 14.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ossuuid                   14.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ossuuid                   14.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ossuuid                   14.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 ossuuid                   15.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 parallel-fastq-dump       15.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 parallel-fastq-dump       15.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 parallel-fastq-dump       15.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 parallel-fastq-dump       15.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl                      15.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl                      15.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 15.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl                      15.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl                      15.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-threaded             15.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-threaded             16.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-threaded             16.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-threaded             16.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-uri                  16.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-uri                  16.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-uri                  16.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-uri                  16.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 16.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-namespacesupport 16.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-namespacesupport 16.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-namespacesupport 16.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-namespacesupport 17.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-sax              17.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-sax              17.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-sax              17.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 perl-xml-sax              17.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pillow                    17.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pillow                    17.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 17.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pillow                    17.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pillow                    17.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pyarrow                   17.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pyarrow                   18.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (61) ━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━━\u001b[0m      55 pyarrow                   18.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (60) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      56 pyarrow                   18.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (60) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      56 pyasn1-modules            18.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (59) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      57 pyasn1-modules            18.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (58) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      58 pyasn1-modules            18.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (58) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      58 pyasn1-modules            18.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 18.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (57) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      59 pydantic-core             18.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (56) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      60 pydantic-core             18.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (56) ━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━━\u001b[0m      60 pydantic-core             18.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (55) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      61 pydantic-core             19.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (55) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      61 pytables                  19.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (54) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      62 pytables                  19.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (53) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      63 pytables                  19.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (53) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      63 pytables                  19.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (52) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      64 python-kaleido            19.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (52) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      64 python-kaleido            19.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 19.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (51) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      65 python-kaleido            19.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (51) ━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━━\u001b[0m      65 python-kaleido            19.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-main                   19.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-main                   20.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-main                   20.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-main                   20.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-webengine              20.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-webengine              20.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-webengine              20.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 qt-webengine              20.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 20.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 rich-click                20.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 rich-click                20.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 rich-click                20.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 rich-click                21.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (50) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      66 salmon                    21.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 salmon                    21.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 salmon                    21.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 salmon                    21.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 spectra                   21.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 spectra                   21.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 21.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 spectra                   21.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 spectra                   21.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 sra-tools                 21.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 sra-tools                 22.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 sra-tools                 22.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 sra-tools                 22.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 trimmomatic               22.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 trimmomatic               22.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 trimmomatic               22.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 trimmomatic               22.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 22.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 wget                      22.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 wget                      22.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 wget                      22.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 wget                      23.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (49) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      67 xcb-util                  23.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (47) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      69 xcb-util                  23.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (47) ━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━━\u001b[0m      69 xcb-util                  23.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (45) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      71 xcb-util                  23.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (44) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      72 xcb-util-renderutil       23.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (43) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      73 xcb-util-renderutil       23.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 23.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (42) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      74 xcb-util-renderutil       23.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (41) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      75 argcomplete               23.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (41) ━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━━\u001b[0m      75 argcomplete               23.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 argcomplete               24.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 argcomplete               24.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 aws-c-cal                 24.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 aws-c-cal                 24.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 aws-c-cal                 24.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 aws-c-cal                 24.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 boost-cpp                 24.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 24.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 boost-cpp                 24.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 boost-cpp                 24.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 boost-cpp                 24.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 cmake                     25.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 cmake                     25.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 cmake                     25.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 cmake                     25.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 curl                      25.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 curl                      25.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 curl                      25.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 25.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 curl                      25.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 entrez-direct             25.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 entrez-direct             25.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 entrez-direct             26.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 entrez-direct             26.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 fastqc                    26.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 fastqc                    26.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 fastqc                    26.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 fastqc                    26.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 glib                      26.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 26.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 glib                      26.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 glib                      26.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (40) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      76 glib                      26.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (39) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      77 gst-plugins-base          27.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (38) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      78 gst-plugins-base          27.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 gst-plugins-base          27.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 gst-plugins-base          27.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 humanize                  27.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 humanize                  27.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 humanize                  27.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 27.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 humanize                  27.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libarrow-acero            27.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libarrow-acero            27.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libarrow-acero            28.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libarrow-acero            28.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost                  28.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost                  28.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost                  28.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost                  28.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-devel            28.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 28.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-devel            28.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-devel            28.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-devel            28.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-headers          29.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-headers          29.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-headers          29.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libboost-headers          29.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libclang13                29.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libclang13                29.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libclang13                29.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 29.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libclang13                29.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libllvm18                 29.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libllvm18                 29.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libllvm18                 30.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libllvm18                 30.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libparquet                30.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (37) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      79 libparquet                30.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (36) ━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━━\u001b[0m      80 libparquet                30.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (35) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      81 libparquet                30.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (35) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      81 libxml2                   30.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 30.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (34) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      82 libxml2                   30.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 libxml2                   30.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 libxml2                   30.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mathjax                   31.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mathjax                   31.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mathjax                   31.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mathjax                   31.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mysql-libs                31.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mysql-libs                31.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mysql-libs                31.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 31.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 mysql-libs                31.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 ncbi-vdb                  31.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 ncbi-vdb                  31.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 ncbi-vdb                  32.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 ncbi-vdb                  32.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 openjdk                   32.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 openjdk                   32.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 openjdk                   32.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 openjdk                   32.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 parallel-fastq-dump       32.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 32.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 parallel-fastq-dump       32.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 parallel-fastq-dump       32.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 parallel-fastq-dump       32.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl                      33.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl                      33.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl                      33.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl                      33.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl-threaded             33.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl-threaded             33.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (33) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      83 perl-threaded             33.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 33.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-threaded             33.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-uri                  33.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-uri                  33.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-uri                  34.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-uri                  34.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-xml-sax              34.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-xml-sax              34.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-xml-sax              34.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 perl-xml-sax              34.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pyasn1-modules            34.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 34.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pyasn1-modules            34.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pyasn1-modules            34.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pyasn1-modules            34.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pytables                  35.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pytables                  35.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pytables                  35.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 pytables                  35.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 python-kaleido            35.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 python-kaleido            35.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 python-kaleido            35.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 35.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 python-kaleido            35.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 qt-webengine              35.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 qt-webengine              35.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 qt-webengine              36.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 qt-webengine              36.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 spectra                   36.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 spectra                   36.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 spectra                   36.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 spectra                   36.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 sra-tools                 36.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 36.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 sra-tools                 36.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (32) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      84 sra-tools                 36.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 wget                      36.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 wget                      37.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 wget                      37.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 wget                      37.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 xcb-util-renderutil       37.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 xcb-util-renderutil       37.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 xcb-util-renderutil       37.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 xcb-util-renderutil       37.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 37.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 argcomplete               37.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 argcomplete               37.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 argcomplete               37.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 argcomplete               38.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 boost-cpp                 38.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 boost-cpp                 38.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 boost-cpp                 38.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 boost-cpp                 38.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 curl                      38.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 curl                      38.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 38.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 curl                      38.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 curl                      38.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 entrez-direct             38.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 entrez-direct             39.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 entrez-direct             39.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 entrez-direct             39.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 glib                      39.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 glib                      39.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 glib                      39.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 glib                      39.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 39.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 gst-plugins-base          39.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 gst-plugins-base          39.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 gst-plugins-base          39.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 gst-plugins-base          40.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 humanize                  40.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 humanize                  40.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 humanize                  40.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 humanize                  40.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libarrow-acero            40.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libarrow-acero            40.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 40.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libarrow-acero            40.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libarrow-acero            40.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libboost                  40.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libboost                  41.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libboost                  41.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libboost                  41.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (31) ━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━━\u001b[0m      85 libboost-devel            41.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libboost-devel            41.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libboost-devel            41.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libboost-devel            41.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 41.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libclang13                41.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libclang13                41.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (30) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      86 libclang13                41.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (29) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      87 libclang13                42.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (28) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      88 libllvm18                 42.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libllvm18                 42.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libllvm18                 42.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libllvm18                 42.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libparquet                42.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libparquet                42.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 42.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libparquet                42.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libparquet                42.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libxml2                   42.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libxml2                   43.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libxml2                   43.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 libxml2                   43.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mathjax                   43.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mathjax                   43.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mathjax                   43.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mathjax                   43.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 43.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mysql-libs                43.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mysql-libs                43.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mysql-libs                43.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 mysql-libs                44.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 ncbi-vdb                  44.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 ncbi-vdb                  44.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 ncbi-vdb                  44.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 ncbi-vdb                  44.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 openjdk                   44.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 openjdk                   44.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 44.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 openjdk                   44.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 openjdk                   44.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 perl                      44.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 perl                      45.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (27) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      89 perl                      45.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (26) ━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━━\u001b[0m      90 perl                      45.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (25) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      91 perl-threaded             45.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (25) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      91 perl-threaded             45.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-threaded             45.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-threaded             45.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 45.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-uri                  45.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-uri                  45.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-uri                  45.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-uri                  46.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-xml-sax              46.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-xml-sax              46.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-xml-sax              46.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 perl-xml-sax              46.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 pyasn1-modules            46.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 pyasn1-modules            46.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 46.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 pyasn1-modules            46.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 pyasn1-modules            46.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 spectra                   46.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 spectra                   47.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 spectra                   47.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 spectra                   47.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 xcb-util-renderutil       47.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 xcb-util-renderutil       47.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 xcb-util-renderutil       47.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 xcb-util-renderutil       47.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 47.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (24) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      92 argcomplete               47.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (23) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      93 argcomplete               47.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (23) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      93 argcomplete               47.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (22) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      94 argcomplete               48.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (22) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      94 boost-cpp                 48.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (21) ━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━━\u001b[0m      95 boost-cpp                 48.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (20) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━\u001b[0m      96 boost-cpp                 48.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (18) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━\u001b[0m      98 boost-cpp                 48.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (18) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━\u001b[0m      98 glib                      48.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (17) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━\u001b[0m      99 glib                      48.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 48.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (16) ━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━━\u001b[0m     100 argcomplete               48.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (15) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     101 argcomplete               48.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (15) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     101 argcomplete               48.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 argcomplete               49.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 boost-cpp                 49.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 boost-cpp                 49.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 boost-cpp                 49.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 boost-cpp                 49.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 humanize                  49.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 humanize                  49.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 49.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 humanize                  49.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 humanize                  49.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libarrow-acero            49.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libarrow-acero            50.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libarrow-acero            50.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libarrow-acero            50.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libclang13                50.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libclang13                50.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (14) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     102 libclang13                50.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 argcomplete               50.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 50.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 argcomplete               50.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 argcomplete               50.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 argcomplete               50.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 boost-cpp                 51.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 boost-cpp                 51.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 boost-cpp                 51.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 boost-cpp                 51.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 humanize                  51.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 humanize                  51.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 humanize                  51.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 51.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 humanize                  51.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 libarrow-acero            51.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 libarrow-acero            51.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (13) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     103 libarrow-acero            52.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (12) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     104 libarrow-acero            52.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (11) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━━\u001b[0m     105 mathjax                   52.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting  (10) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     106 mathjax                   52.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (9) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     107 mathjax                   52.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (8) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     108 mathjax                   52.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (7) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     109 perl                      52.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 52.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (6) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     110 perl                      52.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (6) ━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━━\u001b[0m     110 perl                      52.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 perl                      52.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   53.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   53.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   53.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   53.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               53.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               53.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               53.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 53.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               53.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 libarrow-acero            53.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 libarrow-acero            53.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 libarrow-acero            54.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 libarrow-acero            54.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 mathjax                   54.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 mathjax                   54.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 mathjax                   54.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 mathjax                   54.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 perl                      54.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 54.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 perl                      54.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 perl                      54.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 perl                      54.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   55.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   55.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   55.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 spectra                   55.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               55.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               55.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               55.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 55.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (5) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     111 argcomplete               55.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (4) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     112 libarrow-acero            55.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (3) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     113 libarrow-acero            55.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (2) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     114 libarrow-acero            56.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (2) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     114 libarrow-acero            56.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 56.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   56.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 57.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   57.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 58.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   58.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 59.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━━━ 759.2MB                            5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   59.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:0.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:0.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:1.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:1.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.6s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.4s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.7s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.5s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.8s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.6s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:2.9s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.7s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.0s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.8s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.1s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:2.9s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.2s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:3.0s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.3s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:3.1s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.4s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:3.2s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G[+] 1m:3.5s\n",
      "Downloading      ━━━━━━━━━━━━━━━━━━━━━ 759.2MB                              5.1s\n",
      "Extracting   (1) ━━━━━━━━━━━━━━━━━━━╸\u001b[33m━\u001b[0m     115 mathjax                   1m:3.3s\u001b[2K\u001b[1A\u001b[2K\u001b[1A\u001b[2K\u001b[0G\u001b[?25h\n",
      "Downloading and Extracting Packages\n",
      "\n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n"
     ]
    }
   ],
   "source": [
    "#tell the computer where the mambaforge bin files are located\n",
    "import os\n",
    "os.environ[\"PATH\"] += os.pathsep + os.environ[\"HOME\"]+\"/mambaforge/bin\"\n",
    "\n",
    "#now we can easily use 'mamba' command to install software \n",
    "!mamba install -y -c conda-forge -c bioconda trimmomatic fastqc multiqc salmon gsutil sql-magic entrez-direct gffread parallel-fastq-dump sra-tools sql-magic pyathena -y"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 2: Setup Environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a set of directories to store the reads, reference sequence files, and output files. Notice that first we remove the `data` directory to clean up files from Tutorial_1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/SageMaker/rnaseq-myco-notebook\n"
     ]
    }
   ],
   "source": [
    "! cd $HOMEDIR\n",
    "! echo $PWD\n",
    "! rm -r data/\n",
    "! mkdir -p data\n",
    "! mkdir -p data/raw_fastq\n",
    "! mkdir -p data/trimmed\n",
    "! mkdir -p data/fastqc\n",
    "! mkdir -p data/aligned\n",
    "! mkdir -p data/reference\n",
    "! mkdir -p data/fastqc_samples\n",
    "! mkdir -p data/multiqc_samples\n",
    "! mkdir -p data data/fasterqdump/raw_fastq data/prefetch_fasterqdump/raw_fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set # THREADS depending on your VM size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: THREADS=1\n"
     ]
    }
   ],
   "source": [
    "numthreads=!lscpu | grep '^CPU(s)'| awk '{print $2-1}'\n",
    "\n",
    "#python variable to hold the amount of threads your cpu has,\n",
    "#useful for downstream tools like salmon, trimmomatic, etc\n",
    "threads = int(numthreads[0])\n",
    "\n",
    "#its also good to have a shell version of the variable for commands that use piping, \n",
    "#in jupyter, shell commandds with piping sometimes causes python variables to not work and generally be wonky.\n",
    "%env THREADS=$threads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 3: Downloading relevant FASTQ files using SRA Tools\n",
    "\n",
    "Next we will need to download the relevant fastq files.\n",
    "\n",
    "Because these files can be large, the process of downloading and extracting fastq files can be quite lengthy.\n",
    "\n",
    "The sequence data for this tutorial comes from work by Cushman et al., <em><a href='https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8191103/'>Increased whiB7 expression and antibiotic resistance in Mycobacterium chelonae carrying two prophages</a><em>.\n",
    "\n",
    "We will be downloading the sample runs from this project using SRA tools, downloading from the NCBI's SRA (Sequence Run Archives).\n",
    "\n",
    "However, first we need to find the associated accession numbers in order to download.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.1: Finding run accession numbers.\n",
    "\n",
    "The SRA stores sequence data in terms of runs, (SRR stands for Sequence Read Run). To download runs, we will need the accession ID for each run we wish to download. \n",
    "\n",
    "The Cushman et al., project contains 12 runs. To make it easier, these are the run IDs associated with this project:\n",
    "\n",
    "+ SRR13349122\n",
    "+ SRR13349123\n",
    "+ SRR13349124\n",
    "+ SRR13349125\n",
    "+ SRR13349126\n",
    "+ SRR13349127\n",
    "+ SRR13349128\n",
    "+ SRR13349129\n",
    "+ SRR13349130\n",
    "+ SRR13349131\n",
    "+ SRR13349132\n",
    "+ SRR13349133\n",
    "\n",
    "\n",
    "In this case, all these runs belong to the SRP (Sequence Run Project): SRP300216.\n",
    "\n",
    "Sequence run experiments can be searched for using the SRA database on the NCBI website; and article-specific sample run information can be found in the supplementary section of that article.\n",
    "\n",
    "For instance, here, the the authors posted a link to the sequence data GSE (Gene Series number), <a href='https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE164210'>GSE164210</a>. This leads to the appropriate 'Gene Expression Omnibus' page where, among other useful files and information, the relevant SRA database link can be found. \n",
    "\n",
    "Once the accession numbers are located, one can make a text file containing the list of accession IDs however they like.\n",
    "\n",
    "Once again, to make things easier, we have made a .txt with these IDs that you can simply download here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://rnaseq-myco-bucket/reference/accs.txt...\n",
      "/ [1 files][  144.0 B/  144.0 B]                                                \n",
      "Operation completed over 1 objects/144.0 B.                                      \n",
      "SRR13349122\n",
      "SRR13349123\n",
      "SRR13349124\n",
      "SRR13349125\n",
      "SRR13349126\n",
      "SRR13349127\n",
      "SRR13349128\n",
      "SRR13349129\n",
      "SRR13349130\n",
      "SRR13349131\n",
      "SRR13349132\n",
      "SRR13349133\n"
     ]
    }
   ],
   "source": [
    "!gsutil cp gs://rnaseq-myco-bucket/reference/accs.txt .\n",
    "!cat accs.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "You can can also use BigQuery to generate an accession list following the instructions outlined in [this notebook](https://github.com/STRIDES/NIHCloudLabGCP/blob/main/tutorials/notebooks/SRADownload/SRA-Download.ipynb)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.1.2 (Optional): Generate the accession list file with BigQuery\n",
    "This step uses Python. We will use the BigQuery API. \n",
    "\n",
    "We will create a client the using default project. \n",
    "\n",
    "Then we will query BigQuery using the species name and a range of accession numbers associated with this particular study. \n",
    "\n",
    "Feel free to play around with the query to generate different variations of accession numbers!\n",
    "\n",
    "Please note that if you have errors to make sure you have this API enabled. You can search for BigQuery by navigating back to the Google Cloud Platform dashboard, back to the Google Cloud Platform, and using the search bar at the top, search for 'BigQuery'. On the BigQuery page, click  `Enable`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "aws sts get-caller-identity\n",
    "### STEP 3.2: Using the SRA-toolkit for a single sample.\n",
    "\n",
    "Sequence run accession IDs can be used to download sequence data, using the 'prefetch' tool of the SRA-toolkit.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pyathena import connect\n",
    "import pandas as pd\n",
    "\n",
    "# Use the correct argument name: s3_staging_dir\n",
    "conn = connect(s3_staging_dir='s3://sra-data-athena/', region_name='us-east-1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Crawler sra_crawler started.\n"
     ]
    }
   ],
   "source": [
    "import boto3\n",
    "\n",
    "# Initialize the Glue client\n",
    "glue_client = boto3.client('glue', region_name='us-east-1')\n",
    "\n",
    "# Run the crawler\n",
    "crawler_name = 'sra_crawler'  # Use your crawler's name\n",
    "glue_client.start_crawler(Name=crawler_name)\n",
    "\n",
    "print(f\"Crawler {crawler_name} started.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_7499/3650172437.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df = pd.read_sql(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "      <th>run_file_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13349130</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775150</td>\n",
       "      <td>GSM5004092</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004092}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13349129</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775149</td>\n",
       "      <td>GSM5004091</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004091}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13349125</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775147</td>\n",
       "      <td>GSM5004089</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004089}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13349124</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775147</td>\n",
       "      <td>GSM5004089</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004089}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13349127</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775148</td>\n",
       "      <td>GSM5004090</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004090}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR13349122</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775146</td>\n",
       "      <td>GSM5004088</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004088}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR13349133</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775151</td>\n",
       "      <td>GSM5004093</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004093}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR13349132</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775151</td>\n",
       "      <td>GSM5004093</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004093}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR13349126</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775148</td>\n",
       "      <td>GSM5004090</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004090}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR13349131</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775150</td>\n",
       "      <td>GSM5004092</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004092}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR13349123</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775146</td>\n",
       "      <td>GSM5004088</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004088}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR13349128</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775149</td>\n",
       "      <td>GSM5004091</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004091}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            acc assay_type center_name consent  experiment sample_name  \\\n",
       "0   SRR13349130    RNA-Seq         GEO  public  SRX9775150  GSM5004092   \n",
       "1   SRR13349129    RNA-Seq         GEO  public  SRX9775149  GSM5004091   \n",
       "2   SRR13349125    RNA-Seq         GEO  public  SRX9775147  GSM5004089   \n",
       "3   SRR13349124    RNA-Seq         GEO  public  SRX9775147  GSM5004089   \n",
       "4   SRR13349127    RNA-Seq         GEO  public  SRX9775148  GSM5004090   \n",
       "5   SRR13349122    RNA-Seq         GEO  public  SRX9775146  GSM5004088   \n",
       "6   SRR13349133    RNA-Seq         GEO  public  SRX9775151  GSM5004093   \n",
       "7   SRR13349132    RNA-Seq         GEO  public  SRX9775151  GSM5004093   \n",
       "8   SRR13349126    RNA-Seq         GEO  public  SRX9775148  GSM5004090   \n",
       "9   SRR13349131    RNA-Seq         GEO  public  SRX9775150  GSM5004092   \n",
       "10  SRR13349123    RNA-Seq         GEO  public  SRX9775146  GSM5004088   \n",
       "11  SRR13349128    RNA-Seq         GEO  public  SRX9775149  GSM5004091   \n",
       "\n",
       "             instrument librarylayout libraryselection   librarysource  ...  \\\n",
       "0   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "1   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "2   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "3   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "4   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "5   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "6   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "7   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "8   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "9   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "10  Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "11  Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "\n",
       "   geo_loc_name_sam ena_first_public_run ena_last_update_run sample_name_sam  \\\n",
       "0              None                 None                None            None   \n",
       "1              None                 None                None            None   \n",
       "2              None                 None                None            None   \n",
       "3              None                 None                None            None   \n",
       "4              None                 None                None            None   \n",
       "5              None                 None                None            None   \n",
       "6              None                 None                None            None   \n",
       "7              None                 None                None            None   \n",
       "8              None                 None                None            None   \n",
       "9              None                 None                None            None   \n",
       "10             None                 None                None            None   \n",
       "11             None                 None                None            None   \n",
       "\n",
       "      datastore_filetype datastore_provider  \\\n",
       "0   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "1   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "2   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "3   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "4   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "5   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "6   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "7   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "8   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "9   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "10  [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "11  [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "\n",
       "                            datastore_region  \\\n",
       "0   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "1   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "2   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "3   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "4   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "5   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "6   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "7   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "8   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "9   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "10  [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "11  [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "\n",
       "                                           attributes  \\\n",
       "0   [{k=geo_accession_exp, v=GSM5004092}, {k=bases...   \n",
       "1   [{k=geo_accession_exp, v=GSM5004091}, {k=bases...   \n",
       "2   [{k=geo_accession_exp, v=GSM5004089}, {k=bases...   \n",
       "3   [{k=geo_accession_exp, v=GSM5004089}, {k=bases...   \n",
       "4   [{k=geo_accession_exp, v=GSM5004090}, {k=bases...   \n",
       "5   [{k=geo_accession_exp, v=GSM5004088}, {k=bases...   \n",
       "6   [{k=geo_accession_exp, v=GSM5004093}, {k=bases...   \n",
       "7   [{k=geo_accession_exp, v=GSM5004093}, {k=bases...   \n",
       "8   [{k=geo_accession_exp, v=GSM5004090}, {k=bases...   \n",
       "9   [{k=geo_accession_exp, v=GSM5004092}, {k=bases...   \n",
       "10  [{k=geo_accession_exp, v=GSM5004088}, {k=bases...   \n",
       "11  [{k=geo_accession_exp, v=GSM5004091}, {k=bases...   \n",
       "\n",
       "                                                jattr  run_file_version  \n",
       "0   {\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...                 1  \n",
       "1   {\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...                 1  \n",
       "2   {\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...                 1  \n",
       "3   {\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...                 1  \n",
       "4   {\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...                 1  \n",
       "5   {\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...                 1  \n",
       "6   {\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...                 1  \n",
       "7   {\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...                 1  \n",
       "8   {\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...                 1  \n",
       "9   {\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...                 1  \n",
       "10  {\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...                 1  \n",
       "11  {\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...                 1  \n",
       "\n",
       "[12 rows x 37 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "query = \"\"\"\n",
    "SELECT *\n",
    "FROM AwsDataCatalog.srametadata.metadata\n",
    "WHERE organism = 'Mycobacteroides chelonae' \n",
    "AND acc LIKE '%SRR133491%'\n",
    "\"\"\"\n",
    "df = pd.read_sql(\n",
    "    query, conn\n",
    ")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR13349130\n",
      "SRR13349129\n",
      "SRR13349125\n",
      "SRR13349124\n",
      "SRR13349127\n",
      "SRR13349122\n",
      "SRR13349133\n",
      "SRR13349132\n",
      "SRR13349126\n",
      "SRR13349131\n",
      "SRR13349123\n",
      "SRR13349128"
     ]
    }
   ],
   "source": [
    "#write the SRR column to a text file\n",
    "with open('accs.txt', 'w') as f:\n",
    "    accs = df['acc'].to_string(header=False, index=False)\n",
    "    f.write(accs)\n",
    "    \n",
    "#print the text file\n",
    "!cat accs.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-19T04:57:18 prefetch.3.1.1: 1) Resolving 'SRR13349123'...\n",
      "2024-09-19T04:57:18 prefetch.3.1.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2024-09-19T04:57:18 prefetch.3.1.1: 1) Downloading 'SRR13349123'...\n",
      "2024-09-19T04:57:18 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-19T04:57:18 prefetch.3.1.1:  Downloading via HTTPS...\n"
     ]
    }
   ],
   "source": [
    "#the 'prefetch' command downloads an SRA file.\n",
    "!prefetch SRR13349123 -O data/raw_fastq -f yes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "Here we can see the command for downloading a single SRA file using an acecssion ID 'SRR13349123'\n",
    "\n",
    "Notice the SRA archives sequence files in the SRA format. \n",
    "Typically genome workflows process data in the form of zipped or unzipped .fastq, or .fasta files\n",
    "So before we move on, we need to convert the files from .sra to .fastq.\n",
    "\n",
    "There are multiple ways to do this. \n",
    "\n",
    "Included in the sra toolskit are fastq-dump and fasterq-dump. These convert SRA to FASTQ.\n",
    "\n",
    "If you use fasterq-dump, its recommended to zip your fastq files after they are created.\n",
    "\n",
    "There is also a tool called 'parallel-fastq-dump' which supports zipping the fastq files automatically into fastq.gz files.\n",
    "\n",
    "The below code may take approximately 15 minutes to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "spots read      : 11,165,256\n",
      "reads read      : 22,330,512\n",
      "reads written   : 22,330,512\n"
     ]
    }
   ],
   "source": [
    "#convert sra to fastq\n",
    "!fasterq-dump data/raw_fastq/SRR13349123 -f -O data/raw_fastq/\n",
    "#compress fastq to fastq.gz to save space\n",
    "!gzip data/raw_fastq/SRR13349123_1.fastq\n",
    "!gzip data/raw_fastq/SRR13349123_2.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 3.3 Downloading multiple files using the SRA-toolkit.\n",
    "\n",
    "Often one wants to, as in our case, wish to download multiple runs at once.\n",
    "\n",
    "To aid in this, SRA-tools supports batch downloading. This is why we created the text file earlier.\n",
    "\n",
    "We can download multiple SRA files using a single line of code by using our list SRA IDs, and inputting that into the prefetch command.\n",
    "\n",
    "And then feed that list into the sra-toolkit prefetch command. Note, it may take some time to download all the fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2024-09-18T05:22:37 prefetch.3.1.1: 1) Resolving 'SRR13349128'...\n",
      "2024-09-18T05:22:37 prefetch.3.1.1: Current preference is set to retrieve SRA Normalized Format files with full base quality scores\n",
      "2024-09-18T05:22:37 prefetch.3.1.1: 1) Downloading 'SRR13349128'...\n",
      "2024-09-18T05:22:37 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:22:37 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:23:07 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:23:09 prefetch.3.1.1:  'SRR13349128' is valid: 788638616 bytes were streamed from 788633419\n",
      "2024-09-18T05:23:09 prefetch.3.1.1: 1) 'SRR13349128' was downloaded successfully\n",
      "2024-09-18T05:23:09 prefetch.3.1.1: 'SRR13349128' has 0 dependencies\n",
      "2024-09-18T05:23:09 prefetch.3.1.1: 2) Resolving 'SRR13349122'...\n",
      "2024-09-18T05:23:09 prefetch.3.1.1: 2) Downloading 'SRR13349122'...\n",
      "2024-09-18T05:23:09 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:23:09 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:23:35 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:23:36 prefetch.3.1.1:  'SRR13349122' is valid: 675618606 bytes were streamed from 675611427\n",
      "2024-09-18T05:23:36 prefetch.3.1.1: 2) 'SRR13349122' was downloaded successfully\n",
      "2024-09-18T05:23:36 prefetch.3.1.1: 'SRR13349122' has 0 dependencies\n",
      "2024-09-18T05:23:36 prefetch.3.1.1: 3) Resolving 'SRR13349127'...\n",
      "2024-09-18T05:23:36 prefetch.3.1.1: 3) Downloading 'SRR13349127'...\n",
      "2024-09-18T05:23:36 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:23:36 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:24:04 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:24:05 prefetch.3.1.1:  'SRR13349127' is valid: 733838051 bytes were streamed from 733832418\n",
      "2024-09-18T05:24:05 prefetch.3.1.1: 3) 'SRR13349127' was downloaded successfully\n",
      "2024-09-18T05:24:06 prefetch.3.1.1: 'SRR13349127' has 0 dependencies\n",
      "2024-09-18T05:24:06 prefetch.3.1.1: 4) Resolving 'SRR13349123'...\n",
      "2024-09-18T05:24:06 prefetch.3.1.1: 4) Downloading 'SRR13349123'...\n",
      "2024-09-18T05:24:06 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:24:06 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:24:31 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:24:32 prefetch.3.1.1:  'SRR13349123' is valid: 641070152 bytes were streamed from 641060420\n",
      "2024-09-18T05:24:32 prefetch.3.1.1: 4) 'SRR13349123' was downloaded successfully\n",
      "2024-09-18T05:24:32 prefetch.3.1.1: 'SRR13349123' has 0 dependencies\n",
      "2024-09-18T05:24:32 prefetch.3.1.1: 5) Resolving 'SRR13349132'...\n",
      "2024-09-18T05:24:33 prefetch.3.1.1: 5) Downloading 'SRR13349132'...\n",
      "2024-09-18T05:24:33 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:24:33 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:25:02 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:25:03 prefetch.3.1.1:  'SRR13349132' is valid: 706295513 bytes were streamed from 706292440\n",
      "2024-09-18T05:25:03 prefetch.3.1.1: 5) 'SRR13349132' was downloaded successfully\n",
      "2024-09-18T05:25:03 prefetch.3.1.1: 'SRR13349132' has 0 dependencies\n",
      "2024-09-18T05:25:03 prefetch.3.1.1: 6) Resolving 'SRR13349129'...\n",
      "2024-09-18T05:25:03 prefetch.3.1.1: 6) Downloading 'SRR13349129'...\n",
      "2024-09-18T05:25:03 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:25:03 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:25:32 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:25:33 prefetch.3.1.1:  'SRR13349129' is valid: 743064450 bytes were streamed from 743057440\n",
      "2024-09-18T05:25:33 prefetch.3.1.1: 6) 'SRR13349129' was downloaded successfully\n",
      "2024-09-18T05:25:34 prefetch.3.1.1: 'SRR13349129' has 0 dependencies\n",
      "2024-09-18T05:25:34 prefetch.3.1.1: 7) Resolving 'SRR13349130'...\n",
      "2024-09-18T05:25:34 prefetch.3.1.1: 7) Downloading 'SRR13349130'...\n",
      "2024-09-18T05:25:34 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:25:34 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:25:59 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:26:00 prefetch.3.1.1:  'SRR13349130' is valid: 633040261 bytes were streamed from 633032407\n",
      "2024-09-18T05:26:00 prefetch.3.1.1: 7) 'SRR13349130' was downloaded successfully\n",
      "2024-09-18T05:26:00 prefetch.3.1.1: 'SRR13349130' has 0 dependencies\n",
      "2024-09-18T05:26:00 prefetch.3.1.1: 8) Resolving 'SRR13349131'...\n",
      "2024-09-18T05:26:01 prefetch.3.1.1: 8) Downloading 'SRR13349131'...\n",
      "2024-09-18T05:26:01 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:26:01 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:26:24 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:26:25 prefetch.3.1.1:  'SRR13349131' is valid: 610217382 bytes were streamed from 610208439\n",
      "2024-09-18T05:26:25 prefetch.3.1.1: 8) 'SRR13349131' was downloaded successfully\n",
      "2024-09-18T05:26:25 prefetch.3.1.1: 'SRR13349131' has 0 dependencies\n",
      "2024-09-18T05:26:25 prefetch.3.1.1: 9) Resolving 'SRR13349133'...\n",
      "2024-09-18T05:26:25 prefetch.3.1.1: 9) Downloading 'SRR13349133'...\n",
      "2024-09-18T05:26:25 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:26:25 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:26:50 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:26:51 prefetch.3.1.1:  'SRR13349133' is valid: 665312876 bytes were streamed from 665306440\n",
      "2024-09-18T05:26:51 prefetch.3.1.1: 9) 'SRR13349133' was downloaded successfully\n",
      "2024-09-18T05:26:51 prefetch.3.1.1: 'SRR13349133' has 0 dependencies\n",
      "2024-09-18T05:26:51 prefetch.3.1.1: 10) Resolving 'SRR13349125'...\n",
      "2024-09-18T05:26:51 prefetch.3.1.1: 10) Downloading 'SRR13349125'...\n",
      "2024-09-18T05:26:51 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:26:51 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:27:16 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:27:17 prefetch.3.1.1:  'SRR13349125' is valid: 632540004 bytes were streamed from 632537408\n",
      "2024-09-18T05:27:17 prefetch.3.1.1: 10) 'SRR13349125' was downloaded successfully\n",
      "2024-09-18T05:27:17 prefetch.3.1.1: 'SRR13349125' has 0 dependencies\n",
      "2024-09-18T05:27:17 prefetch.3.1.1: 11) Resolving 'SRR13349124'...\n",
      "2024-09-18T05:27:18 prefetch.3.1.1: 11) Downloading 'SRR13349124'...\n",
      "2024-09-18T05:27:18 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:27:18 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:27:44 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:27:45 prefetch.3.1.1:  'SRR13349124' is valid: 671244393 bytes were streamed from 671237440\n",
      "2024-09-18T05:27:45 prefetch.3.1.1: 11) 'SRR13349124' was downloaded successfully\n",
      "2024-09-18T05:27:46 prefetch.3.1.1: 'SRR13349124' has 0 dependencies\n",
      "2024-09-18T05:27:46 prefetch.3.1.1: 12) Resolving 'SRR13349126'...\n",
      "2024-09-18T05:27:46 prefetch.3.1.1: 12) Downloading 'SRR13349126'...\n",
      "2024-09-18T05:27:46 prefetch.3.1.1:  SRA Normalized Format file is being retrieved\n",
      "2024-09-18T05:27:46 prefetch.3.1.1:  Downloading via HTTPS...\n",
      "2024-09-18T05:28:13 prefetch.3.1.1:  HTTPS download succeed\n",
      "2024-09-18T05:28:15 prefetch.3.1.1:  'SRR13349126' is valid: 773480091 bytes were streamed from 773477408\n",
      "2024-09-18T05:28:15 prefetch.3.1.1: 12) 'SRR13349126' was downloaded successfully\n",
      "2024-09-18T05:28:15 prefetch.3.1.1: 'SRR13349126' has 0 dependencies\n"
     ]
    }
   ],
   "source": [
    "!prefetch --option-file accs.txt -O data/raw_fastq -f yes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.4 Converting Multiple SRA files to Fastq\n",
    "\n",
    "Fasterq-dump does not natively support batch converting of files.\n",
    "\n",
    "There are several ways we can get around this.\n",
    "\n",
    "For instance, one could use loops, or utilize piping.\n",
    "\n",
    "The below code uses a 'for' loop to iterate through all the accession IDs in our accs.txt file.\n",
    "\n",
    "It also adds includes various flags\n",
    "\n",
    "-e for cpu threads\n",
    "-m for maximum memory useage\n",
    "-f to force overwrite\n",
    "-O output directory. \n",
    "\n",
    "This process should take about 35 minutes or so."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n",
      "fasterq-dump quit with error code 3\n"
     ]
    }
   ],
   "source": [
    "!for x in `cat accs.txt`; do fasterq-dump -f -O data/raw_fastq -e $THREADS -m 4G data/raw_fastq/$x/$x.sra; done\n",
    "\n",
    "##example of how to alternatively do the above process with parallel-fastq-dump using piping\n",
    "#!cat accs.txt | xargs -I {} parallel-fastq-dump -O data/raw_fastq/ --tmpdir . --threads $THREADS --gzip --split-files --sra-id {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As before, it is good practice to turn .fastq files into .fastq.gz files to save space.\n",
    "\n",
    "In our case, we will actually need to concatenate the fastq files later on, and so will zip after this.\n",
    "\n",
    "The no redundant SRA files can also be deleted to save more space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#find and delete all SRR subfolders in the raw_fastq directory\n",
    "!find data/raw_fastq -type d -name 'SRR*' -exec rm -rf {} \\;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 4: Copy reference transcriptome files that will be used by Salmon using E-Direct\n",
    "\n",
    "Salmon is a tool that aligns RNA-Seq reads to a transcriptome.\n",
    "\n",
    "So we will need a transcriptome reference file.\n",
    "\n",
    "To get one, we can search through the NCBI assembly database, find an assembly, and download transcriptome reference files from that assembly using FTP links.\n",
    "\n",
    "For instance, we will use the <a href='https://www.ncbi.nlm.nih.gov/assembly/GCF_001632805.1'>ASM163280v1</a> refseq assembly, found by searching through the NCBI assembly database. The FTP links can be accessed through the website in various ways, one way is to click the 'FTP directory for RefSeq assembly' link, found under 'Access the data', section.\n",
    "\n",
    "Alternatively, if one were inclined, one could take the less common route and perform this through the NCBI command line tool suite called 'Entrez Direct' (EDirect).\n",
    "\n",
    "This is an intricate and complicated set of tools, with many ways to do any one thing.\n",
    "\n",
    "Below is an example of using an eDirect search query with a refseq identifier to obtain the relevant FTP directory, and then using that to download desired reference files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 1436k  100 1436k    0     0  1399k      0  0:00:01  0:00:01 --:--:-- 1400k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  397k  100  397k    0     0   909k      0 --:--:-- --:--:-- --:--:--  910k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  225k  100  225k    0     0   329k      0 --:--:-- --:--:-- --:--:--  329k\n"
     ]
    }
   ],
   "source": [
    "#parse for the ftp link and download the genome reference fasta file\n",
    "\n",
    "!esearch -db assembly -query GCF_001632805.1 | efetch -format docsum \\\n",
    "| xtract -pattern DocumentSummary -element FtpPath_RefSeq \\\n",
    "| awk -F\"/\" '{print \"curl -o data/reference/\"$NF\"_genomic.fna.gz \" $0\"/\"$NF\"_genomic.fna.gz\"}' \\\n",
    "| bash\n",
    "\n",
    "#parse for the ftp link and download the gtf reference fasta file\n",
    "\n",
    "!esearch -db assembly -query GCF_001632805.1 | efetch -format docsum \\\n",
    "| xtract -pattern DocumentSummary -element FtpPath_RefSeq \\\n",
    "| awk -F\"/\" '{print \"curl -o data/reference/\"$NF\"_genomic.gff.gz \" $0\"/\"$NF\"_genomic.gff.gz\"}' \\\n",
    "| bash\n",
    "\n",
    "# parse for the ftp link and download the feature-table reference file \n",
    "# (for later use for merging readcounts with gene names in R code).\n",
    "\n",
    "!esearch -db assembly -query GCF_001632805.1 | efetch -format docsum \\\n",
    "| xtract -pattern DocumentSummary -element FtpPath_RefSeq \\\n",
    "| awk -F\"/\" '{print \"curl -o data/reference/\"$NF\"_feature_table.txt.gz \" $0\"/\"$NF\"_feature_table.txt.gz\"}' \\\n",
    "| bash\n",
    "\n",
    "\n",
    "#unzip the compresseed fasta files\n",
    "\n",
    "!gzip -d data/reference/*.gz --force"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we can use a tool called gffread to create a transcriptome reference file using the gtf and genome files we downloaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!gffread -w data/reference/GCF_001632805.1_transcriptome_reference.fa -g data/reference/GCF_001632805.1_ASM163280v1_genomic.fna data/reference/GCF_001632805.1_ASM163280v1_genomic.gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is also recommended to include the full genome at the end of the transcriptome reference file, for the purpose of performing a 'decoy-aware' mapping, more information about which can be found in the Salmon documentation.\n",
    "\n",
    "To alert the tool to the presence of this, we will also create a 'decoy file', which salmon needs pointed towards the full genome sequence in our transcriptome reference file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!cat data/reference/GCF_001632805.1_transcriptome_reference.fa > data/reference/GCF_001632805.1_transcriptome_reference_w_decoy.fa\n",
    "!echo >> data/reference/GCF_001632805.1_transcriptome_reference_w_decoy.fa\n",
    "!cat data/reference/GCF_001632805.1_ASM163280v1_genomic.fna >> data/reference/GCF_001632805.1_transcriptome_reference_w_decoy.fa\n",
    "!echo \"NZ_CP007220.1\" > data/reference/decoys.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 5: Run FastQC\n",
    "FastQC is an invaluable tool that allows you to evaluate whether there are problems with a set of reads. For example, it will provide a report of whether there is any bias in the sequence composition of the reads.\n",
    "\n",
    "The below code may take around 25 minutes to run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "Started analysis of SRR13349128_1.fastq\n",
      "Approx 5% complete for SRR13349128_1.fastq\n",
      "Approx 10% complete for SRR13349128_1.fastq\n",
      "Approx 15% complete for SRR13349128_1.fastq\n",
      "Approx 20% complete for SRR13349128_1.fastq\n",
      "Approx 25% complete for SRR13349128_1.fastq\n",
      "Approx 30% complete for SRR13349128_1.fastq\n",
      "Approx 35% complete for SRR13349128_1.fastq\n",
      "Approx 40% complete for SRR13349128_1.fastq\n",
      "Approx 45% complete for SRR13349128_1.fastq\n",
      "Approx 50% complete for SRR13349128_1.fastq\n",
      "Approx 55% complete for SRR13349128_1.fastq\n",
      "Approx 60% complete for SRR13349128_1.fastq\n",
      "Approx 65% complete for SRR13349128_1.fastq\n",
      "Approx 70% complete for SRR13349128_1.fastq\n",
      "Approx 75% complete for SRR13349128_1.fastq\n",
      "Approx 80% complete for SRR13349128_1.fastq\n",
      "Approx 85% complete for SRR13349128_1.fastq\n",
      "Approx 90% complete for SRR13349128_1.fastq\n",
      "Approx 95% complete for SRR13349128_1.fastq\n",
      "Analysis complete for SRR13349128_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349122_1.fastq\n",
      "Approx 5% complete for SRR13349122_1.fastq\n",
      "Approx 10% complete for SRR13349122_1.fastq\n",
      "Approx 15% complete for SRR13349122_1.fastq\n",
      "Approx 20% complete for SRR13349122_1.fastq\n",
      "Approx 25% complete for SRR13349122_1.fastq\n",
      "Approx 30% complete for SRR13349122_1.fastq\n",
      "Approx 35% complete for SRR13349122_1.fastq\n",
      "Approx 40% complete for SRR13349122_1.fastq\n",
      "Approx 45% complete for SRR13349122_1.fastq\n",
      "Approx 50% complete for SRR13349122_1.fastq\n",
      "Approx 55% complete for SRR13349122_1.fastq\n",
      "Approx 60% complete for SRR13349122_1.fastq\n",
      "Approx 65% complete for SRR13349122_1.fastq\n",
      "Approx 70% complete for SRR13349122_1.fastq\n",
      "Approx 75% complete for SRR13349122_1.fastq\n",
      "Approx 80% complete for SRR13349122_1.fastq\n",
      "Approx 85% complete for SRR13349122_1.fastq\n",
      "Approx 90% complete for SRR13349122_1.fastq\n",
      "Approx 95% complete for SRR13349122_1.fastq\n",
      "Analysis complete for SRR13349122_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349127_1.fastq\n",
      "Approx 5% complete for SRR13349127_1.fastq\n",
      "Approx 10% complete for SRR13349127_1.fastq\n",
      "Approx 15% complete for SRR13349127_1.fastq\n",
      "Approx 20% complete for SRR13349127_1.fastq\n",
      "Approx 25% complete for SRR13349127_1.fastq\n",
      "Approx 30% complete for SRR13349127_1.fastq\n",
      "Approx 35% complete for SRR13349127_1.fastq\n",
      "Approx 40% complete for SRR13349127_1.fastq\n",
      "Approx 45% complete for SRR13349127_1.fastq\n",
      "Approx 50% complete for SRR13349127_1.fastq\n",
      "Approx 55% complete for SRR13349127_1.fastq\n",
      "Approx 60% complete for SRR13349127_1.fastq\n",
      "Approx 65% complete for SRR13349127_1.fastq\n",
      "Approx 70% complete for SRR13349127_1.fastq\n",
      "Approx 75% complete for SRR13349127_1.fastq\n",
      "Approx 80% complete for SRR13349127_1.fastq\n",
      "Approx 85% complete for SRR13349127_1.fastq\n",
      "Approx 90% complete for SRR13349127_1.fastq\n",
      "Approx 95% complete for SRR13349127_1.fastq\n",
      "Analysis complete for SRR13349127_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349123_1.fastq\n",
      "Approx 5% complete for SRR13349123_1.fastq\n",
      "Approx 10% complete for SRR13349123_1.fastq\n",
      "Approx 15% complete for SRR13349123_1.fastq\n",
      "Approx 20% complete for SRR13349123_1.fastq\n",
      "Approx 25% complete for SRR13349123_1.fastq\n",
      "Approx 30% complete for SRR13349123_1.fastq\n",
      "Approx 35% complete for SRR13349123_1.fastq\n",
      "Approx 40% complete for SRR13349123_1.fastq\n",
      "Approx 45% complete for SRR13349123_1.fastq\n",
      "Approx 50% complete for SRR13349123_1.fastq\n",
      "Approx 55% complete for SRR13349123_1.fastq\n",
      "Approx 60% complete for SRR13349123_1.fastq\n",
      "Approx 65% complete for SRR13349123_1.fastq\n",
      "Approx 70% complete for SRR13349123_1.fastq\n",
      "Approx 75% complete for SRR13349123_1.fastq\n",
      "Approx 80% complete for SRR13349123_1.fastq\n",
      "Approx 85% complete for SRR13349123_1.fastq\n",
      "Approx 90% complete for SRR13349123_1.fastq\n",
      "Approx 95% complete for SRR13349123_1.fastq\n",
      "Analysis complete for SRR13349123_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349132_1.fastq\n",
      "Approx 5% complete for SRR13349132_1.fastq\n",
      "Approx 10% complete for SRR13349132_1.fastq\n",
      "Approx 15% complete for SRR13349132_1.fastq\n",
      "Approx 20% complete for SRR13349132_1.fastq\n",
      "Approx 25% complete for SRR13349132_1.fastq\n",
      "Approx 30% complete for SRR13349132_1.fastq\n",
      "Approx 35% complete for SRR13349132_1.fastq\n",
      "Approx 40% complete for SRR13349132_1.fastq\n",
      "Approx 45% complete for SRR13349132_1.fastq\n",
      "Approx 50% complete for SRR13349132_1.fastq\n",
      "Approx 55% complete for SRR13349132_1.fastq\n",
      "Approx 60% complete for SRR13349132_1.fastq\n",
      "Approx 65% complete for SRR13349132_1.fastq\n",
      "Approx 70% complete for SRR13349132_1.fastq\n",
      "Approx 75% complete for SRR13349132_1.fastq\n",
      "Approx 80% complete for SRR13349132_1.fastq\n",
      "Approx 85% complete for SRR13349132_1.fastq\n",
      "Approx 90% complete for SRR13349132_1.fastq\n",
      "Approx 95% complete for SRR13349132_1.fastq\n",
      "Analysis complete for SRR13349132_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349129_1.fastq\n",
      "Approx 5% complete for SRR13349129_1.fastq\n",
      "Approx 10% complete for SRR13349129_1.fastq\n",
      "Approx 15% complete for SRR13349129_1.fastq\n",
      "Approx 20% complete for SRR13349129_1.fastq\n",
      "Approx 25% complete for SRR13349129_1.fastq\n",
      "Approx 30% complete for SRR13349129_1.fastq\n",
      "Approx 35% complete for SRR13349129_1.fastq\n",
      "Approx 40% complete for SRR13349129_1.fastq\n",
      "Approx 45% complete for SRR13349129_1.fastq\n",
      "Approx 50% complete for SRR13349129_1.fastq\n",
      "Approx 55% complete for SRR13349129_1.fastq\n",
      "Approx 60% complete for SRR13349129_1.fastq\n",
      "Approx 65% complete for SRR13349129_1.fastq\n",
      "Approx 70% complete for SRR13349129_1.fastq\n",
      "Approx 75% complete for SRR13349129_1.fastq\n",
      "Approx 80% complete for SRR13349129_1.fastq\n",
      "Approx 85% complete for SRR13349129_1.fastq\n",
      "Approx 90% complete for SRR13349129_1.fastq\n",
      "Approx 95% complete for SRR13349129_1.fastq\n",
      "Analysis complete for SRR13349129_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349130_1.fastq\n",
      "Approx 5% complete for SRR13349130_1.fastq\n",
      "Approx 10% complete for SRR13349130_1.fastq\n",
      "Approx 15% complete for SRR13349130_1.fastq\n",
      "Approx 20% complete for SRR13349130_1.fastq\n",
      "Approx 25% complete for SRR13349130_1.fastq\n",
      "Approx 30% complete for SRR13349130_1.fastq\n",
      "Approx 35% complete for SRR13349130_1.fastq\n",
      "Approx 40% complete for SRR13349130_1.fastq\n",
      "Approx 45% complete for SRR13349130_1.fastq\n",
      "Approx 50% complete for SRR13349130_1.fastq\n",
      "Approx 55% complete for SRR13349130_1.fastq\n",
      "Approx 60% complete for SRR13349130_1.fastq\n",
      "Approx 65% complete for SRR13349130_1.fastq\n",
      "Approx 70% complete for SRR13349130_1.fastq\n",
      "Approx 75% complete for SRR13349130_1.fastq\n",
      "Approx 80% complete for SRR13349130_1.fastq\n",
      "Approx 85% complete for SRR13349130_1.fastq\n",
      "Approx 90% complete for SRR13349130_1.fastq\n",
      "Approx 95% complete for SRR13349130_1.fastq\n",
      "Analysis complete for SRR13349130_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349131_1.fastq\n",
      "Approx 5% complete for SRR13349131_1.fastq\n",
      "Approx 10% complete for SRR13349131_1.fastq\n",
      "Approx 15% complete for SRR13349131_1.fastq\n",
      "Approx 20% complete for SRR13349131_1.fastq\n",
      "Approx 25% complete for SRR13349131_1.fastq\n",
      "Approx 30% complete for SRR13349131_1.fastq\n",
      "Approx 35% complete for SRR13349131_1.fastq\n",
      "Approx 40% complete for SRR13349131_1.fastq\n",
      "Approx 45% complete for SRR13349131_1.fastq\n",
      "Approx 50% complete for SRR13349131_1.fastq\n",
      "Approx 55% complete for SRR13349131_1.fastq\n",
      "Approx 60% complete for SRR13349131_1.fastq\n",
      "Approx 65% complete for SRR13349131_1.fastq\n",
      "Approx 70% complete for SRR13349131_1.fastq\n",
      "Approx 75% complete for SRR13349131_1.fastq\n",
      "Approx 80% complete for SRR13349131_1.fastq\n",
      "Approx 85% complete for SRR13349131_1.fastq\n",
      "Approx 90% complete for SRR13349131_1.fastq\n",
      "Approx 95% complete for SRR13349131_1.fastq\n",
      "Analysis complete for SRR13349131_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349133_1.fastq\n",
      "Approx 5% complete for SRR13349133_1.fastq\n",
      "Approx 10% complete for SRR13349133_1.fastq\n",
      "Approx 15% complete for SRR13349133_1.fastq\n",
      "Approx 20% complete for SRR13349133_1.fastq\n",
      "Approx 25% complete for SRR13349133_1.fastq\n",
      "Approx 30% complete for SRR13349133_1.fastq\n",
      "Approx 35% complete for SRR13349133_1.fastq\n",
      "Approx 40% complete for SRR13349133_1.fastq\n",
      "Approx 45% complete for SRR13349133_1.fastq\n",
      "Approx 50% complete for SRR13349133_1.fastq\n",
      "Approx 55% complete for SRR13349133_1.fastq\n",
      "Approx 60% complete for SRR13349133_1.fastq\n",
      "Approx 65% complete for SRR13349133_1.fastq\n",
      "Approx 70% complete for SRR13349133_1.fastq\n",
      "Approx 75% complete for SRR13349133_1.fastq\n",
      "Approx 80% complete for SRR13349133_1.fastq\n",
      "Approx 85% complete for SRR13349133_1.fastq\n",
      "Approx 90% complete for SRR13349133_1.fastq\n",
      "Approx 95% complete for SRR13349133_1.fastq\n",
      "Analysis complete for SRR13349133_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349125_1.fastq\n",
      "Approx 5% complete for SRR13349125_1.fastq\n",
      "Approx 10% complete for SRR13349125_1.fastq\n",
      "Approx 15% complete for SRR13349125_1.fastq\n",
      "Approx 20% complete for SRR13349125_1.fastq\n",
      "Approx 25% complete for SRR13349125_1.fastq\n",
      "Approx 30% complete for SRR13349125_1.fastq\n",
      "Approx 35% complete for SRR13349125_1.fastq\n",
      "Approx 40% complete for SRR13349125_1.fastq\n",
      "Approx 45% complete for SRR13349125_1.fastq\n",
      "Approx 50% complete for SRR13349125_1.fastq\n",
      "Approx 55% complete for SRR13349125_1.fastq\n",
      "Approx 60% complete for SRR13349125_1.fastq\n",
      "Approx 65% complete for SRR13349125_1.fastq\n",
      "Approx 70% complete for SRR13349125_1.fastq\n",
      "Approx 75% complete for SRR13349125_1.fastq\n",
      "Approx 80% complete for SRR13349125_1.fastq\n",
      "Approx 85% complete for SRR13349125_1.fastq\n",
      "Approx 90% complete for SRR13349125_1.fastq\n",
      "Approx 95% complete for SRR13349125_1.fastq\n",
      "Analysis complete for SRR13349125_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349124_1.fastq\n",
      "Approx 5% complete for SRR13349124_1.fastq\n",
      "Approx 10% complete for SRR13349124_1.fastq\n",
      "Approx 15% complete for SRR13349124_1.fastq\n",
      "Approx 20% complete for SRR13349124_1.fastq\n",
      "Approx 25% complete for SRR13349124_1.fastq\n",
      "Approx 30% complete for SRR13349124_1.fastq\n",
      "Approx 35% complete for SRR13349124_1.fastq\n",
      "Approx 40% complete for SRR13349124_1.fastq\n",
      "Approx 45% complete for SRR13349124_1.fastq\n",
      "Approx 50% complete for SRR13349124_1.fastq\n",
      "Approx 55% complete for SRR13349124_1.fastq\n",
      "Approx 60% complete for SRR13349124_1.fastq\n",
      "Approx 65% complete for SRR13349124_1.fastq\n",
      "Approx 70% complete for SRR13349124_1.fastq\n",
      "Approx 75% complete for SRR13349124_1.fastq\n",
      "Approx 80% complete for SRR13349124_1.fastq\n",
      "Approx 85% complete for SRR13349124_1.fastq\n",
      "Approx 90% complete for SRR13349124_1.fastq\n",
      "Approx 95% complete for SRR13349124_1.fastq\n",
      "Analysis complete for SRR13349124_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349126_1.fastq\n",
      "Approx 5% complete for SRR13349126_1.fastq\n",
      "Approx 10% complete for SRR13349126_1.fastq\n",
      "Approx 15% complete for SRR13349126_1.fastq\n",
      "Approx 20% complete for SRR13349126_1.fastq\n",
      "Approx 25% complete for SRR13349126_1.fastq\n",
      "Approx 30% complete for SRR13349126_1.fastq\n",
      "Approx 35% complete for SRR13349126_1.fastq\n",
      "Approx 40% complete for SRR13349126_1.fastq\n",
      "Approx 45% complete for SRR13349126_1.fastq\n",
      "Approx 50% complete for SRR13349126_1.fastq\n",
      "Approx 55% complete for SRR13349126_1.fastq\n",
      "Approx 60% complete for SRR13349126_1.fastq\n",
      "Approx 65% complete for SRR13349126_1.fastq\n",
      "Approx 70% complete for SRR13349126_1.fastq\n",
      "Approx 75% complete for SRR13349126_1.fastq\n",
      "Approx 80% complete for SRR13349126_1.fastq\n",
      "Approx 85% complete for SRR13349126_1.fastq\n",
      "Approx 90% complete for SRR13349126_1.fastq\n",
      "Approx 95% complete for SRR13349126_1.fastq\n",
      "Analysis complete for SRR13349126_1.fastq\n",
      "null\n",
      "Started analysis of SRR13349128_2.fastq\n",
      "Approx 5% complete for SRR13349128_2.fastq\n",
      "Approx 10% complete for SRR13349128_2.fastq\n",
      "Approx 15% complete for SRR13349128_2.fastq\n",
      "Approx 20% complete for SRR13349128_2.fastq\n",
      "Approx 25% complete for SRR13349128_2.fastq\n",
      "Approx 30% complete for SRR13349128_2.fastq\n",
      "Approx 35% complete for SRR13349128_2.fastq\n",
      "Approx 40% complete for SRR13349128_2.fastq\n",
      "Approx 45% complete for SRR13349128_2.fastq\n",
      "Approx 50% complete for SRR13349128_2.fastq\n",
      "Approx 55% complete for SRR13349128_2.fastq\n",
      "Approx 60% complete for SRR13349128_2.fastq\n",
      "Approx 65% complete for SRR13349128_2.fastq\n",
      "Approx 70% complete for SRR13349128_2.fastq\n",
      "Approx 75% complete for SRR13349128_2.fastq\n",
      "Approx 80% complete for SRR13349128_2.fastq\n",
      "Approx 85% complete for SRR13349128_2.fastq\n",
      "Approx 90% complete for SRR13349128_2.fastq\n",
      "Approx 95% complete for SRR13349128_2.fastq\n",
      "Analysis complete for SRR13349128_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349122_2.fastq\n",
      "Approx 5% complete for SRR13349122_2.fastq\n",
      "Approx 10% complete for SRR13349122_2.fastq\n",
      "Approx 15% complete for SRR13349122_2.fastq\n",
      "Approx 20% complete for SRR13349122_2.fastq\n",
      "Approx 25% complete for SRR13349122_2.fastq\n",
      "Approx 30% complete for SRR13349122_2.fastq\n",
      "Approx 35% complete for SRR13349122_2.fastq\n",
      "Approx 40% complete for SRR13349122_2.fastq\n",
      "Approx 45% complete for SRR13349122_2.fastq\n",
      "Approx 50% complete for SRR13349122_2.fastq\n",
      "Approx 55% complete for SRR13349122_2.fastq\n",
      "Approx 60% complete for SRR13349122_2.fastq\n",
      "Approx 65% complete for SRR13349122_2.fastq\n",
      "Approx 70% complete for SRR13349122_2.fastq\n",
      "Approx 75% complete for SRR13349122_2.fastq\n",
      "Approx 80% complete for SRR13349122_2.fastq\n",
      "Approx 85% complete for SRR13349122_2.fastq\n",
      "Approx 90% complete for SRR13349122_2.fastq\n",
      "Approx 95% complete for SRR13349122_2.fastq\n",
      "Analysis complete for SRR13349122_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349127_2.fastq\n",
      "Approx 5% complete for SRR13349127_2.fastq\n",
      "Approx 10% complete for SRR13349127_2.fastq\n",
      "Approx 15% complete for SRR13349127_2.fastq\n",
      "Approx 20% complete for SRR13349127_2.fastq\n",
      "Approx 25% complete for SRR13349127_2.fastq\n",
      "Approx 30% complete for SRR13349127_2.fastq\n",
      "Approx 35% complete for SRR13349127_2.fastq\n",
      "Approx 40% complete for SRR13349127_2.fastq\n",
      "Approx 45% complete for SRR13349127_2.fastq\n",
      "Approx 50% complete for SRR13349127_2.fastq\n",
      "Approx 55% complete for SRR13349127_2.fastq\n",
      "Approx 60% complete for SRR13349127_2.fastq\n",
      "Approx 65% complete for SRR13349127_2.fastq\n",
      "Approx 70% complete for SRR13349127_2.fastq\n",
      "Approx 75% complete for SRR13349127_2.fastq\n",
      "Approx 80% complete for SRR13349127_2.fastq\n",
      "Approx 85% complete for SRR13349127_2.fastq\n",
      "Approx 90% complete for SRR13349127_2.fastq\n",
      "Approx 95% complete for SRR13349127_2.fastq\n",
      "Analysis complete for SRR13349127_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349123_2.fastq\n",
      "Approx 5% complete for SRR13349123_2.fastq\n",
      "Approx 10% complete for SRR13349123_2.fastq\n",
      "Approx 15% complete for SRR13349123_2.fastq\n",
      "Approx 20% complete for SRR13349123_2.fastq\n",
      "Approx 25% complete for SRR13349123_2.fastq\n",
      "Approx 30% complete for SRR13349123_2.fastq\n",
      "Approx 35% complete for SRR13349123_2.fastq\n",
      "Approx 40% complete for SRR13349123_2.fastq\n",
      "Approx 45% complete for SRR13349123_2.fastq\n",
      "Approx 50% complete for SRR13349123_2.fastq\n",
      "Approx 55% complete for SRR13349123_2.fastq\n",
      "Approx 60% complete for SRR13349123_2.fastq\n",
      "Approx 65% complete for SRR13349123_2.fastq\n",
      "Approx 70% complete for SRR13349123_2.fastq\n",
      "Approx 75% complete for SRR13349123_2.fastq\n",
      "Approx 80% complete for SRR13349123_2.fastq\n",
      "Approx 85% complete for SRR13349123_2.fastq\n",
      "Approx 90% complete for SRR13349123_2.fastq\n",
      "Approx 95% complete for SRR13349123_2.fastq\n",
      "Analysis complete for SRR13349123_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349132_2.fastq\n",
      "Approx 5% complete for SRR13349132_2.fastq\n",
      "Approx 10% complete for SRR13349132_2.fastq\n",
      "Approx 15% complete for SRR13349132_2.fastq\n",
      "Approx 20% complete for SRR13349132_2.fastq\n",
      "Approx 25% complete for SRR13349132_2.fastq\n",
      "Approx 30% complete for SRR13349132_2.fastq\n",
      "Approx 35% complete for SRR13349132_2.fastq\n",
      "Approx 40% complete for SRR13349132_2.fastq\n",
      "Approx 45% complete for SRR13349132_2.fastq\n",
      "Approx 50% complete for SRR13349132_2.fastq\n",
      "Approx 55% complete for SRR13349132_2.fastq\n",
      "Approx 60% complete for SRR13349132_2.fastq\n",
      "Approx 65% complete for SRR13349132_2.fastq\n",
      "Approx 70% complete for SRR13349132_2.fastq\n",
      "Approx 75% complete for SRR13349132_2.fastq\n",
      "Approx 80% complete for SRR13349132_2.fastq\n",
      "Approx 85% complete for SRR13349132_2.fastq\n",
      "Approx 90% complete for SRR13349132_2.fastq\n",
      "Approx 95% complete for SRR13349132_2.fastq\n",
      "Analysis complete for SRR13349132_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349129_2.fastq\n",
      "Approx 5% complete for SRR13349129_2.fastq\n",
      "Approx 10% complete for SRR13349129_2.fastq\n",
      "Approx 15% complete for SRR13349129_2.fastq\n",
      "Approx 20% complete for SRR13349129_2.fastq\n",
      "Approx 25% complete for SRR13349129_2.fastq\n",
      "Approx 30% complete for SRR13349129_2.fastq\n",
      "Approx 35% complete for SRR13349129_2.fastq\n",
      "Approx 40% complete for SRR13349129_2.fastq\n",
      "Approx 45% complete for SRR13349129_2.fastq\n",
      "Approx 50% complete for SRR13349129_2.fastq\n",
      "Approx 55% complete for SRR13349129_2.fastq\n",
      "Approx 60% complete for SRR13349129_2.fastq\n",
      "Approx 65% complete for SRR13349129_2.fastq\n",
      "Approx 70% complete for SRR13349129_2.fastq\n",
      "Approx 75% complete for SRR13349129_2.fastq\n",
      "Approx 80% complete for SRR13349129_2.fastq\n",
      "Approx 85% complete for SRR13349129_2.fastq\n",
      "Approx 90% complete for SRR13349129_2.fastq\n",
      "Approx 95% complete for SRR13349129_2.fastq\n",
      "Analysis complete for SRR13349129_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349130_2.fastq\n",
      "Approx 5% complete for SRR13349130_2.fastq\n",
      "Approx 10% complete for SRR13349130_2.fastq\n",
      "Approx 15% complete for SRR13349130_2.fastq\n",
      "Approx 20% complete for SRR13349130_2.fastq\n",
      "Approx 25% complete for SRR13349130_2.fastq\n",
      "Approx 30% complete for SRR13349130_2.fastq\n",
      "Approx 35% complete for SRR13349130_2.fastq\n",
      "Approx 40% complete for SRR13349130_2.fastq\n",
      "Approx 45% complete for SRR13349130_2.fastq\n",
      "Approx 50% complete for SRR13349130_2.fastq\n",
      "Approx 55% complete for SRR13349130_2.fastq\n",
      "Approx 60% complete for SRR13349130_2.fastq\n",
      "Approx 65% complete for SRR13349130_2.fastq\n",
      "Approx 70% complete for SRR13349130_2.fastq\n",
      "Approx 75% complete for SRR13349130_2.fastq\n",
      "Approx 80% complete for SRR13349130_2.fastq\n",
      "Approx 85% complete for SRR13349130_2.fastq\n",
      "Approx 90% complete for SRR13349130_2.fastq\n",
      "Approx 95% complete for SRR13349130_2.fastq\n",
      "Analysis complete for SRR13349130_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349131_2.fastq\n",
      "Approx 5% complete for SRR13349131_2.fastq\n",
      "Approx 10% complete for SRR13349131_2.fastq\n",
      "Approx 15% complete for SRR13349131_2.fastq\n",
      "Approx 20% complete for SRR13349131_2.fastq\n",
      "Approx 25% complete for SRR13349131_2.fastq\n",
      "Approx 30% complete for SRR13349131_2.fastq\n",
      "Approx 35% complete for SRR13349131_2.fastq\n",
      "Approx 40% complete for SRR13349131_2.fastq\n",
      "Approx 45% complete for SRR13349131_2.fastq\n",
      "Approx 50% complete for SRR13349131_2.fastq\n",
      "Approx 55% complete for SRR13349131_2.fastq\n",
      "Approx 60% complete for SRR13349131_2.fastq\n",
      "Approx 65% complete for SRR13349131_2.fastq\n",
      "Approx 70% complete for SRR13349131_2.fastq\n",
      "Approx 75% complete for SRR13349131_2.fastq\n",
      "Approx 80% complete for SRR13349131_2.fastq\n",
      "Approx 85% complete for SRR13349131_2.fastq\n",
      "Approx 90% complete for SRR13349131_2.fastq\n",
      "Approx 95% complete for SRR13349131_2.fastq\n",
      "Analysis complete for SRR13349131_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349133_2.fastq\n",
      "Approx 5% complete for SRR13349133_2.fastq\n",
      "Approx 10% complete for SRR13349133_2.fastq\n",
      "Approx 15% complete for SRR13349133_2.fastq\n",
      "Approx 20% complete for SRR13349133_2.fastq\n",
      "Approx 25% complete for SRR13349133_2.fastq\n",
      "Approx 30% complete for SRR13349133_2.fastq\n",
      "Approx 35% complete for SRR13349133_2.fastq\n",
      "Approx 40% complete for SRR13349133_2.fastq\n",
      "Approx 45% complete for SRR13349133_2.fastq\n",
      "Approx 50% complete for SRR13349133_2.fastq\n",
      "Approx 55% complete for SRR13349133_2.fastq\n",
      "Approx 60% complete for SRR13349133_2.fastq\n",
      "Approx 65% complete for SRR13349133_2.fastq\n",
      "Approx 70% complete for SRR13349133_2.fastq\n",
      "Approx 75% complete for SRR13349133_2.fastq\n",
      "Approx 80% complete for SRR13349133_2.fastq\n",
      "Approx 85% complete for SRR13349133_2.fastq\n",
      "Approx 90% complete for SRR13349133_2.fastq\n",
      "Approx 95% complete for SRR13349133_2.fastq\n",
      "Analysis complete for SRR13349133_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349125_2.fastq\n",
      "Approx 5% complete for SRR13349125_2.fastq\n",
      "Approx 10% complete for SRR13349125_2.fastq\n",
      "Approx 15% complete for SRR13349125_2.fastq\n",
      "Approx 20% complete for SRR13349125_2.fastq\n",
      "Approx 25% complete for SRR13349125_2.fastq\n",
      "Approx 30% complete for SRR13349125_2.fastq\n",
      "Approx 35% complete for SRR13349125_2.fastq\n",
      "Approx 40% complete for SRR13349125_2.fastq\n",
      "Approx 45% complete for SRR13349125_2.fastq\n",
      "Approx 50% complete for SRR13349125_2.fastq\n",
      "Approx 55% complete for SRR13349125_2.fastq\n",
      "Approx 60% complete for SRR13349125_2.fastq\n",
      "Approx 65% complete for SRR13349125_2.fastq\n",
      "Approx 70% complete for SRR13349125_2.fastq\n",
      "Approx 75% complete for SRR13349125_2.fastq\n",
      "Approx 80% complete for SRR13349125_2.fastq\n",
      "Approx 85% complete for SRR13349125_2.fastq\n",
      "Approx 90% complete for SRR13349125_2.fastq\n",
      "Approx 95% complete for SRR13349125_2.fastq\n",
      "Analysis complete for SRR13349125_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349124_2.fastq\n",
      "Approx 5% complete for SRR13349124_2.fastq\n",
      "Approx 10% complete for SRR13349124_2.fastq\n",
      "Approx 15% complete for SRR13349124_2.fastq\n",
      "Approx 20% complete for SRR13349124_2.fastq\n",
      "Approx 25% complete for SRR13349124_2.fastq\n",
      "Approx 30% complete for SRR13349124_2.fastq\n",
      "Approx 35% complete for SRR13349124_2.fastq\n",
      "Approx 40% complete for SRR13349124_2.fastq\n",
      "Approx 45% complete for SRR13349124_2.fastq\n",
      "Approx 50% complete for SRR13349124_2.fastq\n",
      "Approx 55% complete for SRR13349124_2.fastq\n",
      "Approx 60% complete for SRR13349124_2.fastq\n",
      "Approx 65% complete for SRR13349124_2.fastq\n",
      "Approx 70% complete for SRR13349124_2.fastq\n",
      "Approx 75% complete for SRR13349124_2.fastq\n",
      "Approx 80% complete for SRR13349124_2.fastq\n",
      "Approx 85% complete for SRR13349124_2.fastq\n",
      "Approx 90% complete for SRR13349124_2.fastq\n",
      "Approx 95% complete for SRR13349124_2.fastq\n",
      "Analysis complete for SRR13349124_2.fastq\n",
      "null\n",
      "Started analysis of SRR13349126_2.fastq\n",
      "Approx 5% complete for SRR13349126_2.fastq\n",
      "Approx 10% complete for SRR13349126_2.fastq\n",
      "Approx 15% complete for SRR13349126_2.fastq\n",
      "Approx 20% complete for SRR13349126_2.fastq\n",
      "Approx 25% complete for SRR13349126_2.fastq\n",
      "Approx 30% complete for SRR13349126_2.fastq\n",
      "Approx 35% complete for SRR13349126_2.fastq\n",
      "Approx 40% complete for SRR13349126_2.fastq\n",
      "Approx 45% complete for SRR13349126_2.fastq\n",
      "Approx 50% complete for SRR13349126_2.fastq\n",
      "Approx 55% complete for SRR13349126_2.fastq\n",
      "Approx 60% complete for SRR13349126_2.fastq\n",
      "Approx 65% complete for SRR13349126_2.fastq\n",
      "Approx 70% complete for SRR13349126_2.fastq\n",
      "Approx 75% complete for SRR13349126_2.fastq\n",
      "Approx 80% complete for SRR13349126_2.fastq\n",
      "Approx 85% complete for SRR13349126_2.fastq\n",
      "Approx 90% complete for SRR13349126_2.fastq\n",
      "Approx 95% complete for SRR13349126_2.fastq\n",
      "Analysis complete for SRR13349126_2.fastq\n"
     ]
    }
   ],
   "source": [
    "#run fastqc for forward reads\n",
    "!cat accs.txt | xargs -I {} fastqc \"data/raw_fastq/{}_1.fastq\" -o data/fastqc/\n",
    "#run fastqc for reverse reads\n",
    "!cat accs.txt | xargs -I {} fastqc \"data/raw_fastq/{}_2.fastq\" -o data/fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fastqc will output the results in HTML format, as below, for all forward and reverse reads.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"800\"\n",
       "            height=\"600\"\n",
       "            src=\"./data/fastqc/SRR13349126_1_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f8f03eb6fe0>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(src='./data/fastqc/SRR13349126_1_fastqc.html', width=800, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although its best practice to look over them individually, tools like multiqc allow one to quickly look at a summary of the quality reports of the fastq files.\n",
    "\n",
    "For instance, the below table shows which warnings, passes, or failures, from each fastqc report. There are other summaries created as well by multiqc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=428511;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2mv1.25\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /home/ec2-user/SageMaker/rnaseq-myco-notebook/data/fastqc\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m48/48\u001b[0m  tqc/SRR13349128_2_fastqc.zip\u001b[0m.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 24 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data   (overwritten)\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html   (overwritten)\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Filename</th>\n",
       "      <th>File type</th>\n",
       "      <th>Encoding</th>\n",
       "      <th>Total Sequences</th>\n",
       "      <th>Total Bases</th>\n",
       "      <th>Sequences flagged as poor quality</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>%GC</th>\n",
       "      <th>total_deduplicated_percentage</th>\n",
       "      <th>...</th>\n",
       "      <th>basic_statistics</th>\n",
       "      <th>per_base_sequence_quality</th>\n",
       "      <th>per_sequence_quality_scores</th>\n",
       "      <th>per_base_sequence_content</th>\n",
       "      <th>per_sequence_gc_content</th>\n",
       "      <th>per_base_n_content</th>\n",
       "      <th>sequence_length_distribution</th>\n",
       "      <th>sequence_duplication_levels</th>\n",
       "      <th>overrepresented_sequences</th>\n",
       "      <th>adapter_content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13349122_1</td>\n",
       "      <td>SRR13349122_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10827590.0</td>\n",
       "      <td>552.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>5.503069</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13349122_2</td>\n",
       "      <td>SRR13349122_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10827590.0</td>\n",
       "      <td>552.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9.335313</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13349123_1</td>\n",
       "      <td>SRR13349123_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11165256.0</td>\n",
       "      <td>569.4 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>5.097632</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13349123_2</td>\n",
       "      <td>SRR13349123_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11165256.0</td>\n",
       "      <td>569.4 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9.070512</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13349124_1</td>\n",
       "      <td>SRR13349124_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10727273.0</td>\n",
       "      <td>547 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>5.313842</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR13349124_2</td>\n",
       "      <td>SRR13349124_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10727273.0</td>\n",
       "      <td>547 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9.139844</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR13349125_1</td>\n",
       "      <td>SRR13349125_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10992686.0</td>\n",
       "      <td>560.6 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.918485</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR13349125_2</td>\n",
       "      <td>SRR13349125_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10992686.0</td>\n",
       "      <td>560.6 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>8.541477</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR13349126_1</td>\n",
       "      <td>SRR13349126_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12267497.0</td>\n",
       "      <td>625.6 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.804107</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR13349126_2</td>\n",
       "      <td>SRR13349126_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12267497.0</td>\n",
       "      <td>625.6 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>9.482849</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR13349127_1</td>\n",
       "      <td>SRR13349127_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12563032.0</td>\n",
       "      <td>640.7 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.443504</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR13349127_2</td>\n",
       "      <td>SRR13349127_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12563032.0</td>\n",
       "      <td>640.7 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>9.669627</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRR13349128_1</td>\n",
       "      <td>SRR13349128_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12652387.0</td>\n",
       "      <td>645.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>4.368545</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRR13349128_2</td>\n",
       "      <td>SRR13349128_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12652387.0</td>\n",
       "      <td>645.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>7.965338</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRR13349129_1</td>\n",
       "      <td>SRR13349129_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12961793.0</td>\n",
       "      <td>661 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>3.980868</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRR13349129_2</td>\n",
       "      <td>SRR13349129_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>12961793.0</td>\n",
       "      <td>661 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>7.409335</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRR13349130_1</td>\n",
       "      <td>SRR13349130_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10083015.0</td>\n",
       "      <td>514.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.997743</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRR13349130_2</td>\n",
       "      <td>SRR13349130_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10083015.0</td>\n",
       "      <td>514.2 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>10.136776</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR13349131_1</td>\n",
       "      <td>SRR13349131_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10491160.0</td>\n",
       "      <td>535 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.458165</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRR13349131_2</td>\n",
       "      <td>SRR13349131_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>10491160.0</td>\n",
       "      <td>535 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>10.218760</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRR13349132_1</td>\n",
       "      <td>SRR13349132_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11341357.0</td>\n",
       "      <td>578.4 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>5.233561</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRR13349132_2</td>\n",
       "      <td>SRR13349132_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11341357.0</td>\n",
       "      <td>578.4 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>8.801641</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRR13349133_1</td>\n",
       "      <td>SRR13349133_1.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11603881.0</td>\n",
       "      <td>591.7 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.841224</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR13349133_2</td>\n",
       "      <td>SRR13349133_2.fastq</td>\n",
       "      <td>Conventional base calls</td>\n",
       "      <td>Sanger / Illumina 1.9</td>\n",
       "      <td>11603881.0</td>\n",
       "      <td>591.7 Mbp</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>8.262785</td>\n",
       "      <td>...</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>warn</td>\n",
       "      <td>pass</td>\n",
       "      <td>pass</td>\n",
       "      <td>fail</td>\n",
       "      <td>fail</td>\n",
       "      <td>pass</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample             Filename                File type  \\\n",
       "0   SRR13349122_1  SRR13349122_1.fastq  Conventional base calls   \n",
       "1   SRR13349122_2  SRR13349122_2.fastq  Conventional base calls   \n",
       "2   SRR13349123_1  SRR13349123_1.fastq  Conventional base calls   \n",
       "3   SRR13349123_2  SRR13349123_2.fastq  Conventional base calls   \n",
       "4   SRR13349124_1  SRR13349124_1.fastq  Conventional base calls   \n",
       "5   SRR13349124_2  SRR13349124_2.fastq  Conventional base calls   \n",
       "6   SRR13349125_1  SRR13349125_1.fastq  Conventional base calls   \n",
       "7   SRR13349125_2  SRR13349125_2.fastq  Conventional base calls   \n",
       "8   SRR13349126_1  SRR13349126_1.fastq  Conventional base calls   \n",
       "9   SRR13349126_2  SRR13349126_2.fastq  Conventional base calls   \n",
       "10  SRR13349127_1  SRR13349127_1.fastq  Conventional base calls   \n",
       "11  SRR13349127_2  SRR13349127_2.fastq  Conventional base calls   \n",
       "12  SRR13349128_1  SRR13349128_1.fastq  Conventional base calls   \n",
       "13  SRR13349128_2  SRR13349128_2.fastq  Conventional base calls   \n",
       "14  SRR13349129_1  SRR13349129_1.fastq  Conventional base calls   \n",
       "15  SRR13349129_2  SRR13349129_2.fastq  Conventional base calls   \n",
       "16  SRR13349130_1  SRR13349130_1.fastq  Conventional base calls   \n",
       "17  SRR13349130_2  SRR13349130_2.fastq  Conventional base calls   \n",
       "18  SRR13349131_1  SRR13349131_1.fastq  Conventional base calls   \n",
       "19  SRR13349131_2  SRR13349131_2.fastq  Conventional base calls   \n",
       "20  SRR13349132_1  SRR13349132_1.fastq  Conventional base calls   \n",
       "21  SRR13349132_2  SRR13349132_2.fastq  Conventional base calls   \n",
       "22  SRR13349133_1  SRR13349133_1.fastq  Conventional base calls   \n",
       "23  SRR13349133_2  SRR13349133_2.fastq  Conventional base calls   \n",
       "\n",
       "                 Encoding  Total Sequences Total Bases  \\\n",
       "0   Sanger / Illumina 1.9       10827590.0   552.2 Mbp   \n",
       "1   Sanger / Illumina 1.9       10827590.0   552.2 Mbp   \n",
       "2   Sanger / Illumina 1.9       11165256.0   569.4 Mbp   \n",
       "3   Sanger / Illumina 1.9       11165256.0   569.4 Mbp   \n",
       "4   Sanger / Illumina 1.9       10727273.0     547 Mbp   \n",
       "5   Sanger / Illumina 1.9       10727273.0     547 Mbp   \n",
       "6   Sanger / Illumina 1.9       10992686.0   560.6 Mbp   \n",
       "7   Sanger / Illumina 1.9       10992686.0   560.6 Mbp   \n",
       "8   Sanger / Illumina 1.9       12267497.0   625.6 Mbp   \n",
       "9   Sanger / Illumina 1.9       12267497.0   625.6 Mbp   \n",
       "10  Sanger / Illumina 1.9       12563032.0   640.7 Mbp   \n",
       "11  Sanger / Illumina 1.9       12563032.0   640.7 Mbp   \n",
       "12  Sanger / Illumina 1.9       12652387.0   645.2 Mbp   \n",
       "13  Sanger / Illumina 1.9       12652387.0   645.2 Mbp   \n",
       "14  Sanger / Illumina 1.9       12961793.0     661 Mbp   \n",
       "15  Sanger / Illumina 1.9       12961793.0     661 Mbp   \n",
       "16  Sanger / Illumina 1.9       10083015.0   514.2 Mbp   \n",
       "17  Sanger / Illumina 1.9       10083015.0   514.2 Mbp   \n",
       "18  Sanger / Illumina 1.9       10491160.0     535 Mbp   \n",
       "19  Sanger / Illumina 1.9       10491160.0     535 Mbp   \n",
       "20  Sanger / Illumina 1.9       11341357.0   578.4 Mbp   \n",
       "21  Sanger / Illumina 1.9       11341357.0   578.4 Mbp   \n",
       "22  Sanger / Illumina 1.9       11603881.0   591.7 Mbp   \n",
       "23  Sanger / Illumina 1.9       11603881.0   591.7 Mbp   \n",
       "\n",
       "    Sequences flagged as poor quality  Sequence length   %GC  \\\n",
       "0                                 0.0             51.0  55.0   \n",
       "1                                 0.0             51.0  56.0   \n",
       "2                                 0.0             51.0  56.0   \n",
       "3                                 0.0             51.0  56.0   \n",
       "4                                 0.0             51.0  55.0   \n",
       "5                                 0.0             51.0  56.0   \n",
       "6                                 0.0             51.0  55.0   \n",
       "7                                 0.0             51.0  56.0   \n",
       "8                                 0.0             51.0  55.0   \n",
       "9                                 0.0             51.0  55.0   \n",
       "10                                0.0             51.0  55.0   \n",
       "11                                0.0             51.0  56.0   \n",
       "12                                0.0             51.0  54.0   \n",
       "13                                0.0             51.0  56.0   \n",
       "14                                0.0             51.0  54.0   \n",
       "15                                0.0             51.0  56.0   \n",
       "16                                0.0             51.0  55.0   \n",
       "17                                0.0             51.0  56.0   \n",
       "18                                0.0             51.0  55.0   \n",
       "19                                0.0             51.0  56.0   \n",
       "20                                0.0             51.0  55.0   \n",
       "21                                0.0             51.0  56.0   \n",
       "22                                0.0             51.0  55.0   \n",
       "23                                0.0             51.0  56.0   \n",
       "\n",
       "    total_deduplicated_percentage  ...  basic_statistics  \\\n",
       "0                        5.503069  ...              pass   \n",
       "1                        9.335313  ...              pass   \n",
       "2                        5.097632  ...              pass   \n",
       "3                        9.070512  ...              pass   \n",
       "4                        5.313842  ...              pass   \n",
       "5                        9.139844  ...              pass   \n",
       "6                        4.918485  ...              pass   \n",
       "7                        8.541477  ...              pass   \n",
       "8                        4.804107  ...              pass   \n",
       "9                        9.482849  ...              pass   \n",
       "10                       4.443504  ...              pass   \n",
       "11                       9.669627  ...              pass   \n",
       "12                       4.368545  ...              pass   \n",
       "13                       7.965338  ...              pass   \n",
       "14                       3.980868  ...              pass   \n",
       "15                       7.409335  ...              pass   \n",
       "16                       4.997743  ...              pass   \n",
       "17                      10.136776  ...              pass   \n",
       "18                       4.458165  ...              pass   \n",
       "19                      10.218760  ...              pass   \n",
       "20                       5.233561  ...              pass   \n",
       "21                       8.801641  ...              pass   \n",
       "22                       4.841224  ...              pass   \n",
       "23                       8.262785  ...              pass   \n",
       "\n",
       "    per_base_sequence_quality per_sequence_quality_scores  \\\n",
       "0                        pass                        pass   \n",
       "1                        pass                        pass   \n",
       "2                        pass                        pass   \n",
       "3                        pass                        pass   \n",
       "4                        pass                        pass   \n",
       "5                        pass                        pass   \n",
       "6                        pass                        pass   \n",
       "7                        pass                        pass   \n",
       "8                        pass                        pass   \n",
       "9                        pass                        pass   \n",
       "10                       pass                        pass   \n",
       "11                       pass                        pass   \n",
       "12                       pass                        pass   \n",
       "13                       pass                        pass   \n",
       "14                       pass                        pass   \n",
       "15                       pass                        pass   \n",
       "16                       pass                        pass   \n",
       "17                       pass                        pass   \n",
       "18                       pass                        pass   \n",
       "19                       pass                        pass   \n",
       "20                       pass                        pass   \n",
       "21                       pass                        pass   \n",
       "22                       pass                        pass   \n",
       "23                       pass                        pass   \n",
       "\n",
       "   per_base_sequence_content per_sequence_gc_content per_base_n_content  \\\n",
       "0                       fail                    pass               pass   \n",
       "1                       fail                    pass               pass   \n",
       "2                       fail                    pass               pass   \n",
       "3                       fail                    warn               pass   \n",
       "4                       fail                    pass               pass   \n",
       "5                       fail                    pass               pass   \n",
       "6                       fail                    pass               pass   \n",
       "7                       fail                    warn               pass   \n",
       "8                       fail                    pass               pass   \n",
       "9                       fail                    warn               pass   \n",
       "10                      fail                    pass               pass   \n",
       "11                      fail                    warn               pass   \n",
       "12                      fail                    warn               pass   \n",
       "13                      fail                    pass               pass   \n",
       "14                      fail                    warn               pass   \n",
       "15                      fail                    warn               pass   \n",
       "16                      fail                    warn               pass   \n",
       "17                      fail                    pass               pass   \n",
       "18                      fail                    pass               pass   \n",
       "19                      fail                    warn               pass   \n",
       "20                      fail                    warn               pass   \n",
       "21                      fail                    warn               pass   \n",
       "22                      fail                    warn               pass   \n",
       "23                      fail                    warn               pass   \n",
       "\n",
       "   sequence_length_distribution sequence_duplication_levels  \\\n",
       "0                          pass                        fail   \n",
       "1                          pass                        fail   \n",
       "2                          pass                        fail   \n",
       "3                          pass                        fail   \n",
       "4                          pass                        fail   \n",
       "5                          pass                        fail   \n",
       "6                          pass                        fail   \n",
       "7                          pass                        fail   \n",
       "8                          pass                        fail   \n",
       "9                          pass                        fail   \n",
       "10                         pass                        fail   \n",
       "11                         pass                        fail   \n",
       "12                         pass                        fail   \n",
       "13                         pass                        fail   \n",
       "14                         pass                        fail   \n",
       "15                         pass                        fail   \n",
       "16                         pass                        fail   \n",
       "17                         pass                        fail   \n",
       "18                         pass                        fail   \n",
       "19                         pass                        fail   \n",
       "20                         pass                        fail   \n",
       "21                         pass                        fail   \n",
       "22                         pass                        fail   \n",
       "23                         pass                        fail   \n",
       "\n",
       "   overrepresented_sequences adapter_content  \n",
       "0                       fail            pass  \n",
       "1                       fail            pass  \n",
       "2                       fail            pass  \n",
       "3                       fail            pass  \n",
       "4                       fail            pass  \n",
       "5                       fail            pass  \n",
       "6                       fail            pass  \n",
       "7                       fail            pass  \n",
       "8                       fail            pass  \n",
       "9                       fail            pass  \n",
       "10                      fail            pass  \n",
       "11                      fail            pass  \n",
       "12                      fail            pass  \n",
       "13                      fail            pass  \n",
       "14                      fail            pass  \n",
       "15                      fail            pass  \n",
       "16                      fail            pass  \n",
       "17                      fail            pass  \n",
       "18                      fail            pass  \n",
       "19                      fail            pass  \n",
       "20                      fail            pass  \n",
       "21                      fail            pass  \n",
       "22                      fail            pass  \n",
       "23                      fail            pass  \n",
       "\n",
       "[24 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "!multiqc -f data/fastqc/\n",
    "\n",
    "import pandas as pd\n",
    "dframe = pd.read_csv(\"./multiqc_data/multiqc_fastqc.txt\", sep='\\t')\n",
    "display(dframe)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 5.1 Merging our fastq files\n",
    "\n",
    "The following step may not be necessary -- it depends on the study.\n",
    "\n",
    "In our case, if we look at our SRA files:\n",
    "\n",
    "https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP300216&o=acc_s%3Aa\n",
    "\n",
    "We will notice that, although there are 12 SRA files, coming from 12 SRR runs -- Actually, there are only 6 total samples. \n",
    "\n",
    "In such a case it is possible that, for instance, multiple different lanes in a flowcell may have been used for the same sample.\n",
    "\n",
    "In our analysis will be comparing at the sample level. So, we would like to merge the fastq files that, although were created as separate fastq files by the sequencer, actually came from the sample. \n",
    "\n",
    "It is generally easier to do this merging after an initial fastqc report, as it makes it easier to pinpoint errors that may be lane specific.\n",
    "\n",
    "Combining two FASTQ files is a straightforward process. Remember how FASTQ files are formatted, they are a list of readcounts. Consequently, we can simply 'concatenate' or add one fastq file to the bottom of another to create a merged fastq. Note that header information in a single fastq file may now contain different lane information -- however for our downstream processes this is acceptable. Remember if your fastq files are zipped, you will have to unzip them first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#example of how to concatenate two of our fastq files from the same experiment.\n",
    "!cat data/raw_fastq/SRR13349122_1.fastq data/raw_fastq/SRR13349123_1.fastq > data/raw_fastq/GSM5004088_1.fastq\n",
    "!cat data/raw_fastq/SRR13349122_2.fastq data/raw_fastq/SRR13349123_2.fastq > data/raw_fastq/GSM5004088_2.fastq\n",
    "#notice we concat the forward read with a forward read, and reverse with reverse.\n",
    "#also note here we are naming it after the GSM, which is the sample experiment ID."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We could manually do the above for all 12 of our runs, and it wouldn't be much work.\n",
    "\n",
    "If you are comfortable doing so, that is the best process to use.\n",
    "\n",
    "As always though, there are ways to automate things. For instance, we could make use of our query code from 3.1.2 to obtain a list of the SRX IDs, and take advantage of our Jupyter's ability to write python to write a loop to iterate and concat our list.\n",
    "\n",
    "Specific to our case, each sample contains two paired-end SRRs. \n",
    "\n",
    "Note; running this step will remove the previous unmerged fastq files in order to save space.\n",
    "\n",
    "This will take about one hour."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_5661/1247509149.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df = pd.read_sql(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "      <th>run_file_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13349128</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775149</td>\n",
       "      <td>GSM5004091</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004091}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13349127</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775148</td>\n",
       "      <td>GSM5004090</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004090}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13349129</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775149</td>\n",
       "      <td>GSM5004091</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004091}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13349122</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775146</td>\n",
       "      <td>GSM5004088</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004088}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13349131</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775150</td>\n",
       "      <td>GSM5004092</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004092}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR13349123</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775146</td>\n",
       "      <td>GSM5004088</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004088}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRR13349132</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775151</td>\n",
       "      <td>GSM5004093</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004093}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRR13349125</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775147</td>\n",
       "      <td>GSM5004089</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004089}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRR13349130</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775150</td>\n",
       "      <td>GSM5004092</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004092}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRR13349133</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775151</td>\n",
       "      <td>GSM5004093</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004093}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRR13349126</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775148</td>\n",
       "      <td>GSM5004090</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004090}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRR13349124</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX9775147</td>\n",
       "      <td>GSM5004089</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[fastq, run.zq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.us-east1, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{k=geo_accession_exp, v=GSM5004089}, {k=bases...</td>\n",
       "      <td>{\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            acc assay_type center_name consent  experiment sample_name  \\\n",
       "0   SRR13349128    RNA-Seq         GEO  public  SRX9775149  GSM5004091   \n",
       "1   SRR13349127    RNA-Seq         GEO  public  SRX9775148  GSM5004090   \n",
       "2   SRR13349129    RNA-Seq         GEO  public  SRX9775149  GSM5004091   \n",
       "3   SRR13349122    RNA-Seq         GEO  public  SRX9775146  GSM5004088   \n",
       "4   SRR13349131    RNA-Seq         GEO  public  SRX9775150  GSM5004092   \n",
       "5   SRR13349123    RNA-Seq         GEO  public  SRX9775146  GSM5004088   \n",
       "6   SRR13349132    RNA-Seq         GEO  public  SRX9775151  GSM5004093   \n",
       "7   SRR13349125    RNA-Seq         GEO  public  SRX9775147  GSM5004089   \n",
       "8   SRR13349130    RNA-Seq         GEO  public  SRX9775150  GSM5004092   \n",
       "9   SRR13349133    RNA-Seq         GEO  public  SRX9775151  GSM5004093   \n",
       "10  SRR13349126    RNA-Seq         GEO  public  SRX9775148  GSM5004090   \n",
       "11  SRR13349124    RNA-Seq         GEO  public  SRX9775147  GSM5004089   \n",
       "\n",
       "             instrument librarylayout libraryselection   librarysource  ...  \\\n",
       "0   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "1   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "2   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "3   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "4   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "5   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "6   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "7   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "8   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "9   Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "10  Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "11  Illumina HiSeq 2500        PAIRED             cDNA  TRANSCRIPTOMIC  ...   \n",
       "\n",
       "   geo_loc_name_sam ena_first_public_run ena_last_update_run sample_name_sam  \\\n",
       "0              None                 None                None            None   \n",
       "1              None                 None                None            None   \n",
       "2              None                 None                None            None   \n",
       "3              None                 None                None            None   \n",
       "4              None                 None                None            None   \n",
       "5              None                 None                None            None   \n",
       "6              None                 None                None            None   \n",
       "7              None                 None                None            None   \n",
       "8              None                 None                None            None   \n",
       "9              None                 None                None            None   \n",
       "10             None                 None                None            None   \n",
       "11             None                 None                None            None   \n",
       "\n",
       "      datastore_filetype datastore_provider  \\\n",
       "0   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "1   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "2   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "3   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "4   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "5   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "6   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "7   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "8   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "9   [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "10  [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "11  [fastq, run.zq, sra]     [gs, ncbi, s3]   \n",
       "\n",
       "                            datastore_region  \\\n",
       "0   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "1   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "2   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "3   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "4   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "5   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "6   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "7   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "8   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "9   [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "10  [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "11  [gs.us-east1, ncbi.public, s3.us-east-1]   \n",
       "\n",
       "                                           attributes  \\\n",
       "0   [{k=geo_accession_exp, v=GSM5004091}, {k=bases...   \n",
       "1   [{k=geo_accession_exp, v=GSM5004090}, {k=bases...   \n",
       "2   [{k=geo_accession_exp, v=GSM5004091}, {k=bases...   \n",
       "3   [{k=geo_accession_exp, v=GSM5004088}, {k=bases...   \n",
       "4   [{k=geo_accession_exp, v=GSM5004092}, {k=bases...   \n",
       "5   [{k=geo_accession_exp, v=GSM5004088}, {k=bases...   \n",
       "6   [{k=geo_accession_exp, v=GSM5004093}, {k=bases...   \n",
       "7   [{k=geo_accession_exp, v=GSM5004089}, {k=bases...   \n",
       "8   [{k=geo_accession_exp, v=GSM5004092}, {k=bases...   \n",
       "9   [{k=geo_accession_exp, v=GSM5004093}, {k=bases...   \n",
       "10  [{k=geo_accession_exp, v=GSM5004090}, {k=bases...   \n",
       "11  [{k=geo_accession_exp, v=GSM5004089}, {k=bases...   \n",
       "\n",
       "                                                jattr  run_file_version  \n",
       "0   {\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...                 1  \n",
       "1   {\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...                 1  \n",
       "2   {\"geo_accession_exp\": [\"GSM5004091\"], \"bases\":...                 1  \n",
       "3   {\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...                 1  \n",
       "4   {\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...                 1  \n",
       "5   {\"geo_accession_exp\": [\"GSM5004088\"], \"bases\":...                 1  \n",
       "6   {\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...                 1  \n",
       "7   {\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...                 1  \n",
       "8   {\"geo_accession_exp\": [\"GSM5004092\"], \"bases\":...                 1  \n",
       "9   {\"geo_accession_exp\": [\"GSM5004093\"], \"bases\":...                 1  \n",
       "10  {\"geo_accession_exp\": [\"GSM5004090\"], \"bases\":...                 1  \n",
       "11  {\"geo_accession_exp\": [\"GSM5004089\"], \"bases\":...                 1  \n",
       "\n",
       "[12 rows x 37 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pyathena import connect\n",
    "import pandas as pd\n",
    "\n",
    "# Use the correct argument name: s3_staging_dir\n",
    "conn = connect(s3_staging_dir='s3://sra-data-athena/', region_name='us-east-1')\n",
    "\n",
    "query = \"\"\"\n",
    "SELECT *\n",
    "FROM AwsDataCatalog.srametadata.metadata\n",
    "WHERE organism = 'Mycobacteroides chelonae' \n",
    "AND acc LIKE '%SRR133491%'\n",
    "\"\"\"\n",
    "df = pd.read_sql(\n",
    "    query, conn\n",
    ")\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "# Now get the accession IDs and sample IDs from the created dataframe\n",
    "runs = df['acc'].values\n",
    "samples = list(set(df['sample_name'].values))\n",
    "\n",
    "# Sort them to be in numerical order\n",
    "runs.sort()\n",
    "samples.sort()\n",
    "\n",
    "# Iterate through the samples\n",
    "for index, item in enumerate(samples):\n",
    "    try:\n",
    "        print(f\"Processing sample {samples[index]}...\")\n",
    "\n",
    "        # Concatenate the two SRRs for forward reads\n",
    "        #print(f\"Concatenating forward reads for {samples[index]}...\")\n",
    "        subprocess.run([\"cat\", f\"data/raw_fastq/{runs[index*2]}_1.fastq\", f\"data/raw_fastq/{runs[index*2+1]}_1.fastq\", \">\", f\"data/raw_fastq/{samples[index]}_1.fastq\"], check=True)\n",
    "\n",
    "        # Delete the original fastq files for forward reads\n",
    "        #print(f\"Deleting original forward reads for {samples[index]}...\")\n",
    "        subprocess.run([\"rm\", f\"data/raw_fastq/{runs[index*2]}_1.fastq\"], check=True)\n",
    "        subprocess.run([\"rm\", f\"data/raw_fastq/{runs[index*2+1]}_1.fastq\"], check=True)\n",
    "\n",
    "        # Zip the merged forward fastq file\n",
    "        #print(f\"Zipping forward reads for {samples[index]}...\")\n",
    "        subprocess.run([\"gzip\", f\"data/raw_fastq/{samples[index]}_1.fastq\"], check=True)\n",
    "\n",
    "        # Concatenate the two SRRs for reverse reads\n",
    "        #print(f\"Concatenating reverse reads for {samples[index]}...\")\n",
    "        subprocess.run([\"cat\", f\"data/raw_fastq/{runs[index*2]}_2.fastq\", f\"data/raw_fastq/{runs[index*2+1]}_2.fastq\", \">\", f\"data/raw_fastq/{samples[index]}_2.fastq\"], check=True)\n",
    "\n",
    "        # Delete the original fastq files for reverse reads\n",
    "        #print(f\"Deleting original reverse reads for {samples[index]}...\")\n",
    "        subprocess.run([\"rm\", f\"data/raw_fastq/{runs[index*2]}_2.fastq\"], check=True)\n",
    "        subprocess.run([\"rm\", f\"data/raw_fastq/{runs[index*2+1]}_2.fastq\"], check=True)\n",
    "\n",
    "        # Zip the merged reverse fastq file\n",
    "        #print(f\"Zipping reverse reads for {samples[index]}...\")\n",
    "        subprocess.run([\"gzip\", f\"data/raw_fastq/{samples[index]}_2.fastq\"], check=True)\n",
    "\n",
    "        print(f\"Processing complete for sample {samples[index]}.\\n\")\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error processing sample {samples[index]}: {e}\")\n",
    "        # Optionally, you can exit or skip to the next sample\n",
    "        continue\n",
    "\n",
    "print(\"All samples processed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM5004088\n",
      "GSM5004089\n",
      "GSM5004090\n",
      "GSM5004091\n",
      "GSM5004092\n",
      "GSM5004093"
     ]
    }
   ],
   "source": [
    "#since our files will now be samples, not SRRs we can write a new text file to use for downstream batch processes.\n",
    "#we can use the DF we made in the previous cell.\n",
    "with open('samples.txt', 'w') as f:\n",
    "    df = df.sort_values(by='sample_name', ascending=True)\n",
    "    samples = df['sample_name'].unique()\n",
    "    samples = '\\n'.join(map(str, samples))\n",
    "    f.write(samples)\n",
    "    \n",
    "!cat samples.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 5.3: Copy data file for Trimmomatic\n",
    "\n",
    "One of trimmomatics functions is to trim sequence machine specific adapter sequences. These are usually within the trimmomatic installation directory in a folder called adapters.\n",
    "\n",
    "Directories of packages within conda installations can be confusing, so in the case of using conda with trimmomatic, it may be easier to simply download or create a file with the relevant adapter sequencecs and store it in an easy to find directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying gs://nigms-sandbox/me-inbre-rnaseq-pipelinev2/config/TruSeq3-PE.fa...\n",
      "/ [1/1 files][   95.0 B/   95.0 B] 100% Done                                    \n",
      "Operation completed over 1 objects/95.0 B.                                       \n",
      ">PrefixPE/1\n",
      "TACACTCTTTCCCTACACGACGCTCTTCCGATCT\n",
      ">PrefixPE/2\n",
      "GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!gsutil -m cp -r gs://nigms-sandbox/me-inbre-rnaseq-pipelinev2/config/TruSeq3-PE.fa .\n",
    "!head TruSeq3-PE.fa "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 6: Run Trimmomatic\n",
    "Trimmomatic will trim off any adapter sequences or low quality sequence it detects in the FASTQ files.\n",
    "\n",
    "Using piping and our original list, it is possible to queue up a batch run of trimmomatic for all our files, note that this is a different way to run a loop compared with what we did before.\n",
    "\n",
    "The below code may take approximately 35 minutes to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004088_1.fastq.gz data/raw_fastq/GSM5004088_2.fastq.gz data/trimmed/GSM5004088_1_trimmed.fastq.gz data/trimmed/GSM5004088_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004088_2_trimmed.fastq.gz data/trimmed/GSM5004088_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004089_1.fastq.gz data/raw_fastq/GSM5004089_2.fastq.gz data/trimmed/GSM5004089_1_trimmed.fastq.gz data/trimmed/GSM5004089_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004089_2_trimmed.fastq.gz data/trimmed/GSM5004089_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004090_1.fastq.gz data/raw_fastq/GSM5004090_2.fastq.gz data/trimmed/GSM5004090_1_trimmed.fastq.gz data/trimmed/GSM5004090_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004090_2_trimmed.fastq.gz data/trimmed/GSM5004090_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004091_1.fastq.gz data/raw_fastq/GSM5004091_2.fastq.gz data/trimmed/GSM5004091_1_trimmed.fastq.gz data/trimmed/GSM5004091_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004091_2_trimmed.fastq.gz data/trimmed/GSM5004091_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004092_1.fastq.gz data/raw_fastq/GSM5004092_2.fastq.gz data/trimmed/GSM5004092_1_trimmed.fastq.gz data/trimmed/GSM5004092_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004092_2_trimmed.fastq.gz data/trimmed/GSM5004092_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 1 data/raw_fastq/GSM5004093_1.fastq.gz data/raw_fastq/GSM5004093_2.fastq.gz data/trimmed/GSM5004093_1_trimmed.fastq.gz data/trimmed/GSM5004093_1_trimmed_unpaired.fastq.gz data/trimmed/GSM5004093_2_trimmed.fastq.gz data/trimmed/GSM5004093_2_trimmed_unpaired.fastq.gz ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Error: Unable to detect quality encoding\n"
     ]
    }
   ],
   "source": [
    "!cat samples.txt | xargs -I {} trimmomatic PE -threads $THREADS 'data/raw_fastq/{}_1.fastq.gz' 'data/raw_fastq/{}_2.fastq.gz' 'data/trimmed/{}_1_trimmed.fastq.gz' 'data/trimmed/{}_1_trimmed_unpaired.fastq.gz' 'data/trimmed/{}_2_trimmed.fastq.gz' 'data/trimmed/{}_2_trimmed_unpaired.fastq.gz' ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 7 (Optional): Run FastQC\n",
    "\n",
    "It's best practice to run FastQC after trimming. However, you may decide to run FastQC only once, before or after trimming.\n",
    "\n",
    "We will proceed with only the forward reads -- this is because, looking at trimmomatic, there were very few 'orphaned' reads. That is to say, most forward and reverse reads were successfully paired together. Because we are just trying to map to a transcriptome, the read lengths of the forward reads alone, in this case, around 50~ basepairs, should be sufficient.\n",
    "\n",
    "The below code may take around 10 minutes to run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping 'data/trimmed/GSM5004088_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n",
      "Skipping 'data/trimmed/GSM5004089_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n",
      "Skipping 'data/trimmed/GSM5004090_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n",
      "Skipping 'data/trimmed/GSM5004091_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n",
      "Skipping 'data/trimmed/GSM5004092_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n",
      "Skipping 'data/trimmed/GSM5004093_1_trimmed.fastq.gz' which didn't exist, or couldn't be read\n"
     ]
    }
   ],
   "source": [
    "!cat samples.txt | xargs -I {} fastqc \"data/trimmed/{}_1_trimmed.fastq.gz\" -o data/fastqc_samples/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 8: Run MultiQC\n",
    "MultiQC reads in the FastQC reports and generate a compiled report for all the analyzed FASTQ files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=248969;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2mv1.25\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /home/ec2-user/SageMaker/rnaseq-myco-notebook/data/fastqc_samples\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m  0%\u001b[0m \u001b[32m0/0\u001b[0m  0/0\u001b[0m  \n",
      "\u001b[?25h\u001b[34m           multiqc\u001b[0m | \u001b[33mNo analysis results found. Cleaning up…\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "#!multiqc -f data/fastqc_samples/\n",
    "!multiqc -f -o data/multiqc_samples/ data/fastqc_samples/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 9: Index the Transcriptome so that Trimmed Reads Can Be Mapped Using Salmon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run salmon we must specify the reference transcriptome, and the folder of our created index.\n",
    "\n",
    "Note here, -i does not mean input, but the folder where the index will be created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Server Response: Not Found\n",
      "[2024-09-18 06:16:37.320] [jLog] [info] building index\n",
      "out : data/reference/transcriptome_index\n",
      "\u001b[00m[2024-09-18 06:16:37.320] [puff::index::jointLog] [info] Running fixFasta\n",
      "\u001b[00m\n",
      "[Step 1 of 4] : counting k-mers\n",
      "\n",
      "\u001b[35m[2024-09-18 06:16:37.544] [puff::index::jointLog] [warning] There were 1 transcripts that would need to be removed to avoid duplicates.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:37.544] [puff::index::jointLog] [info] Replaced 0 non-ATCG nucleotides\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:37.544] [puff::index::jointLog] [info] Clipped poly-A tails from 0 transcripts\n",
      "\u001b[00mwrote 4919 cleaned references\n",
      "\u001b[00m[2024-09-18 06:16:37.564] [puff::index::jointLog] [info] Filter size not provided; estimating from number of distinct k-mers\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:37.663] [puff::index::jointLog] [info] ntHll estimated 4966944 distinct k-mers, setting filter size to 2^27\n",
      "\u001b[00mThreads = 1\n",
      "Vertex length = 31\n",
      "Hash functions = 5\n",
      "Filter size = 134217728\n",
      "Capacity = 2\n",
      "Files: \n",
      "data/reference/transcriptome_index/ref_k31_fixed.fa\n",
      "--------------------------------------------------------------------------------\n",
      "Round 0, 0:134217728\n",
      "Pass\tFilling\tFiltering\n",
      "1\t3\t6\t\n",
      "2\t1\t0\n",
      "True junctions count = 10233\n",
      "False junctions count = 6276\n",
      "Hash table size = 16509\n",
      "Candidate marks count = 33476\n",
      "--------------------------------------------------------------------------------\n",
      "Reallocating bifurcations time: 0\n",
      "True marks count: 21486\n",
      "Edges construction time: 0\n",
      "--------------------------------------------------------------------------------\n",
      "Distinct junctions = 10233\n",
      "\n",
      "TwoPaCo::buildGraphMain:: allocated with scalable_malloc; freeing.\n",
      "TwoPaCo::buildGraphMain:: Calling scalable_allocation_command(TBBMALLOC_CLEAN_ALL_BUFFERS, 0);\n",
      "allowedIn: 17\n",
      "Max Junction ID: 10276\n",
      "seen.size():82217 kmerInfo.size():10277\n",
      "approximateContigTotalLength: 4949507\n",
      "counters for complex kmers:\n",
      "(prec>1 & succ>1)=8 | (succ>1 & isStart)=0 | (prec>1 & isEnd)=0 | (isStart & isEnd)=2\n",
      "contig count: 10455 element count: 5331183 complex nodes: 10\n",
      "# of ones in rank vector: 10454\n",
      "\u001b[00m[2024-09-18 06:16:48.180] [puff::index::jointLog] [info] Starting the Pufferfish indexing by reading the GFA binary file.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.180] [puff::index::jointLog] [info] Setting the index/BinaryGfa directory data/reference/transcriptome_index\n",
      "\u001b[00msize = 5331183\n",
      "-----------------------------------------\n",
      "| Loading contigs | Time = 918.46 us\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 480.24 us\n",
      "-----------------------------------------\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "10454\n",
      "\u001b[00m[2024-09-18 06:16:48.189] [puff::index::jointLog] [info] Done wrapping the rank vector with a rank9sel structure.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.189] [puff::index::jointLog] [info] contig count for validation: 10,454\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.192] [puff::index::jointLog] [info] Total # of Contigs : 10,454\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.192] [puff::index::jointLog] [info] Total # of numerical Contigs : 10,454\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.192] [puff::index::jointLog] [info] Total # of contig vec entries: 16,600\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.192] [puff::index::jointLog] [info] bits per offset entry 15\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.192] [puff::index::jointLog] [info] Done constructing the contig vector. 10455\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.194] [puff::index::jointLog] [info] # segments = 10,454\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.194] [puff::index::jointLog] [info] total length = 5,331,183\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.195] [puff::index::jointLog] [info] Reading the reference files ...\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.234] [puff::index::jointLog] [info] positional integer width = 23\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.234] [puff::index::jointLog] [info] seqSize = 5,331,183\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.234] [puff::index::jointLog] [info] rankSize = 5,331,183\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.234] [puff::index::jointLog] [info] edgeVecSize = 0\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.234] [puff::index::jointLog] [info] num keys = 5,017,563\n",
      "\u001b[00mfor info, total work write each  : 2.331    total work inram from level 3 : 4.322  total work raw : 25.000 \n",
      "[Building BooPHF]  99.9 %   elapsed:   0 min 0  sec   remaining:   0 min 0  sec\n",
      "Bitarray        26296128  bits (100.00 %)   (array + ranks )\n",
      "final hash             0  bits (0.00 %) (nb in final hash 0)\n",
      "\u001b[00m[2024-09-18 06:16:48.716] [puff::index::jointLog] [info] mphf size = 3.13474 MB\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.724] [puff::index::jointLog] [info] chunk size = 5,331,183\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:48.724] [puff::index::jointLog] [info] chunk 0 = [0, 5,331,153)\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:49.444] [puff::index::jointLog] [info] finished populating pos vector\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:49.444] [puff::index::jointLog] [info] writing index components\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:49.478] [puff::index::jointLog] [info] finished writing dense pufferfish index\n",
      "\u001b[00m[2024-09-18 06:16:49.482] [jLog] [info] done building index\n"
     ]
    }
   ],
   "source": [
    "!salmon index -t data/reference/GCF_001632805.1_transcriptome_reference_w_decoy.fa -p $THREADS -i data/reference/transcriptome_index --decoys data/reference/decoys.txt -k 31 --keepDuplicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 10: Run Salmon to Map Reads to Transcripts and Quantify Expression Levels\n",
    "Salmon aligns the trimmed reads to the reference transcriptome and generates the read counts per transcript. In this analysis, each gene has a single transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004088_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004088_quant }\n",
      "Logs will be written to data/quants/GSM5004088_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:58.535] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.535] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.535] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.535] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.535] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.536] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 1.7822 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 81.985 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 26.88 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 1.9994 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 9.5596 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 949.28 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "\u001b[00m[2024-09-18 06:16:58.536] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.536] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading positions | Time = 8.6548 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.4956 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 54.925 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:58.561] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004088_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n",
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004089_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004089_quant }\n",
      "Logs will be written to data/quants/GSM5004089_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.746] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 1.8659 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 79.265 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 25.997 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 2.4016 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 8.606 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 893.48 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 9.0149 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.5738 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 52.88 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:58.771] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004089_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n",
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004090_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004090_quant }\n",
      "Logs will be written to data/quants/GSM5004090_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.915] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:58.916] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 1.8766 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 77.047 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 23.29 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 2.0003 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 7.5859 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 857.4 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 8.4648 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.5358 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 53.07 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:58.938] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004090_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n",
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004091_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004091_quant }\n",
      "Logs will be written to data/quants/GSM5004091_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.084] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 1.8515 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 71.779 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 27.356 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 1.9496 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 7.8421 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 889.15 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "\u001b[00m[2024-09-18 06:16:59.085] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.085] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading positions | Time = 8.6598 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.5413 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 51.876 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:59.108] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004091_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n",
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004092_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004092_quant }\n",
      "Logs will be written to data/quants/GSM5004092_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.254] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 1.8959 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 89.545 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 22.878 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 2.0056 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 7.6302 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 911.52 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 9.1928 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.9355 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 47.14 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:59.278] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004092_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n",
      "Version Server Response: Not Found\n",
      "### salmon (selective-alignment-based) v1.10.3\n",
      "### [ program ] => salmon \n",
      "### [ command ] => quant \n",
      "### [ index ] => { data/reference/transcriptome_index }\n",
      "### [ libType ] => { SR }\n",
      "### [ unmatedReads ] => { data/trimmed/GSM5004093_1_trimmed.fastq.gz }\n",
      "### [ threads ] => { 8 }\n",
      "### [ validateMappings ] => { }\n",
      "### [ output ] => { data/quants/GSM5004093_quant }\n",
      "Logs will be written to data/quants/GSM5004093_quant/logs\n",
      "\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] setting maxHashResizeThreads to 8\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] Setting consensusSlack to selective-alignment default of 0.35.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] parsing read library format\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] There is 1 library.\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] Loading pufferfish index\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:16:59.425] [jointLog] [info] Loading dense pufferfish index.\n",
      "\u001b[00m-----------------------------------------\n",
      "| Loading contig table | Time = 2.7585 ms\n",
      "-----------------------------------------\n",
      "size = 10455\n",
      "-----------------------------------------\n",
      "| Loading contig offsets | Time = 83.238 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference lengths | Time = 29.874 us\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading mphf table | Time = 2.3707 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "Number of ones: 10454\n",
      "Number of ones per inventory item: 512\n",
      "Inventory entries filled: 21\n",
      "-----------------------------------------\n",
      "| Loading contig boundaries | Time = 7.7368 ms\n",
      "-----------------------------------------\n",
      "size = 5331183\n",
      "-----------------------------------------\n",
      "| Loading sequence | Time = 859.68 us\n",
      "-----------------------------------------\n",
      "size = 5017563\n",
      "-----------------------------------------\n",
      "| Loading positions | Time = 8.5149 ms\n",
      "-----------------------------------------\n",
      "size = 9694896\n",
      "-----------------------------------------\n",
      "| Loading reference sequence | Time = 1.5117 ms\n",
      "-----------------------------------------\n",
      "-----------------------------------------\n",
      "| Loading reference accumulative lengths | Time = 49.726 us\n",
      "-----------------------------------------\n",
      "\u001b[00m[2024-09-18 06:16:59.450] [jointLog] [info] done\n",
      "\u001b[00mException : [\n",
      "The following errors were detected with the read files\n",
      "======================================================\n",
      "ERROR: file [data/trimmed/GSM5004093_1_trimmed.fastq.gz] does not appear to exist!\n",
      "\n",
      "]\n",
      "salmon quant was invoked improperly.\n",
      "For usage information, try salmon quant --help\n",
      "Exiting.\n"
     ]
    }
   ],
   "source": [
    "!cat samples.txt | xargs -I {} salmon quant -i data/reference/transcriptome_index -l SR -r \"data/trimmed/{}_1_trimmed.fastq.gz\" -p 8 --validateMappings -o \"data/quants/{}_quant\" > dump.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 11: Report the top 10 most highly expressed genes in the samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 10 most highly expressed genes in each wild-type sample.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "head: cannot open ‘data/quants/SRR13349122_quant/quant.sf’ for reading: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004088_quant/quant.sf: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004089_quant/quant.sf: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004090_quant/quant.sf: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!head data/quants/SRR13349122_quant/quant.sf -n 1\n",
    "!sort -nrk 4,4 data/quants/GSM5004088_quant/quant.sf | head -10\n",
    "!sort -nrk 4,4 data/quants/GSM5004089_quant/quant.sf | head -10\n",
    "!sort -nrk 4,4 data/quants/GSM5004090_quant/quant.sf | head -10\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 10 most highly expressed genes in the double lysogen samples.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "head: cannot open ‘data/quants/SRR13349122_quant/quant.sf’ for reading: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004091_quant/quant.sf: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004092_quant/quant.sf: No such file or directory\n",
      "sort: cannot read: data/quants/GSM5004093_quant/quant.sf: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!head data/quants/SRR13349122_quant/quant.sf -n 1\n",
    "!sort -nrk 4,4 data/quants/GSM5004091_quant/quant.sf | head -10\n",
    "!sort -nrk 4,4 data/quants/GSM5004092_quant/quant.sf | head -10\n",
    "!sort -nrk 4,4 data/quants/GSM5004093_quant/quant.sf | head -10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 12: Report the expression of a putative acyl-ACP desaturase (BB28_RS16545) that was downregulated in the double lysogen relative to wild-type\n",
    "A acyl-transferase was reported to be downregulated in the double lysogen as shown in the table of the top 20 upregulated and downregulated genes from the paper describing the study.\n",
    "![RNA-Seq workflow](images/table-cushman.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `grep` to report the expression in the wild-type sample. The fields in the Salmon `quant.sf` file are as follows. The level of expression is reported in the Transcripts Per Million (`TPM`) and number of reads (`NumReads`) fields:  \n",
    "`Name    Length  EffectiveLength TPM     NumReads`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "grep: data/quants/GSM5004088_quant/quant.sf: No such file or directory\n",
      "grep: data/quants/GSM5004089_quant/quant.sf: No such file or directory\n",
      "grep: data/quants/GSM5004090_quant/quant.sf: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!grep 'BB28_RS16545' data/quants/GSM5004088_quant/quant.sf\n",
    "!grep 'BB28_RS16545' data/quants/GSM5004089_quant/quant.sf\n",
    "!grep 'BB28_RS16545' data/quants/GSM5004090_quant/quant.sf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `grep` to report the expression in the double lysogen sample. The fields in the Salmon `quant.sf` file are as follows. The level of expression is reported in the Transcripts Per Million (`TPM`) and number of reads (`NumReads`) fields:  \n",
    "`Name    Length  EffectiveLength TPM     NumReads`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "grep: data/quants/GSM5004091_quant/quant.sf: No such file or directory\n",
      "grep: data/quants/GSM5004092_quant/quant.sf: No such file or directory\n",
      "grep: data/quants/GSM5004093_quant/quant.sf: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!grep 'BB28_RS16545' data/quants/GSM5004091_quant/quant.sf\n",
    "!grep 'BB28_RS16545' data/quants/GSM5004092_quant/quant.sf\n",
    "!grep 'BB28_RS16545' data/quants/GSM5004093_quant/quant.sf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 12: Combine Genecounts to a Single Genecount File\n",
    "Commonly, the readcounts for each sample are combined into a single table, where the rows contain the gene ID, and the columns identify the sample.\n",
    "\n",
    "As before, this can be done in many ways. The quantmerge function outputs a table.\n",
    "\n",
    "However, it is common for readcount tables to have sample headers for the columns, which this does not have.\n",
    "\n",
    "So you could manually add those headers in different ways, for instance using a spreadsheet editor, or using a shell command like sed as below to insert a line at the start of table with the relevant tab-seperated headers.\n",
    "\n",
    "You could also use sed to remove 'gene' or 'rna' prefixes from the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Version Server Response: Not Found\n",
      "\u001b[00m[2024-09-18 06:15:56.885] [mergeLog] [info] samples: [ data/quants/GSM5004088_quant, data/quants/GSM5004089_quant, data/quants/GSM5004090_quant, data/quants/GSM5004091_quant, data/quants/GSM5004092_quant, data/quants/GSM5004093_quant ]\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:15:56.885] [mergeLog] [info] sample names : [ GSM5004088_quant, GSM5004089_quant, GSM5004090_quant, GSM5004091_quant, GSM5004092_quant, GSM5004093_quant ]\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:15:56.885] [mergeLog] [info] output column : NUMREADS\n",
      "\u001b[00m\u001b[00m[2024-09-18 06:15:56.885] [mergeLog] [info] output file : data/quants/merged_quants.txt\n",
      "\u001b[00m\u001b[1m\u001b[41m[2024-09-18 06:15:56.885] [mergeLog] [critical] The sample directory data/quants/GSM5004088_quant either doesn't exist, or doesn't contain a quant.sf file\n",
      "\u001b[00msed: can't read data/quants/merged_quants.txt: No such file or directory\n",
      "sed: can't read data/quants/merged_quants.txt: No such file or directory\n",
      "sed: can't read data/quants/merged_quants.txt: No such file or directory\n",
      "An example of a combined genecount outputfile.\n",
      "head: cannot open ‘data/quants/merged_quants.txt’ for reading: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "##first merge salmon files by number of reads.\n",
    "!salmon quantmerge --column numreads --quants data/quants/*_quant -o data/quants/merged_quants.txt\n",
    "##optinally we can rename the columns\n",
    "!sed -i \"1s/.*/Name\\tGSM5004088\\tGSM5004089\\tGSM5004090\\tGSM5004091\\tGSM5004092\\tGSM5004093/\" data/quants/merged_quants.txt\n",
    "\n",
    "##for further formatting, it may be easier in our r-code to later merge\n",
    "##if we remove the gene- and rna- prefix\n",
    "!sed -i \"s/gene-//\" data/quants/merged_quants.txt\n",
    "!sed -i \"s/rna-//\" data/quants/merged_quants.txt\n",
    "\n",
    "print(\"An example of a combined genecount outputfile.\")\n",
    "!head data/quants/merged_quants.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <a name=\"workflow\">Additional Workflows</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you have read counts per gene, feel free to explore the R workflow which creates plots and analyses using these readcount files, or try other alternate workflows for creating read count files, such as using snakemake."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "[Workflow One:](Tutorial_1.ipynb) A short introduction to downloading and mapping sequences to a transcriptome using Trimmomatic and Salmon. Here is a link to the YouTube video demonstrating the tutorial: <https://youtu.be/ChGfBR4do_Y>.\n",
    "\n",
    "[Workflow One (Extended):](Tutorial_1B_Extended.ipynb) An extended version of workflow one. Once you have got your feet wet, you can retry workflow one with this extended version that covers the entire dataset, and includes elaboration such as using SRA tools for sequence downloading, and examples of running batches of fastq files through the pipeline. This workflow may take around an hour to run.\n",
    "\n",
    "[Workflow One (Using Snakemake):](Tutorial_2_Snakemake.ipynb) Using snakemake to run workflow one.\n",
    "\n",
    "[Workflow Two (DEG Analysis):](Tutorial_3_DEG_Analysis.ipynb) Using Deseq2 and R to conduct clustering and differential gene expression analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![RNA-Seq workflow](images/RNA-Seq_Notebook_Homepage.png)"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": ".m112",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/:m112"
  },
  "kernelspec": {
   "display_name": "conda_tensorflow2_p310",
   "language": "python",
   "name": "conda_tensorflow2_p310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
