{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1972b325-939a-4251-840d-329c61b9121e",
   "metadata": {},
   "source": [
    "# RNA-Seq Analysis Training Demo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18958075-0123-4779-bd9a-c4a354e578fd",
   "metadata": {},
   "source": [
    "## Overview:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd885295-4fb7-4f5a-a22f-9c1033348c6e",
   "metadata": {},
   "source": [
    "This code provides a comprehensive framework for analyzing RNA-Seq data to identify differentially expressed genes and investigate potential regulatory mechanisms. \n",
    "\n",
    "The code will analyze read count data, followed by differential expression analysis utilizing the DESeq2 and edgeR packages to pinpoint genes with statistically significant expression changes between experimental groups. \n",
    "\n",
    "Additionally, the code explores the regulatory landscape by identifying potential transcription factors (TFs) involved in modulating these differentially expressed genes using the NetAct package. \n",
    "\n",
    "It further estimates TF activity levels and constructs networks of interactions among these TFs. \n",
    "\n",
    "Overall, this demo serves to illustrates the essential steps in RNA-Seq analysis, emphasizing the identification of differential expression and the exploration of TF regulatory networks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "316e56a9-e76a-4720-855a-e3bea0c7dca0",
   "metadata": {},
   "source": [
    "## STEP 1: Install Packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "affc41bf-db8f-4319-997d-d1a6632ca7e1",
   "metadata": {},
   "source": [
    "### STEP 1.1 Install Mambaforge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5e417e7-0d00-47e7-8206-b8059c0ed458",
   "metadata": {},
   "source": [
    "First install Miniforge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2b4fea91-148a-44d9-a502-25ff91691a20",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [],
      "text/latex": [],
      "text/markdown": [],
      "text/plain": [
       "character(0)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'PREFIX=/home/ec2-user/miniforge'</li><li>''</li><li>'Transaction'</li><li>''</li><li><span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/miniforge/envs/_virtual_specs_checks'</span></li><li>''</li><li><span style=white-space:pre-wrap>'  All requested packages already installed'</span></li><li>''</li><li>'Dry run. Not executing the transaction.'</li><li>'Unpacking payload ...'</li><li>'Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2'</li><li>'Extracting ca-certificates-2024.8.30-hbcca054_0.conda'</li><li>'Extracting ld_impl_linux-64-2.40-hf3520f5_7.conda'</li><li>'Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2'</li><li>'Extracting python_abi-3.12-5_cp312.conda'</li><li>'Extracting tzdata-2024a-h8827d51_1.conda'</li><li>'Extracting libgomp-14.1.0-h77fa898_1.conda'</li><li>'Extracting _openmp_mutex-4.5-2_gnu.tar.bz2'</li><li>'Extracting libgcc-14.1.0-h77fa898_1.conda'</li><li>'Extracting libexpat-2.6.3-h5888daf_0.conda'</li><li>'Extracting libgcc-ng-14.1.0-h69a702a_1.conda'</li><li>'Extracting libstdcxx-14.1.0-hc0a3c3a_1.conda'</li><li>'Extracting openssl-3.3.2-hb9d3cd8_0.conda'</li><li>'Extracting bzip2-1.0.8-h4bc722e_7.conda'</li><li>'Extracting c-ares-1.32.3-h4bc722e_0.conda'</li><li>'Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2'</li><li>'Extracting libev-4.33-hd590300_2.conda'</li><li>'Extracting libffi-3.4.2-h7f98852_5.tar.bz2'</li><li>'Extracting libiconv-1.17-hd590300_2.conda'</li><li>'Extracting libnsl-2.0.1-hd590300_0.conda'</li><li>'Extracting libstdcxx-ng-14.1.0-h4852527_1.conda'</li><li>'Extracting libuuid-2.38.1-h0b41bf4_0.conda'</li><li>'Extracting libxcrypt-4.4.36-hd590300_1.conda'</li><li>'Extracting libzlib-1.3.1-h4ab18f5_1.conda'</li><li>'Extracting lzo-2.10-hd590300_1001.conda'</li><li>'Extracting ncurses-6.5-he02047a_1.conda'</li><li>'Extracting reproc-14.2.4.post0-hd590300_1.conda'</li><li>'Extracting xz-5.2.6-h166bdaf_0.tar.bz2'</li><li>'Extracting fmt-10.2.1-h00ab1b0_0.conda'</li><li>'Extracting icu-75.1-he02047a_0.conda'</li><li>'Extracting libedit-3.1.20191231-he28a2e2_2.tar.bz2'</li><li>'Extracting libnghttp2-1.58.0-h47da74e_1.conda'</li><li>'Extracting libsolv-0.7.30-h3509ff9_0.conda'</li><li>'Extracting libsqlite-3.46.1-hadc24fc_0.conda'</li><li>'Extracting libssh2-1.11.0-h0841786_0.conda'</li><li>'Extracting lz4-c-1.9.4-hcb278e6_0.conda'</li><li>'Extracting readline-8.2-h8228510_1.conda'</li><li>'Extracting reproc-cpp-14.2.4.post0-h59595ed_1.conda'</li><li>'Extracting tk-8.6.13-noxft_h4845f30_101.conda'</li><li>'Extracting yaml-cpp-0.8.0-h59595ed_0.conda'</li><li>'Extracting zstd-1.5.6-ha6fb4c9_0.conda'</li><li>'Extracting krb5-1.21.3-h659f571_0.conda'</li><li>'Extracting libxml2-2.12.7-he7c6b58_4.conda'</li><li>'Extracting python-3.12.6-hc5c86c4_1_cpython.conda'</li><li>'Extracting libarchive-3.7.4-hfca40fe_0.conda'</li><li>'Extracting libcurl-8.10.1-hbbe4b11_0.conda'</li><li>'Extracting menuinst-2.1.2-py312h7900ff3_1.conda'</li><li>'Extracting archspec-0.2.3-pyhd8ed1ab_0.conda'</li><li>'Extracting boltons-24.0.0-pyhd8ed1ab_0.conda'</li><li>'Extracting brotli-python-1.1.0-py312h2ec8cdc_2.conda'</li><li>'Extracting certifi-2024.8.30-pyhd8ed1ab_0.conda'</li><li>'Extracting charset-normalizer-3.3.2-pyhd8ed1ab_0.conda'</li><li>'Extracting colorama-0.4.6-pyhd8ed1ab_0.tar.bz2'</li><li>'Extracting distro-1.9.0-pyhd8ed1ab_0.conda'</li><li>'Extracting frozendict-2.4.4-py312h66e93f0_1.conda'</li><li>'Extracting hpack-4.0.0-pyh9f0ad1d_0.tar.bz2'</li><li>'Extracting hyperframe-6.0.1-pyhd8ed1ab_0.tar.bz2'</li><li>'Extracting idna-3.10-pyhd8ed1ab_0.conda'</li><li>'Extracting jsonpointer-3.0.0-py312h7900ff3_1.conda'</li><li>'Extracting libmamba-1.5.9-h4cc3d14_0.conda'</li><li>'Extracting packaging-24.1-pyhd8ed1ab_0.conda'</li><li>'Extracting platformdirs-4.3.6-pyhd8ed1ab_0.conda'</li><li>'Extracting pluggy-1.5.0-pyhd8ed1ab_0.conda'</li><li>'Extracting pycosat-0.6.6-py312h98912ed_0.conda'</li><li>'Extracting pycparser-2.22-pyhd8ed1ab_0.conda'</li><li>'Extracting pysocks-1.7.1-pyha2e5f31_6.tar.bz2'</li><li>'Extracting ruamel.yaml.clib-0.2.8-py312h98912ed_0.conda'</li><li>'Extracting setuptools-74.1.2-pyhd8ed1ab_0.conda'</li><li>'Extracting truststore-0.9.2-pyhd8ed1ab_0.conda'</li><li>'Extracting wheel-0.44.0-pyhd8ed1ab_0.conda'</li><li>'Extracting cffi-1.17.1-py312h06ac9bb_0.conda'</li><li>'Extracting h2-4.1.0-pyhd8ed1ab_0.tar.bz2'</li><li>'Extracting jsonpatch-1.33-pyhd8ed1ab_0.conda'</li><li>'Extracting libmambapy-1.5.9-py312h7fb9e8e_0.conda'</li><li>'Extracting pip-24.2-pyh8b19718_1.conda'</li><li>'Extracting ruamel.yaml-0.18.6-py312h98912ed_0.conda'</li><li>'Extracting tqdm-4.66.5-pyhd8ed1ab_0.conda'</li><li>'Extracting zstandard-0.23.0-py312hef9b889_1.conda'</li><li>'Extracting conda-package-streaming-0.10.0-pyhd8ed1ab_0.conda'</li><li>'Extracting urllib3-2.2.3-pyhd8ed1ab_0.conda'</li><li>'Extracting conda-package-handling-2.3.0-pyh7900ff3_0.conda'</li><li>'Extracting requests-2.32.3-pyhd8ed1ab_0.conda'</li><li>'Extracting conda-24.7.1-py312h7900ff3_0.conda'</li><li>'Extracting conda-libmamba-solver-24.7.0-pyhd8ed1ab_0.conda'</li><li>'Extracting mamba-1.5.9-py312h9460a1c_0.conda'</li><li>''</li><li>'Installing base environment...'</li><li>''</li><li>'Transaction'</li><li>''</li><li><span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/miniforge'</span></li><li>''</li><li><span style=white-space:pre-wrap>'  Updating specs:'</span></li><li>''</li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::_libgcc_mutex==0.1=conda_forge[md5=d7c89558ba9fa0495403155b64376d81]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::ca-certificates==2024.8.30=hbcca054_0[md5=c27d1c142233b5bc9ca570c6e2e0c244]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::ld_impl_linux-64==2.40=hf3520f5_7[md5=b80f2f396ca2c28b8c14c437a4ed1e74]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab_3[md5=878f923dd6acc8aeb47a75da6c4098be]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::python_abi==3.12=5_cp312[md5=0424ae29b104430108f5218a66db7260]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::tzdata==2024a=h8827d51_1[md5=8bfdead4e0fff0383ae4c9c50d0531bd]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgomp==14.1.0=h77fa898_1[md5=23c255b008c4f2ae008f81edcabaca89]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::_openmp_mutex==4.5=2_gnu[md5=73aaf86a425cc6e73fcf236a5a46396d]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgcc==14.1.0=h77fa898_1[md5=002ef4463dd1e2b44a94a4ace468f5d2]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libexpat==2.6.3=h5888daf_0[md5=59f4c43bb1b5ef1c71946ff2cbf59524]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgcc-ng==14.1.0=h69a702a_1[md5=1efc0ad219877a73ef977af7dbb51f17]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libstdcxx==14.1.0=hc0a3c3a_1[md5=9dbb9699ea467983ba8a4ba89b08b066]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::openssl==3.3.2=hb9d3cd8_0[md5=4d638782050ab6faa27275bed57e9b4e]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::bzip2==1.0.8=h4bc722e_7[md5=62ee74e96c5ebb0af99386de58cf9553]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::c-ares==1.32.3=h4bc722e_0[md5=7624e34ee6baebfc80d67bac76cc9d9d]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf_0[md5=30186d27e2c9fa62b45fb1476b7200e3]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libev==4.33=hd590300_2[md5=172bf1cd1ff8629f2b1179945ed45055]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libffi==3.4.2=h7f98852_5[md5=d645c6d2ac96843a2bfaccd2d62b3ac3]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libiconv==1.17=hd590300_2[md5=d66573916ffcf376178462f1b61c941e]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libnsl==2.0.1=hd590300_0[md5=30fd6e37fe21f86f4bd26d6ee73eeec7]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libstdcxx-ng==14.1.0=h4852527_1[md5=bd2598399a70bb86d8218e95548d735e]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4_0[md5=40b61aab5c7ba9ff276c41cfffe6b80b]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300_1[md5=5aa797f8787fe7a17d1b0821485b5adc]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libzlib==1.3.1=h4ab18f5_1[md5=57d7dc60e9325e3de37ff8dffd18e814]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::lzo==2.10=hd590300_1001[md5=ec7398d21e2651e0dcb0044d03b9a339]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::ncurses==6.5=he02047a_1[md5=70caf8bb6cf39a0b6b7efc885f51c0fe]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300_1[md5=82ca53502dfd5a64a80dee76dae14685]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::xz==5.2.6=h166bdaf_0[md5=2161070d867d1b1204ea749c8eec4ef0]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0_0[md5=35ef8bc24bd34074ebae3c943d551728]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::icu==75.1=he02047a_0[md5=8b189310083baabfb622af68fd9d3ae3]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2_2[md5=4d331e44109e3f0e19b4cb8f9b82f3e1]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e_1[md5=700ac6ea6d53d5510591c4344d5c989a]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libsolv==0.7.30=h3509ff9_0[md5=02539b77d25aa4f65b20246549e256c3]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libsqlite==3.46.1=hadc24fc_0[md5=36f79405ab16bf271edb55b213836dac]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libssh2==1.11.0=h0841786_0[md5=1f5a58e686b13bcfde88b93f547d23fe]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6_0[md5=318b08df404f9c9be5712aaa5a6f0bb0]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::readline==8.2=h8228510_1[md5=47d31b792659ce70f470b5c82fdfb7a4]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed_1[md5=715e1d720ec1a03715bebd237972fca5]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::tk==8.6.13=noxft_h4845f30_101[md5=d453b98d9c83e71da0741bb0ff4d76bc]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed_0[md5=965eaacd7c18eb8361fd12bb9e7a57d7]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::zstd==1.5.6=ha6fb4c9_0[md5=4d056880988120e29d75bfff282e0f45]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::krb5==1.21.3=h659f571_0[md5=3f43953b7d3fb3aaa1d0d0723d91e368]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libxml2==2.12.7=he7c6b58_4[md5=08a9265c637230c37cb1be4a6cad4536]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::python==3.12.6=hc5c86c4_1_cpython[md5=00836baacdca254f28c54d2543e97514]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libarchive==3.7.4=hfca40fe_0[md5=32ddb97f897740641d8d46a829ce1704]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libcurl==8.10.1=hbbe4b11_0[md5=6e801c50a40301f6978c53976917b277]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::menuinst==2.1.2=py312h7900ff3_1[md5=c6575ae996f2bc0369c73b632db5ca61]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab_0[md5=192278292e20704f663b9c766909d67b]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab_0[md5=61de176bd62041f9cd5bd4fcd09eb0ff]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::brotli-python==1.1.0=py312h2ec8cdc_2[md5=b0b867af6fc74b2a0aa206da29c0f3cf]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::certifi==2024.8.30=pyhd8ed1ab_0[md5=12f7d00853807b0531775e9be891cb11]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab_0[md5=7f4a9e3fcff3f6356ae99244a014da6a]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab_0[md5=3faab06a954c2a04039983f2c4a50d99]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab_0[md5=bbdb409974cd6cb30071b1d978302726]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::frozendict==2.4.4=py312h66e93f0_1[md5=82a14d06575c4023a5df6b0798acff71]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::hpack==4.0.0=pyh9f0ad1d_0[md5=914d6646c4dbb1fd3ff539830a12fd71]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::hyperframe==6.0.1=pyhd8ed1ab_0[md5=9f765cbfab6870c8435b9eefecd7a1f4]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::idna==3.10=pyhd8ed1ab_0[md5=7ba2ede0e7c795ff95088daf0dc59753]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::jsonpointer==3.0.0=py312h7900ff3_1[md5=6b51f7459ea4073eeb5057207e2e1e3d]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libmamba==1.5.9=h4cc3d14_0[md5=896cece5b883ad86e9dd88b1f4d23c99]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::packaging==24.1=pyhd8ed1ab_0[md5=cbe1bb1f21567018ce595d9c2be0f0db]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::platformdirs==4.3.6=pyhd8ed1ab_0[md5=fd8f2b18b65bbf62e8f653100690c8d2]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::pluggy==1.5.0=pyhd8ed1ab_0[md5=d3483c8fc2dc2cc3f5cf43e26d60cabf]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::pycosat==0.6.6=py312h98912ed_0[md5=8f1c372e7b843167be885dc8229931c1]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab_0[md5=844d9eb3b43095b031874477f7d70088]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31_6[md5=2a7de29fb590ca14b5243c4c812c8025]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py312h98912ed_0[md5=05f31c2a79ba61df8d6d903ce4a4ce7b]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::setuptools==74.1.2=pyhd8ed1ab_0[md5=56c9c11d004428e81d02eeb730fc6336]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::truststore==0.9.2=pyhd8ed1ab_0[md5=f14e46d1bf271e748ff556d8b872e28a]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::wheel==0.44.0=pyhd8ed1ab_0[md5=d44e3b085abcaef02983c6305b84b584]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::cffi==1.17.1=py312h06ac9bb_0[md5=a861504bbea4161a9170b85d4d2be840]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::h2==4.1.0=pyhd8ed1ab_0[md5=b748fbf7060927a6e82df7cb5ee8f097]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab_0[md5=bfdb7c5c6ad1077c82a69a8642c87aff]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::libmambapy==1.5.9=py312h7fb9e8e_0[md5=ccaeeb6e3caaf0c744480393791aa366]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::pip==24.2=pyh8b19718_1[md5=6c78fbb8ddfd64bcb55b5cbafd2d2c43]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::ruamel.yaml==0.18.6=py312h98912ed_0[md5=a99a06a875138829ef65f44bbe2c30ca]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::tqdm==4.66.5=pyhd8ed1ab_0[md5=c6e94fc2b2ec71ea33fe7c7da259acb4]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::zstandard==0.23.0=py312hef9b889_1[md5=8b7069e9792ee4e5b4919a7a306d2e67]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-package-streaming==0.10.0=pyhd8ed1ab_0[md5=3480386e00995f7a1dfb3b9aa2fe70fd]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::urllib3==2.2.3=pyhd8ed1ab_0[md5=6b55867f385dd762ed99ea687af32a69]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-package-handling==2.3.0=pyh7900ff3_0[md5=0a7dce281ae2be81acab0aa963e6bb99]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::requests==2.32.3=pyhd8ed1ab_0[md5=5ede4753180c7a550a443c430dc8ab52]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::conda==24.7.1=py312h7900ff3_0[md5=e1bf59014e88eaff036101358f63a496]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-libmamba-solver==24.7.0=pyhd8ed1ab_0[md5=857c9e25f0a77c0bd7eb622d46d9418f]'</span></li><li><span style=white-space:pre-wrap>'   - conda-forge/linux-64::mamba==1.5.9=py312h9460a1c_0[md5=a8525c8a1647b4f5967fa6b552722851]'</span></li><li>''</li><li>''</li><li><span style=white-space:pre-wrap>'  Package                         Version  Build               Channel         Size'</span></li><li>'─────────────────────────────────────────────────────────────────────────────────────'</li><li><span style=white-space:pre-wrap>'  Install:'</span></li><li>'─────────────────────────────────────────────────────────────────────────────────────'</li><li>''</li><li><span style=white-space:pre-wrap>'  + _libgcc_mutex                     0.1  conda_forge         conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + ca-certificates             2024.8.30  hbcca054_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + ld_impl_linux-64                 2.40  hf3520f5_7          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pybind11-abi                        4  hd8ed1ab_3          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + python_abi                       3.12  5_cp312             conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + tzdata                          2024a  h8827d51_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libgomp                        14.1.0  h77fa898_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + _openmp_mutex                     4.5  2_gnu               conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libgcc                         14.1.0  h77fa898_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libexpat                        2.6.3  h5888daf_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libgcc-ng                      14.1.0  h69a702a_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libstdcxx                      14.1.0  hc0a3c3a_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + openssl                         3.3.2  hb9d3cd8_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + bzip2                           1.0.8  h4bc722e_7          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + c-ares                         1.32.3  h4bc722e_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + keyutils                        1.6.1  h166bdaf_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libev                            4.33  hd590300_2          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libffi                          3.4.2  h7f98852_5          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libiconv                         1.17  hd590300_2          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libnsl                          2.0.1  hd590300_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libstdcxx-ng                   14.1.0  h4852527_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libuuid                        2.38.1  h0b41bf4_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libxcrypt                      4.4.36  hd590300_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libzlib                         1.3.1  h4ab18f5_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + lzo                              2.10  hd590300_1001       conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + ncurses                           6.5  he02047a_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + reproc                   14.2.4.post0  hd590300_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + xz                              5.2.6  h166bdaf_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + fmt                            10.2.1  h00ab1b0_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + icu                              75.1  he02047a_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libedit                  3.1.20191231  he28a2e2_2          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libnghttp2                     1.58.0  h47da74e_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libsolv                        0.7.30  h3509ff9_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libsqlite                      3.46.1  hadc24fc_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libssh2                        1.11.0  h0841786_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + lz4-c                           1.9.4  hcb278e6_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + readline                          8.2  h8228510_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + reproc-cpp               14.2.4.post0  h59595ed_1          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + tk                             8.6.13  noxft_h4845f30_101  conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + yaml-cpp                        0.8.0  h59595ed_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + zstd                            1.5.6  ha6fb4c9_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + krb5                           1.21.3  h659f571_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libxml2                        2.12.7  he7c6b58_4          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + python                         3.12.6  hc5c86c4_1_cpython  conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libarchive                      3.7.4  hfca40fe_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libcurl                        8.10.1  hbbe4b11_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + menuinst                        2.1.2  py312h7900ff3_1     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + archspec                        0.2.3  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + boltons                        24.0.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + brotli-python                   1.1.0  py312h2ec8cdc_2     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + certifi                     2024.8.30  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + charset-normalizer              3.3.2  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + colorama                        0.4.6  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + distro                          1.9.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + frozendict                      2.4.4  py312h66e93f0_1     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + hpack                           4.0.0  pyh9f0ad1d_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + hyperframe                      6.0.1  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + idna                             3.10  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + jsonpointer                     3.0.0  py312h7900ff3_1     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libmamba                        1.5.9  h4cc3d14_0          conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + packaging                        24.1  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + platformdirs                    4.3.6  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pluggy                          1.5.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pycosat                         0.6.6  py312h98912ed_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pycparser                        2.22  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pysocks                         1.7.1  pyha2e5f31_6        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + ruamel.yaml.clib                0.2.8  py312h98912ed_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + setuptools                     74.1.2  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + truststore                      0.9.2  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + wheel                          0.44.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + cffi                           1.17.1  py312h06ac9bb_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + h2                              4.1.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + jsonpatch                        1.33  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + libmambapy                      1.5.9  py312h7fb9e8e_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + pip                              24.2  pyh8b19718_1        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + ruamel.yaml                    0.18.6  py312h98912ed_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + tqdm                           4.66.5  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + zstandard                      0.23.0  py312hef9b889_1     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + conda-package-streaming        0.10.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + urllib3                         2.2.3  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + conda-package-handling          2.3.0  pyh7900ff3_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + requests                       2.32.3  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + conda                          24.7.1  py312h7900ff3_0     conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + conda-libmamba-solver          24.7.0  pyhd8ed1ab_0        conda-forge         '</span></li><li><span style=white-space:pre-wrap>'  + mamba                           1.5.9  py312h9460a1c_0     conda-forge         '</span></li><li>''</li><li><span style=white-space:pre-wrap>'  Summary:'</span></li><li>''</li><li><span style=white-space:pre-wrap>'  Install: 85 packages'</span></li><li>''</li><li><span style=white-space:pre-wrap>'  Total download: 0 B'</span></li><li>''</li><li>'─────────────────────────────────────────────────────────────────────────────────────'</li><li>''</li><li>''</li><li>''</li><li>'Transaction starting'</li><li>''</li><li>'Transaction finished'</li><li>''</li><li>'To activate this environment, use:'</li><li>''</li><li><span style=white-space:pre-wrap>'    micromamba activate /home/ec2-user/miniforge'</span></li><li>''</li><li>'Or to execute a single command in this environment, use:'</li><li>''</li><li><span style=white-space:pre-wrap>'    micromamba run -p /home/ec2-user/miniforge mycommand'</span></li><li>''</li><li>'installation finished.'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'PREFIX=/home/ec2-user/miniforge'\n",
       "\\item ''\n",
       "\\item 'Transaction'\n",
       "\\item ''\n",
       "\\item '  Prefix: /home/ec2-user/miniforge/envs/\\_virtual\\_specs\\_checks'\n",
       "\\item ''\n",
       "\\item '  All requested packages already installed'\n",
       "\\item ''\n",
       "\\item 'Dry run. Not executing the transaction.'\n",
       "\\item 'Unpacking payload ...'\n",
       "\\item 'Extracting \\_libgcc\\_mutex-0.1-conda\\_forge.tar.bz2'\n",
       "\\item 'Extracting ca-certificates-2024.8.30-hbcca054\\_0.conda'\n",
       "\\item 'Extracting ld\\_impl\\_linux-64-2.40-hf3520f5\\_7.conda'\n",
       "\\item 'Extracting pybind11-abi-4-hd8ed1ab\\_3.tar.bz2'\n",
       "\\item 'Extracting python\\_abi-3.12-5\\_cp312.conda'\n",
       "\\item 'Extracting tzdata-2024a-h8827d51\\_1.conda'\n",
       "\\item 'Extracting libgomp-14.1.0-h77fa898\\_1.conda'\n",
       "\\item 'Extracting \\_openmp\\_mutex-4.5-2\\_gnu.tar.bz2'\n",
       "\\item 'Extracting libgcc-14.1.0-h77fa898\\_1.conda'\n",
       "\\item 'Extracting libexpat-2.6.3-h5888daf\\_0.conda'\n",
       "\\item 'Extracting libgcc-ng-14.1.0-h69a702a\\_1.conda'\n",
       "\\item 'Extracting libstdcxx-14.1.0-hc0a3c3a\\_1.conda'\n",
       "\\item 'Extracting openssl-3.3.2-hb9d3cd8\\_0.conda'\n",
       "\\item 'Extracting bzip2-1.0.8-h4bc722e\\_7.conda'\n",
       "\\item 'Extracting c-ares-1.32.3-h4bc722e\\_0.conda'\n",
       "\\item 'Extracting keyutils-1.6.1-h166bdaf\\_0.tar.bz2'\n",
       "\\item 'Extracting libev-4.33-hd590300\\_2.conda'\n",
       "\\item 'Extracting libffi-3.4.2-h7f98852\\_5.tar.bz2'\n",
       "\\item 'Extracting libiconv-1.17-hd590300\\_2.conda'\n",
       "\\item 'Extracting libnsl-2.0.1-hd590300\\_0.conda'\n",
       "\\item 'Extracting libstdcxx-ng-14.1.0-h4852527\\_1.conda'\n",
       "\\item 'Extracting libuuid-2.38.1-h0b41bf4\\_0.conda'\n",
       "\\item 'Extracting libxcrypt-4.4.36-hd590300\\_1.conda'\n",
       "\\item 'Extracting libzlib-1.3.1-h4ab18f5\\_1.conda'\n",
       "\\item 'Extracting lzo-2.10-hd590300\\_1001.conda'\n",
       "\\item 'Extracting ncurses-6.5-he02047a\\_1.conda'\n",
       "\\item 'Extracting reproc-14.2.4.post0-hd590300\\_1.conda'\n",
       "\\item 'Extracting xz-5.2.6-h166bdaf\\_0.tar.bz2'\n",
       "\\item 'Extracting fmt-10.2.1-h00ab1b0\\_0.conda'\n",
       "\\item 'Extracting icu-75.1-he02047a\\_0.conda'\n",
       "\\item 'Extracting libedit-3.1.20191231-he28a2e2\\_2.tar.bz2'\n",
       "\\item 'Extracting libnghttp2-1.58.0-h47da74e\\_1.conda'\n",
       "\\item 'Extracting libsolv-0.7.30-h3509ff9\\_0.conda'\n",
       "\\item 'Extracting libsqlite-3.46.1-hadc24fc\\_0.conda'\n",
       "\\item 'Extracting libssh2-1.11.0-h0841786\\_0.conda'\n",
       "\\item 'Extracting lz4-c-1.9.4-hcb278e6\\_0.conda'\n",
       "\\item 'Extracting readline-8.2-h8228510\\_1.conda'\n",
       "\\item 'Extracting reproc-cpp-14.2.4.post0-h59595ed\\_1.conda'\n",
       "\\item 'Extracting tk-8.6.13-noxft\\_h4845f30\\_101.conda'\n",
       "\\item 'Extracting yaml-cpp-0.8.0-h59595ed\\_0.conda'\n",
       "\\item 'Extracting zstd-1.5.6-ha6fb4c9\\_0.conda'\n",
       "\\item 'Extracting krb5-1.21.3-h659f571\\_0.conda'\n",
       "\\item 'Extracting libxml2-2.12.7-he7c6b58\\_4.conda'\n",
       "\\item 'Extracting python-3.12.6-hc5c86c4\\_1\\_cpython.conda'\n",
       "\\item 'Extracting libarchive-3.7.4-hfca40fe\\_0.conda'\n",
       "\\item 'Extracting libcurl-8.10.1-hbbe4b11\\_0.conda'\n",
       "\\item 'Extracting menuinst-2.1.2-py312h7900ff3\\_1.conda'\n",
       "\\item 'Extracting archspec-0.2.3-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting boltons-24.0.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting brotli-python-1.1.0-py312h2ec8cdc\\_2.conda'\n",
       "\\item 'Extracting certifi-2024.8.30-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting charset-normalizer-3.3.2-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting colorama-0.4.6-pyhd8ed1ab\\_0.tar.bz2'\n",
       "\\item 'Extracting distro-1.9.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting frozendict-2.4.4-py312h66e93f0\\_1.conda'\n",
       "\\item 'Extracting hpack-4.0.0-pyh9f0ad1d\\_0.tar.bz2'\n",
       "\\item 'Extracting hyperframe-6.0.1-pyhd8ed1ab\\_0.tar.bz2'\n",
       "\\item 'Extracting idna-3.10-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting jsonpointer-3.0.0-py312h7900ff3\\_1.conda'\n",
       "\\item 'Extracting libmamba-1.5.9-h4cc3d14\\_0.conda'\n",
       "\\item 'Extracting packaging-24.1-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting platformdirs-4.3.6-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting pluggy-1.5.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting pycosat-0.6.6-py312h98912ed\\_0.conda'\n",
       "\\item 'Extracting pycparser-2.22-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting pysocks-1.7.1-pyha2e5f31\\_6.tar.bz2'\n",
       "\\item 'Extracting ruamel.yaml.clib-0.2.8-py312h98912ed\\_0.conda'\n",
       "\\item 'Extracting setuptools-74.1.2-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting truststore-0.9.2-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting wheel-0.44.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting cffi-1.17.1-py312h06ac9bb\\_0.conda'\n",
       "\\item 'Extracting h2-4.1.0-pyhd8ed1ab\\_0.tar.bz2'\n",
       "\\item 'Extracting jsonpatch-1.33-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting libmambapy-1.5.9-py312h7fb9e8e\\_0.conda'\n",
       "\\item 'Extracting pip-24.2-pyh8b19718\\_1.conda'\n",
       "\\item 'Extracting ruamel.yaml-0.18.6-py312h98912ed\\_0.conda'\n",
       "\\item 'Extracting tqdm-4.66.5-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting zstandard-0.23.0-py312hef9b889\\_1.conda'\n",
       "\\item 'Extracting conda-package-streaming-0.10.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting urllib3-2.2.3-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting conda-package-handling-2.3.0-pyh7900ff3\\_0.conda'\n",
       "\\item 'Extracting requests-2.32.3-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting conda-24.7.1-py312h7900ff3\\_0.conda'\n",
       "\\item 'Extracting conda-libmamba-solver-24.7.0-pyhd8ed1ab\\_0.conda'\n",
       "\\item 'Extracting mamba-1.5.9-py312h9460a1c\\_0.conda'\n",
       "\\item ''\n",
       "\\item 'Installing base environment...'\n",
       "\\item ''\n",
       "\\item 'Transaction'\n",
       "\\item ''\n",
       "\\item '  Prefix: /home/ec2-user/miniforge'\n",
       "\\item ''\n",
       "\\item '  Updating specs:'\n",
       "\\item ''\n",
       "\\item '   - conda-forge/linux-64::\\_libgcc\\_mutex==0.1=conda\\_forge{[}md5=d7c89558ba9fa0495403155b64376d81{]}'\n",
       "\\item '   - conda-forge/linux-64::ca-certificates==2024.8.30=hbcca054\\_0{[}md5=c27d1c142233b5bc9ca570c6e2e0c244{]}'\n",
       "\\item '   - conda-forge/linux-64::ld\\_impl\\_linux-64==2.40=hf3520f5\\_7{[}md5=b80f2f396ca2c28b8c14c437a4ed1e74{]}'\n",
       "\\item '   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab\\_3{[}md5=878f923dd6acc8aeb47a75da6c4098be{]}'\n",
       "\\item '   - conda-forge/linux-64::python\\_abi==3.12=5\\_cp312{[}md5=0424ae29b104430108f5218a66db7260{]}'\n",
       "\\item '   - conda-forge/noarch::tzdata==2024a=h8827d51\\_1{[}md5=8bfdead4e0fff0383ae4c9c50d0531bd{]}'\n",
       "\\item '   - conda-forge/linux-64::libgomp==14.1.0=h77fa898\\_1{[}md5=23c255b008c4f2ae008f81edcabaca89{]}'\n",
       "\\item '   - conda-forge/linux-64::\\_openmp\\_mutex==4.5=2\\_gnu{[}md5=73aaf86a425cc6e73fcf236a5a46396d{]}'\n",
       "\\item '   - conda-forge/linux-64::libgcc==14.1.0=h77fa898\\_1{[}md5=002ef4463dd1e2b44a94a4ace468f5d2{]}'\n",
       "\\item '   - conda-forge/linux-64::libexpat==2.6.3=h5888daf\\_0{[}md5=59f4c43bb1b5ef1c71946ff2cbf59524{]}'\n",
       "\\item '   - conda-forge/linux-64::libgcc-ng==14.1.0=h69a702a\\_1{[}md5=1efc0ad219877a73ef977af7dbb51f17{]}'\n",
       "\\item '   - conda-forge/linux-64::libstdcxx==14.1.0=hc0a3c3a\\_1{[}md5=9dbb9699ea467983ba8a4ba89b08b066{]}'\n",
       "\\item '   - conda-forge/linux-64::openssl==3.3.2=hb9d3cd8\\_0{[}md5=4d638782050ab6faa27275bed57e9b4e{]}'\n",
       "\\item '   - conda-forge/linux-64::bzip2==1.0.8=h4bc722e\\_7{[}md5=62ee74e96c5ebb0af99386de58cf9553{]}'\n",
       "\\item '   - conda-forge/linux-64::c-ares==1.32.3=h4bc722e\\_0{[}md5=7624e34ee6baebfc80d67bac76cc9d9d{]}'\n",
       "\\item '   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf\\_0{[}md5=30186d27e2c9fa62b45fb1476b7200e3{]}'\n",
       "\\item '   - conda-forge/linux-64::libev==4.33=hd590300\\_2{[}md5=172bf1cd1ff8629f2b1179945ed45055{]}'\n",
       "\\item '   - conda-forge/linux-64::libffi==3.4.2=h7f98852\\_5{[}md5=d645c6d2ac96843a2bfaccd2d62b3ac3{]}'\n",
       "\\item '   - conda-forge/linux-64::libiconv==1.17=hd590300\\_2{[}md5=d66573916ffcf376178462f1b61c941e{]}'\n",
       "\\item '   - conda-forge/linux-64::libnsl==2.0.1=hd590300\\_0{[}md5=30fd6e37fe21f86f4bd26d6ee73eeec7{]}'\n",
       "\\item '   - conda-forge/linux-64::libstdcxx-ng==14.1.0=h4852527\\_1{[}md5=bd2598399a70bb86d8218e95548d735e{]}'\n",
       "\\item '   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4\\_0{[}md5=40b61aab5c7ba9ff276c41cfffe6b80b{]}'\n",
       "\\item '   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300\\_1{[}md5=5aa797f8787fe7a17d1b0821485b5adc{]}'\n",
       "\\item '   - conda-forge/linux-64::libzlib==1.3.1=h4ab18f5\\_1{[}md5=57d7dc60e9325e3de37ff8dffd18e814{]}'\n",
       "\\item '   - conda-forge/linux-64::lzo==2.10=hd590300\\_1001{[}md5=ec7398d21e2651e0dcb0044d03b9a339{]}'\n",
       "\\item '   - conda-forge/linux-64::ncurses==6.5=he02047a\\_1{[}md5=70caf8bb6cf39a0b6b7efc885f51c0fe{]}'\n",
       "\\item '   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300\\_1{[}md5=82ca53502dfd5a64a80dee76dae14685{]}'\n",
       "\\item '   - conda-forge/linux-64::xz==5.2.6=h166bdaf\\_0{[}md5=2161070d867d1b1204ea749c8eec4ef0{]}'\n",
       "\\item '   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0\\_0{[}md5=35ef8bc24bd34074ebae3c943d551728{]}'\n",
       "\\item '   - conda-forge/linux-64::icu==75.1=he02047a\\_0{[}md5=8b189310083baabfb622af68fd9d3ae3{]}'\n",
       "\\item '   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2\\_2{[}md5=4d331e44109e3f0e19b4cb8f9b82f3e1{]}'\n",
       "\\item '   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e\\_1{[}md5=700ac6ea6d53d5510591c4344d5c989a{]}'\n",
       "\\item '   - conda-forge/linux-64::libsolv==0.7.30=h3509ff9\\_0{[}md5=02539b77d25aa4f65b20246549e256c3{]}'\n",
       "\\item '   - conda-forge/linux-64::libsqlite==3.46.1=hadc24fc\\_0{[}md5=36f79405ab16bf271edb55b213836dac{]}'\n",
       "\\item '   - conda-forge/linux-64::libssh2==1.11.0=h0841786\\_0{[}md5=1f5a58e686b13bcfde88b93f547d23fe{]}'\n",
       "\\item '   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6\\_0{[}md5=318b08df404f9c9be5712aaa5a6f0bb0{]}'\n",
       "\\item '   - conda-forge/linux-64::readline==8.2=h8228510\\_1{[}md5=47d31b792659ce70f470b5c82fdfb7a4{]}'\n",
       "\\item '   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed\\_1{[}md5=715e1d720ec1a03715bebd237972fca5{]}'\n",
       "\\item '   - conda-forge/linux-64::tk==8.6.13=noxft\\_h4845f30\\_101{[}md5=d453b98d9c83e71da0741bb0ff4d76bc{]}'\n",
       "\\item '   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed\\_0{[}md5=965eaacd7c18eb8361fd12bb9e7a57d7{]}'\n",
       "\\item '   - conda-forge/linux-64::zstd==1.5.6=ha6fb4c9\\_0{[}md5=4d056880988120e29d75bfff282e0f45{]}'\n",
       "\\item '   - conda-forge/linux-64::krb5==1.21.3=h659f571\\_0{[}md5=3f43953b7d3fb3aaa1d0d0723d91e368{]}'\n",
       "\\item '   - conda-forge/linux-64::libxml2==2.12.7=he7c6b58\\_4{[}md5=08a9265c637230c37cb1be4a6cad4536{]}'\n",
       "\\item '   - conda-forge/linux-64::python==3.12.6=hc5c86c4\\_1\\_cpython{[}md5=00836baacdca254f28c54d2543e97514{]}'\n",
       "\\item '   - conda-forge/linux-64::libarchive==3.7.4=hfca40fe\\_0{[}md5=32ddb97f897740641d8d46a829ce1704{]}'\n",
       "\\item '   - conda-forge/linux-64::libcurl==8.10.1=hbbe4b11\\_0{[}md5=6e801c50a40301f6978c53976917b277{]}'\n",
       "\\item '   - conda-forge/linux-64::menuinst==2.1.2=py312h7900ff3\\_1{[}md5=c6575ae996f2bc0369c73b632db5ca61{]}'\n",
       "\\item '   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab\\_0{[}md5=192278292e20704f663b9c766909d67b{]}'\n",
       "\\item '   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab\\_0{[}md5=61de176bd62041f9cd5bd4fcd09eb0ff{]}'\n",
       "\\item '   - conda-forge/linux-64::brotli-python==1.1.0=py312h2ec8cdc\\_2{[}md5=b0b867af6fc74b2a0aa206da29c0f3cf{]}'\n",
       "\\item '   - conda-forge/noarch::certifi==2024.8.30=pyhd8ed1ab\\_0{[}md5=12f7d00853807b0531775e9be891cb11{]}'\n",
       "\\item '   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab\\_0{[}md5=7f4a9e3fcff3f6356ae99244a014da6a{]}'\n",
       "\\item '   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab\\_0{[}md5=3faab06a954c2a04039983f2c4a50d99{]}'\n",
       "\\item '   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab\\_0{[}md5=bbdb409974cd6cb30071b1d978302726{]}'\n",
       "\\item '   - conda-forge/linux-64::frozendict==2.4.4=py312h66e93f0\\_1{[}md5=82a14d06575c4023a5df6b0798acff71{]}'\n",
       "\\item '   - conda-forge/noarch::hpack==4.0.0=pyh9f0ad1d\\_0{[}md5=914d6646c4dbb1fd3ff539830a12fd71{]}'\n",
       "\\item '   - conda-forge/noarch::hyperframe==6.0.1=pyhd8ed1ab\\_0{[}md5=9f765cbfab6870c8435b9eefecd7a1f4{]}'\n",
       "\\item '   - conda-forge/noarch::idna==3.10=pyhd8ed1ab\\_0{[}md5=7ba2ede0e7c795ff95088daf0dc59753{]}'\n",
       "\\item '   - conda-forge/linux-64::jsonpointer==3.0.0=py312h7900ff3\\_1{[}md5=6b51f7459ea4073eeb5057207e2e1e3d{]}'\n",
       "\\item '   - conda-forge/linux-64::libmamba==1.5.9=h4cc3d14\\_0{[}md5=896cece5b883ad86e9dd88b1f4d23c99{]}'\n",
       "\\item '   - conda-forge/noarch::packaging==24.1=pyhd8ed1ab\\_0{[}md5=cbe1bb1f21567018ce595d9c2be0f0db{]}'\n",
       "\\item '   - conda-forge/noarch::platformdirs==4.3.6=pyhd8ed1ab\\_0{[}md5=fd8f2b18b65bbf62e8f653100690c8d2{]}'\n",
       "\\item '   - conda-forge/noarch::pluggy==1.5.0=pyhd8ed1ab\\_0{[}md5=d3483c8fc2dc2cc3f5cf43e26d60cabf{]}'\n",
       "\\item '   - conda-forge/linux-64::pycosat==0.6.6=py312h98912ed\\_0{[}md5=8f1c372e7b843167be885dc8229931c1{]}'\n",
       "\\item '   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab\\_0{[}md5=844d9eb3b43095b031874477f7d70088{]}'\n",
       "\\item '   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31\\_6{[}md5=2a7de29fb590ca14b5243c4c812c8025{]}'\n",
       "\\item '   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py312h98912ed\\_0{[}md5=05f31c2a79ba61df8d6d903ce4a4ce7b{]}'\n",
       "\\item '   - conda-forge/noarch::setuptools==74.1.2=pyhd8ed1ab\\_0{[}md5=56c9c11d004428e81d02eeb730fc6336{]}'\n",
       "\\item '   - conda-forge/noarch::truststore==0.9.2=pyhd8ed1ab\\_0{[}md5=f14e46d1bf271e748ff556d8b872e28a{]}'\n",
       "\\item '   - conda-forge/noarch::wheel==0.44.0=pyhd8ed1ab\\_0{[}md5=d44e3b085abcaef02983c6305b84b584{]}'\n",
       "\\item '   - conda-forge/linux-64::cffi==1.17.1=py312h06ac9bb\\_0{[}md5=a861504bbea4161a9170b85d4d2be840{]}'\n",
       "\\item '   - conda-forge/noarch::h2==4.1.0=pyhd8ed1ab\\_0{[}md5=b748fbf7060927a6e82df7cb5ee8f097{]}'\n",
       "\\item '   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab\\_0{[}md5=bfdb7c5c6ad1077c82a69a8642c87aff{]}'\n",
       "\\item '   - conda-forge/linux-64::libmambapy==1.5.9=py312h7fb9e8e\\_0{[}md5=ccaeeb6e3caaf0c744480393791aa366{]}'\n",
       "\\item '   - conda-forge/noarch::pip==24.2=pyh8b19718\\_1{[}md5=6c78fbb8ddfd64bcb55b5cbafd2d2c43{]}'\n",
       "\\item '   - conda-forge/linux-64::ruamel.yaml==0.18.6=py312h98912ed\\_0{[}md5=a99a06a875138829ef65f44bbe2c30ca{]}'\n",
       "\\item '   - conda-forge/noarch::tqdm==4.66.5=pyhd8ed1ab\\_0{[}md5=c6e94fc2b2ec71ea33fe7c7da259acb4{]}'\n",
       "\\item '   - conda-forge/linux-64::zstandard==0.23.0=py312hef9b889\\_1{[}md5=8b7069e9792ee4e5b4919a7a306d2e67{]}'\n",
       "\\item '   - conda-forge/noarch::conda-package-streaming==0.10.0=pyhd8ed1ab\\_0{[}md5=3480386e00995f7a1dfb3b9aa2fe70fd{]}'\n",
       "\\item '   - conda-forge/noarch::urllib3==2.2.3=pyhd8ed1ab\\_0{[}md5=6b55867f385dd762ed99ea687af32a69{]}'\n",
       "\\item '   - conda-forge/noarch::conda-package-handling==2.3.0=pyh7900ff3\\_0{[}md5=0a7dce281ae2be81acab0aa963e6bb99{]}'\n",
       "\\item '   - conda-forge/noarch::requests==2.32.3=pyhd8ed1ab\\_0{[}md5=5ede4753180c7a550a443c430dc8ab52{]}'\n",
       "\\item '   - conda-forge/linux-64::conda==24.7.1=py312h7900ff3\\_0{[}md5=e1bf59014e88eaff036101358f63a496{]}'\n",
       "\\item '   - conda-forge/noarch::conda-libmamba-solver==24.7.0=pyhd8ed1ab\\_0{[}md5=857c9e25f0a77c0bd7eb622d46d9418f{]}'\n",
       "\\item '   - conda-forge/linux-64::mamba==1.5.9=py312h9460a1c\\_0{[}md5=a8525c8a1647b4f5967fa6b552722851{]}'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item '  Package                         Version  Build               Channel         Size'\n",
       "\\item '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "\\item '  Install:'\n",
       "\\item '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "\\item ''\n",
       "\\item '  + \\_libgcc\\_mutex                     0.1  conda\\_forge         conda-forge         '\n",
       "\\item '  + ca-certificates             2024.8.30  hbcca054\\_0          conda-forge         '\n",
       "\\item '  + ld\\_impl\\_linux-64                 2.40  hf3520f5\\_7          conda-forge         '\n",
       "\\item '  + pybind11-abi                        4  hd8ed1ab\\_3          conda-forge         '\n",
       "\\item '  + python\\_abi                       3.12  5\\_cp312             conda-forge         '\n",
       "\\item '  + tzdata                          2024a  h8827d51\\_1          conda-forge         '\n",
       "\\item '  + libgomp                        14.1.0  h77fa898\\_1          conda-forge         '\n",
       "\\item '  + \\_openmp\\_mutex                     4.5  2\\_gnu               conda-forge         '\n",
       "\\item '  + libgcc                         14.1.0  h77fa898\\_1          conda-forge         '\n",
       "\\item '  + libexpat                        2.6.3  h5888daf\\_0          conda-forge         '\n",
       "\\item '  + libgcc-ng                      14.1.0  h69a702a\\_1          conda-forge         '\n",
       "\\item '  + libstdcxx                      14.1.0  hc0a3c3a\\_1          conda-forge         '\n",
       "\\item '  + openssl                         3.3.2  hb9d3cd8\\_0          conda-forge         '\n",
       "\\item '  + bzip2                           1.0.8  h4bc722e\\_7          conda-forge         '\n",
       "\\item '  + c-ares                         1.32.3  h4bc722e\\_0          conda-forge         '\n",
       "\\item '  + keyutils                        1.6.1  h166bdaf\\_0          conda-forge         '\n",
       "\\item '  + libev                            4.33  hd590300\\_2          conda-forge         '\n",
       "\\item '  + libffi                          3.4.2  h7f98852\\_5          conda-forge         '\n",
       "\\item '  + libiconv                         1.17  hd590300\\_2          conda-forge         '\n",
       "\\item '  + libnsl                          2.0.1  hd590300\\_0          conda-forge         '\n",
       "\\item '  + libstdcxx-ng                   14.1.0  h4852527\\_1          conda-forge         '\n",
       "\\item '  + libuuid                        2.38.1  h0b41bf4\\_0          conda-forge         '\n",
       "\\item '  + libxcrypt                      4.4.36  hd590300\\_1          conda-forge         '\n",
       "\\item '  + libzlib                         1.3.1  h4ab18f5\\_1          conda-forge         '\n",
       "\\item '  + lzo                              2.10  hd590300\\_1001       conda-forge         '\n",
       "\\item '  + ncurses                           6.5  he02047a\\_1          conda-forge         '\n",
       "\\item '  + reproc                   14.2.4.post0  hd590300\\_1          conda-forge         '\n",
       "\\item '  + xz                              5.2.6  h166bdaf\\_0          conda-forge         '\n",
       "\\item '  + fmt                            10.2.1  h00ab1b0\\_0          conda-forge         '\n",
       "\\item '  + icu                              75.1  he02047a\\_0          conda-forge         '\n",
       "\\item '  + libedit                  3.1.20191231  he28a2e2\\_2          conda-forge         '\n",
       "\\item '  + libnghttp2                     1.58.0  h47da74e\\_1          conda-forge         '\n",
       "\\item '  + libsolv                        0.7.30  h3509ff9\\_0          conda-forge         '\n",
       "\\item '  + libsqlite                      3.46.1  hadc24fc\\_0          conda-forge         '\n",
       "\\item '  + libssh2                        1.11.0  h0841786\\_0          conda-forge         '\n",
       "\\item '  + lz4-c                           1.9.4  hcb278e6\\_0          conda-forge         '\n",
       "\\item '  + readline                          8.2  h8228510\\_1          conda-forge         '\n",
       "\\item '  + reproc-cpp               14.2.4.post0  h59595ed\\_1          conda-forge         '\n",
       "\\item '  + tk                             8.6.13  noxft\\_h4845f30\\_101  conda-forge         '\n",
       "\\item '  + yaml-cpp                        0.8.0  h59595ed\\_0          conda-forge         '\n",
       "\\item '  + zstd                            1.5.6  ha6fb4c9\\_0          conda-forge         '\n",
       "\\item '  + krb5                           1.21.3  h659f571\\_0          conda-forge         '\n",
       "\\item '  + libxml2                        2.12.7  he7c6b58\\_4          conda-forge         '\n",
       "\\item '  + python                         3.12.6  hc5c86c4\\_1\\_cpython  conda-forge         '\n",
       "\\item '  + libarchive                      3.7.4  hfca40fe\\_0          conda-forge         '\n",
       "\\item '  + libcurl                        8.10.1  hbbe4b11\\_0          conda-forge         '\n",
       "\\item '  + menuinst                        2.1.2  py312h7900ff3\\_1     conda-forge         '\n",
       "\\item '  + archspec                        0.2.3  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + boltons                        24.0.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + brotli-python                   1.1.0  py312h2ec8cdc\\_2     conda-forge         '\n",
       "\\item '  + certifi                     2024.8.30  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + charset-normalizer              3.3.2  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + colorama                        0.4.6  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + distro                          1.9.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + frozendict                      2.4.4  py312h66e93f0\\_1     conda-forge         '\n",
       "\\item '  + hpack                           4.0.0  pyh9f0ad1d\\_0        conda-forge         '\n",
       "\\item '  + hyperframe                      6.0.1  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + idna                             3.10  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + jsonpointer                     3.0.0  py312h7900ff3\\_1     conda-forge         '\n",
       "\\item '  + libmamba                        1.5.9  h4cc3d14\\_0          conda-forge         '\n",
       "\\item '  + packaging                        24.1  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + platformdirs                    4.3.6  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + pluggy                          1.5.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + pycosat                         0.6.6  py312h98912ed\\_0     conda-forge         '\n",
       "\\item '  + pycparser                        2.22  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + pysocks                         1.7.1  pyha2e5f31\\_6        conda-forge         '\n",
       "\\item '  + ruamel.yaml.clib                0.2.8  py312h98912ed\\_0     conda-forge         '\n",
       "\\item '  + setuptools                     74.1.2  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + truststore                      0.9.2  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + wheel                          0.44.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + cffi                           1.17.1  py312h06ac9bb\\_0     conda-forge         '\n",
       "\\item '  + h2                              4.1.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + jsonpatch                        1.33  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + libmambapy                      1.5.9  py312h7fb9e8e\\_0     conda-forge         '\n",
       "\\item '  + pip                              24.2  pyh8b19718\\_1        conda-forge         '\n",
       "\\item '  + ruamel.yaml                    0.18.6  py312h98912ed\\_0     conda-forge         '\n",
       "\\item '  + tqdm                           4.66.5  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + zstandard                      0.23.0  py312hef9b889\\_1     conda-forge         '\n",
       "\\item '  + conda-package-streaming        0.10.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + urllib3                         2.2.3  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + conda-package-handling          2.3.0  pyh7900ff3\\_0        conda-forge         '\n",
       "\\item '  + requests                       2.32.3  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + conda                          24.7.1  py312h7900ff3\\_0     conda-forge         '\n",
       "\\item '  + conda-libmamba-solver          24.7.0  pyhd8ed1ab\\_0        conda-forge         '\n",
       "\\item '  + mamba                           1.5.9  py312h9460a1c\\_0     conda-forge         '\n",
       "\\item ''\n",
       "\\item '  Summary:'\n",
       "\\item ''\n",
       "\\item '  Install: 85 packages'\n",
       "\\item ''\n",
       "\\item '  Total download: 0 B'\n",
       "\\item ''\n",
       "\\item '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'Transaction starting'\n",
       "\\item ''\n",
       "\\item 'Transaction finished'\n",
       "\\item ''\n",
       "\\item 'To activate this environment, use:'\n",
       "\\item ''\n",
       "\\item '    micromamba activate /home/ec2-user/miniforge'\n",
       "\\item ''\n",
       "\\item 'Or to execute a single command in this environment, use:'\n",
       "\\item ''\n",
       "\\item '    micromamba run -p /home/ec2-user/miniforge mycommand'\n",
       "\\item ''\n",
       "\\item 'installation finished.'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'PREFIX=/home/ec2-user/miniforge'\n",
       "2. ''\n",
       "3. 'Transaction'\n",
       "4. ''\n",
       "5. <span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/miniforge/envs/_virtual_specs_checks'</span>\n",
       "6. ''\n",
       "7. <span style=white-space:pre-wrap>'  All requested packages already installed'</span>\n",
       "8. ''\n",
       "9. 'Dry run. Not executing the transaction.'\n",
       "10. 'Unpacking payload ...'\n",
       "11. 'Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2'\n",
       "12. 'Extracting ca-certificates-2024.8.30-hbcca054_0.conda'\n",
       "13. 'Extracting ld_impl_linux-64-2.40-hf3520f5_7.conda'\n",
       "14. 'Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2'\n",
       "15. 'Extracting python_abi-3.12-5_cp312.conda'\n",
       "16. 'Extracting tzdata-2024a-h8827d51_1.conda'\n",
       "17. 'Extracting libgomp-14.1.0-h77fa898_1.conda'\n",
       "18. 'Extracting _openmp_mutex-4.5-2_gnu.tar.bz2'\n",
       "19. 'Extracting libgcc-14.1.0-h77fa898_1.conda'\n",
       "20. 'Extracting libexpat-2.6.3-h5888daf_0.conda'\n",
       "21. 'Extracting libgcc-ng-14.1.0-h69a702a_1.conda'\n",
       "22. 'Extracting libstdcxx-14.1.0-hc0a3c3a_1.conda'\n",
       "23. 'Extracting openssl-3.3.2-hb9d3cd8_0.conda'\n",
       "24. 'Extracting bzip2-1.0.8-h4bc722e_7.conda'\n",
       "25. 'Extracting c-ares-1.32.3-h4bc722e_0.conda'\n",
       "26. 'Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2'\n",
       "27. 'Extracting libev-4.33-hd590300_2.conda'\n",
       "28. 'Extracting libffi-3.4.2-h7f98852_5.tar.bz2'\n",
       "29. 'Extracting libiconv-1.17-hd590300_2.conda'\n",
       "30. 'Extracting libnsl-2.0.1-hd590300_0.conda'\n",
       "31. 'Extracting libstdcxx-ng-14.1.0-h4852527_1.conda'\n",
       "32. 'Extracting libuuid-2.38.1-h0b41bf4_0.conda'\n",
       "33. 'Extracting libxcrypt-4.4.36-hd590300_1.conda'\n",
       "34. 'Extracting libzlib-1.3.1-h4ab18f5_1.conda'\n",
       "35. 'Extracting lzo-2.10-hd590300_1001.conda'\n",
       "36. 'Extracting ncurses-6.5-he02047a_1.conda'\n",
       "37. 'Extracting reproc-14.2.4.post0-hd590300_1.conda'\n",
       "38. 'Extracting xz-5.2.6-h166bdaf_0.tar.bz2'\n",
       "39. 'Extracting fmt-10.2.1-h00ab1b0_0.conda'\n",
       "40. 'Extracting icu-75.1-he02047a_0.conda'\n",
       "41. 'Extracting libedit-3.1.20191231-he28a2e2_2.tar.bz2'\n",
       "42. 'Extracting libnghttp2-1.58.0-h47da74e_1.conda'\n",
       "43. 'Extracting libsolv-0.7.30-h3509ff9_0.conda'\n",
       "44. 'Extracting libsqlite-3.46.1-hadc24fc_0.conda'\n",
       "45. 'Extracting libssh2-1.11.0-h0841786_0.conda'\n",
       "46. 'Extracting lz4-c-1.9.4-hcb278e6_0.conda'\n",
       "47. 'Extracting readline-8.2-h8228510_1.conda'\n",
       "48. 'Extracting reproc-cpp-14.2.4.post0-h59595ed_1.conda'\n",
       "49. 'Extracting tk-8.6.13-noxft_h4845f30_101.conda'\n",
       "50. 'Extracting yaml-cpp-0.8.0-h59595ed_0.conda'\n",
       "51. 'Extracting zstd-1.5.6-ha6fb4c9_0.conda'\n",
       "52. 'Extracting krb5-1.21.3-h659f571_0.conda'\n",
       "53. 'Extracting libxml2-2.12.7-he7c6b58_4.conda'\n",
       "54. 'Extracting python-3.12.6-hc5c86c4_1_cpython.conda'\n",
       "55. 'Extracting libarchive-3.7.4-hfca40fe_0.conda'\n",
       "56. 'Extracting libcurl-8.10.1-hbbe4b11_0.conda'\n",
       "57. 'Extracting menuinst-2.1.2-py312h7900ff3_1.conda'\n",
       "58. 'Extracting archspec-0.2.3-pyhd8ed1ab_0.conda'\n",
       "59. 'Extracting boltons-24.0.0-pyhd8ed1ab_0.conda'\n",
       "60. 'Extracting brotli-python-1.1.0-py312h2ec8cdc_2.conda'\n",
       "61. 'Extracting certifi-2024.8.30-pyhd8ed1ab_0.conda'\n",
       "62. 'Extracting charset-normalizer-3.3.2-pyhd8ed1ab_0.conda'\n",
       "63. 'Extracting colorama-0.4.6-pyhd8ed1ab_0.tar.bz2'\n",
       "64. 'Extracting distro-1.9.0-pyhd8ed1ab_0.conda'\n",
       "65. 'Extracting frozendict-2.4.4-py312h66e93f0_1.conda'\n",
       "66. 'Extracting hpack-4.0.0-pyh9f0ad1d_0.tar.bz2'\n",
       "67. 'Extracting hyperframe-6.0.1-pyhd8ed1ab_0.tar.bz2'\n",
       "68. 'Extracting idna-3.10-pyhd8ed1ab_0.conda'\n",
       "69. 'Extracting jsonpointer-3.0.0-py312h7900ff3_1.conda'\n",
       "70. 'Extracting libmamba-1.5.9-h4cc3d14_0.conda'\n",
       "71. 'Extracting packaging-24.1-pyhd8ed1ab_0.conda'\n",
       "72. 'Extracting platformdirs-4.3.6-pyhd8ed1ab_0.conda'\n",
       "73. 'Extracting pluggy-1.5.0-pyhd8ed1ab_0.conda'\n",
       "74. 'Extracting pycosat-0.6.6-py312h98912ed_0.conda'\n",
       "75. 'Extracting pycparser-2.22-pyhd8ed1ab_0.conda'\n",
       "76. 'Extracting pysocks-1.7.1-pyha2e5f31_6.tar.bz2'\n",
       "77. 'Extracting ruamel.yaml.clib-0.2.8-py312h98912ed_0.conda'\n",
       "78. 'Extracting setuptools-74.1.2-pyhd8ed1ab_0.conda'\n",
       "79. 'Extracting truststore-0.9.2-pyhd8ed1ab_0.conda'\n",
       "80. 'Extracting wheel-0.44.0-pyhd8ed1ab_0.conda'\n",
       "81. 'Extracting cffi-1.17.1-py312h06ac9bb_0.conda'\n",
       "82. 'Extracting h2-4.1.0-pyhd8ed1ab_0.tar.bz2'\n",
       "83. 'Extracting jsonpatch-1.33-pyhd8ed1ab_0.conda'\n",
       "84. 'Extracting libmambapy-1.5.9-py312h7fb9e8e_0.conda'\n",
       "85. 'Extracting pip-24.2-pyh8b19718_1.conda'\n",
       "86. 'Extracting ruamel.yaml-0.18.6-py312h98912ed_0.conda'\n",
       "87. 'Extracting tqdm-4.66.5-pyhd8ed1ab_0.conda'\n",
       "88. 'Extracting zstandard-0.23.0-py312hef9b889_1.conda'\n",
       "89. 'Extracting conda-package-streaming-0.10.0-pyhd8ed1ab_0.conda'\n",
       "90. 'Extracting urllib3-2.2.3-pyhd8ed1ab_0.conda'\n",
       "91. 'Extracting conda-package-handling-2.3.0-pyh7900ff3_0.conda'\n",
       "92. 'Extracting requests-2.32.3-pyhd8ed1ab_0.conda'\n",
       "93. 'Extracting conda-24.7.1-py312h7900ff3_0.conda'\n",
       "94. 'Extracting conda-libmamba-solver-24.7.0-pyhd8ed1ab_0.conda'\n",
       "95. 'Extracting mamba-1.5.9-py312h9460a1c_0.conda'\n",
       "96. ''\n",
       "97. 'Installing base environment...'\n",
       "98. ''\n",
       "99. 'Transaction'\n",
       "100. ''\n",
       "101. <span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/miniforge'</span>\n",
       "102. ''\n",
       "103. <span style=white-space:pre-wrap>'  Updating specs:'</span>\n",
       "104. ''\n",
       "105. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::_libgcc_mutex==0.1=conda_forge[md5=d7c89558ba9fa0495403155b64376d81]'</span>\n",
       "106. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::ca-certificates==2024.8.30=hbcca054_0[md5=c27d1c142233b5bc9ca570c6e2e0c244]'</span>\n",
       "107. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::ld_impl_linux-64==2.40=hf3520f5_7[md5=b80f2f396ca2c28b8c14c437a4ed1e74]'</span>\n",
       "108. <span style=white-space:pre-wrap>'   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab_3[md5=878f923dd6acc8aeb47a75da6c4098be]'</span>\n",
       "109. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::python_abi==3.12=5_cp312[md5=0424ae29b104430108f5218a66db7260]'</span>\n",
       "110. <span style=white-space:pre-wrap>'   - conda-forge/noarch::tzdata==2024a=h8827d51_1[md5=8bfdead4e0fff0383ae4c9c50d0531bd]'</span>\n",
       "111. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgomp==14.1.0=h77fa898_1[md5=23c255b008c4f2ae008f81edcabaca89]'</span>\n",
       "112. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::_openmp_mutex==4.5=2_gnu[md5=73aaf86a425cc6e73fcf236a5a46396d]'</span>\n",
       "113. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgcc==14.1.0=h77fa898_1[md5=002ef4463dd1e2b44a94a4ace468f5d2]'</span>\n",
       "114. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libexpat==2.6.3=h5888daf_0[md5=59f4c43bb1b5ef1c71946ff2cbf59524]'</span>\n",
       "115. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libgcc-ng==14.1.0=h69a702a_1[md5=1efc0ad219877a73ef977af7dbb51f17]'</span>\n",
       "116. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libstdcxx==14.1.0=hc0a3c3a_1[md5=9dbb9699ea467983ba8a4ba89b08b066]'</span>\n",
       "117. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::openssl==3.3.2=hb9d3cd8_0[md5=4d638782050ab6faa27275bed57e9b4e]'</span>\n",
       "118. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::bzip2==1.0.8=h4bc722e_7[md5=62ee74e96c5ebb0af99386de58cf9553]'</span>\n",
       "119. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::c-ares==1.32.3=h4bc722e_0[md5=7624e34ee6baebfc80d67bac76cc9d9d]'</span>\n",
       "120. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf_0[md5=30186d27e2c9fa62b45fb1476b7200e3]'</span>\n",
       "121. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libev==4.33=hd590300_2[md5=172bf1cd1ff8629f2b1179945ed45055]'</span>\n",
       "122. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libffi==3.4.2=h7f98852_5[md5=d645c6d2ac96843a2bfaccd2d62b3ac3]'</span>\n",
       "123. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libiconv==1.17=hd590300_2[md5=d66573916ffcf376178462f1b61c941e]'</span>\n",
       "124. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libnsl==2.0.1=hd590300_0[md5=30fd6e37fe21f86f4bd26d6ee73eeec7]'</span>\n",
       "125. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libstdcxx-ng==14.1.0=h4852527_1[md5=bd2598399a70bb86d8218e95548d735e]'</span>\n",
       "126. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4_0[md5=40b61aab5c7ba9ff276c41cfffe6b80b]'</span>\n",
       "127. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300_1[md5=5aa797f8787fe7a17d1b0821485b5adc]'</span>\n",
       "128. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libzlib==1.3.1=h4ab18f5_1[md5=57d7dc60e9325e3de37ff8dffd18e814]'</span>\n",
       "129. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::lzo==2.10=hd590300_1001[md5=ec7398d21e2651e0dcb0044d03b9a339]'</span>\n",
       "130. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::ncurses==6.5=he02047a_1[md5=70caf8bb6cf39a0b6b7efc885f51c0fe]'</span>\n",
       "131. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300_1[md5=82ca53502dfd5a64a80dee76dae14685]'</span>\n",
       "132. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::xz==5.2.6=h166bdaf_0[md5=2161070d867d1b1204ea749c8eec4ef0]'</span>\n",
       "133. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0_0[md5=35ef8bc24bd34074ebae3c943d551728]'</span>\n",
       "134. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::icu==75.1=he02047a_0[md5=8b189310083baabfb622af68fd9d3ae3]'</span>\n",
       "135. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2_2[md5=4d331e44109e3f0e19b4cb8f9b82f3e1]'</span>\n",
       "136. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e_1[md5=700ac6ea6d53d5510591c4344d5c989a]'</span>\n",
       "137. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libsolv==0.7.30=h3509ff9_0[md5=02539b77d25aa4f65b20246549e256c3]'</span>\n",
       "138. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libsqlite==3.46.1=hadc24fc_0[md5=36f79405ab16bf271edb55b213836dac]'</span>\n",
       "139. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libssh2==1.11.0=h0841786_0[md5=1f5a58e686b13bcfde88b93f547d23fe]'</span>\n",
       "140. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6_0[md5=318b08df404f9c9be5712aaa5a6f0bb0]'</span>\n",
       "141. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::readline==8.2=h8228510_1[md5=47d31b792659ce70f470b5c82fdfb7a4]'</span>\n",
       "142. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed_1[md5=715e1d720ec1a03715bebd237972fca5]'</span>\n",
       "143. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::tk==8.6.13=noxft_h4845f30_101[md5=d453b98d9c83e71da0741bb0ff4d76bc]'</span>\n",
       "144. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed_0[md5=965eaacd7c18eb8361fd12bb9e7a57d7]'</span>\n",
       "145. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::zstd==1.5.6=ha6fb4c9_0[md5=4d056880988120e29d75bfff282e0f45]'</span>\n",
       "146. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::krb5==1.21.3=h659f571_0[md5=3f43953b7d3fb3aaa1d0d0723d91e368]'</span>\n",
       "147. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libxml2==2.12.7=he7c6b58_4[md5=08a9265c637230c37cb1be4a6cad4536]'</span>\n",
       "148. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::python==3.12.6=hc5c86c4_1_cpython[md5=00836baacdca254f28c54d2543e97514]'</span>\n",
       "149. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libarchive==3.7.4=hfca40fe_0[md5=32ddb97f897740641d8d46a829ce1704]'</span>\n",
       "150. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libcurl==8.10.1=hbbe4b11_0[md5=6e801c50a40301f6978c53976917b277]'</span>\n",
       "151. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::menuinst==2.1.2=py312h7900ff3_1[md5=c6575ae996f2bc0369c73b632db5ca61]'</span>\n",
       "152. <span style=white-space:pre-wrap>'   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab_0[md5=192278292e20704f663b9c766909d67b]'</span>\n",
       "153. <span style=white-space:pre-wrap>'   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab_0[md5=61de176bd62041f9cd5bd4fcd09eb0ff]'</span>\n",
       "154. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::brotli-python==1.1.0=py312h2ec8cdc_2[md5=b0b867af6fc74b2a0aa206da29c0f3cf]'</span>\n",
       "155. <span style=white-space:pre-wrap>'   - conda-forge/noarch::certifi==2024.8.30=pyhd8ed1ab_0[md5=12f7d00853807b0531775e9be891cb11]'</span>\n",
       "156. <span style=white-space:pre-wrap>'   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab_0[md5=7f4a9e3fcff3f6356ae99244a014da6a]'</span>\n",
       "157. <span style=white-space:pre-wrap>'   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab_0[md5=3faab06a954c2a04039983f2c4a50d99]'</span>\n",
       "158. <span style=white-space:pre-wrap>'   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab_0[md5=bbdb409974cd6cb30071b1d978302726]'</span>\n",
       "159. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::frozendict==2.4.4=py312h66e93f0_1[md5=82a14d06575c4023a5df6b0798acff71]'</span>\n",
       "160. <span style=white-space:pre-wrap>'   - conda-forge/noarch::hpack==4.0.0=pyh9f0ad1d_0[md5=914d6646c4dbb1fd3ff539830a12fd71]'</span>\n",
       "161. <span style=white-space:pre-wrap>'   - conda-forge/noarch::hyperframe==6.0.1=pyhd8ed1ab_0[md5=9f765cbfab6870c8435b9eefecd7a1f4]'</span>\n",
       "162. <span style=white-space:pre-wrap>'   - conda-forge/noarch::idna==3.10=pyhd8ed1ab_0[md5=7ba2ede0e7c795ff95088daf0dc59753]'</span>\n",
       "163. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::jsonpointer==3.0.0=py312h7900ff3_1[md5=6b51f7459ea4073eeb5057207e2e1e3d]'</span>\n",
       "164. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libmamba==1.5.9=h4cc3d14_0[md5=896cece5b883ad86e9dd88b1f4d23c99]'</span>\n",
       "165. <span style=white-space:pre-wrap>'   - conda-forge/noarch::packaging==24.1=pyhd8ed1ab_0[md5=cbe1bb1f21567018ce595d9c2be0f0db]'</span>\n",
       "166. <span style=white-space:pre-wrap>'   - conda-forge/noarch::platformdirs==4.3.6=pyhd8ed1ab_0[md5=fd8f2b18b65bbf62e8f653100690c8d2]'</span>\n",
       "167. <span style=white-space:pre-wrap>'   - conda-forge/noarch::pluggy==1.5.0=pyhd8ed1ab_0[md5=d3483c8fc2dc2cc3f5cf43e26d60cabf]'</span>\n",
       "168. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::pycosat==0.6.6=py312h98912ed_0[md5=8f1c372e7b843167be885dc8229931c1]'</span>\n",
       "169. <span style=white-space:pre-wrap>'   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab_0[md5=844d9eb3b43095b031874477f7d70088]'</span>\n",
       "170. <span style=white-space:pre-wrap>'   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31_6[md5=2a7de29fb590ca14b5243c4c812c8025]'</span>\n",
       "171. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py312h98912ed_0[md5=05f31c2a79ba61df8d6d903ce4a4ce7b]'</span>\n",
       "172. <span style=white-space:pre-wrap>'   - conda-forge/noarch::setuptools==74.1.2=pyhd8ed1ab_0[md5=56c9c11d004428e81d02eeb730fc6336]'</span>\n",
       "173. <span style=white-space:pre-wrap>'   - conda-forge/noarch::truststore==0.9.2=pyhd8ed1ab_0[md5=f14e46d1bf271e748ff556d8b872e28a]'</span>\n",
       "174. <span style=white-space:pre-wrap>'   - conda-forge/noarch::wheel==0.44.0=pyhd8ed1ab_0[md5=d44e3b085abcaef02983c6305b84b584]'</span>\n",
       "175. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::cffi==1.17.1=py312h06ac9bb_0[md5=a861504bbea4161a9170b85d4d2be840]'</span>\n",
       "176. <span style=white-space:pre-wrap>'   - conda-forge/noarch::h2==4.1.0=pyhd8ed1ab_0[md5=b748fbf7060927a6e82df7cb5ee8f097]'</span>\n",
       "177. <span style=white-space:pre-wrap>'   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab_0[md5=bfdb7c5c6ad1077c82a69a8642c87aff]'</span>\n",
       "178. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::libmambapy==1.5.9=py312h7fb9e8e_0[md5=ccaeeb6e3caaf0c744480393791aa366]'</span>\n",
       "179. <span style=white-space:pre-wrap>'   - conda-forge/noarch::pip==24.2=pyh8b19718_1[md5=6c78fbb8ddfd64bcb55b5cbafd2d2c43]'</span>\n",
       "180. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::ruamel.yaml==0.18.6=py312h98912ed_0[md5=a99a06a875138829ef65f44bbe2c30ca]'</span>\n",
       "181. <span style=white-space:pre-wrap>'   - conda-forge/noarch::tqdm==4.66.5=pyhd8ed1ab_0[md5=c6e94fc2b2ec71ea33fe7c7da259acb4]'</span>\n",
       "182. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::zstandard==0.23.0=py312hef9b889_1[md5=8b7069e9792ee4e5b4919a7a306d2e67]'</span>\n",
       "183. <span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-package-streaming==0.10.0=pyhd8ed1ab_0[md5=3480386e00995f7a1dfb3b9aa2fe70fd]'</span>\n",
       "184. <span style=white-space:pre-wrap>'   - conda-forge/noarch::urllib3==2.2.3=pyhd8ed1ab_0[md5=6b55867f385dd762ed99ea687af32a69]'</span>\n",
       "185. <span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-package-handling==2.3.0=pyh7900ff3_0[md5=0a7dce281ae2be81acab0aa963e6bb99]'</span>\n",
       "186. <span style=white-space:pre-wrap>'   - conda-forge/noarch::requests==2.32.3=pyhd8ed1ab_0[md5=5ede4753180c7a550a443c430dc8ab52]'</span>\n",
       "187. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::conda==24.7.1=py312h7900ff3_0[md5=e1bf59014e88eaff036101358f63a496]'</span>\n",
       "188. <span style=white-space:pre-wrap>'   - conda-forge/noarch::conda-libmamba-solver==24.7.0=pyhd8ed1ab_0[md5=857c9e25f0a77c0bd7eb622d46d9418f]'</span>\n",
       "189. <span style=white-space:pre-wrap>'   - conda-forge/linux-64::mamba==1.5.9=py312h9460a1c_0[md5=a8525c8a1647b4f5967fa6b552722851]'</span>\n",
       "190. ''\n",
       "191. ''\n",
       "192. <span style=white-space:pre-wrap>'  Package                         Version  Build               Channel         Size'</span>\n",
       "193. '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "194. <span style=white-space:pre-wrap>'  Install:'</span>\n",
       "195. '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "196. ''\n",
       "197. <span style=white-space:pre-wrap>'  + _libgcc_mutex                     0.1  conda_forge         conda-forge         '</span>\n",
       "198. <span style=white-space:pre-wrap>'  + ca-certificates             2024.8.30  hbcca054_0          conda-forge         '</span>\n",
       "199. <span style=white-space:pre-wrap>'  + ld_impl_linux-64                 2.40  hf3520f5_7          conda-forge         '</span>\n",
       "200. <span style=white-space:pre-wrap>'  + pybind11-abi                        4  hd8ed1ab_3          conda-forge         '</span>\n",
       "201. <span style=white-space:pre-wrap>'  + python_abi                       3.12  5_cp312             conda-forge         '</span>\n",
       "202. <span style=white-space:pre-wrap>'  + tzdata                          2024a  h8827d51_1          conda-forge         '</span>\n",
       "203. <span style=white-space:pre-wrap>'  + libgomp                        14.1.0  h77fa898_1          conda-forge         '</span>\n",
       "204. <span style=white-space:pre-wrap>'  + _openmp_mutex                     4.5  2_gnu               conda-forge         '</span>\n",
       "205. <span style=white-space:pre-wrap>'  + libgcc                         14.1.0  h77fa898_1          conda-forge         '</span>\n",
       "206. <span style=white-space:pre-wrap>'  + libexpat                        2.6.3  h5888daf_0          conda-forge         '</span>\n",
       "207. <span style=white-space:pre-wrap>'  + libgcc-ng                      14.1.0  h69a702a_1          conda-forge         '</span>\n",
       "208. <span style=white-space:pre-wrap>'  + libstdcxx                      14.1.0  hc0a3c3a_1          conda-forge         '</span>\n",
       "209. <span style=white-space:pre-wrap>'  + openssl                         3.3.2  hb9d3cd8_0          conda-forge         '</span>\n",
       "210. <span style=white-space:pre-wrap>'  + bzip2                           1.0.8  h4bc722e_7          conda-forge         '</span>\n",
       "211. <span style=white-space:pre-wrap>'  + c-ares                         1.32.3  h4bc722e_0          conda-forge         '</span>\n",
       "212. <span style=white-space:pre-wrap>'  + keyutils                        1.6.1  h166bdaf_0          conda-forge         '</span>\n",
       "213. <span style=white-space:pre-wrap>'  + libev                            4.33  hd590300_2          conda-forge         '</span>\n",
       "214. <span style=white-space:pre-wrap>'  + libffi                          3.4.2  h7f98852_5          conda-forge         '</span>\n",
       "215. <span style=white-space:pre-wrap>'  + libiconv                         1.17  hd590300_2          conda-forge         '</span>\n",
       "216. <span style=white-space:pre-wrap>'  + libnsl                          2.0.1  hd590300_0          conda-forge         '</span>\n",
       "217. <span style=white-space:pre-wrap>'  + libstdcxx-ng                   14.1.0  h4852527_1          conda-forge         '</span>\n",
       "218. <span style=white-space:pre-wrap>'  + libuuid                        2.38.1  h0b41bf4_0          conda-forge         '</span>\n",
       "219. <span style=white-space:pre-wrap>'  + libxcrypt                      4.4.36  hd590300_1          conda-forge         '</span>\n",
       "220. <span style=white-space:pre-wrap>'  + libzlib                         1.3.1  h4ab18f5_1          conda-forge         '</span>\n",
       "221. <span style=white-space:pre-wrap>'  + lzo                              2.10  hd590300_1001       conda-forge         '</span>\n",
       "222. <span style=white-space:pre-wrap>'  + ncurses                           6.5  he02047a_1          conda-forge         '</span>\n",
       "223. <span style=white-space:pre-wrap>'  + reproc                   14.2.4.post0  hd590300_1          conda-forge         '</span>\n",
       "224. <span style=white-space:pre-wrap>'  + xz                              5.2.6  h166bdaf_0          conda-forge         '</span>\n",
       "225. <span style=white-space:pre-wrap>'  + fmt                            10.2.1  h00ab1b0_0          conda-forge         '</span>\n",
       "226. <span style=white-space:pre-wrap>'  + icu                              75.1  he02047a_0          conda-forge         '</span>\n",
       "227. <span style=white-space:pre-wrap>'  + libedit                  3.1.20191231  he28a2e2_2          conda-forge         '</span>\n",
       "228. <span style=white-space:pre-wrap>'  + libnghttp2                     1.58.0  h47da74e_1          conda-forge         '</span>\n",
       "229. <span style=white-space:pre-wrap>'  + libsolv                        0.7.30  h3509ff9_0          conda-forge         '</span>\n",
       "230. <span style=white-space:pre-wrap>'  + libsqlite                      3.46.1  hadc24fc_0          conda-forge         '</span>\n",
       "231. <span style=white-space:pre-wrap>'  + libssh2                        1.11.0  h0841786_0          conda-forge         '</span>\n",
       "232. <span style=white-space:pre-wrap>'  + lz4-c                           1.9.4  hcb278e6_0          conda-forge         '</span>\n",
       "233. <span style=white-space:pre-wrap>'  + readline                          8.2  h8228510_1          conda-forge         '</span>\n",
       "234. <span style=white-space:pre-wrap>'  + reproc-cpp               14.2.4.post0  h59595ed_1          conda-forge         '</span>\n",
       "235. <span style=white-space:pre-wrap>'  + tk                             8.6.13  noxft_h4845f30_101  conda-forge         '</span>\n",
       "236. <span style=white-space:pre-wrap>'  + yaml-cpp                        0.8.0  h59595ed_0          conda-forge         '</span>\n",
       "237. <span style=white-space:pre-wrap>'  + zstd                            1.5.6  ha6fb4c9_0          conda-forge         '</span>\n",
       "238. <span style=white-space:pre-wrap>'  + krb5                           1.21.3  h659f571_0          conda-forge         '</span>\n",
       "239. <span style=white-space:pre-wrap>'  + libxml2                        2.12.7  he7c6b58_4          conda-forge         '</span>\n",
       "240. <span style=white-space:pre-wrap>'  + python                         3.12.6  hc5c86c4_1_cpython  conda-forge         '</span>\n",
       "241. <span style=white-space:pre-wrap>'  + libarchive                      3.7.4  hfca40fe_0          conda-forge         '</span>\n",
       "242. <span style=white-space:pre-wrap>'  + libcurl                        8.10.1  hbbe4b11_0          conda-forge         '</span>\n",
       "243. <span style=white-space:pre-wrap>'  + menuinst                        2.1.2  py312h7900ff3_1     conda-forge         '</span>\n",
       "244. <span style=white-space:pre-wrap>'  + archspec                        0.2.3  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "245. <span style=white-space:pre-wrap>'  + boltons                        24.0.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "246. <span style=white-space:pre-wrap>'  + brotli-python                   1.1.0  py312h2ec8cdc_2     conda-forge         '</span>\n",
       "247. <span style=white-space:pre-wrap>'  + certifi                     2024.8.30  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "248. <span style=white-space:pre-wrap>'  + charset-normalizer              3.3.2  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "249. <span style=white-space:pre-wrap>'  + colorama                        0.4.6  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "250. <span style=white-space:pre-wrap>'  + distro                          1.9.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "251. <span style=white-space:pre-wrap>'  + frozendict                      2.4.4  py312h66e93f0_1     conda-forge         '</span>\n",
       "252. <span style=white-space:pre-wrap>'  + hpack                           4.0.0  pyh9f0ad1d_0        conda-forge         '</span>\n",
       "253. <span style=white-space:pre-wrap>'  + hyperframe                      6.0.1  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "254. <span style=white-space:pre-wrap>'  + idna                             3.10  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "255. <span style=white-space:pre-wrap>'  + jsonpointer                     3.0.0  py312h7900ff3_1     conda-forge         '</span>\n",
       "256. <span style=white-space:pre-wrap>'  + libmamba                        1.5.9  h4cc3d14_0          conda-forge         '</span>\n",
       "257. <span style=white-space:pre-wrap>'  + packaging                        24.1  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "258. <span style=white-space:pre-wrap>'  + platformdirs                    4.3.6  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "259. <span style=white-space:pre-wrap>'  + pluggy                          1.5.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "260. <span style=white-space:pre-wrap>'  + pycosat                         0.6.6  py312h98912ed_0     conda-forge         '</span>\n",
       "261. <span style=white-space:pre-wrap>'  + pycparser                        2.22  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "262. <span style=white-space:pre-wrap>'  + pysocks                         1.7.1  pyha2e5f31_6        conda-forge         '</span>\n",
       "263. <span style=white-space:pre-wrap>'  + ruamel.yaml.clib                0.2.8  py312h98912ed_0     conda-forge         '</span>\n",
       "264. <span style=white-space:pre-wrap>'  + setuptools                     74.1.2  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "265. <span style=white-space:pre-wrap>'  + truststore                      0.9.2  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "266. <span style=white-space:pre-wrap>'  + wheel                          0.44.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "267. <span style=white-space:pre-wrap>'  + cffi                           1.17.1  py312h06ac9bb_0     conda-forge         '</span>\n",
       "268. <span style=white-space:pre-wrap>'  + h2                              4.1.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "269. <span style=white-space:pre-wrap>'  + jsonpatch                        1.33  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "270. <span style=white-space:pre-wrap>'  + libmambapy                      1.5.9  py312h7fb9e8e_0     conda-forge         '</span>\n",
       "271. <span style=white-space:pre-wrap>'  + pip                              24.2  pyh8b19718_1        conda-forge         '</span>\n",
       "272. <span style=white-space:pre-wrap>'  + ruamel.yaml                    0.18.6  py312h98912ed_0     conda-forge         '</span>\n",
       "273. <span style=white-space:pre-wrap>'  + tqdm                           4.66.5  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "274. <span style=white-space:pre-wrap>'  + zstandard                      0.23.0  py312hef9b889_1     conda-forge         '</span>\n",
       "275. <span style=white-space:pre-wrap>'  + conda-package-streaming        0.10.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "276. <span style=white-space:pre-wrap>'  + urllib3                         2.2.3  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "277. <span style=white-space:pre-wrap>'  + conda-package-handling          2.3.0  pyh7900ff3_0        conda-forge         '</span>\n",
       "278. <span style=white-space:pre-wrap>'  + requests                       2.32.3  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "279. <span style=white-space:pre-wrap>'  + conda                          24.7.1  py312h7900ff3_0     conda-forge         '</span>\n",
       "280. <span style=white-space:pre-wrap>'  + conda-libmamba-solver          24.7.0  pyhd8ed1ab_0        conda-forge         '</span>\n",
       "281. <span style=white-space:pre-wrap>'  + mamba                           1.5.9  py312h9460a1c_0     conda-forge         '</span>\n",
       "282. ''\n",
       "283. <span style=white-space:pre-wrap>'  Summary:'</span>\n",
       "284. ''\n",
       "285. <span style=white-space:pre-wrap>'  Install: 85 packages'</span>\n",
       "286. ''\n",
       "287. <span style=white-space:pre-wrap>'  Total download: 0 B'</span>\n",
       "288. ''\n",
       "289. '─────────────────────────────────────────────────────────────────────────────────────'\n",
       "290. ''\n",
       "291. ''\n",
       "292. ''\n",
       "293. 'Transaction starting'\n",
       "294. ''\n",
       "295. 'Transaction finished'\n",
       "296. ''\n",
       "297. 'To activate this environment, use:'\n",
       "298. ''\n",
       "299. <span style=white-space:pre-wrap>'    micromamba activate /home/ec2-user/miniforge'</span>\n",
       "300. ''\n",
       "301. 'Or to execute a single command in this environment, use:'\n",
       "302. ''\n",
       "303. <span style=white-space:pre-wrap>'    micromamba run -p /home/ec2-user/miniforge mycommand'</span>\n",
       "304. ''\n",
       "305. 'installation finished.'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "  [1] \"PREFIX=/home/ec2-user/miniforge\"                                                                            \n",
       "  [2] \"\"                                                                                                           \n",
       "  [3] \"Transaction\"                                                                                                \n",
       "  [4] \"\"                                                                                                           \n",
       "  [5] \"  Prefix: /home/ec2-user/miniforge/envs/_virtual_specs_checks\"                                              \n",
       "  [6] \"\"                                                                                                           \n",
       "  [7] \"  All requested packages already installed\"                                                                 \n",
       "  [8] \"\"                                                                                                           \n",
       "  [9] \"Dry run. Not executing the transaction.\"                                                                    \n",
       " [10] \"Unpacking payload ...\"                                                                                      \n",
       " [11] \"Extracting _libgcc_mutex-0.1-conda_forge.tar.bz2\"                                                           \n",
       " [12] \"Extracting ca-certificates-2024.8.30-hbcca054_0.conda\"                                                      \n",
       " [13] \"Extracting ld_impl_linux-64-2.40-hf3520f5_7.conda\"                                                          \n",
       " [14] \"Extracting pybind11-abi-4-hd8ed1ab_3.tar.bz2\"                                                               \n",
       " [15] \"Extracting python_abi-3.12-5_cp312.conda\"                                                                   \n",
       " [16] \"Extracting tzdata-2024a-h8827d51_1.conda\"                                                                   \n",
       " [17] \"Extracting libgomp-14.1.0-h77fa898_1.conda\"                                                                 \n",
       " [18] \"Extracting _openmp_mutex-4.5-2_gnu.tar.bz2\"                                                                 \n",
       " [19] \"Extracting libgcc-14.1.0-h77fa898_1.conda\"                                                                  \n",
       " [20] \"Extracting libexpat-2.6.3-h5888daf_0.conda\"                                                                 \n",
       " [21] \"Extracting libgcc-ng-14.1.0-h69a702a_1.conda\"                                                               \n",
       " [22] \"Extracting libstdcxx-14.1.0-hc0a3c3a_1.conda\"                                                               \n",
       " [23] \"Extracting openssl-3.3.2-hb9d3cd8_0.conda\"                                                                  \n",
       " [24] \"Extracting bzip2-1.0.8-h4bc722e_7.conda\"                                                                    \n",
       " [25] \"Extracting c-ares-1.32.3-h4bc722e_0.conda\"                                                                  \n",
       " [26] \"Extracting keyutils-1.6.1-h166bdaf_0.tar.bz2\"                                                               \n",
       " [27] \"Extracting libev-4.33-hd590300_2.conda\"                                                                     \n",
       " [28] \"Extracting libffi-3.4.2-h7f98852_5.tar.bz2\"                                                                 \n",
       " [29] \"Extracting libiconv-1.17-hd590300_2.conda\"                                                                  \n",
       " [30] \"Extracting libnsl-2.0.1-hd590300_0.conda\"                                                                   \n",
       " [31] \"Extracting libstdcxx-ng-14.1.0-h4852527_1.conda\"                                                            \n",
       " [32] \"Extracting libuuid-2.38.1-h0b41bf4_0.conda\"                                                                 \n",
       " [33] \"Extracting libxcrypt-4.4.36-hd590300_1.conda\"                                                               \n",
       " [34] \"Extracting libzlib-1.3.1-h4ab18f5_1.conda\"                                                                  \n",
       " [35] \"Extracting lzo-2.10-hd590300_1001.conda\"                                                                    \n",
       " [36] \"Extracting ncurses-6.5-he02047a_1.conda\"                                                                    \n",
       " [37] \"Extracting reproc-14.2.4.post0-hd590300_1.conda\"                                                            \n",
       " [38] \"Extracting xz-5.2.6-h166bdaf_0.tar.bz2\"                                                                     \n",
       " [39] \"Extracting fmt-10.2.1-h00ab1b0_0.conda\"                                                                     \n",
       " [40] \"Extracting icu-75.1-he02047a_0.conda\"                                                                       \n",
       " [41] \"Extracting libedit-3.1.20191231-he28a2e2_2.tar.bz2\"                                                         \n",
       " [42] \"Extracting libnghttp2-1.58.0-h47da74e_1.conda\"                                                              \n",
       " [43] \"Extracting libsolv-0.7.30-h3509ff9_0.conda\"                                                                 \n",
       " [44] \"Extracting libsqlite-3.46.1-hadc24fc_0.conda\"                                                               \n",
       " [45] \"Extracting libssh2-1.11.0-h0841786_0.conda\"                                                                 \n",
       " [46] \"Extracting lz4-c-1.9.4-hcb278e6_0.conda\"                                                                    \n",
       " [47] \"Extracting readline-8.2-h8228510_1.conda\"                                                                   \n",
       " [48] \"Extracting reproc-cpp-14.2.4.post0-h59595ed_1.conda\"                                                        \n",
       " [49] \"Extracting tk-8.6.13-noxft_h4845f30_101.conda\"                                                              \n",
       " [50] \"Extracting yaml-cpp-0.8.0-h59595ed_0.conda\"                                                                 \n",
       " [51] \"Extracting zstd-1.5.6-ha6fb4c9_0.conda\"                                                                     \n",
       " [52] \"Extracting krb5-1.21.3-h659f571_0.conda\"                                                                    \n",
       " [53] \"Extracting libxml2-2.12.7-he7c6b58_4.conda\"                                                                 \n",
       " [54] \"Extracting python-3.12.6-hc5c86c4_1_cpython.conda\"                                                          \n",
       " [55] \"Extracting libarchive-3.7.4-hfca40fe_0.conda\"                                                               \n",
       " [56] \"Extracting libcurl-8.10.1-hbbe4b11_0.conda\"                                                                 \n",
       " [57] \"Extracting menuinst-2.1.2-py312h7900ff3_1.conda\"                                                            \n",
       " [58] \"Extracting archspec-0.2.3-pyhd8ed1ab_0.conda\"                                                               \n",
       " [59] \"Extracting boltons-24.0.0-pyhd8ed1ab_0.conda\"                                                               \n",
       " [60] \"Extracting brotli-python-1.1.0-py312h2ec8cdc_2.conda\"                                                       \n",
       " [61] \"Extracting certifi-2024.8.30-pyhd8ed1ab_0.conda\"                                                            \n",
       " [62] \"Extracting charset-normalizer-3.3.2-pyhd8ed1ab_0.conda\"                                                     \n",
       " [63] \"Extracting colorama-0.4.6-pyhd8ed1ab_0.tar.bz2\"                                                             \n",
       " [64] \"Extracting distro-1.9.0-pyhd8ed1ab_0.conda\"                                                                 \n",
       " [65] \"Extracting frozendict-2.4.4-py312h66e93f0_1.conda\"                                                          \n",
       " [66] \"Extracting hpack-4.0.0-pyh9f0ad1d_0.tar.bz2\"                                                                \n",
       " [67] \"Extracting hyperframe-6.0.1-pyhd8ed1ab_0.tar.bz2\"                                                           \n",
       " [68] \"Extracting idna-3.10-pyhd8ed1ab_0.conda\"                                                                    \n",
       " [69] \"Extracting jsonpointer-3.0.0-py312h7900ff3_1.conda\"                                                         \n",
       " [70] \"Extracting libmamba-1.5.9-h4cc3d14_0.conda\"                                                                 \n",
       " [71] \"Extracting packaging-24.1-pyhd8ed1ab_0.conda\"                                                               \n",
       " [72] \"Extracting platformdirs-4.3.6-pyhd8ed1ab_0.conda\"                                                           \n",
       " [73] \"Extracting pluggy-1.5.0-pyhd8ed1ab_0.conda\"                                                                 \n",
       " [74] \"Extracting pycosat-0.6.6-py312h98912ed_0.conda\"                                                             \n",
       " [75] \"Extracting pycparser-2.22-pyhd8ed1ab_0.conda\"                                                               \n",
       " [76] \"Extracting pysocks-1.7.1-pyha2e5f31_6.tar.bz2\"                                                              \n",
       " [77] \"Extracting ruamel.yaml.clib-0.2.8-py312h98912ed_0.conda\"                                                    \n",
       " [78] \"Extracting setuptools-74.1.2-pyhd8ed1ab_0.conda\"                                                            \n",
       " [79] \"Extracting truststore-0.9.2-pyhd8ed1ab_0.conda\"                                                             \n",
       " [80] \"Extracting wheel-0.44.0-pyhd8ed1ab_0.conda\"                                                                 \n",
       " [81] \"Extracting cffi-1.17.1-py312h06ac9bb_0.conda\"                                                               \n",
       " [82] \"Extracting h2-4.1.0-pyhd8ed1ab_0.tar.bz2\"                                                                   \n",
       " [83] \"Extracting jsonpatch-1.33-pyhd8ed1ab_0.conda\"                                                               \n",
       " [84] \"Extracting libmambapy-1.5.9-py312h7fb9e8e_0.conda\"                                                          \n",
       " [85] \"Extracting pip-24.2-pyh8b19718_1.conda\"                                                                     \n",
       " [86] \"Extracting ruamel.yaml-0.18.6-py312h98912ed_0.conda\"                                                        \n",
       " [87] \"Extracting tqdm-4.66.5-pyhd8ed1ab_0.conda\"                                                                  \n",
       " [88] \"Extracting zstandard-0.23.0-py312hef9b889_1.conda\"                                                          \n",
       " [89] \"Extracting conda-package-streaming-0.10.0-pyhd8ed1ab_0.conda\"                                               \n",
       " [90] \"Extracting urllib3-2.2.3-pyhd8ed1ab_0.conda\"                                                                \n",
       " [91] \"Extracting conda-package-handling-2.3.0-pyh7900ff3_0.conda\"                                                 \n",
       " [92] \"Extracting requests-2.32.3-pyhd8ed1ab_0.conda\"                                                              \n",
       " [93] \"Extracting conda-24.7.1-py312h7900ff3_0.conda\"                                                              \n",
       " [94] \"Extracting conda-libmamba-solver-24.7.0-pyhd8ed1ab_0.conda\"                                                 \n",
       " [95] \"Extracting mamba-1.5.9-py312h9460a1c_0.conda\"                                                               \n",
       " [96] \"\"                                                                                                           \n",
       " [97] \"Installing base environment...\"                                                                             \n",
       " [98] \"\"                                                                                                           \n",
       " [99] \"Transaction\"                                                                                                \n",
       "[100] \"\"                                                                                                           \n",
       "[101] \"  Prefix: /home/ec2-user/miniforge\"                                                                         \n",
       "[102] \"\"                                                                                                           \n",
       "[103] \"  Updating specs:\"                                                                                          \n",
       "[104] \"\"                                                                                                           \n",
       "[105] \"   - conda-forge/linux-64::_libgcc_mutex==0.1=conda_forge[md5=d7c89558ba9fa0495403155b64376d81]\"            \n",
       "[106] \"   - conda-forge/linux-64::ca-certificates==2024.8.30=hbcca054_0[md5=c27d1c142233b5bc9ca570c6e2e0c244]\"     \n",
       "[107] \"   - conda-forge/linux-64::ld_impl_linux-64==2.40=hf3520f5_7[md5=b80f2f396ca2c28b8c14c437a4ed1e74]\"         \n",
       "[108] \"   - conda-forge/noarch::pybind11-abi==4=hd8ed1ab_3[md5=878f923dd6acc8aeb47a75da6c4098be]\"                  \n",
       "[109] \"   - conda-forge/linux-64::python_abi==3.12=5_cp312[md5=0424ae29b104430108f5218a66db7260]\"                  \n",
       "[110] \"   - conda-forge/noarch::tzdata==2024a=h8827d51_1[md5=8bfdead4e0fff0383ae4c9c50d0531bd]\"                    \n",
       "[111] \"   - conda-forge/linux-64::libgomp==14.1.0=h77fa898_1[md5=23c255b008c4f2ae008f81edcabaca89]\"                \n",
       "[112] \"   - conda-forge/linux-64::_openmp_mutex==4.5=2_gnu[md5=73aaf86a425cc6e73fcf236a5a46396d]\"                  \n",
       "[113] \"   - conda-forge/linux-64::libgcc==14.1.0=h77fa898_1[md5=002ef4463dd1e2b44a94a4ace468f5d2]\"                 \n",
       "[114] \"   - conda-forge/linux-64::libexpat==2.6.3=h5888daf_0[md5=59f4c43bb1b5ef1c71946ff2cbf59524]\"                \n",
       "[115] \"   - conda-forge/linux-64::libgcc-ng==14.1.0=h69a702a_1[md5=1efc0ad219877a73ef977af7dbb51f17]\"              \n",
       "[116] \"   - conda-forge/linux-64::libstdcxx==14.1.0=hc0a3c3a_1[md5=9dbb9699ea467983ba8a4ba89b08b066]\"              \n",
       "[117] \"   - conda-forge/linux-64::openssl==3.3.2=hb9d3cd8_0[md5=4d638782050ab6faa27275bed57e9b4e]\"                 \n",
       "[118] \"   - conda-forge/linux-64::bzip2==1.0.8=h4bc722e_7[md5=62ee74e96c5ebb0af99386de58cf9553]\"                   \n",
       "[119] \"   - conda-forge/linux-64::c-ares==1.32.3=h4bc722e_0[md5=7624e34ee6baebfc80d67bac76cc9d9d]\"                 \n",
       "[120] \"   - conda-forge/linux-64::keyutils==1.6.1=h166bdaf_0[md5=30186d27e2c9fa62b45fb1476b7200e3]\"                \n",
       "[121] \"   - conda-forge/linux-64::libev==4.33=hd590300_2[md5=172bf1cd1ff8629f2b1179945ed45055]\"                    \n",
       "[122] \"   - conda-forge/linux-64::libffi==3.4.2=h7f98852_5[md5=d645c6d2ac96843a2bfaccd2d62b3ac3]\"                  \n",
       "[123] \"   - conda-forge/linux-64::libiconv==1.17=hd590300_2[md5=d66573916ffcf376178462f1b61c941e]\"                 \n",
       "[124] \"   - conda-forge/linux-64::libnsl==2.0.1=hd590300_0[md5=30fd6e37fe21f86f4bd26d6ee73eeec7]\"                  \n",
       "[125] \"   - conda-forge/linux-64::libstdcxx-ng==14.1.0=h4852527_1[md5=bd2598399a70bb86d8218e95548d735e]\"           \n",
       "[126] \"   - conda-forge/linux-64::libuuid==2.38.1=h0b41bf4_0[md5=40b61aab5c7ba9ff276c41cfffe6b80b]\"                \n",
       "[127] \"   - conda-forge/linux-64::libxcrypt==4.4.36=hd590300_1[md5=5aa797f8787fe7a17d1b0821485b5adc]\"              \n",
       "[128] \"   - conda-forge/linux-64::libzlib==1.3.1=h4ab18f5_1[md5=57d7dc60e9325e3de37ff8dffd18e814]\"                 \n",
       "[129] \"   - conda-forge/linux-64::lzo==2.10=hd590300_1001[md5=ec7398d21e2651e0dcb0044d03b9a339]\"                   \n",
       "[130] \"   - conda-forge/linux-64::ncurses==6.5=he02047a_1[md5=70caf8bb6cf39a0b6b7efc885f51c0fe]\"                   \n",
       "[131] \"   - conda-forge/linux-64::reproc==14.2.4.post0=hd590300_1[md5=82ca53502dfd5a64a80dee76dae14685]\"           \n",
       "[132] \"   - conda-forge/linux-64::xz==5.2.6=h166bdaf_0[md5=2161070d867d1b1204ea749c8eec4ef0]\"                      \n",
       "[133] \"   - conda-forge/linux-64::fmt==10.2.1=h00ab1b0_0[md5=35ef8bc24bd34074ebae3c943d551728]\"                    \n",
       "[134] \"   - conda-forge/linux-64::icu==75.1=he02047a_0[md5=8b189310083baabfb622af68fd9d3ae3]\"                      \n",
       "[135] \"   - conda-forge/linux-64::libedit==3.1.20191231=he28a2e2_2[md5=4d331e44109e3f0e19b4cb8f9b82f3e1]\"          \n",
       "[136] \"   - conda-forge/linux-64::libnghttp2==1.58.0=h47da74e_1[md5=700ac6ea6d53d5510591c4344d5c989a]\"             \n",
       "[137] \"   - conda-forge/linux-64::libsolv==0.7.30=h3509ff9_0[md5=02539b77d25aa4f65b20246549e256c3]\"                \n",
       "[138] \"   - conda-forge/linux-64::libsqlite==3.46.1=hadc24fc_0[md5=36f79405ab16bf271edb55b213836dac]\"              \n",
       "[139] \"   - conda-forge/linux-64::libssh2==1.11.0=h0841786_0[md5=1f5a58e686b13bcfde88b93f547d23fe]\"                \n",
       "[140] \"   - conda-forge/linux-64::lz4-c==1.9.4=hcb278e6_0[md5=318b08df404f9c9be5712aaa5a6f0bb0]\"                   \n",
       "[141] \"   - conda-forge/linux-64::readline==8.2=h8228510_1[md5=47d31b792659ce70f470b5c82fdfb7a4]\"                  \n",
       "[142] \"   - conda-forge/linux-64::reproc-cpp==14.2.4.post0=h59595ed_1[md5=715e1d720ec1a03715bebd237972fca5]\"       \n",
       "[143] \"   - conda-forge/linux-64::tk==8.6.13=noxft_h4845f30_101[md5=d453b98d9c83e71da0741bb0ff4d76bc]\"             \n",
       "[144] \"   - conda-forge/linux-64::yaml-cpp==0.8.0=h59595ed_0[md5=965eaacd7c18eb8361fd12bb9e7a57d7]\"                \n",
       "[145] \"   - conda-forge/linux-64::zstd==1.5.6=ha6fb4c9_0[md5=4d056880988120e29d75bfff282e0f45]\"                    \n",
       "[146] \"   - conda-forge/linux-64::krb5==1.21.3=h659f571_0[md5=3f43953b7d3fb3aaa1d0d0723d91e368]\"                   \n",
       "[147] \"   - conda-forge/linux-64::libxml2==2.12.7=he7c6b58_4[md5=08a9265c637230c37cb1be4a6cad4536]\"                \n",
       "[148] \"   - conda-forge/linux-64::python==3.12.6=hc5c86c4_1_cpython[md5=00836baacdca254f28c54d2543e97514]\"         \n",
       "[149] \"   - conda-forge/linux-64::libarchive==3.7.4=hfca40fe_0[md5=32ddb97f897740641d8d46a829ce1704]\"              \n",
       "[150] \"   - conda-forge/linux-64::libcurl==8.10.1=hbbe4b11_0[md5=6e801c50a40301f6978c53976917b277]\"                \n",
       "[151] \"   - conda-forge/linux-64::menuinst==2.1.2=py312h7900ff3_1[md5=c6575ae996f2bc0369c73b632db5ca61]\"           \n",
       "[152] \"   - conda-forge/noarch::archspec==0.2.3=pyhd8ed1ab_0[md5=192278292e20704f663b9c766909d67b]\"                \n",
       "[153] \"   - conda-forge/noarch::boltons==24.0.0=pyhd8ed1ab_0[md5=61de176bd62041f9cd5bd4fcd09eb0ff]\"                \n",
       "[154] \"   - conda-forge/linux-64::brotli-python==1.1.0=py312h2ec8cdc_2[md5=b0b867af6fc74b2a0aa206da29c0f3cf]\"      \n",
       "[155] \"   - conda-forge/noarch::certifi==2024.8.30=pyhd8ed1ab_0[md5=12f7d00853807b0531775e9be891cb11]\"             \n",
       "[156] \"   - conda-forge/noarch::charset-normalizer==3.3.2=pyhd8ed1ab_0[md5=7f4a9e3fcff3f6356ae99244a014da6a]\"      \n",
       "[157] \"   - conda-forge/noarch::colorama==0.4.6=pyhd8ed1ab_0[md5=3faab06a954c2a04039983f2c4a50d99]\"                \n",
       "[158] \"   - conda-forge/noarch::distro==1.9.0=pyhd8ed1ab_0[md5=bbdb409974cd6cb30071b1d978302726]\"                  \n",
       "[159] \"   - conda-forge/linux-64::frozendict==2.4.4=py312h66e93f0_1[md5=82a14d06575c4023a5df6b0798acff71]\"         \n",
       "[160] \"   - conda-forge/noarch::hpack==4.0.0=pyh9f0ad1d_0[md5=914d6646c4dbb1fd3ff539830a12fd71]\"                   \n",
       "[161] \"   - conda-forge/noarch::hyperframe==6.0.1=pyhd8ed1ab_0[md5=9f765cbfab6870c8435b9eefecd7a1f4]\"              \n",
       "[162] \"   - conda-forge/noarch::idna==3.10=pyhd8ed1ab_0[md5=7ba2ede0e7c795ff95088daf0dc59753]\"                     \n",
       "[163] \"   - conda-forge/linux-64::jsonpointer==3.0.0=py312h7900ff3_1[md5=6b51f7459ea4073eeb5057207e2e1e3d]\"        \n",
       "[164] \"   - conda-forge/linux-64::libmamba==1.5.9=h4cc3d14_0[md5=896cece5b883ad86e9dd88b1f4d23c99]\"                \n",
       "[165] \"   - conda-forge/noarch::packaging==24.1=pyhd8ed1ab_0[md5=cbe1bb1f21567018ce595d9c2be0f0db]\"                \n",
       "[166] \"   - conda-forge/noarch::platformdirs==4.3.6=pyhd8ed1ab_0[md5=fd8f2b18b65bbf62e8f653100690c8d2]\"            \n",
       "[167] \"   - conda-forge/noarch::pluggy==1.5.0=pyhd8ed1ab_0[md5=d3483c8fc2dc2cc3f5cf43e26d60cabf]\"                  \n",
       "[168] \"   - conda-forge/linux-64::pycosat==0.6.6=py312h98912ed_0[md5=8f1c372e7b843167be885dc8229931c1]\"            \n",
       "[169] \"   - conda-forge/noarch::pycparser==2.22=pyhd8ed1ab_0[md5=844d9eb3b43095b031874477f7d70088]\"                \n",
       "[170] \"   - conda-forge/noarch::pysocks==1.7.1=pyha2e5f31_6[md5=2a7de29fb590ca14b5243c4c812c8025]\"                 \n",
       "[171] \"   - conda-forge/linux-64::ruamel.yaml.clib==0.2.8=py312h98912ed_0[md5=05f31c2a79ba61df8d6d903ce4a4ce7b]\"   \n",
       "[172] \"   - conda-forge/noarch::setuptools==74.1.2=pyhd8ed1ab_0[md5=56c9c11d004428e81d02eeb730fc6336]\"             \n",
       "[173] \"   - conda-forge/noarch::truststore==0.9.2=pyhd8ed1ab_0[md5=f14e46d1bf271e748ff556d8b872e28a]\"              \n",
       "[174] \"   - conda-forge/noarch::wheel==0.44.0=pyhd8ed1ab_0[md5=d44e3b085abcaef02983c6305b84b584]\"                  \n",
       "[175] \"   - conda-forge/linux-64::cffi==1.17.1=py312h06ac9bb_0[md5=a861504bbea4161a9170b85d4d2be840]\"              \n",
       "[176] \"   - conda-forge/noarch::h2==4.1.0=pyhd8ed1ab_0[md5=b748fbf7060927a6e82df7cb5ee8f097]\"                      \n",
       "[177] \"   - conda-forge/noarch::jsonpatch==1.33=pyhd8ed1ab_0[md5=bfdb7c5c6ad1077c82a69a8642c87aff]\"                \n",
       "[178] \"   - conda-forge/linux-64::libmambapy==1.5.9=py312h7fb9e8e_0[md5=ccaeeb6e3caaf0c744480393791aa366]\"         \n",
       "[179] \"   - conda-forge/noarch::pip==24.2=pyh8b19718_1[md5=6c78fbb8ddfd64bcb55b5cbafd2d2c43]\"                      \n",
       "[180] \"   - conda-forge/linux-64::ruamel.yaml==0.18.6=py312h98912ed_0[md5=a99a06a875138829ef65f44bbe2c30ca]\"       \n",
       "[181] \"   - conda-forge/noarch::tqdm==4.66.5=pyhd8ed1ab_0[md5=c6e94fc2b2ec71ea33fe7c7da259acb4]\"                   \n",
       "[182] \"   - conda-forge/linux-64::zstandard==0.23.0=py312hef9b889_1[md5=8b7069e9792ee4e5b4919a7a306d2e67]\"         \n",
       "[183] \"   - conda-forge/noarch::conda-package-streaming==0.10.0=pyhd8ed1ab_0[md5=3480386e00995f7a1dfb3b9aa2fe70fd]\"\n",
       "[184] \"   - conda-forge/noarch::urllib3==2.2.3=pyhd8ed1ab_0[md5=6b55867f385dd762ed99ea687af32a69]\"                 \n",
       "[185] \"   - conda-forge/noarch::conda-package-handling==2.3.0=pyh7900ff3_0[md5=0a7dce281ae2be81acab0aa963e6bb99]\"  \n",
       "[186] \"   - conda-forge/noarch::requests==2.32.3=pyhd8ed1ab_0[md5=5ede4753180c7a550a443c430dc8ab52]\"               \n",
       "[187] \"   - conda-forge/linux-64::conda==24.7.1=py312h7900ff3_0[md5=e1bf59014e88eaff036101358f63a496]\"             \n",
       "[188] \"   - conda-forge/noarch::conda-libmamba-solver==24.7.0=pyhd8ed1ab_0[md5=857c9e25f0a77c0bd7eb622d46d9418f]\"  \n",
       "[189] \"   - conda-forge/linux-64::mamba==1.5.9=py312h9460a1c_0[md5=a8525c8a1647b4f5967fa6b552722851]\"              \n",
       "[190] \"\"                                                                                                           \n",
       "[191] \"\"                                                                                                           \n",
       "[192] \"  Package                         Version  Build               Channel         Size\"                        \n",
       "[193] \"─────────────────────────────────────────────────────────────────────────────────────\"                      \n",
       "[194] \"  Install:\"                                                                                                 \n",
       "[195] \"─────────────────────────────────────────────────────────────────────────────────────\"                      \n",
       "[196] \"\"                                                                                                           \n",
       "[197] \"  + _libgcc_mutex                     0.1  conda_forge         conda-forge         \"                        \n",
       "[198] \"  + ca-certificates             2024.8.30  hbcca054_0          conda-forge         \"                        \n",
       "[199] \"  + ld_impl_linux-64                 2.40  hf3520f5_7          conda-forge         \"                        \n",
       "[200] \"  + pybind11-abi                        4  hd8ed1ab_3          conda-forge         \"                        \n",
       "[201] \"  + python_abi                       3.12  5_cp312             conda-forge         \"                        \n",
       "[202] \"  + tzdata                          2024a  h8827d51_1          conda-forge         \"                        \n",
       "[203] \"  + libgomp                        14.1.0  h77fa898_1          conda-forge         \"                        \n",
       "[204] \"  + _openmp_mutex                     4.5  2_gnu               conda-forge         \"                        \n",
       "[205] \"  + libgcc                         14.1.0  h77fa898_1          conda-forge         \"                        \n",
       "[206] \"  + libexpat                        2.6.3  h5888daf_0          conda-forge         \"                        \n",
       "[207] \"  + libgcc-ng                      14.1.0  h69a702a_1          conda-forge         \"                        \n",
       "[208] \"  + libstdcxx                      14.1.0  hc0a3c3a_1          conda-forge         \"                        \n",
       "[209] \"  + openssl                         3.3.2  hb9d3cd8_0          conda-forge         \"                        \n",
       "[210] \"  + bzip2                           1.0.8  h4bc722e_7          conda-forge         \"                        \n",
       "[211] \"  + c-ares                         1.32.3  h4bc722e_0          conda-forge         \"                        \n",
       "[212] \"  + keyutils                        1.6.1  h166bdaf_0          conda-forge         \"                        \n",
       "[213] \"  + libev                            4.33  hd590300_2          conda-forge         \"                        \n",
       "[214] \"  + libffi                          3.4.2  h7f98852_5          conda-forge         \"                        \n",
       "[215] \"  + libiconv                         1.17  hd590300_2          conda-forge         \"                        \n",
       "[216] \"  + libnsl                          2.0.1  hd590300_0          conda-forge         \"                        \n",
       "[217] \"  + libstdcxx-ng                   14.1.0  h4852527_1          conda-forge         \"                        \n",
       "[218] \"  + libuuid                        2.38.1  h0b41bf4_0          conda-forge         \"                        \n",
       "[219] \"  + libxcrypt                      4.4.36  hd590300_1          conda-forge         \"                        \n",
       "[220] \"  + libzlib                         1.3.1  h4ab18f5_1          conda-forge         \"                        \n",
       "[221] \"  + lzo                              2.10  hd590300_1001       conda-forge         \"                        \n",
       "[222] \"  + ncurses                           6.5  he02047a_1          conda-forge         \"                        \n",
       "[223] \"  + reproc                   14.2.4.post0  hd590300_1          conda-forge         \"                        \n",
       "[224] \"  + xz                              5.2.6  h166bdaf_0          conda-forge         \"                        \n",
       "[225] \"  + fmt                            10.2.1  h00ab1b0_0          conda-forge         \"                        \n",
       "[226] \"  + icu                              75.1  he02047a_0          conda-forge         \"                        \n",
       "[227] \"  + libedit                  3.1.20191231  he28a2e2_2          conda-forge         \"                        \n",
       "[228] \"  + libnghttp2                     1.58.0  h47da74e_1          conda-forge         \"                        \n",
       "[229] \"  + libsolv                        0.7.30  h3509ff9_0          conda-forge         \"                        \n",
       "[230] \"  + libsqlite                      3.46.1  hadc24fc_0          conda-forge         \"                        \n",
       "[231] \"  + libssh2                        1.11.0  h0841786_0          conda-forge         \"                        \n",
       "[232] \"  + lz4-c                           1.9.4  hcb278e6_0          conda-forge         \"                        \n",
       "[233] \"  + readline                          8.2  h8228510_1          conda-forge         \"                        \n",
       "[234] \"  + reproc-cpp               14.2.4.post0  h59595ed_1          conda-forge         \"                        \n",
       "[235] \"  + tk                             8.6.13  noxft_h4845f30_101  conda-forge         \"                        \n",
       "[236] \"  + yaml-cpp                        0.8.0  h59595ed_0          conda-forge         \"                        \n",
       "[237] \"  + zstd                            1.5.6  ha6fb4c9_0          conda-forge         \"                        \n",
       "[238] \"  + krb5                           1.21.3  h659f571_0          conda-forge         \"                        \n",
       "[239] \"  + libxml2                        2.12.7  he7c6b58_4          conda-forge         \"                        \n",
       "[240] \"  + python                         3.12.6  hc5c86c4_1_cpython  conda-forge         \"                        \n",
       "[241] \"  + libarchive                      3.7.4  hfca40fe_0          conda-forge         \"                        \n",
       "[242] \"  + libcurl                        8.10.1  hbbe4b11_0          conda-forge         \"                        \n",
       "[243] \"  + menuinst                        2.1.2  py312h7900ff3_1     conda-forge         \"                        \n",
       "[244] \"  + archspec                        0.2.3  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[245] \"  + boltons                        24.0.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[246] \"  + brotli-python                   1.1.0  py312h2ec8cdc_2     conda-forge         \"                        \n",
       "[247] \"  + certifi                     2024.8.30  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[248] \"  + charset-normalizer              3.3.2  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[249] \"  + colorama                        0.4.6  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[250] \"  + distro                          1.9.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[251] \"  + frozendict                      2.4.4  py312h66e93f0_1     conda-forge         \"                        \n",
       "[252] \"  + hpack                           4.0.0  pyh9f0ad1d_0        conda-forge         \"                        \n",
       "[253] \"  + hyperframe                      6.0.1  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[254] \"  + idna                             3.10  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[255] \"  + jsonpointer                     3.0.0  py312h7900ff3_1     conda-forge         \"                        \n",
       "[256] \"  + libmamba                        1.5.9  h4cc3d14_0          conda-forge         \"                        \n",
       "[257] \"  + packaging                        24.1  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[258] \"  + platformdirs                    4.3.6  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[259] \"  + pluggy                          1.5.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[260] \"  + pycosat                         0.6.6  py312h98912ed_0     conda-forge         \"                        \n",
       "[261] \"  + pycparser                        2.22  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[262] \"  + pysocks                         1.7.1  pyha2e5f31_6        conda-forge         \"                        \n",
       "[263] \"  + ruamel.yaml.clib                0.2.8  py312h98912ed_0     conda-forge         \"                        \n",
       "[264] \"  + setuptools                     74.1.2  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[265] \"  + truststore                      0.9.2  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[266] \"  + wheel                          0.44.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[267] \"  + cffi                           1.17.1  py312h06ac9bb_0     conda-forge         \"                        \n",
       "[268] \"  + h2                              4.1.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[269] \"  + jsonpatch                        1.33  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[270] \"  + libmambapy                      1.5.9  py312h7fb9e8e_0     conda-forge         \"                        \n",
       "[271] \"  + pip                              24.2  pyh8b19718_1        conda-forge         \"                        \n",
       "[272] \"  + ruamel.yaml                    0.18.6  py312h98912ed_0     conda-forge         \"                        \n",
       "[273] \"  + tqdm                           4.66.5  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[274] \"  + zstandard                      0.23.0  py312hef9b889_1     conda-forge         \"                        \n",
       "[275] \"  + conda-package-streaming        0.10.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[276] \"  + urllib3                         2.2.3  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[277] \"  + conda-package-handling          2.3.0  pyh7900ff3_0        conda-forge         \"                        \n",
       "[278] \"  + requests                       2.32.3  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[279] \"  + conda                          24.7.1  py312h7900ff3_0     conda-forge         \"                        \n",
       "[280] \"  + conda-libmamba-solver          24.7.0  pyhd8ed1ab_0        conda-forge         \"                        \n",
       "[281] \"  + mamba                           1.5.9  py312h9460a1c_0     conda-forge         \"                        \n",
       "[282] \"\"                                                                                                           \n",
       "[283] \"  Summary:\"                                                                                                 \n",
       "[284] \"\"                                                                                                           \n",
       "[285] \"  Install: 85 packages\"                                                                                     \n",
       "[286] \"\"                                                                                                           \n",
       "[287] \"  Total download: 0 B\"                                                                                      \n",
       "[288] \"\"                                                                                                           \n",
       "[289] \"─────────────────────────────────────────────────────────────────────────────────────\"                      \n",
       "[290] \"\"                                                                                                           \n",
       "[291] \"\"                                                                                                           \n",
       "[292] \"\"                                                                                                           \n",
       "[293] \"Transaction starting\"                                                                                       \n",
       "[294] \"\"                                                                                                           \n",
       "[295] \"Transaction finished\"                                                                                       \n",
       "[296] \"\"                                                                                                           \n",
       "[297] \"To activate this environment, use:\"                                                                         \n",
       "[298] \"\"                                                                                                           \n",
       "[299] \"    micromamba activate /home/ec2-user/miniforge\"                                                           \n",
       "[300] \"\"                                                                                                           \n",
       "[301] \"Or to execute a single command in this environment, use:\"                                                   \n",
       "[302] \"\"                                                                                                           \n",
       "[303] \"    micromamba run -p /home/ec2-user/miniforge mycommand\"                                                   \n",
       "[304] \"\"                                                                                                           \n",
       "[305] \"installation finished.\"                                                                                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Download Miniforge\n",
    "system('curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh', intern = TRUE)\n",
    "\n",
    "# Install Miniforge\n",
    "system('bash Miniforge3-$(uname)-$(uname -m).sh -b -u -p $HOME/miniforge', intern = TRUE)\n",
    "\n",
    "# Add Miniforge bin to the system path\n",
    "Sys.setenv(PATH = paste(Sys.getenv(\"HOME\"), \"/miniforge/bin:\", Sys.getenv(\"PATH\"), sep = \"\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "620961e5-909d-45fa-b538-74897c6b67eb",
   "metadata": {},
   "source": [
    "Next, using mambaforge and bioconda, install the tools that will be used in this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4577d77f-bcf0-4f70-81a8-28689d49628c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Transaction'</li><li>''</li><li><span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/anaconda3/envs/R'</span></li><li>''</li><li><span style=white-space:pre-wrap>'  All requested packages already installed'</span></li><li>''</li><li>''</li><li>'Looking for: [\\'gsutil\\']'</li><li>''</li><li>''</li><li>'Pinned packages:'</li><li><span style=white-space:pre-wrap>'  - python 3.11.*'</span></li><li>''</li><li>''</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Transaction'\n",
       "\\item ''\n",
       "\\item '  Prefix: /home/ec2-user/anaconda3/envs/R'\n",
       "\\item ''\n",
       "\\item '  All requested packages already installed'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'Looking for: {[}\\textbackslash{}'gsutil\\textbackslash{}'{]}'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'Pinned packages:'\n",
       "\\item '  - python 3.11.*'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Transaction'\n",
       "2. ''\n",
       "3. <span style=white-space:pre-wrap>'  Prefix: /home/ec2-user/anaconda3/envs/R'</span>\n",
       "4. ''\n",
       "5. <span style=white-space:pre-wrap>'  All requested packages already installed'</span>\n",
       "6. ''\n",
       "7. ''\n",
       "8. 'Looking for: [\\'gsutil\\']'\n",
       "9. ''\n",
       "10. ''\n",
       "11. 'Pinned packages:'\n",
       "12. <span style=white-space:pre-wrap>'  - python 3.11.*'</span>\n",
       "13. ''\n",
       "14. ''\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Transaction\"                               \n",
       " [2] \"\"                                          \n",
       " [3] \"  Prefix: /home/ec2-user/anaconda3/envs/R\" \n",
       " [4] \"\"                                          \n",
       " [5] \"  All requested packages already installed\"\n",
       " [6] \"\"                                          \n",
       " [7] \"\"                                          \n",
       " [8] \"Looking for: ['gsutil']\"                   \n",
       " [9] \"\"                                          \n",
       "[10] \"\"                                          \n",
       "[11] \"Pinned packages:\"                          \n",
       "[12] \"  - python 3.11.*\"                         \n",
       "[13] \"\"                                          \n",
       "[14] \"\"                                          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Install gsutil and dependencies using mamba\n",
    "system('mamba install -y -c conda-forge -c bioconda gsutil', intern = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f242aa89-48ec-492e-8958-cf9ff2b0090e",
   "metadata": {},
   "source": [
    "### STEP 1.2 Install Bioconductor Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6bff2367-60d7-4bee-9bb0-c1b48f6a6fc7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    BioCsoft: https://bioconductor.org/packages/3.19/bioc\n",
      "    BioCann: https://bioconductor.org/packages/3.19/data/annotation\n",
      "    BioCexp: https://bioconductor.org/packages/3.19/data/experiment\n",
      "    BioCworkflows: https://bioconductor.org/packages/3.19/workflows\n",
      "    BioCbooks: https://bioconductor.org/packages/3.19/books\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n",
      "\n",
      "Installing package(s) 'ComplexHeatmap', 'DESeq2', 'edgeR'\n",
      "\n",
      "also installing the dependencies ‘statmod’, ‘limma’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "Old packages: 'data.table', 'MASS', 'Matrix', 'nlme', 'ps'\n",
      "\n",
      "also installing the dependencies ‘terra’, ‘crosstalk’, ‘whoami’, ‘hunspell’, ‘RMySQL’, ‘RPostgreSQL’, ‘raster’, ‘DT’, ‘foghorn’, ‘gmailr’, ‘mockery’, ‘pingr’, ‘rhub’, ‘spelling’, ‘R.oo’, ‘R.methodsS3’\n",
      "\n",
      "\n",
      "Warning message in install.packages(c(\"dplyr\", \"pheatmap\", \"ggrepel\", \"ggfortify\", :\n",
      "“installation of package ‘terra’ had non-zero exit status”\n",
      "Warning message in install.packages(c(\"dplyr\", \"pheatmap\", \"ggrepel\", \"ggfortify\", :\n",
      "“installation of package ‘RMySQL’ had non-zero exit status”\n",
      "Warning message in install.packages(c(\"dplyr\", \"pheatmap\", \"ggrepel\", \"ggfortify\", :\n",
      "“installation of package ‘RPostgreSQL’ had non-zero exit status”\n",
      "Warning message in install.packages(c(\"dplyr\", \"pheatmap\", \"ggrepel\", \"ggfortify\", :\n",
      "“installation of package ‘raster’ had non-zero exit status”\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Install BiocManager if not installed\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "# Set repositories\n",
    "options(repos = BiocManager::repositories())\n",
    "\n",
    "# Install Bioconductor packages\n",
    "BiocManager::install(c(\"ComplexHeatmap\", \"DESeq2\", \"edgeR\"), force = TRUE)\n",
    "\n",
    "# Install CRAN packages\n",
    "install.packages(c(\"dplyr\", \"pheatmap\", \"ggrepel\", \"ggfortify\", \"devtools\", \"R.utils\"), dependencies = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "086be68c-2bf5-4fcb-8e0b-c63c88bfeed4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n",
      "also installing the dependencies ‘filelock’, ‘BiocFileCache’\n",
      "\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Organism Annotation Package - Zebrafish\n",
    "install.packages(\"org.Dr.eg.db\")\n",
    "library(org.Dr.eg.db)\n",
    "install.packages(\"biomaRt\")\n",
    "library(biomaRt)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d429aded-447b-419e-a4f7-74f235996d1a",
   "metadata": {},
   "source": [
    "### STEP 1.3 Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d764cad0-34c8-4774-bcf8-7a806a509bc6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.20.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: ‘limma’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:DESeq2’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "Loading required package: usethis\n",
      "\n",
      "Loading required package: R.oo\n",
      "\n",
      "Loading required package: R.methodsS3\n",
      "\n",
      "R.methodsS3 v1.8.2 (2022-06-13 22:00:14 UTC) successfully loaded. See ?R.methodsS3 for help.\n",
      "\n",
      "R.oo v1.26.0 (2024-01-24 05:12:50 UTC) successfully loaded. See ?R.oo for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.oo’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:R.methodsS3’:\n",
      "\n",
      "    throw\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:devtools’:\n",
      "\n",
      "    check, unload\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:IRanges’:\n",
      "\n",
      "    trim\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:methods’:\n",
      "\n",
      "    getClasses, getMethods\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    attach, detach, load, save\n",
      "\n",
      "\n",
      "R.utils v2.12.3 (2023-11-18 01:00:02 UTC) successfully loaded. See ?R.utils for help.\n",
      "\n",
      "\n",
      "Attaching package: ‘R.utils’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ComplexHeatmap’:\n",
      "\n",
      "    draw\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    cat, commandArgs, getOption, isOpen, nullfile, parse, use, warnings\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(DESeq2)\n",
    "library(dplyr)\n",
    "library(ComplexHeatmap)\n",
    "library(edgeR)\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(ggfortify)\n",
    "library(devtools)\n",
    "library(Biobase)\n",
    "library(R.utils)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54c0992c-0159-417f-a792-ff9aa26ead5c",
   "metadata": {},
   "source": [
    "### STEP 1.4 Install and Load NetAct from GitHub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dce285fe-633e-4c26-8f62-0588d84b09f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading GitHub repo lusystemsbio/NetAct@HEAD\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biostrings   (NA     -> 2.72.1  ) [CRAN]\n",
      "KEGGREST     (NA     -> 1.44.1  ) [CRAN]\n",
      "umap         (NA     -> 0.2.10.0) [CRAN]\n",
      "visNetwork   (NA     -> 2.1.2   ) [CRAN]\n",
      "Annotatio... (NA     -> 1.66.0  ) [CRAN]\n",
      "fastmatch    (NA     -> 1.1-4   ) [CRAN]\n",
      "cowplot      (NA     -> 1.1.3   ) [CRAN]\n",
      "data.table   (1.15.4 -> 1.16.0  ) [CRAN]\n",
      "sRACIPE      (NA     -> 1.20.0  ) [CRAN]\n",
      "org.Hs.eg.db (NA     -> 3.19.1  ) [CRAN]\n",
      "org.Mm.eg.db (NA     -> 3.19.1  ) [CRAN]\n",
      "infotheo     (NA     -> 1.2.0.1 ) [CRAN]\n",
      "entropy      (NA     -> 1.3.1   ) [CRAN]\n",
      "mclust       (NA     -> 6.1.1   ) [CRAN]\n",
      "fgsea        (NA     -> 1.30.0  ) [CRAN]\n",
      "qvalue       (NA     -> 2.36.0  ) [CRAN]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing 16 packages: Biostrings, KEGGREST, umap, visNetwork, AnnotationDbi, fastmatch, cowplot, data.table, sRACIPE, org.Hs.eg.db, org.Mm.eg.db, infotheo, entropy, mclust, fgsea, qvalue\n",
      "\n",
      "Updating HTML index of packages in '.Library'\n",
      "\n",
      "Making 'packages.html' ...\n",
      " done\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[36m──\u001b[39m \u001b[36mR CMD build\u001b[39m \u001b[36m─────────────────────────────────────────────────────────────────\u001b[39m\n",
      "* checking for file ‘/tmp/RtmpLFr8ZR/remotes3142c04fd62/lusystemsbio-NetAct-f160a23/DESCRIPTION’ ... OK\n",
      "* preparing ‘NetAct’:\n",
      "* checking DESCRIPTION meta-information ... OK\n",
      "* cleaning src\n",
      "* checking for LF line-endings in source and make files and shell scripts\n",
      "* checking for empty or unneeded directories\n",
      "  NB: this package now depends on R (>= 3.5.0)\n",
      "  WARNING: Added dependency on R >= 3.5.0 because serialized objects in\n",
      "  serialize/load version 3 cannot be read in older versions of R.\n",
      "  File(s) containing such objects:\n",
      "    ‘NetAct/data/hDB.rdata’ ‘NetAct/data/mDB.rdata’\n",
      "    ‘NetAct/vignettes/gsearslts_tutorial.RDS’\n",
      "* building ‘NetAct_1.0.7.tar.gz’\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“replacing previous import ‘mclust::logsumexp’ by ‘limma::logsumexp’ when loading ‘NetAct’”\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "devtools::install_github(\"lusystemsbio/NetAct\", dependencies = TRUE, build_vignettes = FALSE)\n",
    "# Load NetAct\n",
    "library(NetAct)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29bfb5ec-1b98-42f0-9364-3c51021b4a85",
   "metadata": {},
   "source": [
    "## STEP 2: Download FASTQ Files from S3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ef5b2d6-a34c-48a7-a7e3-14d4ccec58ca",
   "metadata": {},
   "source": [
    "Download the read counts from the S3 bucket created by STAR and RSEM analysis in Tutorial 1b."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "564694b9-90a3-4cc3-9723-0c63ab9b7653",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define the correct path for your files\n",
    "path <- \"data/counts/\"\n",
    "\n",
    "# Load SRR IDs from accs.txt (if available)\n",
    "accs <- readLines('accs.txt')\n",
    "\n",
    "for (acc in accs) {\n",
    "  # Construct the URL for the S3 bucket\n",
    "  url <- paste0(\"https://sra-data-athena.s3.amazonaws.com/readcounts/\", acc, \".genes_zebrafish.txt\")\n",
    "  \n",
    "  # Construct the wget command\n",
    "  cmd <- paste(\"wget -P data/counts/ \", url)\n",
    "  \n",
    "  # Execute the shell command\n",
    "  system(cmd)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dccb3534-b6ba-4c65-adc0-5243a88c87a1",
   "metadata": {},
   "source": [
    "## STEP 3: Read Count Data and Define Experimental Groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c20552ac-6465-4f8e-a6f8-67a0f92dc8ea",
   "metadata": {},
   "source": [
    "Step 3 is focused on reading in the gene expression data from previously generated files and organizing it for further analysis, specifically for differential expression analysis using the edgeR package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3afb5591-c363-4a95-a5a0-78d2f9b286a1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# List the files with full paths\n",
    "files <- paste0(path, c(\"SRR3371417.genes_zebrafish.txt\", \"SRR3371418.genes_zebrafish.txt\", \n",
    "                        \"SRR3371419.genes_zebrafish.txt\", \"SRR3371420.genes_zebrafish.txt\",\n",
    "                        \"SRR3371421.genes_zebrafish.txt\", \"SRR3371422.genes_zebrafish.txt\"))\n",
    "\n",
    "# Read the DGE data from the files\n",
    "x <- edgeR::readDGE(files, columns=c(\"GeneID\",\"Count\"))\n",
    "x$counts <- round(x$counts)\n",
    "\n",
    "# Define the group factor\n",
    "group <- as.factor(c(\"control\", \"control\", \"control\", \"cortisol\", \"cortisol\", \"cortisol\"))\n",
    "xgroup <- group\n",
    "\n",
    "# Define the sample names and set column names\n",
    "samplenames <- c(\"control1\", \"control2\", \"control3\", \"cortisol1\", \"cortisol2\", \"cortisol3\")\n",
    "colnames(x) <- samplenames"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4a9c7dd-9484-4b08-b18c-485e220f8365",
   "metadata": {},
   "source": [
    "## STEP 4: Data Preprocessing for Differential Expression Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e431ffc4-5cd1-4f7c-ae74-f33d4aa72289",
   "metadata": {},
   "source": [
    "rownames(phenoData)This preprocesses the RNA-seq count data to prepare it for differential expression analysis. It creates a comparison list for the analysis, constructs phenotype data to annotate samples, preprocesses the counts, and removes any duplicate gene entries. These steps ensure the quality and integrity of the data before applying statistical methods to identify differentially expressed genes between the specified groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "19cefccc-6ab1-403d-9289-c403f3a159fc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "compList <- c(\"control-cortisol\")\n",
    "phenoData = new(\"AnnotatedDataFrame\", data = data.frame(celltype = group))\n",
    "rownames(phenoData) = colnames(x$counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "165b08cf-503e-43ba-bca6-1058134bd73e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in .testForValidKeys(x, keys, keytype, fks): None of the keys entered are valid keys for 'ENTREZID'. Please use the keys method to see a listing of valid arguments.\n",
     "output_type": "error",
     "traceback": [
      "Error in .testForValidKeys(x, keys, keytype, fks): None of the keys entered are valid keys for 'ENTREZID'. Please use the keys method to see a listing of valid arguments.\nTraceback:\n",
      "1. AnnotationDbi::select(org.Hs.eg.db, keys = geneid, columns = \"SYMBOL\", \n .     keytype = keytype)",
      "2. AnnotationDbi::select(org.Hs.eg.db, keys = geneid, columns = \"SYMBOL\", \n .     keytype = keytype)",
      "3. .select(x, keys, columns, keytype, jointype = jointype, ...)",
      "4. testSelectArgs(x, keys = keys, cols = cols, keytype = keytype, \n .     fks = fks, skipValidKeysTest = skipValidKeysTest)",
      "5. .testForValidKeys(x, keys, keytype, fks)",
      "6. stop(msg)",
      "7. .handleSimpleError(function (cnd) \n . {\n .     watcher$capture_plot_and_output()\n .     cnd <- sanitize_call(cnd)\n .     watcher$push(cnd)\n .     switch(on_error, continue = invokeRestart(\"eval_continue\"), \n .         stop = invokeRestart(\"eval_stop\"), error = invokeRestart(\"eval_error\", \n .             cnd))\n . }, \"None of the keys entered are valid keys for 'ENTREZID'. Please use the keys method to see a listing of valid arguments.\", \n .     base::quote(.testForValidKeys(x, keys, keytype, fks)))"
     ]
    }
   ],
   "source": [
    "counts <- Preprocess_counts(counts = x$counts, groups = group, mouse = FALSE)\n",
    "# Remove duplicate rows, keeping the first occurrence\n",
    "counts <- counts[!duplicated(rownames(counts)), ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f70c090f-338f-4200-badb-a83a3439317c",
   "metadata": {},
   "source": [
    "### STEP 5: Run Differential Expression Analysis and Create Expression Set"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da4bacb7-35d1-445b-a0a2-664223094111",
   "metadata": {},
   "source": [
    "This step identifies genes that are differentially expressed between the specified experimental groups and prepares the data for downstream analysis by creating a structured expression set object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9f64911-a601-4d6b-acc8-ee430f711971",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DErslt = RNAseqDegs_limma(counts = counts, phenodata = phenoData, \n",
    "                          complist = compList, qval = 0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98cf4dfb-1a82-4cf6-8cff-3666c2fdcafa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "neweset = Biobase::ExpressionSet(assayData = as.matrix(DErslt$e), phenoData = phenoData)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2b7bd39-29f4-499e-8c25-0e9eb2a8a9cd",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DErslt$e"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f83f357c-eed5-4356-aae5-670939fa9b56",
   "metadata": {},
   "source": [
    "### STEP 6: Perform Transcription Factor (TF) Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ab7b1f7-aa13-4293-b6f0-e28cc474421e",
   "metadata": {},
   "source": [
    "This step will identify transcription factors (TFs) that are potentially involved in regulating the differentially expressed genes. By utilizing a database of TFs and employing statistical testing, this analysis aims to uncover crucial regulators that may contribute to the observed changes in gene expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "349f1bc7-6c16-4f75-a293-6e7adfeb23c5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data(\"mDB\")\n",
    "calc <- TRUE\n",
    "\n",
    "if (calc) {\n",
    "  gsearslts <- TF_Selection(GSDB = mDB, DErslt = DErslt, minSize = 5, nperm = 10000,\n",
    "                            qval = 0.05, compList = compList,\n",
    "                            nameFile = \"zebrafish_gsearslts\")\n",
    "} else {\n",
    "  gsearslts <- readRDS(file = \"zebrafish_gsearslts.RDS\")\n",
    "}\n",
    "\n",
    "tfs <- gsearslts$tfs\n",
    "tfs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9b585be-5a0a-4d31-8a50-fdea8b99623b",
   "metadata": {},
   "source": [
    "### STEP 7: Reselect Transcription Factors (TFs) by Applying a Stricter q-value Threshold of 0.01"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5001fde-0711-475b-995a-c094e0045939",
   "metadata": {},
   "source": [
    "The revised text accurately conveys the purpose of this step. By applying a stricter q-value threshold of 0.01, the code refines the selection of TFs, ensuring that only those with a higher level of statistical significance are retained. This refinement is crucial for downstream analyses as it helps to prioritize the most relevant TFs that may be driving the observed gene expression changes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50fa9931-69d2-4f40-8993-625a4309419f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Reselect_TFs(GSEArslt = gsearslts$GSEArslt, qval = 0.01)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5789efa1-601e-440e-95ae-3c02a6a432a9",
   "metadata": {},
   "source": [
    "### STEP 8: Calculate TF Activity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9b432ea-9ec2-44bc-9924-16f922a0f62f",
   "metadata": {},
   "source": [
    "Calculating TF activity is crucial for understanding the regulatory mechanisms underlying the differential expression of genes. By assessing how active specific transcription factors are in different experimental groups, researchers can gain insights into the regulatory networks that may influence biological processes and pathways of interest. The heatmap visualization further aids in the interpretation of these results, highlighting the relationships between TF activity and gene expression patterns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f01319ac-fb54-4fc1-aa39-89a468452ea3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "act.me <- TF_Activity(tfs, mDB, neweset, DErslt)\n",
    "acts_mat = act.me$all_activities\n",
    "Activity_heatmap(acts_mat, neweset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7351491c-c4ff-4787-bcc8-8ab74525df0c",
   "metadata": {},
   "source": [
    "### STEP 9: Build TF Network and Simulate Circuit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8bb6266-9e14-4ab2-a514-080e60cd2948",
   "metadata": {},
   "source": [
    "TF_Filter function constructs a filtered network of transcription factor interactions based on activity levels and known regulatory links, setting the stage for subsequent analysis and the transcriptional regulatory network plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94601b12-47c8-436a-a7e7-9d93a0a0bc89",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "tf_links = TF_Filter(acts_mat, mDB, miTh = .05, nbins = 8, corMethod = \"spearman\", DPI = T)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e0762b0-da02-4f14-8107-b02234842d6f",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 10: Plot the transcription factor network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a583c03-eea1-4573-b20d-d6e8560ec06c",
   "metadata": {},
   "source": [
    "After the TF analysis, the plot_network function effectively visualizes the transcription factor network, enabling researchers to intuitively understand the complex regulatory interactions among TFs and their target genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcdf312b-1048-47e1-86bf-275280bb9e2b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "plot_network(tf_links)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bee814d4-59ad-4cd0-9cfe-7ed15618a88b",
   "metadata": {},
   "source": [
    "### STEP 11: Run RACIPE simulation on the constructed GRN"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ff99700-4474-4ddc-b9ec-c654481c1f42",
   "metadata": {},
   "source": [
    "Lastly, the RACIPE simulation is executed on the constructed gene regulatory network, allowing for the exploration of gene expression dynamics under varying regulatory parameters. The results can help to derive conclusions about the potential regulatory mechanisms at play and to identify key transcription factors that might be crucial in the system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56845776-a7bb-40cf-b04f-9775012fc3ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "racipe_results <- sRACIPE::sracipeSimulate(circuit = tf_links, numModels = 200, plots = TRUE)"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": ".m112",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/:m112"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
