{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Bulk RNA-Seq Analysis Training Demo"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This short tutorial demonstrates how to run an bulk RNA-Seq analysis workflow using a subset of reads from a zebrafish (*Danio rerio*) data set. Steps in the workflow include read trimming, read QC, read mapping, and counting mapped reads per gene to estimate gene expression. This tutorial workflow uses a subset of sequence reads from a study published by Hartig et al (PubMed ID: [2746894](https://pubmed.ncbi.nlm.nih.gov/32746894/); Gene Expression Omnibus: [GSE137987](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE137987)). In this study, the authors were interested in examining how chronic early life stress alters gene expression in adulthood. RNA samples were extracted from the blood of adult zebrafish who were exposed to cortisol or control conditions during just the first 5 days of development. There were three control (DMSO-treated) samples, and three cortisol-treated samples.\n",
    "\n",
    "This tutorial only analyzes a subset of the read data (1 million reads per sample) from this experiment and aligns them to just chromosome 14 rather than the whole genome because analyzing the entire data set takes much longer to run. For those who are interested, we do provide another tutorial where the entire data set is analyzed but it takes several hours to run.\n",
    "\n",
    "![RNA-Seq workflow](../../images/Tutorial_1_Workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## STEP 1: Install Miniforge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Miniforge is a lightweight Conda distribution that offers a streamlined installation process and efficient package management. It provides access to a vast repository of packages.\n",
    "\n",
    "Conda packages and environments are useful for several reasons. Conda packages contain metadata. This metadata includes information about what other programs the given software needs to be installed, in order to run. When installing a package with Conda, those other packages are automatically also installed. In this way, the user does not have to worry about manually installing each dependency. This makes installation quick and simple.\n",
    "\n",
    "These packages are installed inside of environments, which are simply folders within the local installation of Conda. This has several benefits. Local installation means easier installation for non-admin users who may not have access to all system directories. Each environment can hold specific software with specific versions, and it easy to swap to different environments. In addition, the environments themselves are portable, as each environment contains a manifest on how to recreate that environment.\n",
    "\n",
    "Miniforge itself is a Conda package manager, this means it requires Conda in order to work. It is used to install and update Conda packages, which it gets from a ‘channel’, or repository. It is an alternative to the native Conda package manager. It is often used for reasons of speed.\n",
    "\n",
    "Bioconda is a ‘channel’, or repository, that the Mambaforge package manager can download packages from. It is a repository of Conda packages that are related to biology. These packages are versions of popular biology software that are curated and uploaded by contributing users."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\n",
      "100 74.3M  100 74.3M    0     0  89.1M      0 --:--:-- --:--:-- --:--:--  182M\n",
      "19:28:20\n"
     ]
    }
   ],
   "source": [
    "# Download Miniforge or Mambaforge (you can use either based on preference)\n",
    "!curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh > /dev/null\n",
    "\n",
    "# Install Miniforge (or Mambaforge) - no need to install conda since mamba will be available immediately\n",
    "!bash Miniforge3-$(uname)-$(uname -m).sh -b -u -p $HOME/miniforge > /dev/null\n",
    "!date +\"%T\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, using miniforge and bioconda, install the tools that will be used in this tutorial."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n",
      "\u001b[33m\u001b[1mwarning  libmamba\u001b[m Could not parse mod/etag header\n"
     ]
    }
   ],
   "source": [
    "# Update PATH to point to the Miniforge (or Mambaforge) bin files\n",
    "import os\n",
    "os.environ[\"PATH\"] += os.pathsep + os.environ[\"HOME\"]+\"/miniforge/bin\" \n",
    "\n",
    "# Now use mamba directly to install your software packages\n",
    "!mamba install -y -c conda-forge -c bioconda trimmomatic fastqc multiqc sql-magic entrez-direct gffread parallel-fastq-dump sra-tools pyathena samtools star rsem subread pigz -y > /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## STEP 2: Setup Environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a set of directories to store the reads, reference sequence files, and output files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/SageMaker\n"
     ]
    }
   ],
   "source": [
    "!cd $HOMEDIR\n",
    "!echo $PWD\n",
    "!mkdir -p data_14\n",
    "!mkdir -p data_14/trimmed\n",
    "!mkdir -p data_14/fastqc/\n",
    "!mkdir -p data_14/fastqc_samples/\n",
    "!mkdir -p data_14/raw_fastq/\n",
    "!mkdir -p data_14/reference\n",
    "!mkdir -p data_14/aligned_bam\n",
    "!mkdir -p data_14/rsem_output\n",
    "!mkdir -p data_14/reference/STAR_index_14"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the number of available threads based on the VM.\n",
    "\n",
    "This is useful for later tools such as trimmomatic, or STAR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of threads: 15\n"
     ]
    }
   ],
   "source": [
    "import multiprocessing\n",
    "\n",
    "num_cores = multiprocessing.cpu_count()\n",
    "THREADS = max(1, num_cores - 1)\n",
    "\n",
    "print(\"Number of threads:\", THREADS)\n",
    "os.environ[\"THREADS\"] = str(THREADS)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## STEP 3: Downloading relevant files\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.1: Finding run accession numbers.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code retrieves the SRR IDs associated with project PRJNA892075 from the NCBI database and saves them to a .txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR3371417\n",
      "SRR3371418\n",
      "SRR3371419\n",
      "SRR3371420\n",
      "SRR3371421\n",
      "SRR3371422\n"
     ]
    }
   ],
   "source": [
    "!esearch -db sra -query \"PRJNA318296\" | efetch -format runinfo | cut -d',' -f1 | tail -n +2 > accs.txt\n",
    "!cat accs.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3.2: Download subsampling FASTQ files from S3 bucket"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order for this tutorial to run quickly, we will only analyze 1,000,000 reads from each of the eight samples in PRJNA892075. These subsampled files have been posted on a publicly accessible S3 bucket.\n",
    "\n",
    "The code will connect to the S3 bucket and download the paired-end FASTQ files for each SRR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Load SRR IDs from accs.txt (if available)\n",
    "with open('accs.txt', 'r') as f:\n",
    "    accs = [line.strip() for line in f.readlines()]\n",
    "    \n",
    "# Define the path to S3 bucket\n",
    "\n",
    "s3_path = 's3://sra-data-athena/fastqfiles/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-01-02 19:33:17--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371417_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.217.254.33, 54.231.160.89, 3.5.27.211, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.217.254.33|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124264970 (119M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371417_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.51M  54.9MB/s    in 2.2s    \n",
      "\n",
      "2025-01-02 19:33:20 (54.9 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371417_1.fastq.4’ saved [124264970/124264970]\n",
      "\n",
      "--2025-01-02 19:33:20--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371417_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 3.5.28.215, 52.216.59.209, 52.216.219.153, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|3.5.28.215|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124085987 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371417_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.34M  48.2MB/s    in 2.5s    \n",
      "\n",
      "2025-01-02 19:33:23 (48.2 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371417_2.fastq.4’ saved [124085987/124085987]\n",
      "\n",
      "--2025-01-02 19:33:23--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371418_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 3.5.30.61, 3.5.24.197, 54.231.228.201, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|3.5.30.61|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124194532 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371418_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.44M  52.3MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:25 (52.3 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371418_1.fastq.4’ saved [124194532/124194532]\n",
      "\n",
      "--2025-01-02 19:33:25--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371418_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.217.173.161, 52.217.171.17, 54.231.234.129, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.217.173.161|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124174013 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371418_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.42M  53.6MB/s    in 2.2s    \n",
      "\n",
      "2025-01-02 19:33:28 (53.6 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371418_2.fastq.4’ saved [124174013/124174013]\n",
      "\n",
      "--2025-01-02 19:33:28--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371419_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 3.5.8.196, 3.5.29.162, 52.217.40.44, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|3.5.8.196|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124362880 (119M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371419_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.60M  51.8MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:30 (51.8 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371419_1.fastq.4’ saved [124362880/124362880]\n",
      "\n",
      "--2025-01-02 19:33:30--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371419_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.216.33.97, 3.5.0.84, 3.5.28.69, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.216.33.97|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124085978 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371419_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.34M  49.8MB/s    in 2.4s    \n",
      "\n",
      "2025-01-02 19:33:33 (49.8 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371419_2.fastq.4’ saved [124085978/124085978]\n",
      "\n",
      "--2025-01-02 19:33:33--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371420_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 16.182.65.193, 3.5.19.1, 52.217.162.81, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|16.182.65.193|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124150136 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371420_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.40M  51.5MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:36 (51.5 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371420_1.fastq.4’ saved [124150136/124150136]\n",
      "\n",
      "--2025-01-02 19:33:36--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371420_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.217.197.161, 54.231.200.57, 3.5.1.120, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.217.197.161|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124144324 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371420_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.39M  50.4MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:38 (50.4 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371420_2.fastq.4’ saved [124144324/124144324]\n",
      "\n",
      "--2025-01-02 19:33:39--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371421_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 54.231.137.201, 52.217.65.180, 16.182.64.73, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|54.231.137.201|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124212548 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371421_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.46M  52.4MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:41 (52.4 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371421_1.fastq.4’ saved [124212548/124212548]\n",
      "\n",
      "--2025-01-02 19:33:41--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371421_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.217.167.177, 54.231.165.81, 52.217.202.81, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.217.167.177|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124235014 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371421_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.48M  51.1MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:44 (51.1 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371421_2.fastq.4’ saved [124235014/124235014]\n",
      "\n",
      "--2025-01-02 19:33:44--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371422_1.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.216.57.241, 52.216.12.44, 52.217.91.108, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.216.57.241|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124262143 (119M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371422_1.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.50M  52.2MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:46 (52.2 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371422_1.fastq.4’ saved [124262143/124262143]\n",
      "\n",
      "--2025-01-02 19:33:46--  https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_SRR3371422_2.fastq\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 52.216.217.137, 3.5.25.32, 52.217.234.73, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|52.216.217.137|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 124192811 (118M) [binary/octet-stream]\n",
      "Saving to: ‘data_14/raw_fastq/subsampled_chr14_SRR3371422_2.fastq.4’\n",
      "\n",
      "subsampled_chr14_SR 100%[===================>] 118.44M  51.9MB/s    in 2.3s    \n",
      "\n",
      "2025-01-02 19:33:49 (51.9 MB/s) - ‘data_14/raw_fastq/subsampled_chr14_SRR3371422_2.fastq.4’ saved [124192811/124192811]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for acc in accs:\n",
    "    !wget https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_{acc}_1.fastq -P data_14/raw_fastq/\n",
    "    !wget https://sra-data-athena.s3.amazonaws.com/fastqfiles/subsampled_chr14_{acc}_2.fastq -P data_14/raw_fastq/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 3.3: Download reference genome and annotation files that will be used by STAR and RSEM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This step downloads and unzips reference files for the zebrafish genome and annotations needed by STAR. These reference files will be used by STAR to align RNA-seq reads during the analysis. To expedite this tutorial, we'll filter the zebrafish genome and annotation to include only data from chromosome 19."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-01-02 19:35:03--  ftp://ftp.ensembl.org/pub/release-104/fasta/danio_rerio/dna/Danio_rerio.GRCz11.dna.primary_assembly.fa.gz\n",
      "           => ‘data_14/reference/zebrafish_genome.fa.gz’\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-104/fasta/danio_rerio/dna ... done.\n",
      "==> SIZE Danio_rerio.GRCz11.dna.primary_assembly.fa.gz ... 410230731\n",
      "==> PASV ... done.    ==> RETR Danio_rerio.GRCz11.dna.primary_assembly.fa.gz ... done.\n",
      "Length: 410230731 (391M) (unauthoritative)\n",
      "\n",
      "Danio_rerio.GRCz11. 100%[===================>] 391.23M  33.9MB/s    in 12s     \n",
      "\n",
      "2025-01-02 19:35:16 (31.7 MB/s) - ‘data_14/reference/zebrafish_genome.fa.gz’ saved [410230731]\n",
      "\n",
      "--2025-01-02 19:35:16--  ftp://ftp.ensembl.org/pub/release-104/gtf/danio_rerio/Danio_rerio.GRCz11.104.gtf.gz\n",
      "           => ‘data_14/reference/zebrafish_annotation.gtf.gz’\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/release-104/gtf/danio_rerio ... done.\n",
      "==> SIZE Danio_rerio.GRCz11.104.gtf.gz ... 18042498\n",
      "==> PASV ... done.    ==> RETR Danio_rerio.GRCz11.104.gtf.gz ... done.\n",
      "Length: 18042498 (17M) (unauthoritative)\n",
      "\n",
      "Danio_rerio.GRCz11. 100%[===================>]  17.21M  15.0MB/s    in 1.1s    \n",
      "\n",
      "2025-01-02 19:35:18 (15.0 MB/s) - ‘data_14/reference/zebrafish_annotation.gtf.gz’ saved [18042498]\n",
      "\n",
      "--2025-01-02 19:35:18--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11/GCF_000002035.6_GRCz11_feature_table.txt.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.7, 130.14.250.31, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4834794 (4.6M) [application/x-gzip]\n",
      "Saving to: ‘data_14/reference/zebrafish_feature_table.txt.gz’\n",
      "\n",
      "data_14/reference/z 100%[===================>]   4.61M  27.6MB/s    in 0.2s    \n",
      "\n",
      "2025-01-02 19:35:19 (27.6 MB/s) - ‘data_14/reference/zebrafish_feature_table.txt.gz’ saved [4834794/4834794]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget ftp://ftp.ensembl.org/pub/release-104/fasta/danio_rerio/dna/Danio_rerio.GRCz11.dna.primary_assembly.fa.gz -O data_14/reference/zebrafish_genome.fa.gz\n",
    "! wget ftp://ftp.ensembl.org/pub/release-104/gtf/danio_rerio/Danio_rerio.GRCz11.104.gtf.gz -O data_14/reference/zebrafish_annotation.gtf.gz\n",
    "! wget -O data_14/reference/zebrafish_feature_table.txt.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.6_GRCz11/GCF_000002035.6_GRCz11_feature_table.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!gunzip -f data_14/reference/zebrafish_genome.fa.gz \n",
    "!gunzip -f data_14/reference/zebrafish_annotation.gtf.gz\n",
    "!gunzip -f data_14/reference/zebrafish_feature_table.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Use samtools to extract only chr14 from the reference genome\n",
    "!samtools faidx data_14/reference/zebrafish_genome.fa 14 > data_14/reference/zebrafish_genome_chr14.fa \n",
    "\n",
    "# Subset the GTF file to include only chr14 annotations\n",
    "!grep \"^14\" data_14/reference/zebrafish_annotation.gtf > data_14/reference/zebrafish_annotation_chr14.gtf "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 3.4: Copy data file for Trimmomatic\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of trimmomatics functions is to trim sequence machine specific adapter sequences. These are usually within the trimmomatic installation directory in a folder called adapters.\n",
    "\n",
    "Directories of packages within conda installations can be confusing, so in the case of using conda with trimmomatic, it may be easier to simply download or create a file with the relevant adapter sequencecs and store it in an easy to find directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-01-02 19:35:50--  https://sra-data-athena.s3.amazonaws.com/reference/TruSeq3-PE.fa\n",
      "Resolving sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)... 16.15.185.171, 52.217.71.220, 3.5.3.159, ...\n",
      "Connecting to sra-data-athena.s3.amazonaws.com (sra-data-athena.s3.amazonaws.com)|16.15.185.171|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 95 [binary/octet-stream]\n",
      "Saving to: ‘data_14/trimmed/TruSeq3-PE.fa.4’\n",
      "\n",
      "TruSeq3-PE.fa.4     100%[===================>]      95  --.-KB/s    in 0s      \n",
      "\n",
      "2025-01-02 19:35:50 (4.11 MB/s) - ‘data_14/trimmed/TruSeq3-PE.fa.4’ saved [95/95]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget -P data_14/trimmed/ https://sra-data-athena.s3.amazonaws.com/reference/TruSeq3-PE.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 4: Run Trimmomatic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trimmomatic will trim off any adapter sequences or low quality sequence it detects in the FASTQ files.\n",
    "\n",
    "Trimmomatic takes an input, in this case a forward and reverse FASTQ file, and the user names the outputs. In this case the first two outputs are the ‘paired’ and ‘orphaned’ trimmed files for the forward reads. The second two are the ‘paired’ and ‘orphaned’ trimmed files for the reverse reads.\n",
    "\n",
    "Paired in this case means the forward and reverse sequences were able to be aligned to each other after trimming. Orphaned, or unpaired, typically means one of the reads was discarded as a result of trimming, and only the forward or reverse read survived.\n",
    "\n",
    "Here we take only the file with paired reads, as there are only a few orphaned reads, and including orphaned reads can complicate different downstream analyses. Unless there is a significant amount of them, or a specific reason to use them, it is generally easier to discard unpaired reads. Also in the interest of simplicity and speed, we proceed in further steps with just using the paired-end reads of just the forward-end read files, which in this context is sufficient, however in different contexts, using both forward and reverse is often preferable.\n",
    "\n",
    "The last part of the command specifies how the trimming is done: ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
    "\n",
    "In greater detail:\n",
    "\n",
    "‘ILLUMINACLIP:TruSeq3-PE.fa’ refers to which adapters should be cut from the reads.\n",
    "\n",
    "‘2:30:10:2’ refers to various metrics, which are recommended defaults.\n",
    "\n",
    "‘2’ refers to the seed mismatch. This refers to the amount of mismatches a 'seed' may have in aligning to a possible adapter.\n",
    "\n",
    "‘30’ refers to the palindrome clip threshold. This refers to the similarity score. If forward and reverse reads, after having an adapter attached to them, are greater than this score, trimming of adapter fragments will be performed. Forward reads are clipped, and reverse reads dropped.\n",
    "\n",
    "‘10’ refers to the simple clip threshold. The basic, and alternative method to palindromic searching for adapters. Adapters are tested against reads and if sufficiently matched (above the threshold, in this case, 10), they are clipped.\n",
    "\n",
    "‘2’ refers to the minimum adapter fragment length in palindrome mode.\n",
    "\n",
    "‘LEADING:3’ refers to trimming bases from the start of a read. Bases at the start of a read will continue to be trimmed, sequentially, as long as the bases remain below a PHRED score of 3.\n",
    "\n",
    "‘TRAILING:3’ refers to the same as above, but at the end of a read.\n",
    "\n",
    "‘MINLEN:36’, the read is discarded if below this length.\n",
    "\n",
    "Greater information about parameters can be found in the trimmomatic documentation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371417_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371417_2.fastq data_14/trimmed/SRR3371417_1_trimmed.fastq data_14/trimmed/SRR3371417_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371417_2_trimmed.fastq data_14/trimmed/SRR3371417_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 1000000 (100.00%) Forward Only Surviving: 0 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371418_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371418_2.fastq data_14/trimmed/SRR3371418_1_trimmed.fastq data_14/trimmed/SRR3371418_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371418_2_trimmed.fastq data_14/trimmed/SRR3371418_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 1000000 (100.00%) Forward Only Surviving: 0 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371419_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371419_2.fastq data_14/trimmed/SRR3371419_1_trimmed.fastq data_14/trimmed/SRR3371419_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371419_2_trimmed.fastq data_14/trimmed/SRR3371419_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 1000000 (100.00%) Forward Only Surviving: 0 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371420_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371420_2.fastq data_14/trimmed/SRR3371420_1_trimmed.fastq data_14/trimmed/SRR3371420_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371420_2_trimmed.fastq data_14/trimmed/SRR3371420_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 1000000 (100.00%) Forward Only Surviving: 0 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371421_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371421_2.fastq data_14/trimmed/SRR3371421_1_trimmed.fastq data_14/trimmed/SRR3371421_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371421_2_trimmed.fastq data_14/trimmed/SRR3371421_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 1000000 (100.00%) Forward Only Surviving: 0 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n",
      "TrimmomaticPE: Started with arguments:\n",
      " -threads 15 data_14/raw_fastq/subsampled_chr14_SRR3371422_1.fastq data_14/raw_fastq/subsampled_chr14_SRR3371422_2.fastq data_14/trimmed/SRR3371422_1_trimmed.fastq data_14/trimmed/SRR3371422_1_trimmed_unpaired.fastq data_14/trimmed/SRR3371422_2_trimmed.fastq data_14/trimmed/SRR3371422_2_trimmed_unpaired.fastq ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36\n",
      "Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'\n",
      "ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences\n",
      "Quality encoding detected as phred33\n",
      "Input Read Pairs: 1000000 Both Surviving: 999999 (100.00%) Forward Only Surviving: 1 (0.00%) Reverse Only Surviving: 0 (0.00%) Dropped: 0 (0.00%)\n",
      "TrimmomaticPE: Completed successfully\n"
     ]
    }
   ],
   "source": [
    "!cat accs.txt | xargs -I {} \\\n",
    "trimmomatic PE -threads $THREADS \\\n",
    "'data_14/raw_fastq/subsampled_chr14_{}_1.fastq' 'data_14/raw_fastq/subsampled_chr14_{}_2.fastq' \\\n",
    "'data_14/trimmed/{}_1_trimmed.fastq' 'data_14/trimmed/{}_1_trimmed_unpaired.fastq' \\\n",
    "'data_14/trimmed/{}_2_trimmed.fastq' 'data_14/trimmed/{}_2_trimmed_unpaired.fastq' \\\n",
    "ILLUMINACLIP:data_14/trimmed/TruSeq3-PE.fa:2:30:10:2:keepBothReads LEADING:3 TRAILING:3 MINLEN:36"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 5: Run FastQC\n",
    "FastQC is an invaluable tool that allows you to evaluate whether there are problems with a set of reads. For example, it will provide a report of whether there is any bias in the sequence composition of the reads.\n",
    "\n",
    "Because jupyter is at its core a python editor, we can use python code and html support to display results in-line.\n",
    "\n",
    "FastQC looks for different characteristics of quality in reads. It is very rare that every metric will pass. In many cases, they serve as warnings, which should be compared to the context of the experiment. For instance, here, per base sequence content, sequence length distribution, sequence duplication levels, and overrepresented sequences all throw warnings. Per base sequence content routinely fails in RNA-sequencing in the first 15~ or so bases due to biased fragmentation. In most of our samples, this is where we see the failure (20% or more difference between A/T or G/C), and so is not unexpected. The overrepresented sequences can be BLASTed to show the majority of them are ribosomal RNA. Ribosomal RNA contamination is also common and will not be indexed later, and so not a large concern. Other metrics look good."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.001s][warning][perf,memops] Cannot use file /tmp/hsperfdata_ec2-user/13677 because it is locked by another process (errno = 11)\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "null\n",
      "Started analysis of SRR3371419_1_trimmed.fastq\n",
      "null\n",
      "Started analysis of SRR3371422_1_trimmed.fastq\n",
      "null\n",
      "null\n",
      "Started analysis of SRR3371417_1_trimmed.fastq\n",
      "null\n",
      "Started analysis of SRR3371420_1_trimmed.fastq\n",
      "null\n",
      "Started analysis of SRR3371421_1_trimmed.fastq\n",
      "null\n",
      "Started analysis of SRR3371418_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 5% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 10% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 15% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 20% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 25% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 30% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 35% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 40% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 45% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 50% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 55% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 60% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 65% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 70% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 75% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 80% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 85% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 100% complete for SRR3371420_1_trimmed.fastq\n",
      "Analysis complete for SRR3371420_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371419_1_trimmed.fastq\n",
      "Approx 90% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 100% complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371418_1_trimmed.fastq\n",
      "Analysis complete for SRR3371417_1_trimmed.fastq\n",
      "Approx 100% complete for SRR3371419_1_trimmed.fastq\n",
      "Analysis complete for SRR3371419_1_trimmed.fastq\n",
      "Analysis complete for SRR3371422_1_trimmed.fastq\n",
      "Approx 95% complete for SRR3371421_1_trimmed.fastq\n",
      "Approx 100% complete for SRR3371418_1_trimmed.fastq\n",
      "Analysis complete for SRR3371418_1_trimmed.fastq\n",
      "Approx 100% complete for SRR3371421_1_trimmed.fastq\n",
      "Analysis complete for SRR3371421_1_trimmed.fastq\n",
      "Started analysis of SRR3371422_2_trimmed.fastq\n",
      "Started analysis of SRR3371417_2_trimmed.fastq\n",
      "Started analysis of SRR3371419_2_trimmed.fastq\n",
      "Started analysis of SRR3371420_2_trimmed.fastq\n",
      "Started analysis of SRR3371418_2_trimmed.fastq\n",
      "Started analysis of SRR3371421_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 5% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 10% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 15% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 20% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 25% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 30% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 35% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 40% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 45% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 50% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 55% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 60% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 65% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 70% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 75% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 80% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 85% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 100% complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371422_2_trimmed.fastq\n",
      "Analysis complete for SRR3371420_2_trimmed.fastq\n",
      "Approx 90% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 100% complete for SRR3371417_2_trimmed.fastq\n",
      "Analysis complete for SRR3371417_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371418_2_trimmed.fastq\n",
      "Analysis complete for SRR3371422_2_trimmed.fastq\n",
      "Approx 100% complete for SRR3371419_2_trimmed.fastq\n",
      "Analysis complete for SRR3371419_2_trimmed.fastq\n",
      "Approx 95% complete for SRR3371421_2_trimmed.fastq\n",
      "Approx 100% complete for SRR3371418_2_trimmed.fastq\n",
      "Analysis complete for SRR3371418_2_trimmed.fastq\n",
      "Approx 100% complete for SRR3371421_2_trimmed.fastq\n",
      "Analysis complete for SRR3371421_2_trimmed.fastq\n"
     ]
    }
   ],
   "source": [
    "# Run FastQC\n",
    "!cat accs.txt | xargs -P $THREADS -I {} fastqc data_14/trimmed/{}_1_trimmed.fastq data_14/trimmed/{}_2_trimmed.fastq -o data_14/fastqc_samples/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"800\"\n",
       "            height=\"600\"\n",
       "            src=\"./data_14/fastqc_samples/SRR3371417_1_trimmed_fastqc.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f3c6f5862f0>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame(src='./data_14/fastqc_samples/SRR3371417_1_trimmed_fastqc.html', width=800, height=600)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 6: MultiQC\n",
    "MultiQC reads in the FastQC reports and generate a compiled report for all the analyzed FASTQ files.\n",
    "\n",
    "Being able to use python with bash also means we can seamlessly use popular python packages, such as pandas, to interact with or view the files we create."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[91m///\u001b[0m \u001b]8;id=158268;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🍾 \u001b[2mv1.26\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /home/ec2-user/SageMaker/data_14/fastqc_samples\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m39/39\u001b[0m  l\u001b[0mm  \n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 14 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : data_14/multiqc_samples/multiqc_data   (overwritten)\n",
      "\u001b[34m     write_results\u001b[0m | Report      : data_14/multiqc_samples/multiqc_report.html   (overwritten)\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "#!multiqc -f data_14/fastqc_samples/\n",
    "!multiqc -f -o data_14/multiqc_samples/ data_14/fastqc_samples/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: jupyterquiz in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (2.8.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install jupyterquiz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=\"jyEcrGZbGFPh\" data-shufflequestions=\"False\"\n",
       "               data-shuffleanswers=\"True\"\n",
       "               data-preserveresponses=\"false\"\n",
       "               data-numquestions=\"1000000\"\n",
       "               data-maxwidth=\"600\"\n",
       "               style=\"border-radius: 10px; text-align: left\"> <style>\n",
       "#jyEcrGZbGFPh {\n",
       "   --jq-multiple-choice-bg: #6f78ffff;\n",
       "   --jq-mc-button-bg: #fafafa;\n",
       "   --jq-mc-button-border: #e0e0e0e0;\n",
       "   --jq-mc-button-inset-shadow: #555555;\n",
       "   --jq-many-choice-bg: #f75c03ff;\n",
       "   --jq-numeric-bg: #392061ff;\n",
       "   --jq-numeric-input-bg: #c0c0c0;\n",
       "   --jq-numeric-input-label: #101010;\n",
       "   --jq-numeric-input-shadow: #999999;\n",
       "   --jq-incorrect-color: #c80202;\n",
       "   --jq-correct-color: #009113;\n",
       "   --jq-text-color: #fafafa;\n",
       "}\n",
       "\n",
       ".Quiz {\n",
       "    max-width: 600px;\n",
       "    margin-top: 15px;\n",
       "    margin-left: auto;\n",
       "    margin-right: auto;\n",
       "/*    margin-bottom: 15px;*/\n",
       "/*    padding-bottom: 4px;*/\n",
       "    padding-top: 4px;\n",
       "    line-height: 1.1;\n",
       "    font-size: 16pt;\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".QuizCode {\n",
       "    font-size: 14pt;\n",
       "    margin-top: 10px;\n",
       "    margin-left: 20px;\n",
       "    margin-right: 20px;\n",
       "}\n",
       "\n",
       ".QuizCode>pre {\n",
       "    padding: 4px;\n",
       "}\n",
       "\n",
       ".Quiz code {\n",
       "    background-color: lightgray;\n",
       "    color: black;\n",
       "}\n",
       "\n",
       ".Quiz .QuizCode code {\n",
       "    background-color: inherit;\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       "\n",
       ".Quiz .MCButton code {\n",
       "    background-color: inherit;\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".MCButton .QuizCode {\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "\n",
       ".Answer {\n",
       "    border-radius: inherit;\n",
       "    display: grid;\n",
       "    grid-gap: 10px;\n",
       "    grid-template-columns: 1fr 1fr;\n",
       "    margin: 10px 0;\n",
       "}\n",
       "\n",
       "@media only screen and (max-width:480px) {\n",
       "    .Answer {\n",
       "        grid-template-columns: 1fr;\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       ".Feedback {\n",
       "    font-size: 16pt;\n",
       "    text-align: center;\n",
       "/*    min-height: 2em;*/\n",
       "}\n",
       "\n",
       ".Input {\n",
       "    align: left;\n",
       "    font-size: 20pt;\n",
       "}\n",
       "\n",
       ".Input-text {\n",
       "    display: block;\n",
       "    margin: 10px;\n",
       "    color: inherit;\n",
       "    width: 140px;\n",
       "    background-color: var(--jq-numeric-input-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    padding: 5px;\n",
       "    padding-left: 10px;\n",
       "    font-family: inherit;\n",
       "    font-size: 20px;\n",
       "    font-weight: inherit;\n",
       "    line-height: 20pt;\n",
       "    border: none;\n",
       "    border-radius: 0.2rem;\n",
       "    transition: box-shadow 0.1s);\n",
       "}\n",
       "\n",
       ".Input-text:focus {\n",
       "    /*outline: none;*/\n",
       "    background-color: var(--jq-numeric-input-bg);\n",
       "    box-shadow: 0.6rem 0.8rem 1.4rem -0.5rem var(--jq-numeric-input-shadow);\n",
       "}\n",
       "\n",
       ".MCButton {\n",
       "    background: var(--jq-mc-button-bg);\n",
       "    border: 1px solid var(--jq-mc-button-border);\n",
       "    border-radius: inherit;\n",
       "    color: #333333;\n",
       "    padding: 10px;\n",
       "    font-size: 16px;\n",
       "    cursor: pointer;\n",
       "    text-align: center;\n",
       "    display: flex;\n",
       "    align-items: center;\n",
       "    justify-content: center;\n",
       "}\n",
       "\n",
       ".MCButton p {\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".MultipleChoiceQn {\n",
       "    padding: 10px;\n",
       "    background: var(--jq-multiple-choice-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".ManyChoiceQn {\n",
       "    padding: 10px;\n",
       "    background: var(--jq-many-choice-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".NumericQn {\n",
       "    background: var(--jq-numeric-bg);\n",
       "    border-radius: inherit;\n",
       "    color: var(--jq-text-color);\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       ".NumericQn p {\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".InpLabel {\n",
       "    color: var(--jq-numeric-input-label);\n",
       "    float: left;\n",
       "    font-size: 15pt;\n",
       "    line-height: 34px;\n",
       "    margin-right: 10px;\n",
       "}\n",
       "\n",
       ".incorrect {\n",
       "    color: var(--jq-incorrect-color);\n",
       "}\n",
       "\n",
       ".correct {\n",
       "    color: var(--jq-correct-color);\n",
       "}\n",
       "\n",
       ".correctButton {\n",
       "    /*\n",
       "    background: var(--jq-correct-color);\n",
       "   */\n",
       "    animation: correct-anim 0.6s ease;\n",
       "    animation-fill-mode: forwards;\n",
       "    box-shadow: inset 0 0 5px var(--jq-mc-button-inset-shadow);\n",
       "    color: var(--jq-text-color);\n",
       "    /*outline: none;*/\n",
       "}\n",
       "\n",
       ".incorrectButton {\n",
       "    animation: incorrect-anim 0.8s ease;\n",
       "    animation-fill-mode: forwards;\n",
       "    box-shadow: inset 0 0 5px var(--jq-mc-button-inset-shadow);\n",
       "    color: var(--jq-text-color);\n",
       "    /*outline: none;*/\n",
       "}\n",
       "\n",
       "@keyframes incorrect-anim {\n",
       "    100% {\n",
       "        background-color: var(--jq-incorrect-color);\n",
       "    }\n",
       "}\n",
       "\n",
       "@keyframes correct-anim {\n",
       "    100% {\n",
       "        background-color: var(--jq-correct-color);\n",
       "    }\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "var questionsjyEcrGZbGFPh=[\n",
       "    {\n",
       "        \"question\": \"Select a single answer. How many reads passed trimming for the sample, SRR3371417?\",\n",
       "        \"type\": \"multiple_choice\",\n",
       "        \"answers\": [\n",
       "            {\n",
       "                \"answer\": \"1,000,000,000\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"False\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"100,000\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"False\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"1,000,000\",\n",
       "                \"correct\": true,\n",
       "                \"feedback\": \"Correct\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"10,000\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"False\"\n",
       "            }\n",
       "        ]\n",
       "    },\n",
       "    {\n",
       "        \"question\": \"Select a single answer. For the most over-represented trimmed sequence for sample, SRR3371417, how many times does it appear in the trimmed reads?\",\n",
       "        \"type\": \"multiple_choice\",\n",
       "        \"answers\": [\n",
       "            {\n",
       "                \"answer\": \"2,088\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is the top over-represented sequence across all samples in the MultiQC report.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"1,082\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is the top over-represented sequence in the reverse SRR3371422 sample.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"None\",\n",
       "                \"correct\": true,\n",
       "                \"feedback\": \"Correct. None of the reads in this sequence were over-represented.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"1,000,000\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is the total number of reads in the SRR3371417 sample.\"\n",
       "            }\n",
       "        ]\n",
       "    }\n",
       "]\n",
       ";\n",
       "    // Make a random ID\n",
       "function makeid(length) {\n",
       "    var result = [];\n",
       "    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';\n",
       "    var charactersLength = characters.length;\n",
       "    for (var i = 0; i < length; i++) {\n",
       "        result.push(characters.charAt(Math.floor(Math.random() * charactersLength)));\n",
       "    }\n",
       "    return result.join('');\n",
       "}\n",
       "\n",
       "// Choose a random subset of an array. Can also be used to shuffle the array\n",
       "function getRandomSubarray(arr, size) {\n",
       "    var shuffled = arr.slice(0), i = arr.length, temp, index;\n",
       "    while (i--) {\n",
       "        index = Math.floor((i + 1) * Math.random());\n",
       "        temp = shuffled[index];\n",
       "        shuffled[index] = shuffled[i];\n",
       "        shuffled[i] = temp;\n",
       "    }\n",
       "    return shuffled.slice(0, size);\n",
       "}\n",
       "\n",
       "function printResponses(responsesContainer) {\n",
       "    var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "    var stringResponses='<B>IMPORTANT!</B>To preserve this answer sequence for submission, when you have finalized your answers: <ol> <li> Copy the text in this cell below \"Answer String\"</li> <li> Double click on the cell directly below the Answer String, labeled \"Replace Me\"</li> <li> Select the whole \"Replace Me\" text</li> <li> Paste in your answer string and press shift-Enter.</li><li>Save the notebook using the save icon or File->Save Notebook menu item</li></ul><br><br><br><b>Answer String:</b><br> ';\n",
       "    console.log(responses);\n",
       "    responses.forEach((response, index) => {\n",
       "        if (response) {\n",
       "            console.log(index + ': ' + response);\n",
       "            stringResponses+= index + ': ' + response +\"<BR>\";\n",
       "        }\n",
       "    });\n",
       "    responsesContainer.innerHTML=stringResponses;\n",
       "}\n",
       "/* Callback function to determine whether a selected multiple-choice\n",
       "   button corresponded to a correct answer and to provide feedback\n",
       "   based on the answer */\n",
       "function check_mc() {\n",
       "    var id = this.id.split('-')[0];\n",
       "    //var response = this.id.split('-')[1];\n",
       "    //console.log(response);\n",
       "    //console.log(\"In check_mc(), id=\"+id);\n",
       "    //console.log(event.srcElement.id)           \n",
       "    //console.log(event.srcElement.dataset.correct)   \n",
       "    //console.log(event.srcElement.dataset.feedback)\n",
       "\n",
       "    var label = event.srcElement;\n",
       "    //console.log(label, label.nodeName);\n",
       "    var depth = 0;\n",
       "    while ((label.nodeName != \"LABEL\") && (depth < 20)) {\n",
       "        label = label.parentElement;\n",
       "        console.log(depth, label);\n",
       "        depth++;\n",
       "    }\n",
       "\n",
       "\n",
       "\n",
       "    var answers = label.parentElement.children;\n",
       "    //console.log(answers);\n",
       "\n",
       "    // Split behavior based on multiple choice vs many choice:\n",
       "    var fb = document.getElementById(\"fb\" + id);\n",
       "\n",
       "\n",
       "\n",
       "    /* Multiple choice (1 answer). Allow for 0 correct\n",
       "       answers as an edge case */\n",
       "    if (fb.dataset.numcorrect <= 1) {\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            //console.log(responsesContainer);\n",
       "            var response = label.firstChild.innerText;\n",
       "            if (label.querySelector(\".QuizCode\")){\n",
       "                response+= label.querySelector(\".QuizCode\").firstChild.innerText;\n",
       "            }\n",
       "            console.log(response);\n",
       "            //console.log(document.getElementById(\"quizWrap\"+id));\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            console.log(responses);\n",
       "            responses[qnum]= response;\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End code to preserve responses\n",
       "\n",
       "        for (var i = 0; i < answers.length; i++) {\n",
       "            var child = answers[i];\n",
       "            //console.log(child);\n",
       "            child.className = \"MCButton\";\n",
       "        }\n",
       "\n",
       "\n",
       "\n",
       "        if (label.dataset.correct == \"true\") {\n",
       "            // console.log(\"Correct action\");\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                fb.innerHTML = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                fb.innerHTML = \"Correct!\";\n",
       "            }\n",
       "            label.classList.add(\"correctButton\");\n",
       "\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"correct\");\n",
       "\n",
       "        } else {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                fb.innerHTML = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                fb.innerHTML = \"Incorrect -- try again.\";\n",
       "            }\n",
       "            //console.log(\"Error action\");\n",
       "            label.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "    }\n",
       "    else { /* Many choice (more than 1 correct answer) */\n",
       "        var reset = false;\n",
       "        var feedback;\n",
       "         if (label.dataset.correct == \"true\") {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                feedback = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                feedback = \"Correct!\";\n",
       "            }\n",
       "            if (label.dataset.answered <= 0) {\n",
       "                if (fb.dataset.answeredcorrect < 0) {\n",
       "                    fb.dataset.answeredcorrect = 1;\n",
       "                    reset = true;\n",
       "                } else {\n",
       "                    fb.dataset.answeredcorrect++;\n",
       "                }\n",
       "                if (reset) {\n",
       "                    for (var i = 0; i < answers.length; i++) {\n",
       "                        var child = answers[i];\n",
       "                        child.className = \"MCButton\";\n",
       "                        child.dataset.answered = 0;\n",
       "                    }\n",
       "                }\n",
       "                label.classList.add(\"correctButton\");\n",
       "                label.dataset.answered = 1;\n",
       "                fb.className = \"Feedback\";\n",
       "                fb.classList.add(\"correct\");\n",
       "\n",
       "            }\n",
       "        } else {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                feedback = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                feedback = \"Incorrect -- try again.\";\n",
       "            }\n",
       "            if (fb.dataset.answeredcorrect > 0) {\n",
       "                fb.dataset.answeredcorrect = -1;\n",
       "                reset = true;\n",
       "            } else {\n",
       "                fb.dataset.answeredcorrect--;\n",
       "            }\n",
       "\n",
       "            if (reset) {\n",
       "                for (var i = 0; i < answers.length; i++) {\n",
       "                    var child = answers[i];\n",
       "                    child.className = \"MCButton\";\n",
       "                    child.dataset.answered = 0;\n",
       "                }\n",
       "            }\n",
       "            label.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            //console.log(responsesContainer);\n",
       "            var response = label.firstChild.innerText;\n",
       "            if (label.querySelector(\".QuizCode\")){\n",
       "                response+= label.querySelector(\".QuizCode\").firstChild.innerText;\n",
       "            }\n",
       "            console.log(response);\n",
       "            //console.log(document.getElementById(\"quizWrap\"+id));\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            if (label.dataset.correct == \"true\") {\n",
       "                if (typeof(responses[qnum]) == \"object\"){\n",
       "                    if (!responses[qnum].includes(response))\n",
       "                        responses[qnum].push(response);\n",
       "                } else{\n",
       "                    responses[qnum]= [ response ];\n",
       "                }\n",
       "            } else {\n",
       "                responses[qnum]= response;\n",
       "            }\n",
       "            console.log(responses);\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End save responses stuff\n",
       "\n",
       "\n",
       "\n",
       "        var numcorrect = fb.dataset.numcorrect;\n",
       "        var answeredcorrect = fb.dataset.answeredcorrect;\n",
       "        if (answeredcorrect >= 0) {\n",
       "            fb.innerHTML = feedback + \" [\" + answeredcorrect + \"/\" + numcorrect + \"]\";\n",
       "        } else {\n",
       "            fb.innerHTML = feedback + \" [\" + 0 + \"/\" + numcorrect + \"]\";\n",
       "        }\n",
       "\n",
       "\n",
       "    }\n",
       "\n",
       "    if (typeof MathJax != 'undefined') {\n",
       "        var version = MathJax.version;\n",
       "        console.log('MathJax version', version);\n",
       "        if (version[0] == \"2\") {\n",
       "            MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "        } else if (version[0] == \"3\") {\n",
       "            MathJax.typeset([fb]);\n",
       "        }\n",
       "    } else {\n",
       "        console.log('MathJax not detected');\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "/* Function to produce the HTML buttons for a multiple choice/\n",
       "   many choice question  and to update the CSS tags based on\n",
       "   the question type */\n",
       "function make_mc(qa, shuffle_answers, outerqDiv, qDiv, aDiv, id) {\n",
       "\n",
       "    var shuffled;\n",
       "    if (shuffle_answers == \"True\") {\n",
       "        //console.log(shuffle_answers+\" read as true\");\n",
       "        shuffled = getRandomSubarray(qa.answers, qa.answers.length);\n",
       "    } else {\n",
       "        //console.log(shuffle_answers+\" read as false\");\n",
       "        shuffled = qa.answers;\n",
       "    }\n",
       "\n",
       "\n",
       "    var num_correct = 0;\n",
       "\n",
       "    shuffled.forEach((item, index, ans_array) => {\n",
       "        //console.log(answer);\n",
       "\n",
       "        // Make input element\n",
       "        var inp = document.createElement(\"input\");\n",
       "        inp.type = \"radio\";\n",
       "        inp.id = \"quizo\" + id + index;\n",
       "        inp.style = \"display:none;\";\n",
       "        aDiv.append(inp);\n",
       "\n",
       "        //Make label for input element\n",
       "        var lab = document.createElement(\"label\");\n",
       "        lab.className = \"MCButton\";\n",
       "        lab.id = id + '-' + index;\n",
       "        lab.onclick = check_mc;\n",
       "        var aSpan = document.createElement('span');\n",
       "        aSpan.classsName = \"\";\n",
       "        //qDiv.id=\"quizQn\"+id+index;\n",
       "        if (\"answer\" in item) {\n",
       "            aSpan.innerHTML = jaxify(item.answer);\n",
       "            //aSpan.innerHTML=item.answer;\n",
       "        }\n",
       "        lab.append(aSpan);\n",
       "\n",
       "        // Create div for code inside question\n",
       "        var codeSpan;\n",
       "        if (\"code\" in item) {\n",
       "            codeSpan = document.createElement('span');\n",
       "            codeSpan.id = \"code\" + id + index;\n",
       "            codeSpan.className = \"QuizCode\";\n",
       "            var codePre = document.createElement('pre');\n",
       "            codeSpan.append(codePre);\n",
       "            var codeCode = document.createElement('code');\n",
       "            codePre.append(codeCode);\n",
       "            codeCode.innerHTML = item.code;\n",
       "            lab.append(codeSpan);\n",
       "            //console.log(codeSpan);\n",
       "        }\n",
       "\n",
       "        //lab.textContent=item.answer;\n",
       "\n",
       "        // Set the data attributes for the answer\n",
       "        lab.setAttribute('data-correct', item.correct);\n",
       "        if (item.correct) {\n",
       "            num_correct++;\n",
       "        }\n",
       "        if (\"feedback\" in item) {\n",
       "            lab.setAttribute('data-feedback', item.feedback);\n",
       "        }\n",
       "        lab.setAttribute('data-answered', 0);\n",
       "\n",
       "        aDiv.append(lab);\n",
       "\n",
       "    });\n",
       "\n",
       "    if (num_correct > 1) {\n",
       "        outerqDiv.className = \"ManyChoiceQn\";\n",
       "    } else {\n",
       "        outerqDiv.className = \"MultipleChoiceQn\";\n",
       "    }\n",
       "\n",
       "    return num_correct;\n",
       "\n",
       "}\n",
       "function check_numeric(ths, event) {\n",
       "\n",
       "    if (event.keyCode === 13) {\n",
       "        ths.blur();\n",
       "\n",
       "        var id = ths.id.split('-')[0];\n",
       "\n",
       "        var submission = ths.value;\n",
       "        if (submission.indexOf('/') != -1) {\n",
       "            var sub_parts = submission.split('/');\n",
       "            //console.log(sub_parts);\n",
       "            submission = sub_parts[0] / sub_parts[1];\n",
       "        }\n",
       "        //console.log(\"Reader entered\", submission);\n",
       "\n",
       "        if (\"precision\" in ths.dataset) {\n",
       "            var precision = ths.dataset.precision;\n",
       "            submission = Number(Number(submission).toPrecision(precision));\n",
       "        }\n",
       "\n",
       "\n",
       "        //console.log(\"In check_numeric(), id=\"+id);\n",
       "        //console.log(event.srcElement.id)           \n",
       "        //console.log(event.srcElement.dataset.feedback)\n",
       "\n",
       "        var fb = document.getElementById(\"fb\" + id);\n",
       "        fb.style.display = \"none\";\n",
       "        fb.innerHTML = \"Incorrect -- try again.\";\n",
       "\n",
       "        var answers = JSON.parse(ths.dataset.answers);\n",
       "        //console.log(answers);\n",
       "\n",
       "        var defaultFB = \"Incorrect. Try again.\";\n",
       "        var correct;\n",
       "        var done = false;\n",
       "        answers.every(answer => {\n",
       "            //console.log(answer.type);\n",
       "\n",
       "            correct = false;\n",
       "            // if (answer.type==\"value\"){\n",
       "            if ('value' in answer) {\n",
       "                if (submission == answer.value) {\n",
       "                    if (\"feedback\" in answer) {\n",
       "                        fb.innerHTML = jaxify(answer.feedback);\n",
       "                    } else {\n",
       "                        fb.innerHTML = jaxify(\"Correct\");\n",
       "                    }\n",
       "                    correct = answer.correct;\n",
       "                    //console.log(answer.correct);\n",
       "                    done = true;\n",
       "                }\n",
       "                // } else if (answer.type==\"range\") {\n",
       "            } else if ('range' in answer) {\n",
       "                console.log(answer.range);\n",
       "                console.log(submission, submission >=answer.range[0], submission < answer.range[1])\n",
       "                if ((submission >= answer.range[0]) && (submission < answer.range[1])) {\n",
       "                    fb.innerHTML = jaxify(answer.feedback);\n",
       "                    correct = answer.correct;\n",
       "                    console.log(answer.correct);\n",
       "                    done = true;\n",
       "                }\n",
       "            } else if (answer.type == \"default\") {\n",
       "                if (\"feedback\" in answer) {\n",
       "                    defaultFB = answer.feedback;\n",
       "                } \n",
       "            }\n",
       "            if (done) {\n",
       "                return false; // Break out of loop if this has been marked correct\n",
       "            } else {\n",
       "                return true; // Keep looking for case that includes this as a correct answer\n",
       "            }\n",
       "        });\n",
       "        console.log(\"done:\", done);\n",
       "\n",
       "        if ((!done) && (defaultFB != \"\")) {\n",
       "            fb.innerHTML = jaxify(defaultFB);\n",
       "            //console.log(\"Default feedback\", defaultFB);\n",
       "        }\n",
       "\n",
       "        fb.style.display = \"block\";\n",
       "        if (correct) {\n",
       "            ths.className = \"Input-text\";\n",
       "            ths.classList.add(\"correctButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"correct\");\n",
       "        } else {\n",
       "            ths.className = \"Input-text\";\n",
       "            ths.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            console.log(submission);\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            //console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            console.log(responses);\n",
       "            if (submission == ths.value){\n",
       "                responses[qnum]= submission;\n",
       "            } else {\n",
       "                responses[qnum]= ths.value + \"(\" + submission +\")\";\n",
       "            }\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End code to preserve responses\n",
       "\n",
       "        if (typeof MathJax != 'undefined') {\n",
       "            var version = MathJax.version;\n",
       "            console.log('MathJax version', version);\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                MathJax.typeset([fb]);\n",
       "            }\n",
       "        } else {\n",
       "            console.log('MathJax not detected');\n",
       "        }\n",
       "        return false;\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "function isValid(el, charC) {\n",
       "    //console.log(\"Input char: \", charC);\n",
       "    if (charC == 46) {\n",
       "        if (el.value.indexOf('.') === -1) {\n",
       "            return true;\n",
       "        } else if (el.value.indexOf('/') != -1) {\n",
       "            var parts = el.value.split('/');\n",
       "            if (parts[1].indexOf('.') === -1) {\n",
       "                return true;\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 47) {\n",
       "        if (el.value.indexOf('/') === -1) {\n",
       "            if ((el.value != \"\") && (el.value != \".\")) {\n",
       "                return true;\n",
       "            } else {\n",
       "                return false;\n",
       "            }\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 45) {\n",
       "        var edex = el.value.indexOf('e');\n",
       "        if (edex == -1) {\n",
       "            edex = el.value.indexOf('E');\n",
       "        }\n",
       "\n",
       "        if (el.value == \"\") {\n",
       "            return true;\n",
       "        } else if (edex == (el.value.length - 1)) { // If just after e or E\n",
       "            return true;\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 101) { // \"e\"\n",
       "        if ((el.value.indexOf('e') === -1) && (el.value.indexOf('E') === -1) && (el.value.indexOf('/') == -1)) {\n",
       "            // Prev symbol must be digit or decimal point:\n",
       "            if (el.value.slice(-1).search(/\\d/) >= 0) {\n",
       "                return true;\n",
       "            } else if (el.value.slice(-1).search(/\\./) >= 0) {\n",
       "                return true;\n",
       "            } else {\n",
       "                return false;\n",
       "            }\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else {\n",
       "        if (charC > 31 && (charC < 48 || charC > 57))\n",
       "            return false;\n",
       "    }\n",
       "    return true;\n",
       "}\n",
       "\n",
       "function numeric_keypress(evnt) {\n",
       "    var charC = (evnt.which) ? evnt.which : evnt.keyCode;\n",
       "\n",
       "    if (charC == 13) {\n",
       "        check_numeric(this, evnt);\n",
       "    } else {\n",
       "        return isValid(this, charC);\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "function make_numeric(qa, outerqDiv, qDiv, aDiv, id) {\n",
       "\n",
       "\n",
       "\n",
       "    //console.log(answer);\n",
       "\n",
       "\n",
       "    outerqDiv.className = \"NumericQn\";\n",
       "    aDiv.style.display = 'block';\n",
       "\n",
       "    var lab = document.createElement(\"label\");\n",
       "    lab.className = \"InpLabel\";\n",
       "    lab.innerHTML = \"Type numeric answer here:\";\n",
       "    aDiv.append(lab);\n",
       "\n",
       "    var inp = document.createElement(\"input\");\n",
       "    inp.type = \"text\";\n",
       "    //inp.id=\"input-\"+id;\n",
       "    inp.id = id + \"-0\";\n",
       "    inp.className = \"Input-text\";\n",
       "    inp.setAttribute('data-answers', JSON.stringify(qa.answers));\n",
       "    if (\"precision\" in qa) {\n",
       "        inp.setAttribute('data-precision', qa.precision);\n",
       "    }\n",
       "    aDiv.append(inp);\n",
       "    //console.log(inp);\n",
       "\n",
       "    //inp.addEventListener(\"keypress\", check_numeric);\n",
       "    //inp.addEventListener(\"keypress\", numeric_keypress);\n",
       "    /*\n",
       "    inp.addEventListener(\"keypress\", function(event) {\n",
       "        return numeric_keypress(this, event);\n",
       "    }\n",
       "                        );\n",
       "                        */\n",
       "    //inp.onkeypress=\"return numeric_keypress(this, event)\";\n",
       "    inp.onkeypress = numeric_keypress;\n",
       "    inp.onpaste = event => false;\n",
       "\n",
       "    inp.addEventListener(\"focus\", function (event) {\n",
       "        this.value = \"\";\n",
       "        return false;\n",
       "    }\n",
       "    );\n",
       "\n",
       "\n",
       "}\n",
       "function jaxify(string) {\n",
       "    var mystring = string;\n",
       "\n",
       "    var count = 0;\n",
       "    var loc = mystring.search(/([^\\\\]|^)(\\$)/);\n",
       "\n",
       "    var count2 = 0;\n",
       "    var loc2 = mystring.search(/([^\\\\]|^)(\\$\\$)/);\n",
       "\n",
       "    //console.log(loc);\n",
       "\n",
       "    while ((loc >= 0) || (loc2 >= 0)) {\n",
       "\n",
       "        /* Have to replace all the double $$ first with current implementation */\n",
       "        if (loc2 >= 0) {\n",
       "            if (count2 % 2 == 0) {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$\\$)/, \"$1\\\\[\");\n",
       "            } else {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$\\$)/, \"$1\\\\]\");\n",
       "            }\n",
       "            count2++;\n",
       "        } else {\n",
       "            if (count % 2 == 0) {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$)/, \"$1\\\\(\");\n",
       "            } else {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$)/, \"$1\\\\)\");\n",
       "            }\n",
       "            count++;\n",
       "        }\n",
       "        loc = mystring.search(/([^\\\\]|^)(\\$)/);\n",
       "        loc2 = mystring.search(/([^\\\\]|^)(\\$\\$)/);\n",
       "        //console.log(mystring,\", loc:\",loc,\", loc2:\",loc2);\n",
       "    }\n",
       "\n",
       "    // repace markdown style links with actual links\n",
       "    mystring = mystring.replace(/<(.*?)>/, '<a href=\"$1\" target=\"_blank\" class=\"Link\">$1</a>');\n",
       "    mystring = mystring.replace(/\\[(.*?)\\]\\((.*?)\\)/, '<a href=\"$2\" target=\"_blank\" class=\"Link\">$1</a>');\n",
       "\n",
       "    //console.log(mystring);\n",
       "    return mystring;\n",
       "}\n",
       "\n",
       "\n",
       "function show_questions(json, mydiv) {\n",
       "    console.log('show_questions');\n",
       "    //var mydiv=document.getElementById(myid);\n",
       "    var shuffle_questions = mydiv.dataset.shufflequestions;\n",
       "    var num_questions = mydiv.dataset.numquestions;\n",
       "    var shuffle_answers = mydiv.dataset.shuffleanswers;\n",
       "    var max_width = mydiv.dataset.maxwidth;\n",
       "\n",
       "    if (num_questions > json.length) {\n",
       "        num_questions = json.length;\n",
       "    }\n",
       "\n",
       "    var questions;\n",
       "    if ((num_questions < json.length) || (shuffle_questions == \"True\")) {\n",
       "        //console.log(num_questions+\",\"+json.length);\n",
       "        questions = getRandomSubarray(json, num_questions);\n",
       "    } else {\n",
       "        questions = json;\n",
       "    }\n",
       "\n",
       "    //console.log(\"SQ: \"+shuffle_questions+\", NQ: \" + num_questions + \", SA: \", shuffle_answers);\n",
       "\n",
       "    // Iterate over questions\n",
       "    questions.forEach((qa, index, array) => {\n",
       "        //console.log(qa.question); \n",
       "\n",
       "        var id = makeid(8);\n",
       "        //console.log(id);\n",
       "\n",
       "\n",
       "        // Create Div to contain question and answers\n",
       "        var iDiv = document.createElement('div');\n",
       "        //iDiv.id = 'quizWrap' + id + index;\n",
       "        iDiv.id = 'quizWrap' + id;\n",
       "        iDiv.className = 'Quiz';\n",
       "        iDiv.setAttribute('data-qnum', index);\n",
       "        iDiv.style.maxWidth  =max_width+\"px\";\n",
       "        mydiv.appendChild(iDiv);\n",
       "        // iDiv.innerHTML=qa.question;\n",
       "        \n",
       "        var outerqDiv = document.createElement('div');\n",
       "        outerqDiv.id = \"OuterquizQn\" + id + index;\n",
       "        // Create div to contain question part\n",
       "        var qDiv = document.createElement('div');\n",
       "        qDiv.id = \"quizQn\" + id + index;\n",
       "        \n",
       "        if (qa.question) {\n",
       "            iDiv.append(outerqDiv);\n",
       "\n",
       "            //qDiv.textContent=qa.question;\n",
       "            qDiv.innerHTML = jaxify(qa.question);\n",
       "            outerqDiv.append(qDiv);\n",
       "        }\n",
       "\n",
       "        // Create div for code inside question\n",
       "        var codeDiv;\n",
       "        if (\"code\" in qa) {\n",
       "            codeDiv = document.createElement('div');\n",
       "            codeDiv.id = \"code\" + id + index;\n",
       "            codeDiv.className = \"QuizCode\";\n",
       "            var codePre = document.createElement('pre');\n",
       "            codeDiv.append(codePre);\n",
       "            var codeCode = document.createElement('code');\n",
       "            codePre.append(codeCode);\n",
       "            codeCode.innerHTML = qa.code;\n",
       "            outerqDiv.append(codeDiv);\n",
       "            //console.log(codeDiv);\n",
       "        }\n",
       "\n",
       "\n",
       "        // Create div to contain answer part\n",
       "        var aDiv = document.createElement('div');\n",
       "        aDiv.id = \"quizAns\" + id + index;\n",
       "        aDiv.className = 'Answer';\n",
       "        iDiv.append(aDiv);\n",
       "\n",
       "        //console.log(qa.type);\n",
       "\n",
       "        var num_correct;\n",
       "        if ((qa.type == \"multiple_choice\") || (qa.type == \"many_choice\") ) {\n",
       "            num_correct = make_mc(qa, shuffle_answers, outerqDiv, qDiv, aDiv, id);\n",
       "            if (\"answer_cols\" in qa) {\n",
       "                //aDiv.style.gridTemplateColumns = 'auto '.repeat(qa.answer_cols);\n",
       "                aDiv.style.gridTemplateColumns = 'repeat(' + qa.answer_cols + ', 1fr)';\n",
       "            }\n",
       "        } else if (qa.type == \"numeric\") {\n",
       "            //console.log(\"numeric\");\n",
       "            make_numeric(qa, outerqDiv, qDiv, aDiv, id);\n",
       "        }\n",
       "\n",
       "\n",
       "        //Make div for feedback\n",
       "        var fb = document.createElement(\"div\");\n",
       "        fb.id = \"fb\" + id;\n",
       "        //fb.style=\"font-size: 20px;text-align:center;\";\n",
       "        fb.className = \"Feedback\";\n",
       "        fb.setAttribute(\"data-answeredcorrect\", 0);\n",
       "        fb.setAttribute(\"data-numcorrect\", num_correct);\n",
       "        iDiv.append(fb);\n",
       "\n",
       "\n",
       "    });\n",
       "    var preserveResponses = mydiv.dataset.preserveresponses;\n",
       "    console.log(preserveResponses);\n",
       "    console.log(preserveResponses == \"true\");\n",
       "    if (preserveResponses == \"true\") {\n",
       "        console.log(preserveResponses);\n",
       "        // Create Div to contain record of answers\n",
       "        var iDiv = document.createElement('div');\n",
       "        iDiv.id = 'responses' + mydiv.id;\n",
       "        iDiv.className = 'JCResponses';\n",
       "        // Create a place to store responses as an empty array\n",
       "        iDiv.setAttribute('data-responses', '[]');\n",
       "\n",
       "        // Dummy Text\n",
       "        iDiv.innerHTML=\"<b>Select your answers and then follow the directions that will appear here.</b>\"\n",
       "        //iDiv.className = 'Quiz';\n",
       "        mydiv.appendChild(iDiv);\n",
       "    }\n",
       "//console.log(\"At end of show_questions\");\n",
       "    if (typeof MathJax != 'undefined') {\n",
       "        console.log(\"MathJax version\", MathJax.version);\n",
       "        var version = MathJax.version;\n",
       "        setTimeout(function(){\n",
       "            var version = MathJax.version;\n",
       "            console.log('After sleep, MathJax version', version);\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                if (MathJax.hasOwnProperty('typeset') ) {\n",
       "                    MathJax.typeset([mydiv]);\n",
       "                } else {\n",
       "                    console.log('WARNING: Trying to force load MathJax 3');\n",
       "                    window.MathJax = {\n",
       "                        tex: {\n",
       "                            inlineMath: [['$', '$'], ['\\\\(', '\\\\)']]\n",
       "                        },\n",
       "                        svg: {\n",
       "                            fontCache: 'global'\n",
       "                        }\n",
       "                    };\n",
       "\n",
       "                    (function () {\n",
       "                        var script = document.createElement('script');\n",
       "                        script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js';\n",
       "                        script.async = true;\n",
       "                        document.head.appendChild(script);\n",
       "                    })();\n",
       "                }\n",
       "            }\n",
       "        }, 500);\n",
       "if (typeof version == 'undefined') {\n",
       "        } else\n",
       "        {\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                if (MathJax.hasOwnProperty('typeset') ) {\n",
       "                    MathJax.typeset([mydiv]);\n",
       "                } else {\n",
       "                    console.log('WARNING: Trying to force load MathJax 3');\n",
       "                    window.MathJax = {\n",
       "                        tex: {\n",
       "                            inlineMath: [['$', '$'], ['\\\\(', '\\\\)']]\n",
       "                        },\n",
       "                        svg: {\n",
       "                            fontCache: 'global'\n",
       "                        }\n",
       "                    };\n",
       "\n",
       "                    (function () {\n",
       "                        var script = document.createElement('script');\n",
       "                        script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js';\n",
       "                        script.async = true;\n",
       "                        document.head.appendChild(script);\n",
       "                    })();\n",
       "                }\n",
       "            } else {\n",
       "                console.log(\"MathJax not found\");\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    // stop event propagation for the .Link class\n",
       "    var links = document.getElementsByClassName('Link')\n",
       "    for (var i = 0; i < links.length; i++) {\n",
       "        links[i].addEventListener('click', function(e){\n",
       "            e.stopPropagation();\n",
       "        });\n",
       "    }\n",
       "\n",
       "    return false;\n",
       "}\n",
       "/* This is to handle asynchrony issues in loading Jupyter notebooks\n",
       "           where the quiz has been previously run. The Javascript was generally\n",
       "           being run before the div was added to the DOM. I tried to do this\n",
       "           more elegantly using Mutation Observer, but I didn't get it to work.\n",
       "\n",
       "           Someone more knowledgeable could make this better ;-) */\n",
       "\n",
       "        function try_show() {\n",
       "          if(document.getElementById(\"jyEcrGZbGFPh\")) {\n",
       "            show_questions(questionsjyEcrGZbGFPh,  jyEcrGZbGFPh); \n",
       "          } else {\n",
       "             setTimeout(try_show, 200);\n",
       "          }\n",
       "        };\n",
       "    \n",
       "        {\n",
       "        // console.log(element);\n",
       "\n",
       "        //console.log(\"jyEcrGZbGFPh\");\n",
       "        // console.log(document.getElementById(\"jyEcrGZbGFPh\"));\n",
       "\n",
       "        try_show();\n",
       "        }\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import json\n",
    "from jupyterquiz import display_quiz\n",
    "\n",
    "#display Multiple Choice quiz\n",
    "display_quiz(\"quiz/bulkRNADR1.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 7: Preparing the STAR-Compatible RSEM Reference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The provided code snippet demonstrates the process of preparing a reference genome and annotation files for RNA-Seq analysis using RSEM and STAR. The initial command, rsem-prepare-reference, creates a reference index for the zebrafish genome based on the provided GTF annotation file (zebrafish_annotation_chr14.gtf). This index, along with the reference genome sequence (zebrafish_genome_chr14.fa), is essential for quantifying gene and isoform expression using RSEM. The --star option ensures compatibility with STAR, a popular transcriptome aligner.\n",
    "\n",
    "However, the subsequent STAR genome generation step encounters a warning due to an excessively large --genomeSAindexNbases parameter. This parameter controls the size of the genome's suffix array index, which is used for efficient alignment. If the parameter is too large for the given genome size, it can lead to segmentation faults during the mapping process. To address this issue, the code adjusts the --genomeSAindexNbases parameter to 11, which is more appropriate for the zebrafish genome. This modification ensures that the STAR genome generation process completes successfully without errors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!!!!! WARNING: --genomeSAindexNbases 14 is too large for the genome size=52660232, which may cause seg-fault at the mapping step. Re-run genome generation with recommended --genomeSAindexNbases 11\n"
     ]
    }
   ],
   "source": [
    "# Preparing the reference genome\n",
    "!rsem-prepare-reference --gtf data_14/reference/zebrafish_annotation_chr14.gtf --star -p $THREADS data_14/reference/zebrafish_genome_chr14.fa zebrafish_reference > /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!STAR  --runThreadN $THREADS  --runMode genomeGenerate  --genomeDir .  --genomeFastaFiles data_14/reference/zebrafish_genome_chr14.fa  --sjdbGTFfile data_14/reference/zebrafish_annotation_chr14.gtf  --sjdbOverhang 100  --outFileNamePrefix zebrafish_reference  --genomeSAindexNbases 11 > /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 8: Automated RNA-seq Quantification with RSEM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This code automates the process of quantifying gene expression from RNA-seq data using RSEM. It reads a list of SRR accession numbers from a text file, iterates over each accession, and executes the RSEM command to calculate gene expression levels. The code assumes that the trimmed FASTQ files for each SRR are located in a specified directory (data_14/trimmed/) and uses a previously generated reference index (zebrafish_reference). The output files are saved in the data_14/rsem_output directory, with filenames based on the corresponding SRR accession number. This script simplifies the workflow for analyzing multiple RNA-seq samples using RSEM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# Ensure you've set the path to the RSEM binary\n",
    "# Read the SRR accessions from the file\n",
    "with open('accs.txt', 'r') as f:\n",
    "    srr_accessions = [line.strip() for line in f.readlines()]\n",
    "\n",
    "# Define the output directory\n",
    "output_dir = \"data_14/rsem_output\"\n",
    "\n",
    "# Loop through each SRR accession and run rsem-calculate-expression\n",
    "for srr in srr_accessions:\n",
    "    !rsem-calculate-expression -p $THREADS --paired-end --star \\\n",
    "    data_14/trimmed/{srr}_1_trimmed.fastq data_14/trimmed/{srr}_2_trimmed.fastq zebrafish_reference data_14/rsem_output/{srr} > /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 9: Report the top 10 most highly expressed genes in the samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 10 most highly expressed genes in each wild-type sample.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 10 Genes by TPM for SRR3371417:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  124247.07\n",
      "195  ENSDARG00000036316  103500.44\n",
      "778  ENSDARG00000102317   94779.48\n",
      "272  ENSDARG00000041619   60956.18\n",
      "320  ENSDARG00000056119   49629.96\n",
      "309  ENSDARG00000053684   47847.78\n",
      "166  ENSDARG00000032929   32511.65\n",
      "100  ENSDARG00000019353   26246.46\n",
      "829  ENSDARG00000104068   19439.28\n",
      "39   ENSDARG00000007576   18041.34\n",
      "Top 10 Genes by TPM for SRR3371418:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  118843.46\n",
      "195  ENSDARG00000036316   96563.21\n",
      "778  ENSDARG00000102317   90832.66\n",
      "272  ENSDARG00000041619   57510.67\n",
      "309  ENSDARG00000053684   47467.12\n",
      "320  ENSDARG00000056119   46283.91\n",
      "166  ENSDARG00000032929   28985.29\n",
      "100  ENSDARG00000019353   25631.34\n",
      "39   ENSDARG00000007576   17977.82\n",
      "829  ENSDARG00000104068   17177.21\n",
      "Top 10 Genes by TPM for SRR3371419:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  121331.73\n",
      "195  ENSDARG00000036316  104500.53\n",
      "778  ENSDARG00000102317   94279.49\n",
      "272  ENSDARG00000041619   58951.66\n",
      "309  ENSDARG00000053684   54128.40\n",
      "320  ENSDARG00000056119   48126.36\n",
      "100  ENSDARG00000019353   25645.05\n",
      "166  ENSDARG00000032929   21808.82\n",
      "829  ENSDARG00000104068   17664.92\n",
      "469  ENSDARG00000076667   17077.79\n",
      "Top 10 Genes by TPM for SRR3371420:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  119090.85\n",
      "195  ENSDARG00000036316   95957.33\n",
      "778  ENSDARG00000102317   88618.73\n",
      "272  ENSDARG00000041619   56339.81\n",
      "309  ENSDARG00000053684   54119.64\n",
      "320  ENSDARG00000056119   45114.81\n",
      "166  ENSDARG00000032929   42733.81\n",
      "39   ENSDARG00000007576   27601.34\n",
      "100  ENSDARG00000019353   24348.31\n",
      "112  ENSDARG00000021124   15511.78\n",
      "Top 10 Genes by TPM for SRR3371421:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  119892.68\n",
      "195  ENSDARG00000036316  101044.00\n",
      "778  ENSDARG00000102317   90521.38\n",
      "309  ENSDARG00000053684   58134.55\n",
      "272  ENSDARG00000041619   56018.63\n",
      "320  ENSDARG00000056119   45200.38\n",
      "166  ENSDARG00000032929   30330.23\n",
      "100  ENSDARG00000019353   23327.46\n",
      "39   ENSDARG00000007576   18813.20\n",
      "829  ENSDARG00000104068   16859.02\n",
      "Top 10 Genes by TPM for SRR3371422:\n",
      "                gene_id        TPM\n",
      "616  ENSDARG00000092553  132616.49\n",
      "195  ENSDARG00000036316  109633.86\n",
      "778  ENSDARG00000102317   96331.50\n",
      "272  ENSDARG00000041619   63232.06\n",
      "309  ENSDARG00000053684   57006.84\n",
      "320  ENSDARG00000056119   51779.95\n",
      "166  ENSDARG00000032929   27441.16\n",
      "100  ENSDARG00000019353   27345.06\n",
      "469  ENSDARG00000076667   17030.34\n",
      "829  ENSDARG00000104068   16685.10\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Path to RSEM results directory\n",
    "rsem_results_dir = 'data_14/rsem_output'\n",
    "\n",
    "# Loop through each file in accs.txt\n",
    "for srr_id in open('accs.txt'):\n",
    "    srr_id = srr_id.strip()  # Remove newline character\n",
    "    rsem_result_file = f'{rsem_results_dir}/{srr_id}.genes.results'\n",
    "\n",
    "    # Load the RSEM results into a Pandas DataFrame\n",
    "    df = pd.read_csv(rsem_result_file, sep='\\t')\n",
    "\n",
    "    # Sort the DataFrame by TPM values in descending order and get the top 10 genes\n",
    "    top_10_genes = df.sort_values(by='TPM', ascending=False).head(10)\n",
    "\n",
    "    # Print the top 10 genes with their TPM values\n",
    "    print(f\"Top 10 Genes by TPM for {srr_id}:\")\n",
    "    print(top_10_genes[['gene_id', 'TPM']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div id=\"gQYrhhgfYdSG\" data-shufflequestions=\"False\"\n",
       "               data-shuffleanswers=\"True\"\n",
       "               data-preserveresponses=\"false\"\n",
       "               data-numquestions=\"1000000\"\n",
       "               data-maxwidth=\"600\"\n",
       "               style=\"border-radius: 10px; text-align: left\"> <style>\n",
       "#gQYrhhgfYdSG {\n",
       "   --jq-multiple-choice-bg: #6f78ffff;\n",
       "   --jq-mc-button-bg: #fafafa;\n",
       "   --jq-mc-button-border: #e0e0e0e0;\n",
       "   --jq-mc-button-inset-shadow: #555555;\n",
       "   --jq-many-choice-bg: #f75c03ff;\n",
       "   --jq-numeric-bg: #392061ff;\n",
       "   --jq-numeric-input-bg: #c0c0c0;\n",
       "   --jq-numeric-input-label: #101010;\n",
       "   --jq-numeric-input-shadow: #999999;\n",
       "   --jq-incorrect-color: #c80202;\n",
       "   --jq-correct-color: #009113;\n",
       "   --jq-text-color: #fafafa;\n",
       "}\n",
       "\n",
       ".Quiz {\n",
       "    max-width: 600px;\n",
       "    margin-top: 15px;\n",
       "    margin-left: auto;\n",
       "    margin-right: auto;\n",
       "/*    margin-bottom: 15px;*/\n",
       "/*    padding-bottom: 4px;*/\n",
       "    padding-top: 4px;\n",
       "    line-height: 1.1;\n",
       "    font-size: 16pt;\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".QuizCode {\n",
       "    font-size: 14pt;\n",
       "    margin-top: 10px;\n",
       "    margin-left: 20px;\n",
       "    margin-right: 20px;\n",
       "}\n",
       "\n",
       ".QuizCode>pre {\n",
       "    padding: 4px;\n",
       "}\n",
       "\n",
       ".Quiz code {\n",
       "    background-color: lightgray;\n",
       "    color: black;\n",
       "}\n",
       "\n",
       ".Quiz .QuizCode code {\n",
       "    background-color: inherit;\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       "\n",
       ".Quiz .MCButton code {\n",
       "    background-color: inherit;\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".MCButton .QuizCode {\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "\n",
       ".Answer {\n",
       "    border-radius: inherit;\n",
       "    display: grid;\n",
       "    grid-gap: 10px;\n",
       "    grid-template-columns: 1fr 1fr;\n",
       "    margin: 10px 0;\n",
       "}\n",
       "\n",
       "@media only screen and (max-width:480px) {\n",
       "    .Answer {\n",
       "        grid-template-columns: 1fr;\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       ".Feedback {\n",
       "    font-size: 16pt;\n",
       "    text-align: center;\n",
       "/*    min-height: 2em;*/\n",
       "}\n",
       "\n",
       ".Input {\n",
       "    align: left;\n",
       "    font-size: 20pt;\n",
       "}\n",
       "\n",
       ".Input-text {\n",
       "    display: block;\n",
       "    margin: 10px;\n",
       "    color: inherit;\n",
       "    width: 140px;\n",
       "    background-color: var(--jq-numeric-input-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    padding: 5px;\n",
       "    padding-left: 10px;\n",
       "    font-family: inherit;\n",
       "    font-size: 20px;\n",
       "    font-weight: inherit;\n",
       "    line-height: 20pt;\n",
       "    border: none;\n",
       "    border-radius: 0.2rem;\n",
       "    transition: box-shadow 0.1s);\n",
       "}\n",
       "\n",
       ".Input-text:focus {\n",
       "    /*outline: none;*/\n",
       "    background-color: var(--jq-numeric-input-bg);\n",
       "    box-shadow: 0.6rem 0.8rem 1.4rem -0.5rem var(--jq-numeric-input-shadow);\n",
       "}\n",
       "\n",
       ".MCButton {\n",
       "    background: var(--jq-mc-button-bg);\n",
       "    border: 1px solid var(--jq-mc-button-border);\n",
       "    border-radius: inherit;\n",
       "    color: #333333;\n",
       "    padding: 10px;\n",
       "    font-size: 16px;\n",
       "    cursor: pointer;\n",
       "    text-align: center;\n",
       "    display: flex;\n",
       "    align-items: center;\n",
       "    justify-content: center;\n",
       "}\n",
       "\n",
       ".MCButton p {\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".MultipleChoiceQn {\n",
       "    padding: 10px;\n",
       "    background: var(--jq-multiple-choice-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".ManyChoiceQn {\n",
       "    padding: 10px;\n",
       "    background: var(--jq-many-choice-bg);\n",
       "    color: var(--jq-text-color);\n",
       "    border-radius: inherit;\n",
       "}\n",
       "\n",
       ".NumericQn {\n",
       "    background: var(--jq-numeric-bg);\n",
       "    border-radius: inherit;\n",
       "    color: var(--jq-text-color);\n",
       "    padding: 10px;\n",
       "}\n",
       "\n",
       ".NumericQn p {\n",
       "    color: inherit;\n",
       "}\n",
       "\n",
       ".InpLabel {\n",
       "    color: var(--jq-numeric-input-label);\n",
       "    float: left;\n",
       "    font-size: 15pt;\n",
       "    line-height: 34px;\n",
       "    margin-right: 10px;\n",
       "}\n",
       "\n",
       ".incorrect {\n",
       "    color: var(--jq-incorrect-color);\n",
       "}\n",
       "\n",
       ".correct {\n",
       "    color: var(--jq-correct-color);\n",
       "}\n",
       "\n",
       ".correctButton {\n",
       "    /*\n",
       "    background: var(--jq-correct-color);\n",
       "   */\n",
       "    animation: correct-anim 0.6s ease;\n",
       "    animation-fill-mode: forwards;\n",
       "    box-shadow: inset 0 0 5px var(--jq-mc-button-inset-shadow);\n",
       "    color: var(--jq-text-color);\n",
       "    /*outline: none;*/\n",
       "}\n",
       "\n",
       ".incorrectButton {\n",
       "    animation: incorrect-anim 0.8s ease;\n",
       "    animation-fill-mode: forwards;\n",
       "    box-shadow: inset 0 0 5px var(--jq-mc-button-inset-shadow);\n",
       "    color: var(--jq-text-color);\n",
       "    /*outline: none;*/\n",
       "}\n",
       "\n",
       "@keyframes incorrect-anim {\n",
       "    100% {\n",
       "        background-color: var(--jq-incorrect-color);\n",
       "    }\n",
       "}\n",
       "\n",
       "@keyframes correct-anim {\n",
       "    100% {\n",
       "        background-color: var(--jq-correct-color);\n",
       "    }\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "var questionsgQYrhhgfYdSG=[\n",
       "    {\n",
       "        \"question\": \"Select a single answer. What is the Ensembl Gene ID for the most highly expressed gene across the samples?\",\n",
       "        \"type\": \"multiple_choice\",\n",
       "        \"answers\": [\n",
       "            {\n",
       "                \"answer\": \"ENSDARG00000007576\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"False. This was the least expressed gene among the top 10 in the SRR3371417 sample.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"ENSDARG00000092553\",\n",
       "                \"correct\": true,\n",
       "                \"feedback\": \"True. ENSDARG00000092553 was the most highly expressed gene among all samples.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"ENSDARG00000036316\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Almost. ENSDARG00000036316 was highly expressed in all samples, but not the most highly expressed one.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"ENSDARG00000015653\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"False. This gene was not listed in any of the top 10 gene lists.\"\n",
       "            }\n",
       "        ]\n",
       "    },\n",
       "    {\n",
       "        \"question\": \"Select a single answer. How many reads did that gene have in the sample SRR21972723?\",\n",
       "        \"type\": \"multiple_choice\",\n",
       "        \"answers\": [\n",
       "            {\n",
       "                \"answer\": \"121331.73\",\n",
       "                \"correct\": true,\n",
       "                \"feedback\": \"True. The ENSDARG00000092553 gene has 121,331.73 reads in the SRR3371419 sample.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"124247.07\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is the number of reads for ENSDARG00000092553 in the SRR3371417 sample.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"118843.46\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is the number of reads for ENSDARG00000092553 in the SRR3371418 sample.\"\n",
       "            },\n",
       "            {\n",
       "                \"answer\": \"211331.73\",\n",
       "                \"correct\": false,\n",
       "                \"feedback\": \"Incorrect. This is not the correct number of reads for the ENSDARG00000092553 gene in the SRR3371419 sample.\"\n",
       "            }\n",
       "        ]\n",
       "    }\n",
       "]\n",
       ";\n",
       "    // Make a random ID\n",
       "function makeid(length) {\n",
       "    var result = [];\n",
       "    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';\n",
       "    var charactersLength = characters.length;\n",
       "    for (var i = 0; i < length; i++) {\n",
       "        result.push(characters.charAt(Math.floor(Math.random() * charactersLength)));\n",
       "    }\n",
       "    return result.join('');\n",
       "}\n",
       "\n",
       "// Choose a random subset of an array. Can also be used to shuffle the array\n",
       "function getRandomSubarray(arr, size) {\n",
       "    var shuffled = arr.slice(0), i = arr.length, temp, index;\n",
       "    while (i--) {\n",
       "        index = Math.floor((i + 1) * Math.random());\n",
       "        temp = shuffled[index];\n",
       "        shuffled[index] = shuffled[i];\n",
       "        shuffled[i] = temp;\n",
       "    }\n",
       "    return shuffled.slice(0, size);\n",
       "}\n",
       "\n",
       "function printResponses(responsesContainer) {\n",
       "    var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "    var stringResponses='<B>IMPORTANT!</B>To preserve this answer sequence for submission, when you have finalized your answers: <ol> <li> Copy the text in this cell below \"Answer String\"</li> <li> Double click on the cell directly below the Answer String, labeled \"Replace Me\"</li> <li> Select the whole \"Replace Me\" text</li> <li> Paste in your answer string and press shift-Enter.</li><li>Save the notebook using the save icon or File->Save Notebook menu item</li></ul><br><br><br><b>Answer String:</b><br> ';\n",
       "    console.log(responses);\n",
       "    responses.forEach((response, index) => {\n",
       "        if (response) {\n",
       "            console.log(index + ': ' + response);\n",
       "            stringResponses+= index + ': ' + response +\"<BR>\";\n",
       "        }\n",
       "    });\n",
       "    responsesContainer.innerHTML=stringResponses;\n",
       "}\n",
       "/* Callback function to determine whether a selected multiple-choice\n",
       "   button corresponded to a correct answer and to provide feedback\n",
       "   based on the answer */\n",
       "function check_mc() {\n",
       "    var id = this.id.split('-')[0];\n",
       "    //var response = this.id.split('-')[1];\n",
       "    //console.log(response);\n",
       "    //console.log(\"In check_mc(), id=\"+id);\n",
       "    //console.log(event.srcElement.id)           \n",
       "    //console.log(event.srcElement.dataset.correct)   \n",
       "    //console.log(event.srcElement.dataset.feedback)\n",
       "\n",
       "    var label = event.srcElement;\n",
       "    //console.log(label, label.nodeName);\n",
       "    var depth = 0;\n",
       "    while ((label.nodeName != \"LABEL\") && (depth < 20)) {\n",
       "        label = label.parentElement;\n",
       "        console.log(depth, label);\n",
       "        depth++;\n",
       "    }\n",
       "\n",
       "\n",
       "\n",
       "    var answers = label.parentElement.children;\n",
       "    //console.log(answers);\n",
       "\n",
       "    // Split behavior based on multiple choice vs many choice:\n",
       "    var fb = document.getElementById(\"fb\" + id);\n",
       "\n",
       "\n",
       "\n",
       "    /* Multiple choice (1 answer). Allow for 0 correct\n",
       "       answers as an edge case */\n",
       "    if (fb.dataset.numcorrect <= 1) {\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            //console.log(responsesContainer);\n",
       "            var response = label.firstChild.innerText;\n",
       "            if (label.querySelector(\".QuizCode\")){\n",
       "                response+= label.querySelector(\".QuizCode\").firstChild.innerText;\n",
       "            }\n",
       "            console.log(response);\n",
       "            //console.log(document.getElementById(\"quizWrap\"+id));\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            console.log(responses);\n",
       "            responses[qnum]= response;\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End code to preserve responses\n",
       "\n",
       "        for (var i = 0; i < answers.length; i++) {\n",
       "            var child = answers[i];\n",
       "            //console.log(child);\n",
       "            child.className = \"MCButton\";\n",
       "        }\n",
       "\n",
       "\n",
       "\n",
       "        if (label.dataset.correct == \"true\") {\n",
       "            // console.log(\"Correct action\");\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                fb.innerHTML = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                fb.innerHTML = \"Correct!\";\n",
       "            }\n",
       "            label.classList.add(\"correctButton\");\n",
       "\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"correct\");\n",
       "\n",
       "        } else {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                fb.innerHTML = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                fb.innerHTML = \"Incorrect -- try again.\";\n",
       "            }\n",
       "            //console.log(\"Error action\");\n",
       "            label.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "    }\n",
       "    else { /* Many choice (more than 1 correct answer) */\n",
       "        var reset = false;\n",
       "        var feedback;\n",
       "         if (label.dataset.correct == \"true\") {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                feedback = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                feedback = \"Correct!\";\n",
       "            }\n",
       "            if (label.dataset.answered <= 0) {\n",
       "                if (fb.dataset.answeredcorrect < 0) {\n",
       "                    fb.dataset.answeredcorrect = 1;\n",
       "                    reset = true;\n",
       "                } else {\n",
       "                    fb.dataset.answeredcorrect++;\n",
       "                }\n",
       "                if (reset) {\n",
       "                    for (var i = 0; i < answers.length; i++) {\n",
       "                        var child = answers[i];\n",
       "                        child.className = \"MCButton\";\n",
       "                        child.dataset.answered = 0;\n",
       "                    }\n",
       "                }\n",
       "                label.classList.add(\"correctButton\");\n",
       "                label.dataset.answered = 1;\n",
       "                fb.className = \"Feedback\";\n",
       "                fb.classList.add(\"correct\");\n",
       "\n",
       "            }\n",
       "        } else {\n",
       "            if (\"feedback\" in label.dataset) {\n",
       "                feedback = jaxify(label.dataset.feedback);\n",
       "            } else {\n",
       "                feedback = \"Incorrect -- try again.\";\n",
       "            }\n",
       "            if (fb.dataset.answeredcorrect > 0) {\n",
       "                fb.dataset.answeredcorrect = -1;\n",
       "                reset = true;\n",
       "            } else {\n",
       "                fb.dataset.answeredcorrect--;\n",
       "            }\n",
       "\n",
       "            if (reset) {\n",
       "                for (var i = 0; i < answers.length; i++) {\n",
       "                    var child = answers[i];\n",
       "                    child.className = \"MCButton\";\n",
       "                    child.dataset.answered = 0;\n",
       "                }\n",
       "            }\n",
       "            label.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            //console.log(responsesContainer);\n",
       "            var response = label.firstChild.innerText;\n",
       "            if (label.querySelector(\".QuizCode\")){\n",
       "                response+= label.querySelector(\".QuizCode\").firstChild.innerText;\n",
       "            }\n",
       "            console.log(response);\n",
       "            //console.log(document.getElementById(\"quizWrap\"+id));\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            if (label.dataset.correct == \"true\") {\n",
       "                if (typeof(responses[qnum]) == \"object\"){\n",
       "                    if (!responses[qnum].includes(response))\n",
       "                        responses[qnum].push(response);\n",
       "                } else{\n",
       "                    responses[qnum]= [ response ];\n",
       "                }\n",
       "            } else {\n",
       "                responses[qnum]= response;\n",
       "            }\n",
       "            console.log(responses);\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End save responses stuff\n",
       "\n",
       "\n",
       "\n",
       "        var numcorrect = fb.dataset.numcorrect;\n",
       "        var answeredcorrect = fb.dataset.answeredcorrect;\n",
       "        if (answeredcorrect >= 0) {\n",
       "            fb.innerHTML = feedback + \" [\" + answeredcorrect + \"/\" + numcorrect + \"]\";\n",
       "        } else {\n",
       "            fb.innerHTML = feedback + \" [\" + 0 + \"/\" + numcorrect + \"]\";\n",
       "        }\n",
       "\n",
       "\n",
       "    }\n",
       "\n",
       "    if (typeof MathJax != 'undefined') {\n",
       "        var version = MathJax.version;\n",
       "        console.log('MathJax version', version);\n",
       "        if (version[0] == \"2\") {\n",
       "            MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "        } else if (version[0] == \"3\") {\n",
       "            MathJax.typeset([fb]);\n",
       "        }\n",
       "    } else {\n",
       "        console.log('MathJax not detected');\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "\n",
       "/* Function to produce the HTML buttons for a multiple choice/\n",
       "   many choice question  and to update the CSS tags based on\n",
       "   the question type */\n",
       "function make_mc(qa, shuffle_answers, outerqDiv, qDiv, aDiv, id) {\n",
       "\n",
       "    var shuffled;\n",
       "    if (shuffle_answers == \"True\") {\n",
       "        //console.log(shuffle_answers+\" read as true\");\n",
       "        shuffled = getRandomSubarray(qa.answers, qa.answers.length);\n",
       "    } else {\n",
       "        //console.log(shuffle_answers+\" read as false\");\n",
       "        shuffled = qa.answers;\n",
       "    }\n",
       "\n",
       "\n",
       "    var num_correct = 0;\n",
       "\n",
       "    shuffled.forEach((item, index, ans_array) => {\n",
       "        //console.log(answer);\n",
       "\n",
       "        // Make input element\n",
       "        var inp = document.createElement(\"input\");\n",
       "        inp.type = \"radio\";\n",
       "        inp.id = \"quizo\" + id + index;\n",
       "        inp.style = \"display:none;\";\n",
       "        aDiv.append(inp);\n",
       "\n",
       "        //Make label for input element\n",
       "        var lab = document.createElement(\"label\");\n",
       "        lab.className = \"MCButton\";\n",
       "        lab.id = id + '-' + index;\n",
       "        lab.onclick = check_mc;\n",
       "        var aSpan = document.createElement('span');\n",
       "        aSpan.classsName = \"\";\n",
       "        //qDiv.id=\"quizQn\"+id+index;\n",
       "        if (\"answer\" in item) {\n",
       "            aSpan.innerHTML = jaxify(item.answer);\n",
       "            //aSpan.innerHTML=item.answer;\n",
       "        }\n",
       "        lab.append(aSpan);\n",
       "\n",
       "        // Create div for code inside question\n",
       "        var codeSpan;\n",
       "        if (\"code\" in item) {\n",
       "            codeSpan = document.createElement('span');\n",
       "            codeSpan.id = \"code\" + id + index;\n",
       "            codeSpan.className = \"QuizCode\";\n",
       "            var codePre = document.createElement('pre');\n",
       "            codeSpan.append(codePre);\n",
       "            var codeCode = document.createElement('code');\n",
       "            codePre.append(codeCode);\n",
       "            codeCode.innerHTML = item.code;\n",
       "            lab.append(codeSpan);\n",
       "            //console.log(codeSpan);\n",
       "        }\n",
       "\n",
       "        //lab.textContent=item.answer;\n",
       "\n",
       "        // Set the data attributes for the answer\n",
       "        lab.setAttribute('data-correct', item.correct);\n",
       "        if (item.correct) {\n",
       "            num_correct++;\n",
       "        }\n",
       "        if (\"feedback\" in item) {\n",
       "            lab.setAttribute('data-feedback', item.feedback);\n",
       "        }\n",
       "        lab.setAttribute('data-answered', 0);\n",
       "\n",
       "        aDiv.append(lab);\n",
       "\n",
       "    });\n",
       "\n",
       "    if (num_correct > 1) {\n",
       "        outerqDiv.className = \"ManyChoiceQn\";\n",
       "    } else {\n",
       "        outerqDiv.className = \"MultipleChoiceQn\";\n",
       "    }\n",
       "\n",
       "    return num_correct;\n",
       "\n",
       "}\n",
       "function check_numeric(ths, event) {\n",
       "\n",
       "    if (event.keyCode === 13) {\n",
       "        ths.blur();\n",
       "\n",
       "        var id = ths.id.split('-')[0];\n",
       "\n",
       "        var submission = ths.value;\n",
       "        if (submission.indexOf('/') != -1) {\n",
       "            var sub_parts = submission.split('/');\n",
       "            //console.log(sub_parts);\n",
       "            submission = sub_parts[0] / sub_parts[1];\n",
       "        }\n",
       "        //console.log(\"Reader entered\", submission);\n",
       "\n",
       "        if (\"precision\" in ths.dataset) {\n",
       "            var precision = ths.dataset.precision;\n",
       "            submission = Number(Number(submission).toPrecision(precision));\n",
       "        }\n",
       "\n",
       "\n",
       "        //console.log(\"In check_numeric(), id=\"+id);\n",
       "        //console.log(event.srcElement.id)           \n",
       "        //console.log(event.srcElement.dataset.feedback)\n",
       "\n",
       "        var fb = document.getElementById(\"fb\" + id);\n",
       "        fb.style.display = \"none\";\n",
       "        fb.innerHTML = \"Incorrect -- try again.\";\n",
       "\n",
       "        var answers = JSON.parse(ths.dataset.answers);\n",
       "        //console.log(answers);\n",
       "\n",
       "        var defaultFB = \"Incorrect. Try again.\";\n",
       "        var correct;\n",
       "        var done = false;\n",
       "        answers.every(answer => {\n",
       "            //console.log(answer.type);\n",
       "\n",
       "            correct = false;\n",
       "            // if (answer.type==\"value\"){\n",
       "            if ('value' in answer) {\n",
       "                if (submission == answer.value) {\n",
       "                    if (\"feedback\" in answer) {\n",
       "                        fb.innerHTML = jaxify(answer.feedback);\n",
       "                    } else {\n",
       "                        fb.innerHTML = jaxify(\"Correct\");\n",
       "                    }\n",
       "                    correct = answer.correct;\n",
       "                    //console.log(answer.correct);\n",
       "                    done = true;\n",
       "                }\n",
       "                // } else if (answer.type==\"range\") {\n",
       "            } else if ('range' in answer) {\n",
       "                console.log(answer.range);\n",
       "                console.log(submission, submission >=answer.range[0], submission < answer.range[1])\n",
       "                if ((submission >= answer.range[0]) && (submission < answer.range[1])) {\n",
       "                    fb.innerHTML = jaxify(answer.feedback);\n",
       "                    correct = answer.correct;\n",
       "                    console.log(answer.correct);\n",
       "                    done = true;\n",
       "                }\n",
       "            } else if (answer.type == \"default\") {\n",
       "                if (\"feedback\" in answer) {\n",
       "                    defaultFB = answer.feedback;\n",
       "                } \n",
       "            }\n",
       "            if (done) {\n",
       "                return false; // Break out of loop if this has been marked correct\n",
       "            } else {\n",
       "                return true; // Keep looking for case that includes this as a correct answer\n",
       "            }\n",
       "        });\n",
       "        console.log(\"done:\", done);\n",
       "\n",
       "        if ((!done) && (defaultFB != \"\")) {\n",
       "            fb.innerHTML = jaxify(defaultFB);\n",
       "            //console.log(\"Default feedback\", defaultFB);\n",
       "        }\n",
       "\n",
       "        fb.style.display = \"block\";\n",
       "        if (correct) {\n",
       "            ths.className = \"Input-text\";\n",
       "            ths.classList.add(\"correctButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"correct\");\n",
       "        } else {\n",
       "            ths.className = \"Input-text\";\n",
       "            ths.classList.add(\"incorrectButton\");\n",
       "            fb.className = \"Feedback\";\n",
       "            fb.classList.add(\"incorrect\");\n",
       "        }\n",
       "\n",
       "        // What follows is for the saved responses stuff\n",
       "        var outerContainer = fb.parentElement.parentElement;\n",
       "        var responsesContainer = document.getElementById(\"responses\" + outerContainer.id);\n",
       "        if (responsesContainer) {\n",
       "            console.log(submission);\n",
       "            var qnum = document.getElementById(\"quizWrap\"+id).dataset.qnum;\n",
       "            //console.log(\"Question \" + qnum);\n",
       "            //console.log(id, \", got numcorrect=\",fb.dataset.numcorrect);\n",
       "            var responses=JSON.parse(responsesContainer.dataset.responses);\n",
       "            console.log(responses);\n",
       "            if (submission == ths.value){\n",
       "                responses[qnum]= submission;\n",
       "            } else {\n",
       "                responses[qnum]= ths.value + \"(\" + submission +\")\";\n",
       "            }\n",
       "            responsesContainer.setAttribute('data-responses', JSON.stringify(responses));\n",
       "            printResponses(responsesContainer);\n",
       "        }\n",
       "        // End code to preserve responses\n",
       "\n",
       "        if (typeof MathJax != 'undefined') {\n",
       "            var version = MathJax.version;\n",
       "            console.log('MathJax version', version);\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                MathJax.typeset([fb]);\n",
       "            }\n",
       "        } else {\n",
       "            console.log('MathJax not detected');\n",
       "        }\n",
       "        return false;\n",
       "    }\n",
       "\n",
       "}\n",
       "\n",
       "function isValid(el, charC) {\n",
       "    //console.log(\"Input char: \", charC);\n",
       "    if (charC == 46) {\n",
       "        if (el.value.indexOf('.') === -1) {\n",
       "            return true;\n",
       "        } else if (el.value.indexOf('/') != -1) {\n",
       "            var parts = el.value.split('/');\n",
       "            if (parts[1].indexOf('.') === -1) {\n",
       "                return true;\n",
       "            }\n",
       "        }\n",
       "        else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 47) {\n",
       "        if (el.value.indexOf('/') === -1) {\n",
       "            if ((el.value != \"\") && (el.value != \".\")) {\n",
       "                return true;\n",
       "            } else {\n",
       "                return false;\n",
       "            }\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 45) {\n",
       "        var edex = el.value.indexOf('e');\n",
       "        if (edex == -1) {\n",
       "            edex = el.value.indexOf('E');\n",
       "        }\n",
       "\n",
       "        if (el.value == \"\") {\n",
       "            return true;\n",
       "        } else if (edex == (el.value.length - 1)) { // If just after e or E\n",
       "            return true;\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else if (charC == 101) { // \"e\"\n",
       "        if ((el.value.indexOf('e') === -1) && (el.value.indexOf('E') === -1) && (el.value.indexOf('/') == -1)) {\n",
       "            // Prev symbol must be digit or decimal point:\n",
       "            if (el.value.slice(-1).search(/\\d/) >= 0) {\n",
       "                return true;\n",
       "            } else if (el.value.slice(-1).search(/\\./) >= 0) {\n",
       "                return true;\n",
       "            } else {\n",
       "                return false;\n",
       "            }\n",
       "        } else {\n",
       "            return false;\n",
       "        }\n",
       "    } else {\n",
       "        if (charC > 31 && (charC < 48 || charC > 57))\n",
       "            return false;\n",
       "    }\n",
       "    return true;\n",
       "}\n",
       "\n",
       "function numeric_keypress(evnt) {\n",
       "    var charC = (evnt.which) ? evnt.which : evnt.keyCode;\n",
       "\n",
       "    if (charC == 13) {\n",
       "        check_numeric(this, evnt);\n",
       "    } else {\n",
       "        return isValid(this, charC);\n",
       "    }\n",
       "}\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "function make_numeric(qa, outerqDiv, qDiv, aDiv, id) {\n",
       "\n",
       "\n",
       "\n",
       "    //console.log(answer);\n",
       "\n",
       "\n",
       "    outerqDiv.className = \"NumericQn\";\n",
       "    aDiv.style.display = 'block';\n",
       "\n",
       "    var lab = document.createElement(\"label\");\n",
       "    lab.className = \"InpLabel\";\n",
       "    lab.innerHTML = \"Type numeric answer here:\";\n",
       "    aDiv.append(lab);\n",
       "\n",
       "    var inp = document.createElement(\"input\");\n",
       "    inp.type = \"text\";\n",
       "    //inp.id=\"input-\"+id;\n",
       "    inp.id = id + \"-0\";\n",
       "    inp.className = \"Input-text\";\n",
       "    inp.setAttribute('data-answers', JSON.stringify(qa.answers));\n",
       "    if (\"precision\" in qa) {\n",
       "        inp.setAttribute('data-precision', qa.precision);\n",
       "    }\n",
       "    aDiv.append(inp);\n",
       "    //console.log(inp);\n",
       "\n",
       "    //inp.addEventListener(\"keypress\", check_numeric);\n",
       "    //inp.addEventListener(\"keypress\", numeric_keypress);\n",
       "    /*\n",
       "    inp.addEventListener(\"keypress\", function(event) {\n",
       "        return numeric_keypress(this, event);\n",
       "    }\n",
       "                        );\n",
       "                        */\n",
       "    //inp.onkeypress=\"return numeric_keypress(this, event)\";\n",
       "    inp.onkeypress = numeric_keypress;\n",
       "    inp.onpaste = event => false;\n",
       "\n",
       "    inp.addEventListener(\"focus\", function (event) {\n",
       "        this.value = \"\";\n",
       "        return false;\n",
       "    }\n",
       "    );\n",
       "\n",
       "\n",
       "}\n",
       "function jaxify(string) {\n",
       "    var mystring = string;\n",
       "\n",
       "    var count = 0;\n",
       "    var loc = mystring.search(/([^\\\\]|^)(\\$)/);\n",
       "\n",
       "    var count2 = 0;\n",
       "    var loc2 = mystring.search(/([^\\\\]|^)(\\$\\$)/);\n",
       "\n",
       "    //console.log(loc);\n",
       "\n",
       "    while ((loc >= 0) || (loc2 >= 0)) {\n",
       "\n",
       "        /* Have to replace all the double $$ first with current implementation */\n",
       "        if (loc2 >= 0) {\n",
       "            if (count2 % 2 == 0) {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$\\$)/, \"$1\\\\[\");\n",
       "            } else {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$\\$)/, \"$1\\\\]\");\n",
       "            }\n",
       "            count2++;\n",
       "        } else {\n",
       "            if (count % 2 == 0) {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$)/, \"$1\\\\(\");\n",
       "            } else {\n",
       "                mystring = mystring.replace(/([^\\\\]|^)(\\$)/, \"$1\\\\)\");\n",
       "            }\n",
       "            count++;\n",
       "        }\n",
       "        loc = mystring.search(/([^\\\\]|^)(\\$)/);\n",
       "        loc2 = mystring.search(/([^\\\\]|^)(\\$\\$)/);\n",
       "        //console.log(mystring,\", loc:\",loc,\", loc2:\",loc2);\n",
       "    }\n",
       "\n",
       "    // repace markdown style links with actual links\n",
       "    mystring = mystring.replace(/<(.*?)>/, '<a href=\"$1\" target=\"_blank\" class=\"Link\">$1</a>');\n",
       "    mystring = mystring.replace(/\\[(.*?)\\]\\((.*?)\\)/, '<a href=\"$2\" target=\"_blank\" class=\"Link\">$1</a>');\n",
       "\n",
       "    //console.log(mystring);\n",
       "    return mystring;\n",
       "}\n",
       "\n",
       "\n",
       "function show_questions(json, mydiv) {\n",
       "    console.log('show_questions');\n",
       "    //var mydiv=document.getElementById(myid);\n",
       "    var shuffle_questions = mydiv.dataset.shufflequestions;\n",
       "    var num_questions = mydiv.dataset.numquestions;\n",
       "    var shuffle_answers = mydiv.dataset.shuffleanswers;\n",
       "    var max_width = mydiv.dataset.maxwidth;\n",
       "\n",
       "    if (num_questions > json.length) {\n",
       "        num_questions = json.length;\n",
       "    }\n",
       "\n",
       "    var questions;\n",
       "    if ((num_questions < json.length) || (shuffle_questions == \"True\")) {\n",
       "        //console.log(num_questions+\",\"+json.length);\n",
       "        questions = getRandomSubarray(json, num_questions);\n",
       "    } else {\n",
       "        questions = json;\n",
       "    }\n",
       "\n",
       "    //console.log(\"SQ: \"+shuffle_questions+\", NQ: \" + num_questions + \", SA: \", shuffle_answers);\n",
       "\n",
       "    // Iterate over questions\n",
       "    questions.forEach((qa, index, array) => {\n",
       "        //console.log(qa.question); \n",
       "\n",
       "        var id = makeid(8);\n",
       "        //console.log(id);\n",
       "\n",
       "\n",
       "        // Create Div to contain question and answers\n",
       "        var iDiv = document.createElement('div');\n",
       "        //iDiv.id = 'quizWrap' + id + index;\n",
       "        iDiv.id = 'quizWrap' + id;\n",
       "        iDiv.className = 'Quiz';\n",
       "        iDiv.setAttribute('data-qnum', index);\n",
       "        iDiv.style.maxWidth  =max_width+\"px\";\n",
       "        mydiv.appendChild(iDiv);\n",
       "        // iDiv.innerHTML=qa.question;\n",
       "        \n",
       "        var outerqDiv = document.createElement('div');\n",
       "        outerqDiv.id = \"OuterquizQn\" + id + index;\n",
       "        // Create div to contain question part\n",
       "        var qDiv = document.createElement('div');\n",
       "        qDiv.id = \"quizQn\" + id + index;\n",
       "        \n",
       "        if (qa.question) {\n",
       "            iDiv.append(outerqDiv);\n",
       "\n",
       "            //qDiv.textContent=qa.question;\n",
       "            qDiv.innerHTML = jaxify(qa.question);\n",
       "            outerqDiv.append(qDiv);\n",
       "        }\n",
       "\n",
       "        // Create div for code inside question\n",
       "        var codeDiv;\n",
       "        if (\"code\" in qa) {\n",
       "            codeDiv = document.createElement('div');\n",
       "            codeDiv.id = \"code\" + id + index;\n",
       "            codeDiv.className = \"QuizCode\";\n",
       "            var codePre = document.createElement('pre');\n",
       "            codeDiv.append(codePre);\n",
       "            var codeCode = document.createElement('code');\n",
       "            codePre.append(codeCode);\n",
       "            codeCode.innerHTML = qa.code;\n",
       "            outerqDiv.append(codeDiv);\n",
       "            //console.log(codeDiv);\n",
       "        }\n",
       "\n",
       "\n",
       "        // Create div to contain answer part\n",
       "        var aDiv = document.createElement('div');\n",
       "        aDiv.id = \"quizAns\" + id + index;\n",
       "        aDiv.className = 'Answer';\n",
       "        iDiv.append(aDiv);\n",
       "\n",
       "        //console.log(qa.type);\n",
       "\n",
       "        var num_correct;\n",
       "        if ((qa.type == \"multiple_choice\") || (qa.type == \"many_choice\") ) {\n",
       "            num_correct = make_mc(qa, shuffle_answers, outerqDiv, qDiv, aDiv, id);\n",
       "            if (\"answer_cols\" in qa) {\n",
       "                //aDiv.style.gridTemplateColumns = 'auto '.repeat(qa.answer_cols);\n",
       "                aDiv.style.gridTemplateColumns = 'repeat(' + qa.answer_cols + ', 1fr)';\n",
       "            }\n",
       "        } else if (qa.type == \"numeric\") {\n",
       "            //console.log(\"numeric\");\n",
       "            make_numeric(qa, outerqDiv, qDiv, aDiv, id);\n",
       "        }\n",
       "\n",
       "\n",
       "        //Make div for feedback\n",
       "        var fb = document.createElement(\"div\");\n",
       "        fb.id = \"fb\" + id;\n",
       "        //fb.style=\"font-size: 20px;text-align:center;\";\n",
       "        fb.className = \"Feedback\";\n",
       "        fb.setAttribute(\"data-answeredcorrect\", 0);\n",
       "        fb.setAttribute(\"data-numcorrect\", num_correct);\n",
       "        iDiv.append(fb);\n",
       "\n",
       "\n",
       "    });\n",
       "    var preserveResponses = mydiv.dataset.preserveresponses;\n",
       "    console.log(preserveResponses);\n",
       "    console.log(preserveResponses == \"true\");\n",
       "    if (preserveResponses == \"true\") {\n",
       "        console.log(preserveResponses);\n",
       "        // Create Div to contain record of answers\n",
       "        var iDiv = document.createElement('div');\n",
       "        iDiv.id = 'responses' + mydiv.id;\n",
       "        iDiv.className = 'JCResponses';\n",
       "        // Create a place to store responses as an empty array\n",
       "        iDiv.setAttribute('data-responses', '[]');\n",
       "\n",
       "        // Dummy Text\n",
       "        iDiv.innerHTML=\"<b>Select your answers and then follow the directions that will appear here.</b>\"\n",
       "        //iDiv.className = 'Quiz';\n",
       "        mydiv.appendChild(iDiv);\n",
       "    }\n",
       "//console.log(\"At end of show_questions\");\n",
       "    if (typeof MathJax != 'undefined') {\n",
       "        console.log(\"MathJax version\", MathJax.version);\n",
       "        var version = MathJax.version;\n",
       "        setTimeout(function(){\n",
       "            var version = MathJax.version;\n",
       "            console.log('After sleep, MathJax version', version);\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                if (MathJax.hasOwnProperty('typeset') ) {\n",
       "                    MathJax.typeset([mydiv]);\n",
       "                } else {\n",
       "                    console.log('WARNING: Trying to force load MathJax 3');\n",
       "                    window.MathJax = {\n",
       "                        tex: {\n",
       "                            inlineMath: [['$', '$'], ['\\\\(', '\\\\)']]\n",
       "                        },\n",
       "                        svg: {\n",
       "                            fontCache: 'global'\n",
       "                        }\n",
       "                    };\n",
       "\n",
       "                    (function () {\n",
       "                        var script = document.createElement('script');\n",
       "                        script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js';\n",
       "                        script.async = true;\n",
       "                        document.head.appendChild(script);\n",
       "                    })();\n",
       "                }\n",
       "            }\n",
       "        }, 500);\n",
       "if (typeof version == 'undefined') {\n",
       "        } else\n",
       "        {\n",
       "            if (version[0] == \"2\") {\n",
       "                MathJax.Hub.Queue([\"Typeset\", MathJax.Hub]);\n",
       "            } else if (version[0] == \"3\") {\n",
       "                if (MathJax.hasOwnProperty('typeset') ) {\n",
       "                    MathJax.typeset([mydiv]);\n",
       "                } else {\n",
       "                    console.log('WARNING: Trying to force load MathJax 3');\n",
       "                    window.MathJax = {\n",
       "                        tex: {\n",
       "                            inlineMath: [['$', '$'], ['\\\\(', '\\\\)']]\n",
       "                        },\n",
       "                        svg: {\n",
       "                            fontCache: 'global'\n",
       "                        }\n",
       "                    };\n",
       "\n",
       "                    (function () {\n",
       "                        var script = document.createElement('script');\n",
       "                        script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-svg.js';\n",
       "                        script.async = true;\n",
       "                        document.head.appendChild(script);\n",
       "                    })();\n",
       "                }\n",
       "            } else {\n",
       "                console.log(\"MathJax not found\");\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    // stop event propagation for the .Link class\n",
       "    var links = document.getElementsByClassName('Link')\n",
       "    for (var i = 0; i < links.length; i++) {\n",
       "        links[i].addEventListener('click', function(e){\n",
       "            e.stopPropagation();\n",
       "        });\n",
       "    }\n",
       "\n",
       "    return false;\n",
       "}\n",
       "/* This is to handle asynchrony issues in loading Jupyter notebooks\n",
       "           where the quiz has been previously run. The Javascript was generally\n",
       "           being run before the div was added to the DOM. I tried to do this\n",
       "           more elegantly using Mutation Observer, but I didn't get it to work.\n",
       "\n",
       "           Someone more knowledgeable could make this better ;-) */\n",
       "\n",
       "        function try_show() {\n",
       "          if(document.getElementById(\"gQYrhhgfYdSG\")) {\n",
       "            show_questions(questionsgQYrhhgfYdSG,  gQYrhhgfYdSG); \n",
       "          } else {\n",
       "             setTimeout(try_show, 200);\n",
       "          }\n",
       "        };\n",
       "    \n",
       "        {\n",
       "        // console.log(element);\n",
       "\n",
       "        //console.log(\"gQYrhhgfYdSG\");\n",
       "        // console.log(document.getElementById(\"gQYrhhgfYdSG\"));\n",
       "\n",
       "        try_show();\n",
       "        }\n",
       "        "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#display Multiple Choice quiz\n",
    "display_quiz(\"quiz/bulkRNADR2.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 10 most highly expressed genes in the double lysogen samples.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### STEP 10: Report the expression of irg1l for each file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `grep` to report the expression in the wild-type sample. The fields in the RSEM `genes.results` file are as follows. The level of expression is reported in the Transcripts Per Million (`TPM`) and number of reads (`NumReads`) fields:  \n",
    "`Name    Length  EffectiveLength TPM     NumReads`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TPM for ENSDARG00000092553 in SRR3371417: 124247.07\n",
      "TPM for ENSDARG00000092553 in SRR3371418: 118843.46\n",
      "TPM for ENSDARG00000092553 in SRR3371419: 121331.73\n",
      "TPM for ENSDARG00000092553 in SRR3371420: 119090.85\n",
      "TPM for ENSDARG00000092553 in SRR3371421: 119892.68\n",
      "TPM for ENSDARG00000092553 in SRR3371422: 132616.49\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Path to RSEM results directory\n",
    "rsem_results_dir = 'data_14/rsem_output'\n",
    "\n",
    "# Target gene ID\n",
    "target_gene = 'ENSDARG00000092553'\n",
    "\n",
    "# Loop through each file in accs.txt\n",
    "for srr_id in open('accs.txt'):\n",
    "    srr_id = srr_id.strip()  # Remove newline character\n",
    "    rsem_result_file = f'{rsem_results_dir}/{srr_id}.genes.results'\n",
    "\n",
    "    # Load the RSEM results into a Pandas DataFrame\n",
    "    df = pd.read_csv(rsem_result_file, sep='\\t')\n",
    "\n",
    "    # Filter for the target gene\n",
    "    target_gene_data = df[df['gene_id'] == target_gene]\n",
    "\n",
    "    # Print the target gene's TPM value for the SRR ID\n",
    "    print(f\"TPM for {target_gene} in {srr_id}: {target_gene_data['TPM'].values[0]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## <a name=\"workflow\">Additional Workflows</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that you have read counts per gene, feel free to explore the R workflow which creates plots and analyses using these readcount files, or try other alternate workflows for creating read count files, such as using snakemake."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "[Workflow One (Extended):](Tutorial_1B_Extended_zebrafish.ipynb) An extended version of workflow one. Once you have got your feet wet, you can retry workflow one with this extended version that covers the entire dataset.\n",
    "\n",
    "[Workflow Two (DEG Analysis):](Tutorial_2_DEG_Analysis_zebrafish.ipynb) Using Deseq2 and R to conduct clustering and differential gene expression analysis.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": ".m112",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/:m112"
  },
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
